% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllGenerics.R, R/plotGenderMarkers-methods.R
\docType{methods}
\name{plotGenderMarkers}
\alias{plotGenderMarkers}
\alias{plotGenderMarkers,SummarizedExperiment-method}
\alias{plotGenderMarkers,SingleCellExperiment-method}
\title{Plot sexually dimorphic gender marker genes}
\usage{
plotGenderMarkers(object, ...)

\S4method{plotGenderMarkers}{SummarizedExperiment}(object, assay = 1L,
  interestingGroups = NULL, countsAxisLabel = "counts",
  medianLine = TRUE, color = getOption("basejump.discrete.color",
  NULL), legend = getOption("basejump.legend", TRUE), style = "wide")
}
\arguments{
\item{object}{Object.}

\item{...}{Additional arguments.}

\item{assay}{\code{vector(1)}.
Name or index of count matrix slotted in \code{assays}. When passing in a
string, the name must be defined in \code{assayNames}.}

\item{interestingGroups}{\code{character}.
Groups of interest that define the samples. If left unset, defaults to
\code{sampleName}.}

\item{countsAxisLabel}{\code{character(1)}.
Label to use for the counts axis.}

\item{medianLine}{\code{logical(1)}.
Include median line for each group. Disabled if samples are colored by
sample name.}

\item{color}{\code{ScaleDiscrete}.
Desired ggplot2 color scale. Must supply discrete values. When set \code{NULL},
the default ggplot2 color palette will be used. If manual color definitions
are desired, we recommend using \code{ggplot2::scale_color_manual}.

To set the discrete color palette globally, use:\preformatted{options(basejump.color.discrete = ggplot2::scale_color_viridis_d())
}}

\item{legend}{\code{logical(1)}.
Show plot legend.}

\item{style}{\code{character(1)}.
Plot style.}
}
\value{
\code{ggplot}.
}
\description{
This is a convenience function that wraps \code{plotGene} to quickly plot known
sexually dimorphic genes, and overall gene experssion for the X and Y
chromosomes.
}
\details{
Currently only \emph{Homo sapiens} and \emph{Mus musculus} genomes are supported.
}
\seealso{
\code{plotGene}.
}
