---
title: "basejump"
author: "Michael Steinbaugh"
date: "`r BiocStyle::doc_date()`"
output: BiocStyle::html_document
bibliography: ../inst/rmarkdown/shared/bibliography.bib
vignette: >
  %\VignetteIndexEntry{basejump}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
abstract: >
  The basejump package is an infrastructure toolkit that extends the base
  functionality of Bioconductor [@Huber2015-nw]. The package leverages the S4
  object system for object-oriented programming, and defines multiple additional
  generic functions for use in genomics research. basejump provides simple,
  user-friendly functions for the acquisition of genome annotations from
  multiple online databases, including native support for Ensembl
  [@Hubbard2002-ah] and websites supporting standard FASTA and GTF/GFF file
  formats. Consistent handling of sample metadata remains a challenge for many
  bioinformatic analyses, and basejump aims to address this by providing a suite
  of sanitization functions to help standardize these variable inputs.
  Additionally, interactive read/write operations in R can be cumbersome and
  non-trivial when working with multiple data objects; here we provide
  additional functions designed for interactive use that aim to reduce friction
  and provide consistent handling of multiple common file formats used in
  genomics research.
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(message = FALSE)
```

# Introduction

```{r}
library(basejump)
library(SummarizedExperiment)
options(acid.test = TRUE)
data(RangedSummarizedExperiment, package = "acidtest")
rse <- RangedSummarizedExperiment
```

# The S4 object system

The [basejump][] package defines generics and methods for object-oriented programming with the S4 object system, which is used extensively by the [Bioconductor][] project. These resources describe the S4 object system in detail:

- [S4 classes and methods](https://www.bioconductor.org/help/course-materials/2017/Zurich/S4-classes-and-methods.html), Martin Morgan and Hervé Pagès, 2017.
- [A quick overview of the S4 class system](https://bioconductor.org/packages/devel/bioc/vignettes/S4Vectors/inst/doc/S4QuickOverview.pdf), Hervé Pagès, 2016.
- [The S4 object system](http://adv-r.had.co.nz/S4.html), Hadley Wickham.
- [Bioconductor for Genomic Data Science](http://kasperdanielhansen.github.io/genbioconductor), Kasper Hansen, 2016.
- [S4 System Development in Bioconductor](https://www.bioconductor.org/help/course-materials/2010/AdvancedR/S4InBioconductor.pdf), Patrick Aboyoun, 2010.

In [R][], you can check whether a function is using the S4 object system with the `isS4()` function. Additionally, `showMethods()` and `getMethod()` are useful for exploring source code of S4 methods. In contrast, to obtain information on functions using the [S3][] object system, use `methods()` instead.

# R session information

```{r}
utils::sessionInfo()
```

# References

The papers and software cited in our workflows are available as a [shared library](https://paperpile.com/shared/agxufd) on [Paperpile][].

[annotables]: https://github.com/stephenturner/annotables/
[AnnotationHub]: https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html
[basejump]: https://basejump.acidgenomics.com/
[Bioconductor]: https://bioconductor.org/
[EnsDb.Hsapiens.v75]: https://bioconductor.org/packages/EnsDb.Hsapiens.v75/
[Ensembl]: https://www.ensembl.org/
[ensembldb]: https://bioconductor.org/packages/ensembldb/
[FASTA]: https://www.ncbi.nlm.nih.gov/BLAST/fasta.shtml
[FlyBase]: http://flybase.org/
[GenomeInfoDb]: https://bioconductor.org/packages/GenomeInfoDb/
[GFF]: https://useast.ensembl.org/info/website/upload/gff.html  "General Feature Format"
[GTF]: http://mblab.wustl.edu/GTF22.html  "General Transfer Format"
[Gencode]: https://www.gencodegenes.org/
[GenomicRanges]: https://bioconductor.org/packages/GenomicRanges/
[ggplot2]: https://ggplot2.tidyverse.org/
[GRCh37]: https://grch37.ensembl.org/
[hg19]: https://genome.ucsc.edu/cgi-bin/hgGateway?db=hg19
[IRanges]: https://www.bioconductor.org/packages/IRanges/
[Paperpile]: https://paperpile.com/
[pheatmap]: https://cran.r-project.org/package=pheatmap
[R]: https://www.r-project.org/
[R Markdown]: https://rmarkdown.rstudio.com/
[r-wormbase]: https://wormbase.acidgenomics.com/
[RefSeq]: https://www.ncbi.nlm.nih.gov/refseq/
[rio]: https://cran.r-project.org/package=rio
[rtracklayer]: https://bioconductor.org/packages/rtracklayer/
[S3]: http://adv-r.had.co.nz/S3.html
[SummarizedExperiment]: https://bioconductor.org/packages/SummarizedExperiment/
[UCSC]: http://genome.ucsc.edu/
[WormBase]: https://wormbase.org/
