<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Changelog • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Changelog" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small></small></h1>
      <small>Source: <a href='https://github.com/steinbaugh/basejump/blob/master/NEWS.md'><code>NEWS.md</code></a></small>
    </div>

    <div id="basejump-0-9-5-2019-01-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-5-2019-01-22" class="anchor"></a>basejump 0.9.5 (2019-01-22)</h2>
</div>
    <div id="basejump-0-9-4-2019-01-12" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-4-2019-01-12" class="anchor"></a>basejump 0.9.4 (2019-01-12)</h2>
</div>
    <div id="basejump-0-9-3-2019-01-08" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-3-2019-01-08" class="anchor"></a>basejump 0.9.3 (2019-01-08)</h2>
</div>
    <div id="basejump-0-9-2-2019-01-07" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-2-2019-01-07" class="anchor"></a>basejump 0.9.2 (2019-01-07)</h2>
</div>
    <div id="basejump-0-9-1-2018-12-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-1-2018-12-22" class="anchor"></a>basejump 0.9.1 (2018-12-22)</h2>
</div>
    <div id="basejump-0-9-0-2018-12-12" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-0-2018-12-12" class="anchor"></a>basejump 0.9.0 (2018-12-12)</h2>
</div>
    <div id="basejump-0-8-8-2018-12-03" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-8-2018-12-03" class="anchor"></a>basejump 0.8.8 (2018-12-03)</h2>
</div>
    <div id="basejump-0-8-6-2018-11-30" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-6-2018-11-30" class="anchor"></a>basejump 0.8.6 (2018-11-30)</h2>
</div>
    <div id="basejump-0-8-5-2018-11-29" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-5-2018-11-29" class="anchor"></a>basejump 0.8.5 (2018-11-29)</h2>
</div>
    <div id="basejump-0-8-4-2018-11-28" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-4-2018-11-28" class="anchor"></a>basejump 0.8.4 (2018-11-28)</h2>
</div>
    <div id="basejump-0-8-3-2018-11-25" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-3-2018-11-25" class="anchor"></a>basejump 0.8.3 (2018-11-25)</h2>
</div>
    <div id="basejump-0-8-2-2018-11-19" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-2-2018-11-19" class="anchor"></a>basejump 0.8.2 (2018-11-19)</h2>
</div>
    <div id="basejump-0-8-1-2018-11-15" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-1-2018-11-15" class="anchor"></a>basejump 0.8.1 (2018-11-15)</h2>
</div>
    <div id="basejump-0-8-0-2018-11-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-0-2018-11-11" class="anchor"></a>basejump 0.8.0 (2018-11-11)</h2>
</div>
    <div id="basejump-0-7-2-2018-08-29" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-7-2-2018-08-29" class="anchor"></a>basejump 0.7.2 (2018-08-29)</h2>
<div id="new-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>assertAllAreURL</code>.</li>
<li>
<code>assertAllAreValidNames</code>.</li>
<li>
<code>validNames</code>.</li>
</ul>
</div>
<div id="major-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes" class="anchor"></a>Major changes</h3>
<ul>
<li>Added a draft vignette explaining the functions available in the package.</li>
<li>
<code>gene2symbol</code>, <code>makeGene2symbolFromEnsembl</code>, and <code>makeGene2symbolFromGFF</code> functions now support the <code>unique</code> argument, which returns sanitized values in the <code>geneName</code> column, ensuring there are no duplicates. This is enabled by default (recommended) but can be disabled using <code>unique = FALSE</code>. This functionality was added to ensure consistent gene name handling in single-cell RNA-seq analyses.</li>
<li>
<code>saveData</code> now supports <code>basejump.save.ext</code> and <code>basejump.save.compress</code> global options, so the desired file type (e.g. RDS instead of RDA) and compression (e.g. xz instead of gzip) can be easily specified for an entire project.</li>
</ul>
</div>
<div id="minor-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>sampleNames</code> now supports assignment for <code>SummarizedExperiment</code> method.</li>
<li>Now exporting <code>lanePattern</code> regular expression pattern as a global, which was previously defined in the <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package.</li>
<li>Bug fix for <code>as</code> coercion method support. Need to ensure <code>exportMethods(coerce)</code> is included in <code>NAMESPACE</code> file, otherwise <code>tibble</code> coercion using <code>as(x, "tbl_df")</code> won’t work when called from an Rscript without the package library loaded. Thanks <a href='https://github.com/roryk'>@roryk</a> for noticing this.</li>
<li>Updated <code>gene2symbol</code> generic to use <code>...</code>, since we’ve added the <code>unique = TRUE</code> argument in this release.</li>
<li>
<code>annotable</code>: Moved to <code>makeGRanges.R</code> file, and improved the internal code to export supported formals also used in <code>makeGRangesFromEnsembl</code>. The function should work exactly the same as previous releases, but now with clearer supported arguments in the documentation.</li>
<li>Skipping code coverage for <code>cleanSystemLibrary</code>, since <a href="https://travis-ci.com/">Travis CI</a> installs packages into the system library, and causes this check to return <code>FALSE</code>.</li>
<li>Consistently using <code>as(from, "tbl_df")</code> for internal tibble coercion in all functions.</li>
<li>
<code>geneSynonyms</code> and <code>panther</code>: <code>organism</code> argument matching no longer suggests a default. The current list of supported organisms is in the documentation, and described in the internal <code>match.arg</code> call.</li>
<li>All <code>SummarizedExperiment</code> methods use <code>validObject</code> validity checks, where applicable.</li>
<li>Consolidated documentation for all <code>makeGRanges</code>, <code>makeGene2symbol</code>, and <code>makeTx2gene</code> functions.</li>
<li>Heatmap functions: simplified the internal code responsible for defining <code>annotationCol</code> and <code>annotationColors</code> automatically.</li>
<li>
<code>sampleData</code>: Made validity check stricter, requiring <code>sampleName</code> column to be defined, otherwise the function will intentionally error.</li>
<li>Now using <code>formals</code> internall to keep ggplot2 theme formals consistent.</li>
<li>Updated example data scripts and resaved internal data.</li>
<li>Updated contribution guidelines in <code>CONTRIBUTING.md</code> file.</li>
</ul>
</div>
</div>
    <div id="basejump-0-7-1-2018-08-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-7-1-2018-08-18" class="anchor"></a>basejump 0.7.1 (2018-08-18)</h2>
<div id="new-functions-1" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-1" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>cleanSystemLibrary</code>: Utility function to check whether a user has installed packages into the <a href="https://www.r-project.org/">R</a> system library. Refer to <code>.libPaths</code> documentation for more information on library paths.</li>
</ul>
</div>
<div id="major-changes-1" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-1" class="anchor"></a>Major changes</h3>
<ul>
<li>Now using <code>build</code> instead of <code>genomeBuild</code> for <a href="https://www.ensembl.org/">Ensembl</a> annotation functions. The <code>genomeBuild</code> argument still works but now will inform the user about the change.</li>
</ul>
</div>
<div id="minor-changes-1" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-1" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrated <code>prepareTemplate</code> from <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package. Simplifed this function to copy all files inside <code>extdata/rmarkdown/shared</code> within a specified package. Currently in use for <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a>, <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a>, and the new <a href="https://steinbaugh.com/pointillism/">pointillism</a> clustering package.</li>
<li>Made <a href="https://www.ensembl.org/">Ensembl</a> release matching stricter, based on the metadata columns.</li>
</ul>
</div>
</div>
    <div id="basejump-0-7-0-2018-08-07" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-7-0-2018-08-07" class="anchor"></a>basejump 0.7.0 (2018-08-07)</h2>
<p>In this release, we are migrating some of the S4 generics previously exported in the <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package. We are consolidating these functions here for simplicity and stability.</p>
<div id="new-functions-2" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-2" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>makeSummarizedExperiment</code>: Renamed <code>prepareSummarizedExperiment</code>, previously exported in the <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package. We are using the <code>make</code> prefix here for consistency (see other gene annotation functions).</li>
<li>Migrated S4 generics and methods from <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a>: <code>flatFiles</code>, <code>metrics</code>, <code>plotCorrleationHeatmap</code>, <code>plotGene</code>, <code>plotHeatmap</code>, <code>plotQC</code>, and <code>plotQuantileHeatmap</code>.</li>
</ul>
</div>
<div id="major-changes-2" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-2" class="anchor"></a>Major changes</h3>
<ul>
<li>Now using <code><a href="https://www.rdocumentation.org/packages/curl/topics/nslookup">curl::has_internet</a></code> internally to check for Internet connection. This applies to the annotation functions that query web databases.</li>
<li>Added coercion method support for converting a <code>SummarizedExperiment</code> to an unstructured <code>list</code>. This is the method used internally for <code>flatFiles</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-6-4-2018-08-03" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-6-4-2018-08-03" class="anchor"></a>basejump 0.6.4 (2018-08-03)</h2>
<div id="new-functions-3" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-3" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>matchInterestingGroups</code>: New developer function to automatically handle <code>interestingGroups</code> argument used across various plotting functions and in the <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> infrastructure packges.</li>
</ul>
</div>
</div>
    <div id="basejump-0-6-3-2018-08-01" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-6-3-2018-08-01" class="anchor"></a>basejump 0.6.3 (2018-08-01)</h2>
<div id="minor-changes-2" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-2" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrated <code>separatorBar</code> and <code>updateMessage</code> global export from <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a>. Improved <code>separatorBar</code> appeareance to automatically scale to current session width, using <code><a href="https://www.rdocumentation.org/packages/base/topics/options">getOption("width")</a></code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-6-2-2018-07-31" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-6-2-2018-07-31" class="anchor"></a>basejump 0.6.2 (2018-07-31)</h2>
<div id="new-functions-4" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-4" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>convertSymbolsToGenes</code>: provides <code>SummarizedExperiment</code> method support for converting objects containing gene symbols (“geneName”) as rownames back to gene identifiers (“geneID”).</li>
<li>
<code>eggnog</code>: quickly download current annotations from <a href="http://eggnogdb.embl.de/">EggNOG</a> database. Useful for annotating gene-to-protein matches; currently in use with the <a href="https://steinbaugh.com/brightworm/">brightworm</a> RNAi screening package, which contains <a href="http://www.wormbase.org/">WormBase</a> gene ID and <a href="http://eggnogdb.embl.de/">EggNOG</a> ID annotations.</li>
</ul>
</div>
<div id="major-changes-3" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-3" class="anchor"></a>Major changes</h3>
<ul>
<li>Now suggesting <a href="https://cran.r-project.org/package=BiocManager">BiocManager</a> instead of <a href="https://bioconductor.org/packages/BiocInstaller/">BiocInstaller</a> for installation.</li>
<li>
<code>broadClass</code> now supports <code>GRanges</code> and <code>SummarizedExperiment</code>. Support for <code>data.frame</code> and/or <code>DataFrame</code> class objects has been removed.</li>
</ul>
</div>
<div id="minor-changes-3" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-3" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>convertGenesToSymbols</code> and <code>convertTranscriptsToGenes</code> now have <code>organism</code> and <code>gene2symbol</code> arguments set <code>NULL</code> by default.</li>
<li>Upgraded to <a href="https://cran.r-project.org/package=roxygen2">roxygen2</a> v6.1 for documentation, which improves handling of aliases in the Rd manual files.</li>
<li>Added <code><a href="https://dplyr.tidyverse.org/reference/pull.html">dplyr::pull</a></code> to reexported functions.</li>
<li>Improved package documentation by declaring the supported class(es) for each function argument.</li>
<li>Moved <code>foldChangeToLogRatio</code> and <code>logRatioToFoldChange</code> constructors into <code>numeric</code> method declarations.</li>
<li>Simplified internal code for <code>gene2symbol</code> <code>SummarizedExperiment</code> method.</li>
<li>
<code>toStringUnique</code> now uses <code>x</code> instead of <code>atomic</code> as primary argument.</li>
</ul>
</div>
</div>
    <div id="basejump-0-6-1-2018-07-21" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-6-1-2018-07-21" class="anchor"></a>basejump 0.6.1 (2018-07-21)</h2>
<div id="minor-changes-4" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-4" class="anchor"></a>Minor changes</h3>
<ul>
<li>Bug fix for <code>convertGenesToSymbols</code> method for <code>SummarizedExperiment</code>. Previously, if <code>geneName</code> column was a factor, this function would error. This issue has been fixed by ensuring that the symbols provided in <code>geneName</code> are coerced to a character vector.</li>
<li>Improved <a href="https://conda.io/">conda</a> installation instructions.</li>
</ul>
</div>
</div>
    <div id="basejump-0-6-0-2018-07-17" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-6-0-2018-07-17" class="anchor"></a>basejump 0.6.0 (2018-07-17)</h2>
<div id="major-changes-4" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-4" class="anchor"></a>Major changes</h3>
<ul>
<li>Now importing <a href="https://bioconductor.org/packages/SummarizedExperiment/">SummarizedExperiment</a> package and providing basic method support for generics that were previously used in the <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package.</li>
<li>Improved GFFv3 handling in <code>makeGRangesFromGFF</code> and other GFF utility functions, including <code>makeGene2symbolFromGFF</code> and <code>makeTx2geneFromGFF</code>. Note that <code>makeGRangesFromGFF</code> now returns additional metadata columns accessible with <code><a href="https://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class">S4Vectors::mcols</a></code>, and that these columns are now sorted alphabetically.</li>
</ul>
</div>
<div id="migrated-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#migrated-functions" class="anchor"></a>Migrated functions</h3>
<p>Previously, these functions were exported in the <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package, but they provide non-<a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a>-specific functionality, and should be included here in the <a href="https://steinbaugh.com/basejump/">basejump</a> package instead:</p>
<ul>
<li>
<code>assertFormalInterestingGroups</code>.</li>
<li>
<code>gene2symbol</code>.</li>
<li>
<code>interestingGroups</code>, <code>uniteInterestingGroups</code>.</li>
<li>
<code>sampleData</code>, <code>sanitizeSampleData</code>.</li>
<li>
<code>sampleNames</code>.</li>
<li>
<code>selectSamples</code>.</li>
</ul>
<p>Providing basic <code>SummarizedExperiment</code> class method support for <code>counts</code>.</p>
</div>
<div id="minor-changes-5" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-5" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>geometricMean</code> generic was not exported correctly.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-11-2018-07-09" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-11-2018-07-09" class="anchor"></a>basejump 0.5.11 (2018-07-09)</h2>
<div id="minor-changes-6" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-6" class="anchor"></a>Minor changes</h3>
<ul>
<li>Now requiring <a href="http://ggplot2.tidyverse.org/">ggplot2</a> v3.0 internally.</li>
<li>
<code>theme_midnight</code> and <code>theme_paperwhite</code> now extend <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">ggplot2::theme_linedraw</a></code>, improving the consistency between these themes.</li>
<li>Example data is now consistenly formatted using snake case: <code>rnaseq_counts</code> and <code>single_cell_counts</code>, instead of the previous camel case conventions: <code>rnaseqCounts</code>, <code>singleCellCounts</code>.</li>
<li>Camel variants of the ggplot themes are now deprecated.</li>
<li>Updated internal gene synonyms data from <a href="https://www.nih.gov/">NIH</a>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-10-2018-06-28" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-10-2018-06-28" class="anchor"></a>basejump 0.5.10 (2018-06-28)</h2>
<div id="minor-changes-7" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-7" class="anchor"></a>Minor changes</h3>
<ul>
<li>Removed <code>makeNames</code> argument from <code>readFileByExtension</code> function. Use the <code>makeNames</code> family of functions manually after data import instead. This helps avoid unwanted sanitization of data.</li>
<li>Simplified assert checks for internal load function.</li>
<li>Improved code coverage.</li>
<li>
<a href="https://ci.appveyor.com/">AppVeyor CI</a> updates to work with <a href="https://bioconductor.org/">Bioconductor</a> 3.8 devel.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-9-2018-06-12" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-9-2018-06-12" class="anchor"></a>basejump 0.5.9 (2018-06-12)</h2>
<div id="minor-changes-8" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-8" class="anchor"></a>Minor changes</h3>
<ul>
<li>Markdown function consistency improvements. Now all relevant <a href="https://daringfireball.net/projects/markdown/">Markdown</a> functions use <code>text</code> as the primary argument, instead of <code>object</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-8-2018-06-05" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-8-2018-06-05" class="anchor"></a>basejump 0.5.8 (2018-06-05)</h2>
<div id="minor-changes-9" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-9" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>makeGRangesFromEnsembl</code> now supports remapping of <a href="https://genome.ucsc.edu/">UCSC</a> genome build to Ensembl. However, this isn’t recommended, and will warn the user.</li>
<li>
<code>convertUCSCBuildToEnsembl</code> now returns <code>NULL</code> instead of erroring on genome build match failure.</li>
<li>
<code>stripTranscriptVersions</code> now matches “<code>.</code>”, “<code>-</code>”, and “<code>_</code>” version delimiters.</li>
<li>Made unused <code>dynamicPlotlist</code> function defunct.</li>
<li>Consider deprecating <code>assertIsCharacterOrNULL</code> and <code>assertIsDataFrameOrNULL</code> in a future release.</li>
</ul>
</div>
<div id="infrastructure-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#infrastructure-changes" class="anchor"></a>Infrastructure changes</h3>
<ul>
<li>Reenable <a href="https://travis-ci.com/">Travis CI</a> blocklist, excluding develop branch from checks.</li>
<li>Reorganized documentation of deprecated and defunct functions.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-7-2018-05-24" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-7-2018-05-24" class="anchor"></a>basejump 0.5.7 (2018-05-24)</h2>
<div id="minor-changes-10" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-10" class="anchor"></a>Minor changes</h3>
<ul>
<li>Tweaked gray color accents for <code>theme_midnight</code> and <code>theme_paperwhite</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-6-2018-05-19" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-6-2018-05-19" class="anchor"></a>basejump 0.5.6 (2018-05-19)</h2>
<div id="minor-changes-11" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-11" class="anchor"></a>Minor changes</h3>
<ul>
<li>Fixed NAMESPACE issue with <code><a href="https://www.rdocumentation.org/packages/GenomeInfoDb/topics/seqinfo">GenomeInfoDb::seqnames</a></code>.</li>
<li>Improved <code>readGFF</code> working example to reflect switch to <code>GRanges</code> return.</li>
<li>Added <a href="https://www.apple.com/macos/">macOS</a> bioc-release image to <a href="https://travis-ci.com/">Travis CI</a> build checks.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-5-2018-05-15" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-5-2018-05-15" class="anchor"></a>basejump 0.5.5 (2018-05-15)</h2>
<div id="major-changes-5" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-5" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>readGFF</code> now uses <code><a href="https://www.rdocumentation.org/packages/rtracklayer/topics/export">rtracklayer::import</a></code> internally to return GFF file as a <code>GRanges</code> object instead of a <code>data.frame</code>.</li>
</ul>
</div>
<div id="minor-changes-12" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-12" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>assertIsGFF</code> and <code>parseGFFAttributes</code> functions are now defunct.</li>
<li>Simplified internal GFF handling code for <code>makeGRangesFromGFF</code>, <code>makeGene2symbolFromGFF</code>, and <code>makeTx2geneFromGFF</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-4-2018-05-08" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-4-2018-05-08" class="anchor"></a>basejump 0.5.4 (2018-05-08)</h2>
<ul>
<li>Migrated <code>sanitizeSampleData</code> to <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package.</li>
<li>Updated <a href="https://bioconductor.org/">Bioconductor</a> install method for 3.7.</li>
</ul>
</div>
    <div id="basejump-0-5-3-2018-04-30" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-3-2018-04-30" class="anchor"></a>basejump 0.5.3 (2018-04-30)</h2>
<div id="minor-changes-13" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-13" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved internal S4 method code for <code>fixNA</code> and <code>removeNA</code>.</li>
<li>Tweaked gray accent colors in <code>theme_midnight</code> and <code>theme_paperwhite</code>. Now using British spelling internally for ggplot code.</li>
<li>Improved <code>strip.background</code> for <code>theme_paperwhite</code>, removing the black box around the labels when facet wrapping is enabled.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-2-2018-04-26" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-2-2018-04-26" class="anchor"></a>basejump 0.5.2 (2018-04-26)</h2>
<div id="minor-changes-14" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-14" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved documentation for assert check functions.</li>
<li>Deprecated <code>geomean</code> in favor of <code>geometricMean</code>.</li>
<li>Simplified internal code for <code>grepString</code>.</li>
<li>Added message during <code>hgnc2gene</code> call.</li>
<li>Miscellaneous documentation fixes.</li>
<li>Moved internal constructors into the S4 method definitions, where applicable.</li>
<li>Simplified default parameter definition for <code>panther(organism = "XXX")</code>.</li>
<li>Improved code coverage, using <code>nocov</code> where appropriate.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-1-2018-04-16" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-1-2018-04-16" class="anchor"></a>basejump 0.5.1 (2018-04-16)</h2>
<div id="minor-changes-15" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-15" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>emptyRanges</code>: Now using <code>match.arg</code> internally to capture <code>seqname</code> argument.</li>
<li>Removed legacy <code>.assignCamelArgs</code> and <code>.assignCamelFormals</code> internal functions.</li>
<li>Improved internal handling of XLSX files in <code>localOrRemoteFile</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-5-0-2018-04-13" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-5-0-2018-04-13" class="anchor"></a>basejump 0.5.0 (2018-04-13)</h2>
<div id="new-functions-5" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-5" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>emptyRanges</code> enables easy creation of placeholder ranges for <code>GRanges</code> objects, where transgene and FASTA spike-ins are needed.</li>
<li>
<code>hgnc2gene</code> enables easy mapping of <a href="https://www.genenames.org/">HGNC</a> to <a href="https://www.ensembl.org/">Ensembl</a> gene identifiers.</li>
<li>
<code>mgi2gene</code> enables easy mapping of <a href="http://www.informatics.jax.org/">MGI</a> to <a href="https://www.ensembl.org/">Ensembl</a> gene identifiers.</li>
<li>
<code>panther</code> function enables easy querying of the <a href="http://pantherdb.org/">PANTHER</a> website. Human, mouse, nematode worm, and fruit fly are currently supported. The specific <a href="http://pantherdb.org/">PANTHER</a> release (e.g. 13) can be declared using the <code>release</code> argument. Otherwise, the function will return the most recent annotations from the <a href="http://pantherdb.org/">PANTHER</a> website.</li>
<li>Added <code>isURL</code> check function.</li>
<li>
<code>readJSON</code> adds support for JSON files. Like the other read functions, it supports both local files and remote URLs.</li>
</ul>
</div>
<div id="ggplot2-themes" class="section level3">
<h3 class="hasAnchor">
<a href="#ggplot2-themes" class="anchor"></a>ggplot2 themes</h3>
<ul>
<li>
<code>theme_midnight</code> and <code>theme_paperwhite</code> provide minimal, high contrast <a href="http://ggplot2.tidyverse.org/">ggplot2</a> themes with bold sans serif labels.</li>
</ul>
</div>
<div id="major-changes-6" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-6" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>loadData</code> now supports <code>.rda</code>, <code>.rds</code>, and <code>.RData</code> files. The function will error by design if multiple data extensions are detected inside the directory specified with the <code>dir</code> argument.</li>
</ul>
</div>
<div id="minor-changes-16" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-16" class="anchor"></a>Minor changes</h3>
<ul>
<li>Consolidated assert check function code.</li>
<li>Moved assertive imports to <code>basejump-package.R</code> file.</li>
<li>Consolidated globals inside package to <code>globals.R</code> file.</li>
<li>Removed internal <code>.biocLite</code> function. Now using <code>requireNamespace</code> instead, without attempting to install automatically.</li>
<li>Added internal support for safe loading RDS files.</li>
<li>Switched back to using <code>message</code>, <code>warning</code>, and <code>stop</code> instead of the <a href="http://rlang.tidyverse.org/">rlang</a> equivalents.</li>
<li>Improved internal method declaration using <code>getMethod</code> where applicable.</li>
<li>
<code>multiassignAsEnvir</code> is now recommended in place of <code>multiassignAsNewEnvir</code>.</li>
<li>
<code>readFileByExtension</code> will now attempt to use the <a href="https://cran.r-project.org/package=rio">rio</a> package for file extensions that are not natively supported.</li>
<li>
<code>writeCounts</code> now uses <code>mapply</code> internally.</li>
<li>Migrated <code>assertFormalAnnotationCol</code> to <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package.</li>
</ul>
</div>
</div>
    <div id="basejump-0-4-0-2018-03-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-4-0-2018-03-22" class="anchor"></a>basejump 0.4.0 (2018-03-22)</h2>
<div id="major-changes-7" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-7" class="anchor"></a>Major changes</h3>
<ul>
<li>Introducing new functions for the acquistion of gene and transcript annotations from <a href="https://www.ensembl.org/">Ensembl</a>: <code>ensembl</code>, <code>genes</code>, and <code>transcripts</code>. These functions allow the return of <code>GRanges</code>, <code>DataFrame</code>, and <code>data.frame</code> class objects from <a href="https://bioconductor.org/packages/AnnotationHub/">AnnotationHub</a> using <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a>.</li>
<li>Improved internal <code>broadClass</code> definition code to match against chromosome from <a href="https://www.ensembl.org/">Ensembl</a> if available.</li>
<li>
<code>loadDataAsName</code> now works with unquoted names, improving consistency with <code>loadData</code> (non-standard evaluation).</li>
</ul>
</div>
<div id="minor-changes-17" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-17" class="anchor"></a>Minor changes</h3>
<ul>
<li>Added new <code>convertUCSCBuildToEnsembl</code> function, for easy remapping of <a href="https://genome.ucsc.edu/">UCSC</a> to <a href="https://www.ensembl.org/">Ensembl</a> genome build names (e.g. <code>hg38</code> to <code>GRCh38</code>).</li>
<li>Migrated matrix methods for <code>plotCorrelationHeatmap</code> here from <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a>, for improved consistency with other heatmap functions.</li>
<li>Exporting <code>makeNames</code> variant of <code><a href="https://www.rdocumentation.org/packages/base/topics/make.names">base::make.names</a></code> that sanitizes using underscores (“_“) rather than dots (”.“).</li>
<li>Converted <code>readYAML</code> from a generic to standard function.</li>
<li>Added support for <a href="https://ci.appveyor.com/">AppVeyor CI</a> code testing on <a href="https://www.microsoft.com/en-us/windows/">Windows</a>.</li>
<li>Made <a href="https://travis-ci.com/">Travis CI</a> build checks stricter, adding support for <code>BiocCheck</code>.</li>
<li>Added new assert checks: <code>assertAreGeneAnnotations</code>, <code>assertAreTranscriptAnnotations</code>, <code>isAnImplicitInteger</code>.</li>
<li>Simplified working examples for assert checks to just show successes.</li>
</ul>
</div>
<div id="deprecations" class="section level3">
<h3 class="hasAnchor">
<a href="#deprecations" class="anchor"></a>Deprecations</h3>
<ul>
<li>
<code>annotable</code> function has been deprecated in favor of the new <code>ensembl</code> function.</li>
<li>
<code>checkAnnotable</code> deprecated in favor of <code>assertIsAnnotable</code>.</li>
<li>
<code>checkGene2symbol</code> deprecated in favor of <code>assertIsGene2symbol</code>.</li>
<li>
<code>checkTx2gene</code> deprecated in favor of <code>assertIsTx2gene</code>.</li>
<li>
<code>assertFormalColorFunction</code> deprecated in favor of <code>assertIsHexColorFunctionOrNULL</code>.</li>
<li>
<code>initializeDir</code> deprecated in favor of <code>initializeDirectory</code>.</li>
<li>Defunct: <code>summarizeRows</code>, <code>wash</code>, <code>packageSE</code>, <code>prepareSE</code>, <code>metadataTable</code>, <code>comp</code>, <code>revcomp</code>, <code>symbol2gene</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-3-1-2018-02-19" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-3-1-2018-02-19" class="anchor"></a>basejump 0.3.1 (2018-02-19)</h2>
<ul>
<li>Now exporting all assert checks in camel case instead of snake case, to match consistency in the rest of the package.</li>
<li>Added <code>sanitizeColData</code> function.</li>
<li>Added <code>assertAllAreNonExisting</code> function.</li>
<li>Now exporting <code>midnightTheme</code> as a <code>theme_midnight</code> alias to match the syntax in the <a href="http://ggplot2.tidyverse.org/">ggplot2</a> package.</li>
<li>Added working examples and code coverage for all assert check functions.</li>
<li>Simplified the internal collapse code for <code>annotable</code> to simply work on the <a href="https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html">Entrez</a> identifier column (<code>entrez</code>). If a manually passed in data frame still has duplicates, the function will now abort instead of attempting to use <code>collapseToString</code>.</li>
<li>Added <a href="http://ggplot2.tidyverse.org/">ggplot2</a> color palette assert checks: <code>assertColorScaleContinuousOrNULL</code>, <code>assertColorScaleDiscreteOrNULL</code>, <code>assertFillScaleContinuousOrNULL</code>, <code>assertFillScaleDiscreteOrNULL</code>.</li>
</ul>
</div>
    <div id="basejump-0-3-0-2018-02-16" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-3-0-2018-02-16" class="anchor"></a>basejump 0.3.0 (2018-02-16)</h2>
<ul>
<li>Switch to using <a href="https://cran.r-project.org/package=assertive">assertive</a> internally for assert checks.</li>
<li>Now exporting these assert check functions: <code>assert_formal_annotation_col</code>, <code>assert_formal_color_function</code>, <code>assert_formal_compress</code>, <code>assert_formal_gene2symbol</code>, <code>assert_formal_header_level</code>, <code>assert_has_rownames</code>, <code>assert_is_a_number_or_null</code>, <code>assert_is_a_string_or_null</code>, <code>assert_is_an_implicit_integer</code>, <code>assert_is_an_implicit_integer_or_null</code> <code>assert_is_an_integer_or_null</code>, <code>assert_is_annotable</code>, <code>assert_is_character_or_null</code>, <code>assert_is_data.frame_or_null</code>, <code>assert_is_gene2symbol</code>, <code>assert_is_implicit_integer</code>, <code>assert_is_implicit_integer_or_null</code>, <code>assert_is_tx2gene</code>, <code>has_rownames</code>, <code>initializeDirectory</code>, <code>is_implicit_integer</code>.</li>
<li>Renamed <code>md*</code> functions to <code>markdown*</code>.</li>
</ul>
</div>
    <div id="basejump-0-2-1-2018-02-07" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-2-1-2018-02-07" class="anchor"></a>basejump 0.2.1 (2018-02-07)</h2>
<ul>
<li>Added <code>convertGenesToSymbols</code> and <code>convertTranscriptsToGenes</code> functions. Previously some of this functionality was contained within the <code>gene2symbol</code> and <code>tx2gene</code> generics for the character method. This behavior was inconsistent with <code>gene2symbol</code> and <code>tx2gene</code> usage in the <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> R packages, so I decided to split these out into separate functions. Now <code>gene2symbol</code> and <code>tx2gene</code> work consistently with the <code>annotable</code> function to return gene-to-symbol and transcript-to-gene identifier mappings in a <code>data.frame</code>.</li>
<li>
<code>markdownHeader</code>, <code>markdownList</code>, and <code>markdownPlotlist</code> are now exported as S4 generics. The <code>md*</code> function variants are now exported as aliases.</li>
<li>
<code>geomean</code> has been renamed to <code>geometricMean</code>.</li>
<li>Miscellaneous improvements to error messages and warnings.</li>
</ul>
</div>
    <div id="basejump-0-2-0-2018-01-28" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-2-0-2018-01-28" class="anchor"></a>basejump 0.2.0 (2018-01-28)</h2>
<ul>
<li>Offloaded internal <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a>-specific code to new package named <a href="http://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a>. Consequently, this makes the <a href="https://steinbaugh.com/basejump/">basejump</a> package leaner, meaner, and easier to manage. The following functions are now exported in that package: <code>bcbio</code>, <code>checkInterestingGroups</code>, <code>flatFiles</code>, <code>interestingGroups</code>, <code>metrics</code>, <code>plotDot</code>, <code>plotGene</code>, <code>plotQC</code>, <code>plotViolin</code>, <code>prepareSummarizedExperiment</code>, <code>prepareTemplate</code>, <code>readDataVersions</code>, <code>readLogFile</code>, <code>readProgramVersions</code>, <code>readSampleMetadataFile</code>, <code>sampleMetadata</code>, <code>sampleYAML</code>, <code>sampleYAMLMetadata</code>, <code>sampleYAMLMetrics</code>, and <code>selectSamples</code>. These functions are now deprecated here in <a href="https://steinbaugh.com/basejump/">basejump</a> (see <code>deprecated.R</code> file for more information).</li>
<li>
<code>comp</code> and <code>revcomp</code> have been deprecated in favor of <code>complement</code> and <code>reverseComplement</code> from the <a href="https://bioconductor.org/packages/Biostrings/">Biostrings</a> package.</li>
<li>Internally, errors, messages, and warnings now use methods from the <a href="http://rlang.tidyverse.org/">rlang</a> package: <code>abort</code>, <code>inform</code>, and <code>warn</code>, in place of <code>stop</code>, <code>message</code>, and <code>warning</code>, respectively.</li>
<li>Improved error handing for missing files in <code>loadData</code>. Additionally, the file name must match the internal name in the RData file, otherwise <code>loadData</code> will warn the user. This is more strict than the default behavior of <code><a href="https://www.rdocumentation.org/packages/base/topics/load">base::load</a></code>, but helps prevent accidental overwrite in the current working environment.</li>
<li>
<code>localOrRemoteFile</code>, previously an internal function, is now exported.</li>
<li>
<code>annotable</code> now uses internal <a href="https://grch37.ensembl.org/">GRCh37</a> annotations from the <a href="https://github.com/stephenturner/annotables/">annotables</a> package, which is saved in the <code>extdata/</code> directory internally. Previously, these genome annotations were accessed from lazy loaded data saved in the <code>data/</code> directory of the package repository.</li>
<li>
<code>annotables</code> now checks for all packages attached by <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a> and <a href="https://bioconductor.org/packages/AnnotationHub/">AnnotationHub</a> and forces detachment at the end of the function call. Otherwise, this can result in the unwanted effect of <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a> masking other user-loaded functions, such as the <a href="http://www.tidyverse.org/">tidyverse</a> suite (e.g. <code><a href="https://dplyr.tidyverse.org/reference/select.html">dplyr::select</a></code>).</li>
<li>Consistently reformatted code indents to be a multple of 4 spaces, as recommended by <a href="https://bioconductor.org/">Bioconductor</a>.</li>
<li>
<code>camel</code> now handles delimited numbers a little differently. Previously, delimiters in between numbers (e.g. the commas in “1,000,000”) were stripped. Sometimes this can result in confusing names. For example, if we have a column formatted in dotted case containing a decimal (e.g. “resolution.1.6”), the decimal would be stripped (e.g. “resolution16” in camel). Now, we sanitize a numeric delimiter as a lower case “x” character (e.g. “resolution1x6”). This ensures that numbers containing decimals remain semantically meaningful when sanitized with <code>camel</code>.</li>
<li>Internally, <code>gsub</code> (and <code>grepl</code>) calls have been simplified to use the default order of “pattern, replacement, x”.</li>
<li>Internally, all implicit integers have been converted to explicit integers, where applicable.</li>
</ul>
</div>
    <div id="basejump-0-1-10-2017-12-20" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-10-2017-12-20" class="anchor"></a>basejump 0.1.10 (2017-12-20)</h2>
<ul>
<li>Bug fix for multiplexed sample input into <code>readSampleMetadataFile</code>. We were detecting the presence of <code>index</code> column but should instead check against <code>sequence</code> column.</li>
<li>Added <code>dynamicPlotlist</code> and <code>mdPlotlist</code> plotting utilities.</li>
<li>Added <code>uniqueSymbols</code> parameter to <code>annotable</code> function.</li>
</ul>
</div>
    <div id="basejump-0-1-9-2017-12-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-9-2017-12-18" class="anchor"></a>basejump 0.1.9 (2017-12-18)</h2>
<ul>
<li>Added <code>plotHeatmap</code> functionality.</li>
<li>Migrated <code>tpm</code> generic from <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a>, for future use in <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a>.</li>
<li>Added matrix method support for <code>plotHeatmap</code>.</li>
<li>Added matrix method support for <code>plotQuantileHeatmap</code>, which works similarly as <code>plotHeatmap</code>.</li>
</ul>
</div>
    <div id="basejump-0-1-8-2017-12-14" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-8-2017-12-14" class="anchor"></a>basejump 0.1.8 (2017-12-14)</h2>
<ul>
<li>Improved <code>matrix</code> and <code>dgCMatrix</code> method support in <code>aggregateReplicates</code> and <code>aggregateFeatures</code> functions. Both of these functions now use a consistent <code>groupings</code> parameter, which uses a named factor to define the mappings of either samples (columns) for <code>aggregateReplicates</code> or genes/transcripts (rows) for <code>aggregateFeatures</code>.</li>
<li>Update for <code>makeNames</code> sanitization functions. Now they will work on <code><a href="https://www.rdocumentation.org/packages/base/topics/names">names(x)</a></code> for vectors by default.</li>
<li>Improved <code>detectOrganism</code> to match against “H. sapiens”, etc.</li>
<li>Added internal <a href="https://grch37.ensembl.org/">GRCh37</a> transcript to gene mapping.</li>
<li>Improved organism matching to detect “Homo_sapiens” and “H. sapiens”.</li>
<li>Factors are now supported in the <code>makeNames</code> utilities: <code>camel</code>, <code>dotted</code>, <code>snake</code>, and <code>upperCamel</code>.</li>
<li>Improved handling of <code>NA</code> values from <a href="https://www.libreoffice.org/">LibreOffice</a> and <a href="https://products.office.com/en-us/excel/">Microsoft Excel</a> output in <code>readFileByExtension</code>. This function now sets <code>""</code>, <code>NA</code>, and <code>#N/A</code> strings as <code>NA</code> correctly.</li>
</ul>
</div>
    <div id="basejump-0-1-7-2017-12-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-7-2017-12-11" class="anchor"></a>basejump 0.1.7 (2017-12-11)</h2>
<ul>
<li>Renamed <code>fc2lr</code> to <code>foldChangeToLogRatio</code> and <code>lr2fc</code> and <code>logRatioToFoldChange</code>.</li>
<li>Moved <code>plotDot</code> and <code>plotViolin</code> generics here from <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a>.</li>
<li>Added internal <a href="https://grch37.ensembl.org/">GRCh37</a> gene annotations.</li>
</ul>
</div>
    <div id="basejump-0-1-6-2017-12-06" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-6-2017-12-06" class="anchor"></a>basejump 0.1.6 (2017-12-06)</h2>
<ul>
<li>Moved <code>microplate</code> code from the <a href="https://steinbaugh.com/wormbase/">wormbase</a> package here, since it’s of general interest.</li>
</ul>
</div>
    <div id="basejump-0-1-5-2017-12-05" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-5-2017-12-05" class="anchor"></a>basejump 0.1.5 (2017-12-05)</h2>
<ul>
<li>Added <code>checkAnnotable</code>, <code>checkGene2symbol</code>, <code>checkTx2gene</code>, and <code>sanitizeAnnotable</code> utility functions that will be used in the bcbio R packages.</li>
</ul>
</div>
    <div id="basejump-0-1-4-2017-12-04" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-4-2017-12-04" class="anchor"></a>basejump 0.1.4 (2017-12-04)</h2>
<ul>
<li>Added <code>midnightTheme</code> <a href="http://ggplot2.tidyverse.org/">ggplot2</a> theme. Originally this was defined as <code>darkTheme</code> in the <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a> package, but can be useful for other plots and has been moved here for general bioinformatics usage. The theme now uses <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">ggplot2::theme_minimal</a></code> as the base, with some color tweaks, namely dark gray axes without white axis lines.</li>
<li>Improve NAMESPACE imports to include <code><a href="https://www.rdocumentation.org/packages/stats/topics/formula">stats::formula</a></code> and <code><a href="https://www.rdocumentation.org/packages/utils/topics/capture.output">utils::capture.output</a></code>.</li>
</ul>
</div>
    <div id="basejump-0-1-3-2017-12-01" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-3-2017-12-01" class="anchor"></a>basejump 0.1.3 (2017-12-01)</h2>
<ul>
<li>
<code>loadData</code> and <code>loadDataAsName</code> now default to <code>replace = TRUE</code>. If an object with the same name exists in the destination environment, then a warning is generated.</li>
<li>
<code>collapseToString</code> only attempts to dynamically return the original object class on objects that aren’t class <code>data.frame</code>. I updated this code to behave more nicely with grouped tibbles (<code>grouped_df</code>), which are a virtual class of <code>data.frame</code> and therefore can’t be coerced using <code>as(object, "grouped_df")</code>.</li>
<li>DNA sequence utility functions <code>comp</code> and <code>revcomp</code> now return <code>NULL</code> for integers and numerics.</li>
<li>For <code>prepareSummarizedExperiment</code>, added support for dropping <code>NULL</code> objects in assays list. This is useful for handling output from <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a> when <code>transformLimit</code> is reached. In this case, the <code>rlog</code> and <code>vst</code> matrices aren’t generated and set <code>NULL</code> in the assays list. Using <code><a href="https://www.rdocumentation.org/packages/base/topics/funprog">Filter(Negate(is.null), assays)</a></code> we can drop these <code>NULL</code> objects and prevent a downstream dimension mismatch in the <code><a href="https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class">SummarizedExperiment::SummarizedExperiment</a></code> call.</li>
<li>Improved support for multiplexed files in <code>readSampleMetadataFile</code>. This now checks for a sequence column containing ACGT nucleotides. When those are detected, the <code>revcomp</code> column is generated. Otherwise this step is skipped. This is useful for handling multiplexed sample metadata from 10X Genomics <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> single-cell RNA-seq samples.</li>
<li>Updated <code>annotable</code> function to include nested <a href="https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html">Entrez</a> identifiers in the <code>entrez</code> column. This is useful for downstream functional analysis.</li>
</ul>
</div>
    <div id="basejump-0-1-2-2017-11-30" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-2-2017-11-30" class="anchor"></a>basejump 0.1.2 (2017-11-30)</h2>
<ul>
<li>Added <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> <code>plotQC</code> generic.</li>
<li>Added back <code>toStringUnique</code> code, which is still in use in the <a href="https://steinbaugh.com/wormbase/">wormbase</a> package.</li>
<li>Added deprecations for <code>summarizeRows</code> (now <code>collapseToString</code>) and <code>wash</code> functions.</li>
<li>Updated installation method to include manual installation of <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a>. Otherwise, <a href="https://steinbaugh.com/basejump/">basejump</a> installation fails due to <a href="https://bioconductor.org/packages/GenomeInfoDb/">GenomeInfoDb</a> and <a href="https://bioconductor.org/packages/GenomeInfoDbData/">GenomeInfoDbData</a> not getting installed completely.</li>
<li>Now suggesting that the user installs suggested packages in the README.</li>
<li>Updated <a href="http://pantherdb.org/">PANTHER</a> annotation scripts.</li>
<li>Bug fix for <code>detectOrganism</code>. Now allowing <code>NULL</code> return for unsupported organism, with a warning.</li>
</ul>
</div>
    <div id="basejump-0-1-1-2017-11-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-1-2017-11-11" class="anchor"></a>basejump 0.1.1 (2017-11-11)</h2>
<ul>
<li>Added overwrite support for <code>saveData</code>. Now will skip on existing files when <code>overwrite = FALSE</code>.</li>
<li>Bug fix for <code>readDataVersions</code>, which shouldn’t have the column types defined, using <code>col_types = "ccT"</code>.</li>
<li>Improved key value pair method for <code>loadDataAsName</code>. Now rather than using a named character vector for the <code>mappings</code> argument, the user can simply pass the key value pairs in as dots. For example, <code>newName1 =   "oldName1", newName2 = "oldName2"</code>. The legacy <code>mappings</code> method will still work, as long as the dots argument is a length of 1.</li>
<li>
<a href="https://www.ensembl.org/">Ensembl</a> release version now defaults to <code>NULL</code> instead of <code>current</code> for <code>annotable</code>, <code>gene2symbol</code>, <code>symbol2gene</code> and <code>tx2gene</code> functions.</li>
<li>Allow <code>rowData</code> to be left unset in <code>prepareSummarizedExperiment</code>. This is useful for setting up objects that don’t contain gene annotations.</li>
<li>Removed sample selection by pattern matching (<code>pattern</code>, <code>patternCol</code> arguments) in <code>readSampleMetadata</code>. This feature wasn’t fully baked and doesn’t offer enough functionality to the user.</li>
</ul>
</div>
    <div id="basejump-0-1-0-2017-10-23" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-0-2017-10-23" class="anchor"></a>basejump 0.1.0 (2017-10-23)</h2>
<ul>
<li>Bump version to match <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a> package.</li>
<li>Improved unit testing coverage of <code>prepareSummarizedExperiment</code>.</li>
<li>Added quiet mode support to functions that output messages, where applicable.</li>
<li>Consolidated <a href="https://cran.r-project.org/package=roxygen2">roxygen2</a> function imports to <code>basejump-package.R</code> file.</li>
<li>Deprecated <code>sampleDirs</code> generic.</li>
<li>Improved organism detection in <code>detectOrganism</code> and added support for chicken genome.</li>
<li>Clarified warning messages in <code>prepareSummarizedExperiment</code> to make sample loading with <code>loadRNASeq</code> and <code>loadSingleCell</code> in the <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> packages less confusing.</li>
<li>Improved <code>NULL</code> returns in <code>readDataVersions</code>, <code>readLogFile</code>, and <code>readProgramVersions</code> utility functions.</li>
<li>Fixed export of <code>*GTF</code> alias functions to simply wrap the <code>*GFF</code> functions with S4 methods support.</li>
<li>Improved lane split technical replicate handling in <code>readSampleMetadataFile</code>.</li>
<li>Improved <code>camel</code> syntax for both lax and strict modes. Added <code>upperCamel</code> function.</li>
<li>Switched <code>str_</code> to base <code>grep</code> and <code>gsub</code> in internal functions.</li>
</ul>
</div>
    <div id="basejump-0-0-25-2017-10-13" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-25-2017-10-13" class="anchor"></a>basejump 0.0.25 (2017-10-13)</h2>
<ul>
<li>Improved consistency of <code>setMethod</code> calls using <code>signature</code>.</li>
<li>Converted <code>loadRemoteData</code> to a standard function instead of using S4 dispatch, allowing the <code>envir</code> argument to be set properly.</li>
</ul>
</div>
    <div id="basejump-0-0-24-2017-10-10" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-24-2017-10-10" class="anchor"></a>basejump 0.0.24 (2017-10-10)</h2>
<ul>
<li>Added additional package version requirements in <code>DESCRIPTION</code> file.</li>
<li>Implicit integers are allowed.</li>
<li>Using <a href="https://github.com/">GitHub</a> version of <a href="https://cran.r-project.org/package=covr">covr</a> package for unit tests.</li>
<li>Renamed <code>multiassignAsNewEnv</code> to <code>multiassignAsNewEnvir</code>
</li>
<li>Added <code>*GFF</code> function variants for <code>gene2symbolFromGTF</code> and <code>tx2geneFromGTF</code>.</li>
<li>Added shared <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> generic functions: <code>aggregateReplicates</code>, <code>bcbio</code>, <code>bcbio&lt;-</code>, <code>interestingGroups</code>, <code>metrics</code>, <code>plotGene</code>, <code>sampleDirs</code>, <code>sampleMetadata</code>, <code>selectSamples</code>. These functions are saved in <code>bcbioGenerics.R</code> file.</li>
<li>Incorporated <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> utility functions shared across <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a> and <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a>: <code>readDataVersions</code> (deprecated <code>.dataVersions</code>), <code>readLogFile</code> (deprecated <code>.logFile</code>), <code>readProgramVersions</code> (deprecated <code>.programs</code>), <code>sampleYAML</code> (deprecated <code>.sampleYAML</code>), <code>sampleYAMLMetadata</code> (deprecated <code>sampleYAMLMetadata</code>), <code>sampleYAMLMetrics</code> (deprecated .sampleYAMLMetrics`).</li>
<li>Deprecated <code>metadataTable</code> function.</li>
<li>Moved all <a href="https://cran.r-project.org/package=roxygen2">roxygen2</a> documentation to <code>methods-*.R</code> files where applicable.</li>
<li>Now using <code>assays</code> in <code>prepareSummarizedExperiment</code> generic definition as primary object.</li>
<li>Improved <code>assignAndSaveData</code> to add silent return of file path.</li>
<li>Now consistently using clear function definitions (`<code>) in chain   operations with magrittr pipe (</code>%&gt;%`).</li>
<li>Added <code>.prepareSampleMetadata</code> utility function, for use with loading sample metadata from an external CSV, <a href="https://products.office.com/en-us/excel/">Excel</a>, or YAML file.</li>
<li>Added <code>loadData</code> functionality back to the package.</li>
<li>Initial commit of <code>loadDataAsName</code> function.</li>
<li>Improved <code>annotable</code> function documentation and support for <a href="https://www.ensembl.org/">Ensembl</a> release versions.</li>
<li>Improved sanitization rules for <code>camel</code>, <code>dotted</code>, and <code>snake</code> name functions. Added the <code>strict</code> argument to <code>camel</code> and <code>dotted</code>.</li>
<li>Improved the documentation for the <code>makeNames</code> functions, by splitting each into their own separate methods file.</li>
<li>Improved S4 method support for the logRatio functions.</li>
<li>Added integer support for <code>geomean</code> function. Also improved internal code of <code>geomean</code> based on Paul McMurdie’s Stack Overflow post. See function documentation for more information.</li>
<li>Added release support for <code>tx2gene</code> functions.</li>
<li>Reduced the number of reexported functions (for documentation) down to the <a href="https://magrittr.tidyverse.org/">magrittr</a> pipe (<code>%&gt;%</code>), <code>Matrix</code>, <code>DataFrame</code>, and <code>tibble</code>.</li>
<li>Added silent return of file path to <code>saveData</code> function.</li>
<li>Improved documentation for <a href="https://tibble.tidyverse.org/">tibble</a> coercion using <code>as(object, "tibble")</code>
</li>
<li>Renamed <code>collapse</code> function to <code>collapseToString</code>, to avoid NAMESPACE collisions with tidyverse packages (<a href="http://dplyr.tidyverse.org/">dplyr</a>, <a href="https://glue.tidyverse.org/">glue</a>).</li>
</ul>
</div>
    <div id="basejump-0-0-23-2017-09-14" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-23-2017-09-14" class="anchor"></a>basejump 0.0.23 (2017-09-14)</h2>
<ul>
<li>Upgraded <code>annotable</code> function to query <a href="https://www.ensembl.org/">Ensembl</a> using the <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a> package rather than <a href="https://github.com/stephenturner/annotables/">annotables</a>.</li>
</ul>
</div>
    <div id="basejump-0-0-22-2017-09-08" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-22-2017-09-08" class="anchor"></a>basejump 0.0.22 (2017-09-08)</h2>
<ul>
<li>Improved unit testing coverage.</li>
<li>Renamed <code>prepareSE</code> to <code>prepareSummarizedExperiment</code>. Improved row and column name handling in the function. It now outputs more helpful diagnostic messages on error.</li>
<li>Reworked and simplified <code>detectHPC</code> function to allow for unit testing.</li>
</ul>
</div>
    <div id="basejump-0-0-21-2017-09-01" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-21-2017-09-01" class="anchor"></a>basejump 0.0.21 (2017-09-01)</h2>
<ul>
<li>NAMESPACE improvements. Reduced the number of re-exported functions to simplify the package.</li>
<li>Improved code coverage and unit testing with additional <a href="https://cran.r-project.org/package=testthat">testthat</a> checks. Specifically, added unit testing for remote download functions and improved testing for GTF file utilities.</li>
<li>Code coverage now above 90%!</li>
<li>Renamed <code>packageSE</code> to <code>prepareSE</code> for better semantic meaning.</li>
<li>Made multiple generics more flexible by inclusion of passthrough (<code>…</code>).</li>
<li>Reduced the number of deprecated functions.</li>
<li>Initial commit of internal <code>localOrRemote</code> utility function.</li>
<li>Initial commit of <code>prepareTemplate</code> function.</li>
<li>Added additional <code>data-raw/</code> scripts.</li>
<li>Added <code>onLoad.R</code> script back to ensure proper attachment of <a href="https://github.com/stephenturner/annotables/">annotables</a> data package.</li>
<li>Removed <a href="http://www.tidyverse.org/">tidyverse</a> S4 method support.</li>
<li>Improved remote file handling for <code>readFileByExtension</code>, <code>readGTF</code>, and <code>readYAML</code> functions.</li>
</ul>
</div>
    <div id="basejump-0-0-20-2017-08-16" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-20-2017-08-16" class="anchor"></a>basejump 0.0.20 (2017-08-16)</h2>
<ul>
<li>Offloaded <a href="https://cran.r-project.org/package=devtools">devtools</a> functions to personal package.</li>
<li>Upgraded all functions to S4 functions where possible.</li>
<li>Assign utilities were kept as S3 functions since S4 dispatch makes <code>parent.frame</code> assignment not work correctly.</li>
<li>Deprecated snake case and British spelling variants to reduce the number of exported functions.</li>
<li>Added more working examples.</li>
<li>Added unit testing for <a href="https://github.com/stephenturner/annotables/">annotables</a> functions.</li>
</ul>
</div>
    <div id="basejump-0-0-19-2017-07-24" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-19-2017-07-24" class="anchor"></a>basejump 0.0.19 (2017-07-24)</h2>
<ul>
<li>Improved documentation and consistency with <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> packages.</li>
<li>Improved integration of gene annotation calls using <a href="https://github.com/stephenturner/annotables/">annotables</a> package.</li>
</ul>
</div>
    <div id="basejump-0-0-18-2017-07-24" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-18-2017-07-24" class="anchor"></a>basejump 0.0.18 (2017-07-24)</h2>
<ul>
<li>Initial support for <a href="https://bioconductor.org/packages/SummarizedExperiment/">SummarizedExperiment</a> creation with <code>packageSE</code>.</li>
</ul>
</div>
    <div id="basejump-0-0-17-2017-07-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-17-2017-07-11" class="anchor"></a>basejump 0.0.17 (2017-07-11)</h2>
<ul>
<li>Added <a href="https://github.com/stephenturner/annotables/">annotables</a> common code from <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> packages.</li>
<li>Added automatic file reading using <a href="http://readr.tidyverse.org/">readr</a> package.</li>
<li>Combined write counts functions from <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a> and <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSinglecell</a>.</li>
<li>Initial commit of <code>assign_data</code> for use in <a href="http://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a> sample loops.</li>
</ul>
</div>
    <div id="basejump-0-0-16-2017-06-28" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-16-2017-06-28" class="anchor"></a>basejump 0.0.16 (2017-06-28)</h2>
<ul>
<li>Minor NAMESPACE updates while working on <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> packages.</li>
<li>Tweaks for <a href="http://www.tidyverse.org/">tidyverse</a> S4 generic verbs. In particular, <code>as_tibble</code> now provides better consistency for rowname conversion.</li>
</ul>
</div>
    <div id="basejump-0-0-15-2017-06-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-15-2017-06-18" class="anchor"></a>basejump 0.0.15 (2017-06-18)</h2>
<ul>
<li>Added <a href="https://cran.r-project.org/package=testthat">testthat</a> support for <a href="https://github.com/jimhester/lintr/">lintr</a> checks.</li>
<li>Added S4 generic for <code>as_tibble</code>.</li>
</ul>
</div>
    <div id="basejump-0-0-14-2017-06-13" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-14-2017-06-13" class="anchor"></a>basejump 0.0.14 (2017-06-13)</h2>
<ul>
<li>
<a href="http://dplyr.tidyverse.org/">dplyr</a> 0.7 NAMESPACE fixes and function tweaks.</li>
</ul>
</div>
    <div id="basejump-0-0-13-2017-06-12" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-13-2017-06-12" class="anchor"></a>basejump 0.0.13 (2017-06-12)</h2>
<ul>
<li>
<code>setMethod</code> for <a href="http://www.tidyverse.org/">tidyverse</a> NAMESPACE collisons (<a href="http://dplyr.tidyverse.org/">dplyr</a>, <a href="http://tidyr.tidyverse.org/">tidyr</a>) using `signature(“data.frame”).</li>
</ul>
</div>
    <div id="basejump-0-0-12-2017-05-25" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-12-2017-05-25" class="anchor"></a>basejump 0.0.12 (2017-05-25)</h2>
<ul>
<li>Updated exports based on <a href="https://steinbaugh.com/wormbase/">wormbase</a> package.</li>
</ul>
</div>
    <div id="basejump-0-0-11-2017-05-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-11-2017-05-18" class="anchor"></a>basejump 0.0.11 (2017-05-18)</h2>
<ul>
<li>Improved naming functions to dynamically handle character vectors and objects that support naming assignments.</li>
<li>Added <code>removeNA</code> utility function.</li>
</ul>
</div>
    <div id="basejump-0-0-10-2017-05-14" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-10-2017-05-14" class="anchor"></a>basejump 0.0.10 (2017-05-14)</h2>
<ul>
<li>Added NAMESPACE utilities to deal with <a href="http://www.tidyverse.org/">tidyverse</a> generic verbs.</li>
<li>Switched package documentation method to use <a href="https://cran.r-project.org/package=roxygen2">roxygen2</a> with pkgapi.</li>
</ul>
</div>
    <div id="basejump-0-0-9-2017-05-13" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-9-2017-05-13" class="anchor"></a>basejump 0.0.9 (2017-05-13)</h2>
<ul>
<li>Added snake case function variants.</li>
</ul>
</div>
    <div id="basejump-0-0-8-2017-05-13" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-8-2017-05-13" class="anchor"></a>basejump 0.0.8 (2017-05-13)</h2>
<ul>
<li>Added back <code>saveData</code> utility functions.</li>
</ul>
</div>
    <div id="basejump-0-0-7-2017-0422" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-7-2017-0422" class="anchor"></a>basejump 0.0.7 (2017-0422)</h2>
<ul>
<li>Bug fixes for <a href="http://dplyr.tidyverse.org/">dplyr</a> 0.6.0 update and improved kable handling.</li>
</ul>
</div>
    <div id="basejump-0-0-6-2017-04-14" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-6-2017-04-14" class="anchor"></a>basejump 0.0.6 (2017-04-14)</h2>
<ul>
<li>Dependency fix for successful compilation on the <a href="https://rc.hms.harvard.edu/">HMS RC</a> Orchestra cluster.</li>
</ul>
</div>
    <div id="basejump-0-0-5-2017-04-12" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-5-2017-04-12" class="anchor"></a>basejump 0.0.5 (2017-04-12)</h2>
<ul>
<li>Consolidated functions in the documentation.</li>
</ul>
</div>
    <div id="basejump-0-0-4-2017-04-07" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-4-2017-04-07" class="anchor"></a>basejump 0.0.4 (2017-04-07)</h2>
<ul>
<li>Improved documentation.</li>
</ul>
</div>
    <div id="basejump-0-0-3-2017-03-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-3-2017-03-22" class="anchor"></a>basejump 0.0.3 (2017-03-22)</h2>
<ul>
<li>Removed dependencies and transfer functions to <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a>.</li>
</ul>
</div>
    <div id="basejump-0-0-2-2017-03-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-2-2017-03-22" class="anchor"></a>basejump 0.0.2 (2017-03-22)</h2>
<ul>
<li>Added <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> data import functions.</li>
<li>Added <a href="http://ggplot2.tidyverse.org/">ggplot2</a> wrapper functions for quality control.</li>
</ul>
</div>
    <div id="basejump-0-0-1-2017-03-17" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-0-1-2017-03-17" class="anchor"></a>basejump 0.0.1 (2017-03-17)</h2>
<ul>
<li>Initial draft release.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#basejump-0-9-5-2019-01-22">0.9.5</a></li>
        <li><a href="#basejump-0-9-4-2019-01-12">0.9.4</a></li>
        <li><a href="#basejump-0-9-3-2019-01-08">0.9.3</a></li>
        <li><a href="#basejump-0-9-2-2019-01-07">0.9.2</a></li>
        <li><a href="#basejump-0-9-1-2018-12-22">0.9.1</a></li>
        <li><a href="#basejump-0-9-0-2018-12-12">0.9.0</a></li>
        <li><a href="#basejump-0-8-8-2018-12-03">0.8.8</a></li>
        <li><a href="#basejump-0-8-6-2018-11-30">0.8.6</a></li>
        <li><a href="#basejump-0-8-5-2018-11-29">0.8.5</a></li>
        <li><a href="#basejump-0-8-4-2018-11-28">0.8.4</a></li>
        <li><a href="#basejump-0-8-3-2018-11-25">0.8.3</a></li>
        <li><a href="#basejump-0-8-2-2018-11-19">0.8.2</a></li>
        <li><a href="#basejump-0-8-1-2018-11-15">0.8.1</a></li>
        <li><a href="#basejump-0-8-0-2018-11-11">0.8.0</a></li>
        <li><a href="#basejump-0-7-2-2018-08-29">0.7.2</a></li>
        <li><a href="#basejump-0-7-1-2018-08-18">0.7.1</a></li>
        <li><a href="#basejump-0-7-0-2018-08-07">0.7.0</a></li>
        <li><a href="#basejump-0-6-4-2018-08-03">0.6.4</a></li>
        <li><a href="#basejump-0-6-3-2018-08-01">0.6.3</a></li>
        <li><a href="#basejump-0-6-2-2018-07-31">0.6.2</a></li>
        <li><a href="#basejump-0-6-1-2018-07-21">0.6.1</a></li>
        <li><a href="#basejump-0-6-0-2018-07-17">0.6.0</a></li>
        <li><a href="#basejump-0-5-11-2018-07-09">0.5.11</a></li>
        <li><a href="#basejump-0-5-10-2018-06-28">0.5.10</a></li>
        <li><a href="#basejump-0-5-9-2018-06-12">0.5.9</a></li>
        <li><a href="#basejump-0-5-8-2018-06-05">0.5.8</a></li>
        <li><a href="#basejump-0-5-7-2018-05-24">0.5.7</a></li>
        <li><a href="#basejump-0-5-6-2018-05-19">0.5.6</a></li>
        <li><a href="#basejump-0-5-5-2018-05-15">0.5.5</a></li>
        <li><a href="#basejump-0-5-4-2018-05-08">0.5.4</a></li>
        <li><a href="#basejump-0-5-3-2018-04-30">0.5.3</a></li>
        <li><a href="#basejump-0-5-2-2018-04-26">0.5.2</a></li>
        <li><a href="#basejump-0-5-1-2018-04-16">0.5.1</a></li>
        <li><a href="#basejump-0-5-0-2018-04-13">0.5.0</a></li>
        <li><a href="#basejump-0-4-0-2018-03-22">0.4.0</a></li>
        <li><a href="#basejump-0-3-1-2018-02-19">0.3.1</a></li>
        <li><a href="#basejump-0-3-0-2018-02-16">0.3.0</a></li>
        <li><a href="#basejump-0-2-1-2018-02-07">0.2.1</a></li>
        <li><a href="#basejump-0-2-0-2018-01-28">0.2.0</a></li>
        <li><a href="#basejump-0-1-10-2017-12-20">0.1.10</a></li>
        <li><a href="#basejump-0-1-9-2017-12-18">0.1.9</a></li>
        <li><a href="#basejump-0-1-8-2017-12-14">0.1.8</a></li>
        <li><a href="#basejump-0-1-7-2017-12-11">0.1.7</a></li>
        <li><a href="#basejump-0-1-6-2017-12-06">0.1.6</a></li>
        <li><a href="#basejump-0-1-5-2017-12-05">0.1.5</a></li>
        <li><a href="#basejump-0-1-4-2017-12-04">0.1.4</a></li>
        <li><a href="#basejump-0-1-3-2017-12-01">0.1.3</a></li>
        <li><a href="#basejump-0-1-2-2017-11-30">0.1.2</a></li>
        <li><a href="#basejump-0-1-1-2017-11-11">0.1.1</a></li>
        <li><a href="#basejump-0-1-0-2017-10-23">0.1.0</a></li>
        <li><a href="#basejump-0-0-25-2017-10-13">0.0.25</a></li>
        <li><a href="#basejump-0-0-24-2017-10-10">0.0.24</a></li>
        <li><a href="#basejump-0-0-23-2017-09-14">0.0.23</a></li>
        <li><a href="#basejump-0-0-22-2017-09-08">0.0.22</a></li>
        <li><a href="#basejump-0-0-21-2017-09-01">0.0.21</a></li>
        <li><a href="#basejump-0-0-20-2017-08-16">0.0.20</a></li>
        <li><a href="#basejump-0-0-19-2017-07-24">0.0.19</a></li>
        <li><a href="#basejump-0-0-18-2017-07-24">0.0.18</a></li>
        <li><a href="#basejump-0-0-17-2017-07-11">0.0.17</a></li>
        <li><a href="#basejump-0-0-16-2017-06-28">0.0.16</a></li>
        <li><a href="#basejump-0-0-15-2017-06-18">0.0.15</a></li>
        <li><a href="#basejump-0-0-14-2017-06-13">0.0.14</a></li>
        <li><a href="#basejump-0-0-13-2017-06-12">0.0.13</a></li>
        <li><a href="#basejump-0-0-12-2017-05-25">0.0.12</a></li>
        <li><a href="#basejump-0-0-11-2017-05-18">0.0.11</a></li>
        <li><a href="#basejump-0-0-10-2017-05-14">0.0.10</a></li>
        <li><a href="#basejump-0-0-9-2017-05-13">0.0.9</a></li>
        <li><a href="#basejump-0-0-8-2017-05-13">0.0.8</a></li>
        <li><a href="#basejump-0-0-7-2017-0422">0.0.7</a></li>
        <li><a href="#basejump-0-0-6-2017-04-14">0.0.6</a></li>
        <li><a href="#basejump-0-0-5-2017-04-12">0.0.5</a></li>
        <li><a href="#basejump-0-0-4-2017-04-07">0.0.4</a></li>
        <li><a href="#basejump-0-0-3-2017-03-22">0.0.3</a></li>
        <li><a href="#basejump-0-0-2-2017-03-22">0.0.2</a></li>
        <li><a href="#basejump-0-0-1-2017-03-17">0.0.1</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

