<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.10 • basejump</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Version 0.10" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small>0.10</small></h1>
      
    </div>

    <div id="basejump-0-10-0-2019-03-28" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-0-2019-03-28" class="anchor"></a>basejump 0.10.0 (2019-03-28)</h2>
<div id="major-changes-9" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-9" class="anchor"></a>Major changes</h3>
<ul>
<li>Split out Ensembl (AnnotationHub/ensembldb) annotation processing and GFF/GTF file loading utilites to new <a href="https://freerange.acidgenomics.com/">freerange</a> package. All of these functions remain re-exported here in <a href="https://basejump.acidgenomics.com/">basejump</a>. This includes: <code>annotable</code>, <code>convertUCSCBuildToEnsembl</code>, <code>detectOrganism</code>, <code>emptyRanges</code>, <code>makeGRangesFromEnsDb</code>, <code>makeGRangesFromEnsembl</code>, <code>makeGRangesFromGFF</code>, <code>makeGRangesFromGTF</code>.</li>
<li>Split out heatmap functions to firestarter package. This includes <code>plotHeatmap</code>, <code>plotCorrelationHeatmap</code>, and <code>plotQuantileHeatmap</code>.</li>
<li>Split out all ggplot2 functions to new minimalism package.</li>
</ul>
</div>
<div id="minor-changes-27" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-27" class="anchor"></a>Minor changes</h3>
<ul>
<li>Offloaded <code>removeNA</code> and <code>sanitizeNA</code> code to <a href="https://brio.acidgenomics.com/">brio</a> package, so these functions can be imported in thew new <a href="https://freerange.acidgenomics.com/">freerange</a> package.</li>
<li>Moved <code>organism_mappings</code> internal dataset to <a href="https://freerange.acidgenomics.com/">freerange</a> package.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-1-2019-04-01" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-1-2019-04-01" class="anchor"></a>basejump 0.10.1 (2019-04-01)</h2>
<div id="new-functions-6" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-6" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>rankedMatrix</code>: New utility function for quickly performing ranked matrix calculations. Particularly useful for differential expression comparison across studies using log2 fold change or Wald test statistic.</li>
</ul>
</div>
<div id="minor-changes-26" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-26" class="anchor"></a>Minor changes</h3>
<ul>
<li>Simplified reexport methods for functions that define S4 methods.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-2-2019-04-07" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-2-2019-04-07" class="anchor"></a>basejump 0.10.2 (2019-04-07)</h2>
<div id="major-changes-8" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-8" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>aggregateRows</code>, <code>aggregateCols</code>, <code>aggregateCellsToSamples</code>: Improved internal code for SummarizedExperiment metadata handling. Applies primarily to colData and rowData juggling for these methods.</li>
<li>
<code>interestingGroups</code>: Reworked to define method against <code>Annotated</code> class. <code>SummarizedExperiment</code> inherits from this class, supporting <code>metadata</code>.</li>
<li>
<code>mapCellsToSamples</code>: Tightened up match assert checks.</li>
<li>
<code>mcolnames</code> now uses S4 methods, primarily against <code>Vector</code> class.</li>
<li>
<code>organism</code>: Reworked S4 methods. Added support for <code>Annotated</code> and <code>DataTable</code> classes from <a href="https://bioconductor.org/packages/S4Vectors/">S4Vectors</a>.</li>
</ul>
</div>
<div id="minor-changes-25" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-25" class="anchor"></a>Minor changes</h3>
<ul>
<li>Added additional unit tests, to improve code coverage.</li>
<li>Now covering the aggregation functions in better detail, using a minimal SingleCellExperiment object that works with <code>aggregateCols</code>.</li>
<li>Consistently use “acid” prefix instead of “basejump” for global options.</li>
<li>Miscellaneous working example improvements.</li>
<li>
<code>detectLanes</code>: Renamed primary argument from <code>object</code> to <code>path</code>. Improved <code>pattern</code> formal to evaluate <code>lanePattern</code> global.</li>
<li>For S4 generators, renamed test paramter from <code>basejump.test</code> to <code>acid.test</code>.</li>
<li>
<code>sampleData</code>: Improved error message when <code>sampleName</code> factor column is missing.</li>
<li>
<code>zeroVsDepth</code> now returns <code>depth</code> column as <code>integer</code> instead of <code>numeric</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-3-2019-04-07" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-3-2019-04-07" class="anchor"></a>basejump 0.10.3 (2019-04-07)</h2>
<div id="minor-changes-24" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-24" class="anchor"></a>Minor changes</h3>
<ul>
<li>Bug fix release for <a href="https://freerange.acidgenomics.com/">freerange</a> update. <code>emptyRanges</code> format has been renamed from <code>mcolsNames</code> to <code>mcolnames</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-4-2019-04-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-4-2019-04-22" class="anchor"></a>basejump 0.10.4 (2019-04-22)</h2>
<div id="major-changes-7" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-7" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>meltCounts</code>: Switched from using <code>nonzeroGenes</code> formal approach to <code>minCounts</code> and <code>minCountsMethod</code>, which is more flexible.</li>
</ul>
</div>
<div id="minor-changes-23" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-23" class="anchor"></a>Minor changes</h3>
<ul>
<li>Consolidated minimalism and firestarter code into acidplots package.</li>
<li>Improve global options defined in <code>formalsList</code>.</li>
<li>Added back basejump vignette.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-5-2019-04-25" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-5-2019-04-25" class="anchor"></a>basejump 0.10.5 (2019-04-25)</h2>
<div id="minor-changes-22" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-22" class="anchor"></a>Minor changes</h3>
<ul>
<li>Reworked S4 generic reexport method, in an attempt to get pkgdown to build vignette correctly. Otherwise, <code>sampleData</code> is erroring.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-6-2019-04-29" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-6-2019-04-29" class="anchor"></a>basejump 0.10.6 (2019-04-29)</h2>
<div id="new-functions-5" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-5" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>showHeader</code>: Utility function for <code>show</code> methods defined in other packages.</li>
</ul>
</div>
<div id="minor-changes-21" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-21" class="anchor"></a>Minor changes</h3>
<ul>
<li>Updated basejump dependencies (see <code>DESCRIPTION</code> for details).</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-7-2019-04-30" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-7-2019-04-30" class="anchor"></a>basejump 0.10.7 (2019-04-30)</h2>
<div id="new-functions-4" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-4" class="anchor"></a>New functions</h3>
<ul>
<li>Added method support for <code>alphaThreshold</code> and <code>lfcThreshold</code> against <code>Annotated</code> class. These values get stored in the <code>metadata</code> slot of the object.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-8-2019-05-05" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-8-2019-05-05" class="anchor"></a>basejump 0.10.8 (2019-05-05)</h2>
<div id="major-changes-6" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-6" class="anchor"></a>Major changes</h3>
<ul>
<li>Now pinned to R &gt;= 3.5.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-9-2019-05-29" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-9-2019-05-29" class="anchor"></a>basejump 0.10.9 (2019-05-29)</h2>
<div id="minor-changes-20" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-20" class="anchor"></a>Minor changes</h3>
<ul>
<li>Relaxed the deprecations on some functions to provide backward compatibility support for bcbioBase and bcbioRNASeq packages: <code>readFileByExtension</code>, <code>readYAML</code>, <code>fixNA</code>.</li>
<li>Now ensuring <code>theme_midnight</code> and <code>theme_paperwhite</code> are deprecated but exported with support, by suggesting acidplots package.</li>
<li>Added back defunct function warnings: <code>assertHasRownames</code>, <code>tx2geneFromGFF</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-10-2019-06-08" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-10-2019-06-08" class="anchor"></a>basejump 0.10.10 (2019-06-08)</h2>
<div id="new-functions-3" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-3" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>matchEnsemblReleaseToURL</code>: Takes an Ensembl release version (e.g. <code>96</code>) as input and returns the corresponding archive URL.</li>
<li>
<code>matchHumanOrthologs</code>: Convenience function that wraps biomaRt package to map model system gene identifiers to HGNC IDs and symbols. This is particularly useful for running orthologus GSEA with our pfgsea package.</li>
</ul>
</div>
<div id="major-changes-5" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-5" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>combine</code> method for <code>SummarizedExperiment</code> now includes all matrices defined in <code>assays</code> slot. Also improved support for <code>colData</code> handling on subsets where <code>NA</code> values have been removed.</li>
</ul>
</div>
<div id="minor-changes-19" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-19" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>Gene2Symbol</code>: Modified <code>format</code> formal to use “unmodified” instead of “long”, which is more intuitive.</li>
<li>
<code>makeGene2Symbol</code>: Added support for <code>format</code> argument, similar to <code>Gene2Symbol</code> generator function.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-11-2019-07-17" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-11-2019-07-17" class="anchor"></a>basejump 0.10.11 (2019-07-17)</h2>
<div id="minor-changes-18" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-18" class="anchor"></a>Minor changes</h3>
<ul>
<li>Made <code>theme_midnight</code> and <code>theme_paperwhite</code> defunct, in favor if variants in new <a href="https://acidplots.acidgenomics.com/">acidplots</a> package.</li>
<li>Decreased the number of functions reexported from goalie package.</li>
<li>Updated basejump dependency package versions.</li>
<li>Improved Travis CI Docker configuration.</li>
</ul>
</div>
</div>
    <div id="basejump-0-10-12-2019-07-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-10-12-2019-07-18" class="anchor"></a>basejump 0.10.12 (2019-07-18)</h2>
<div id="minor-changes-17" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-17" class="anchor"></a>Minor changes</h3>
<ul>
<li>Bug fix for <code>combine</code> method on SummarizedExperiment. Needed to ensure row names are assigned on rowData to provide backward compatibility for Bioc 3.7.</li>
<li>Improved unit test exceptions on Docker and AppVeyor.</li>
<li>Improved installation instructions.</li>
<li>
<code>mapGenesToRownames</code>: Improved matching for <code>SummarizedExperiment</code> objects that don’t contain gene-to-symbol mappings defined in <code>rowData</code>.</li>
<li>
<code>meltCounts</code>: Improved factor handling. Also added <code>matrix</code> method support. Added advanced option to disable <code>minCounts</code> filtering, by setting as <code>NULL</code>.</li>
</ul>
</div>
<div id="deprecations-1" class="section level3">
<h3 class="hasAnchor">
<a href="#deprecations-1" class="anchor"></a>Deprecations</h3>
<ul>
<li>Tightened up the list of deprecated functions.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#basejump-0-10-0-2019-03-28">0.10.0</a></li>
        <li><a href="#basejump-0-10-1-2019-04-01">0.10.1</a></li>
        <li><a href="#basejump-0-10-2-2019-04-07">0.10.2</a></li>
        <li><a href="#basejump-0-10-3-2019-04-07">0.10.3</a></li>
        <li><a href="#basejump-0-10-4-2019-04-22">0.10.4</a></li>
        <li><a href="#basejump-0-10-5-2019-04-25">0.10.5</a></li>
        <li><a href="#basejump-0-10-6-2019-04-29">0.10.6</a></li>
        <li><a href="#basejump-0-10-7-2019-04-30">0.10.7</a></li>
        <li><a href="#basejump-0-10-8-2019-05-05">0.10.8</a></li>
        <li><a href="#basejump-0-10-9-2019-05-29">0.10.9</a></li>
        <li><a href="#basejump-0-10-10-2019-06-08">0.10.10</a></li>
        <li><a href="#basejump-0-10-11-2019-07-17">0.10.11</a></li>
        <li><a href="#basejump-0-10-12-2019-07-18">0.10.12</a></li>
      </ul>
    </div>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


