<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.1 • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Version 0.1" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.10.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small>0.1</small></h1>
      
    </div>

    <div id="basejump-0-1-0-2017-10-23" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-0-2017-10-23" class="anchor"></a>basejump 0.1.0 (2017-10-23)</h2>
<ul>
<li>Bump version to match <a href="https://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a> package.</li>
<li>Improved unit testing coverage of <code>prepareSummarizedExperiment</code>.</li>
<li>Added quiet mode support to functions that output messages, where applicable.</li>
<li>Consolidated <a href="https://cran.r-project.org/package=roxygen2">roxygen2</a> function imports to <code>basejump-package.R</code> file.</li>
<li>Deprecated <code>sampleDirs</code> generic.</li>
<li>Improved organism detection in <code>detectOrganism</code> and added support for chicken genome.</li>
<li>Clarified warning messages in <code>prepareSummarizedExperiment</code> to make sample loading with <code>loadRNASeq</code> and <code>loadSingleCell</code> in the <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> packages less confusing.</li>
<li>Improved <code>NULL</code> returns in <code>readDataVersions</code>, <code>readLogFile</code>, and <code>readProgramVersions</code> utility functions.</li>
<li>Fixed export of <code>*GTF</code> alias functions to simply wrap the <code>*GFF</code> functions with S4 methods support.</li>
<li>Improved lane split technical replicate handling in <code>readSampleMetadataFile</code>.</li>
<li>Improved <code>camel</code> syntax for both lax and strict modes. Added <code>upperCamel</code> function.</li>
<li>Switched <code>str_</code> to base <code>grep</code> and <code>gsub</code> in internal functions.</li>
</ul>
</div>
    <div id="basejump-0-1-1-2017-11-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-1-2017-11-11" class="anchor"></a>basejump 0.1.1 (2017-11-11)</h2>
<ul>
<li>Added overwrite support for <code>saveData</code>. Now will skip on existing files when <code>overwrite = FALSE</code>.</li>
<li>Bug fix for <code>readDataVersions</code>, which shouldn’t have the column types defined, using <code>col_types = "ccT"</code>.</li>
<li>Improved key value pair method for <code>loadDataAsName</code>. Now rather than using a named character vector for the <code>mappings</code> argument, the user can simply pass the key value pairs in as dots. For example, <code>newName1 = "oldName1", newName2 = "oldName2"</code>. The legacy <code>mappings</code> method will still work, as long as the dots argument is a length of 1.</li>
<li>
<a href="https://www.ensembl.org/">Ensembl</a> release version now defaults to <code>NULL</code> instead of <code>current</code> for <code>annotable</code>, <code>gene2symbol</code>, <code>symbol2gene</code> and <code>tx2gene</code> functions.</li>
<li>Allow <code>rowData</code> to be left unset in <code>prepareSummarizedExperiment</code>. This is useful for setting up objects that don’t contain gene annotations.</li>
<li>Removed sample selection by pattern matching (<code>pattern</code>, <code>patternCol</code> arguments) in <code>readSampleMetadata</code>. This feature wasn’t fully baked and doesn’t offer enough functionality to the user.</li>
</ul>
</div>
    <div id="basejump-0-1-2-2017-11-30" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-2-2017-11-30" class="anchor"></a>basejump 0.1.2 (2017-11-30)</h2>
<ul>
<li>Added <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> <code>plotQC</code> generic.</li>
<li>Added back <code>toStringUnique</code> code, which is still in use in the <a href="https://wormbase.acidgenomics.com/">wormbase</a> package.</li>
<li>Added deprecations for <code>summarizeRows</code> (now <code>collapseToString</code>) and <code>wash</code> functions.</li>
<li>Updated installation method to include manual installation of <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a>. Otherwise, <a href="https://basejump.acidgenomics.com/">basejump</a> installation fails due to <a href="https://bioconductor.org/packages/GenomeInfoDb/">GenomeInfoDb</a> and <a href="https://bioconductor.org/packages/GenomeInfoDbData/">GenomeInfoDbData</a> not getting installed completely.</li>
<li>Now suggesting that the user installs suggested packages in the README.</li>
<li>Updated <a href="http://pantherdb.org/">PANTHER</a> annotation scripts.</li>
<li>Bug fix for <code>detectOrganism</code>. Now allowing <code>NULL</code> return for unsupported organism, with a warning.</li>
</ul>
</div>
    <div id="basejump-0-1-3-2017-12-01" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-3-2017-12-01" class="anchor"></a>basejump 0.1.3 (2017-12-01)</h2>
<ul>
<li>
<code>loadData</code> and <code>loadDataAsName</code> now default to <code>replace = TRUE</code>. If an object with the same name exists in the destination environment, then a warning is generated.</li>
<li>
<code>collapseToString</code> only attempts to dynamically return the original object class on objects that aren’t class <code>data.frame</code>. I updated this code to behave more nicely with grouped tibbles (<code>grouped_df</code>), which are a virtual class of <code>data.frame</code> and therefore can’t be coerced using <code>as(object, "grouped_df")</code>.</li>
<li>DNA sequence utility functions <code>comp</code> and <code>revcomp</code> now return <code>NULL</code> for integers and numerics.</li>
<li>For <code>prepareSummarizedExperiment</code>, added support for dropping <code>NULL</code> objects in assays list. This is useful for handling output from <a href="https://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a> when <code>transformLimit</code> is reached. In this case, the <code>rlog</code> and <code>vst</code> matrices aren’t generated and set <code>NULL</code> in the assays list. Using <code><a href="https://rdrr.io/r/base/funprog.html">Filter(Negate(is.null), assays)</a></code> we can drop these <code>NULL</code> objects and prevent a downstream dimension mismatch in the <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html">SummarizedExperiment::SummarizedExperiment</a></code> call.</li>
<li>Improved support for multiplexed files in <code>readSampleMetadataFile</code>. This now checks for a sequence column containing ACGT nucleotides. When those are detected, the <code>revcomp</code> column is generated. Otherwise this step is skipped. This is useful for handling multiplexed sample metadata from 10X Genomics <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> single-cell RNA-seq samples.</li>
<li>Updated <code>annotable</code> function to include nested <a href="https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html">Entrez</a> identifiers in the <code>entrez</code> column. This is useful for downstream functional analysis.</li>
</ul>
</div>
    <div id="basejump-0-1-4-2017-12-04" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-4-2017-12-04" class="anchor"></a>basejump 0.1.4 (2017-12-04)</h2>
<ul>
<li>Added <code>midnightTheme</code> <a href="http://ggplot2.tidyverse.org/">ggplot2</a> theme. Originally this was defined as <code>darkTheme</code> in the <a href="https://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a> package, but can be useful for other plots and has been moved here for general bioinformatics usage. The theme now uses <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">ggplot2::theme_minimal</a></code> as the base, with some color tweaks, namely dark gray axes without white axis lines.</li>
<li>Improve NAMESPACE imports to include <code><a href="https://rdrr.io/r/stats/formula.html">stats::formula</a></code> and <code><a href="https://rdrr.io/r/utils/capture.output.html">utils::capture.output</a></code>.</li>
</ul>
</div>
    <div id="basejump-0-1-5-2017-12-05" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-5-2017-12-05" class="anchor"></a>basejump 0.1.5 (2017-12-05)</h2>
<ul>
<li>Added <code>checkAnnotable</code>, <code>checkGene2symbol</code>, <code>checkTx2gene</code>, and <code>sanitizeAnnotable</code> utility functions that will be used in the bcbio R packages.</li>
</ul>
</div>
    <div id="basejump-0-1-6-2017-12-06" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-6-2017-12-06" class="anchor"></a>basejump 0.1.6 (2017-12-06)</h2>
<ul>
<li>Moved <code>microplate</code> code from the <a href="https://wormbase.acidgenomics.com/">wormbase</a> package here, since it’s of general interest.</li>
</ul>
</div>
    <div id="basejump-0-1-7-2017-12-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-7-2017-12-11" class="anchor"></a>basejump 0.1.7 (2017-12-11)</h2>
<ul>
<li>Renamed <code>fc2lr</code> to <code>foldChangeToLogRatio</code> and <code>lr2fc</code> and <code>logRatioToFoldChange</code>.</li>
<li>Moved <code>plotDot</code> and <code>plotViolin</code> generics here from <a href="https://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a>.</li>
<li>Added internal <a href="https://grch37.ensembl.org/">GRCh37</a> gene annotations.</li>
</ul>
</div>
    <div id="basejump-0-1-8-2017-12-14" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-8-2017-12-14" class="anchor"></a>basejump 0.1.8 (2017-12-14)</h2>
<ul>
<li>Improved <code>matrix</code> and <code>dgCMatrix</code> method support in <code>aggregateReplicates</code> and <code>aggregateFeatures</code> functions. Both of these functions now use a consistent <code>groupings</code> parameter, which uses a named factor to define the mappings of either samples (columns) for <code>aggregateReplicates</code> or genes/transcripts (rows) for <code>aggregateFeatures</code>.</li>
<li>Update for <code>makeNames</code> sanitization functions. Now they will work on <code><a href="https://rdrr.io/r/base/names.html">names(x)</a></code> for vectors by default.</li>
<li>Improved <code>detectOrganism</code> to match against “H. sapiens”, etc.</li>
<li>Added internal <a href="https://grch37.ensembl.org/">GRCh37</a> transcript to gene mapping.</li>
<li>Improved organism matching to detect “Homo_sapiens” and “H. sapiens”.</li>
<li>Factors are now supported in the <code>makeNames</code> utilities: <code>camel</code>, <code>dotted</code>, <code>snake</code>, and <code>upperCamel</code>.</li>
<li>Improved handling of <code>NA</code> values from <a href="https://www.libreoffice.org/">LibreOffice</a> and <a href="https://products.office.com/en-us/excel/">Microsoft Excel</a> output in <code>readFileByExtension</code>. This function now sets <code>""</code>, <code>NA</code>, and <code>#N/A</code> strings as <code>NA</code> correctly.</li>
</ul>
</div>
    <div id="basejump-0-1-9-2017-12-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-9-2017-12-18" class="anchor"></a>basejump 0.1.9 (2017-12-18)</h2>
<ul>
<li>Added <code>plotHeatmap</code> functionality.</li>
<li>Migrated <code>tpm</code> generic from <a href="https://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a>, for future use in <a href="https://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a>.</li>
<li>Added matrix method support for <code>plotHeatmap</code>.</li>
<li>Added matrix method support for <code>plotQuantileHeatmap</code>, which works similarly as <code>plotHeatmap</code>.</li>
</ul>
</div>
    <div id="basejump-0-1-10-2017-12-20" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-1-10-2017-12-20" class="anchor"></a>basejump 0.1.10 (2017-12-20)</h2>
<ul>
<li>Bug fix for multiplexed sample input into <code>readSampleMetadataFile</code>. We were detecting the presence of <code>index</code> column but should instead check against <code>sequence</code> column.</li>
<li>Added <code>dynamicPlotlist</code> and <code>mdPlotlist</code> plotting utilities.</li>
<li>Added <code>uniqueSymbols</code> parameter to <code>annotable</code> function.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#basejump-0-1-0-2017-10-23">0.1.0</a></li>
        <li><a href="#basejump-0-1-1-2017-11-11">0.1.1</a></li>
        <li><a href="#basejump-0-1-2-2017-11-30">0.1.2</a></li>
        <li><a href="#basejump-0-1-3-2017-12-01">0.1.3</a></li>
        <li><a href="#basejump-0-1-4-2017-12-04">0.1.4</a></li>
        <li><a href="#basejump-0-1-5-2017-12-05">0.1.5</a></li>
        <li><a href="#basejump-0-1-6-2017-12-06">0.1.6</a></li>
        <li><a href="#basejump-0-1-7-2017-12-11">0.1.7</a></li>
        <li><a href="#basejump-0-1-8-2017-12-14">0.1.8</a></li>
        <li><a href="#basejump-0-1-9-2017-12-18">0.1.9</a></li>
        <li><a href="#basejump-0-1-10-2017-12-20">0.1.10</a></li>
      </ul>
    </div>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


