<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.7 • basejump</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Version 0.7" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.12.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small>0.7</small></h1>
      
    </div>

    <div id="basejump-070-2018-08-07" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-070-2018-08-07" class="anchor"></a>basejump 0.7.0 (2018-08-07)</h2>
<p>In this release, we are migrating some of the S4 generics previously exported in the <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package. We are consolidating these functions here for simplicity and stability.</p>
<div id="new-functions-17" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-17" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>makeSummarizedExperiment</code>: Renamed <code>prepareSummarizedExperiment</code>, previously exported in the <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package. We are using the <code>make</code> prefix here for consistency (see other gene annotation functions).</li>
<li>Migrated S4 generics and methods from <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a>: <code>flatFiles</code>, <code>metrics</code>, <code>plotCorrleationHeatmap</code>, <code>plotGene</code>, <code>plotHeatmap</code>, <code>plotQC</code>, and <code>plotQuantileHeatmap</code>.</li>
</ul>
</div>
<div id="major-changes-22" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-22" class="anchor"></a>Major changes</h3>
<ul>
<li>Now using <code><a href="https://jeroen.cran.dev/curl/reference/nslookup.html">curl::has_internet</a></code> internally to check for Internet connection. This applies to the annotation functions that query web databases.</li>
<li>Added coercion method support for converting a <code>SummarizedExperiment</code> to an unstructured <code>list</code>. This is the method used internally for <code>flatFiles</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-071-2018-08-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-071-2018-08-18" class="anchor"></a>basejump 0.7.1 (2018-08-18)</h2>
<div id="new-functions-16" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-16" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>cleanSystemLibrary</code>: Utility function to check whether a user has installed packages into the <a href="https://www.r-project.org/">R</a> system library. Refer to <code>.libPaths</code> documentation for more information on library paths.</li>
</ul>
</div>
<div id="major-changes-21" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-21" class="anchor"></a>Major changes</h3>
<ul>
<li>Now using <code>build</code> instead of <code>genomeBuild</code> for <a href="https://www.ensembl.org/">Ensembl</a> annotation functions. The <code>genomeBuild</code> argument still works but now will inform the user about the change.</li>
</ul>
</div>
<div id="minor-changes-67" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-67" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrated <code>prepareTemplate</code> from <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package. Simplifed this function to copy all files inside <code>extdata/rmarkdown/shared</code> within a specified package. Currently in use for <a href="https://bioinformatics.sph.harvard.edu/bcbioRNASeq/">bcbioRNASeq</a>, <a href="https://bioinformatics.sph.harvard.edu/bcbioSingleCell/">bcbioSingleCell</a>, and the new <a href="https://pointillism.acidgenomics.com/">pointillism</a> clustering package.</li>
<li>Made <a href="https://www.ensembl.org/">Ensembl</a> release matching stricter, based on the metadata columns.</li>
</ul>
</div>
</div>
    <div id="basejump-072-2018-08-29" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-072-2018-08-29" class="anchor"></a>basejump 0.7.2 (2018-08-29)</h2>
<div id="new-functions-15" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-15" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>assertAllAreURL</code>.</li>
<li>
<code>assertAllAreValidNames</code>.</li>
<li>
<code>validNames</code>.</li>
</ul>
</div>
<div id="major-changes-20" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-20" class="anchor"></a>Major changes</h3>
<ul>
<li>Added a draft vignette explaining the functions available in the package.</li>
<li>
<code>gene2symbol</code>, <code>makeGene2symbolFromEnsembl</code>, and <code>makeGene2symbolFromGFF</code> functions now support the <code>unique</code> argument, which returns sanitized values in the <code>geneName</code> column, ensuring there are no duplicates. This is enabled by default (recommended) but can be disabled using <code>unique = FALSE</code>. This functionality was added to ensure consistent gene name handling in single-cell RNA-seq analyses.</li>
<li>
<code>saveData</code> now supports <code>basejump.save.ext</code> and <code>basejump.save.compress</code> global options, so the desired file type (e.g. RDS instead of RDA) and compression (e.g. xz instead of gzip) can be easily specified for an entire project.</li>
</ul>
</div>
<div id="minor-changes-66" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-66" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>sampleNames</code> now supports assignment for <code>SummarizedExperiment</code> method.</li>
<li>Now exporting <code>lanePattern</code> regular expression pattern as a global, which was previously defined in the <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package.</li>
<li>Bug fix for <code>as</code> coercion method support. Need to ensure <code>exportMethods(coerce)</code> is included in <code>NAMESPACE</code> file, otherwise <code>tibble</code> coercion using <code>as(x, "tbl_df")</code> won’t work when called from an Rscript without the package library loaded. Thanks <a href='https://github.com/roryk'>@roryk</a> for noticing this.</li>
<li>Updated <code>gene2symbol</code> generic to use <code>...</code>, since we’ve added the <code>unique = TRUE</code> argument in this release.</li>
<li>
<code>annotable</code>: Moved to <code>makeGRanges.R</code> file, and improved the internal code to export supported formals also used in <code>makeGRangesFromEnsembl</code>. The function should work exactly the same as previous releases, but now with clearer supported arguments in the documentation.</li>
<li>Skipping code coverage for <code>cleanSystemLibrary</code>, since <a href="https://travis-ci.com/">Travis CI</a> installs packages into the system library, and causes this check to return <code>FALSE</code>.</li>
<li>Consistently using <code>as(from, "tbl_df")</code> for internal tibble coercion in all functions.</li>
<li>
<code>geneSynonyms</code> and <code>panther</code>: <code>organism</code> argument matching no longer suggests a default. The current list of supported organisms is in the documentation, and described in the internal <code>match.arg</code> call.</li>
<li>All <code>SummarizedExperiment</code> methods use <code>validObject</code> validity checks, where applicable.</li>
<li>Consolidated documentation for all <code>makeGRanges</code>, <code>makeGene2symbol</code>, and <code>makeTx2gene</code> functions.</li>
<li>Heatmap functions: simplified the internal code responsible for defining <code>annotationCol</code> and <code>annotationColors</code> automatically.</li>
<li>
<code>sampleData</code>: Made validity check stricter, requiring <code>sampleName</code> column to be defined, otherwise the function will intentionally error.</li>
<li>Now using <code>formals</code> internall to keep ggplot2 theme formals consistent.</li>
<li>Updated example data scripts and resaved internal data.</li>
<li>Updated contribution guidelines in <code>CONTRIBUTING.md</code> file.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


