<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.6 • basejump</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Version 0.6" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.12.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://acidgenomics.com/">
    <span class="fas fa fas fa-vial fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/acidgenomics/basejump/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small>0.6</small></h1>
      
    </div>

    <div id="basejump-060-2018-07-17" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-060-2018-07-17" class="anchor"></a>basejump 0.6.0 (2018-07-17)</h2>
<div id="major-changes-24" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-24" class="anchor"></a>Major changes</h3>
<ul>
<li>Now importing <a href="https://bioconductor.org/packages/SummarizedExperiment/">SummarizedExperiment</a> package and providing basic method support for generics that were previously used in the <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package.</li>
<li>Improved GFFv3 handling in <code>makeGRangesFromGFF</code> and other GFF utility functions, including <code>makeGene2symbolFromGFF</code> and <code>makeTx2geneFromGFF</code>. Note that <code>makeGRangesFromGFF</code> now returns additional metadata columns accessible with <code><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">S4Vectors::mcols</a></code>, and that these columns are now sorted alphabetically.</li>
</ul>
</div>
<div id="migrated-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#migrated-functions" class="anchor"></a>Migrated functions</h3>
<p>Previously, these functions were exported in the <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package, but they provide non-<a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a>-specific functionality, and should be included here in the <a href="https://basejump.acidgenomics.com/">basejump</a> package instead:</p>
<ul>
<li>
<code>assertFormalInterestingGroups</code>.</li>
<li>
<code>gene2symbol</code>.</li>
<li>
<code>interestingGroups</code>, <code>uniteInterestingGroups</code>.</li>
<li>
<code>sampleData</code>, <code>sanitizeSampleData</code>.</li>
<li>
<code>sampleNames</code>.</li>
<li>
<code>selectSamples</code>.</li>
</ul>
<p>Providing basic <code>SummarizedExperiment</code> class method support for <code>counts</code>.</p>
</div>
<div id="minor-changes-75" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-75" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>geometricMean</code> generic was not exported correctly.</li>
</ul>
</div>
</div>
    <div id="basejump-061-2018-07-21" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-061-2018-07-21" class="anchor"></a>basejump 0.6.1 (2018-07-21)</h2>
<div id="minor-changes-74" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-74" class="anchor"></a>Minor changes</h3>
<ul>
<li>Bug fix for <code>convertGenesToSymbols</code> method for <code>SummarizedExperiment</code>. Previously, if <code>geneName</code> column was a factor, this function would error. This issue has been fixed by ensuring that the symbols provided in <code>geneName</code> are coerced to a character vector.</li>
<li>Improved <a href="https://conda.io/">conda</a> installation instructions.</li>
</ul>
</div>
</div>
    <div id="basejump-062-2018-07-31" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-062-2018-07-31" class="anchor"></a>basejump 0.6.2 (2018-07-31)</h2>
<div id="new-functions-21" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-21" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>convertSymbolsToGenes</code>: provides <code>SummarizedExperiment</code> method support for converting objects containing gene symbols (“geneName”) as rownames back to gene identifiers (“geneID”).</li>
<li>
<code>eggnog</code>: quickly download current annotations from <a href="http://eggnogdb.embl.de/">EggNOG</a> database. Useful for annotating gene-to-protein matches; currently in use with the <a href="https://brightworm.acidgenomics.com/">brightworm</a> RNAi screening package, which contains <a href="http://www.wormbase.org/">WormBase</a> gene ID and <a href="http://eggnogdb.embl.de/">EggNOG</a> ID annotations.</li>
</ul>
</div>
<div id="major-changes-23" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-23" class="anchor"></a>Major changes</h3>
<ul>
<li>Now suggesting <a href="https://cran.r-project.org/package=BiocManager">BiocManager</a> instead of <a href="https://bioconductor.org/packages/BiocInstaller/">BiocInstaller</a> for installation.</li>
<li>
<code>broadClass</code> now supports <code>GRanges</code> and <code>SummarizedExperiment</code>. Support for <code>data.frame</code> and/or <code>DataFrame</code> class objects has been removed.</li>
</ul>
</div>
<div id="minor-changes-73" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-73" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>convertGenesToSymbols</code> and <code>convertTranscriptsToGenes</code> now have <code>organism</code> and <code>gene2symbol</code> arguments set <code>NULL</code> by default.</li>
<li>Upgraded to <a href="https://cran.r-project.org/package=roxygen2">roxygen2</a> v6.1 for documentation, which improves handling of aliases in the Rd manual files.</li>
<li>Added <code><a href="https://dplyr.tidyverse.org/reference/pull.html">dplyr::pull</a></code> to reexported functions.</li>
<li>Improved package documentation by declaring the supported class(es) for each function argument.</li>
<li>Moved <code>foldChangeToLogRatio</code> and <code>logRatioToFoldChange</code> constructors into <code>numeric</code> method declarations.</li>
<li>Simplified internal code for <code>gene2symbol</code> <code>SummarizedExperiment</code> method.</li>
<li>
<code>toStringUnique</code> now uses <code>x</code> instead of <code>atomic</code> as primary argument.</li>
</ul>
</div>
</div>
    <div id="basejump-063-2018-08-01" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-063-2018-08-01" class="anchor"></a>basejump 0.6.3 (2018-08-01)</h2>
<div id="minor-changes-72" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-72" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrated <code>separatorBar</code> and <code>updateMessage</code> global export from <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a>. Improved <code>separatorBar</code> appeareance to automatically scale to current session width, using <code><a href="https://rdrr.io/r/base/options.html">getOption("width")</a></code>.</li>
</ul>
</div>
</div>
    <div id="basejump-064-2018-08-03" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-064-2018-08-03" class="anchor"></a>basejump 0.6.4 (2018-08-03)</h2>
<div id="new-functions-20" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-20" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>matchInterestingGroups</code>: New developer function to automatically handle <code>interestingGroups</code> argument used across various plotting functions and in the <a href="https://github.com/bcbio/bcbio-nextgen/">bcbio</a> infrastructure packges.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


