<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.11 • basejump</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Version 0.11" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.22</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small>0.11</small></h1>
      
    </div>

    <div id="basejump-0-11-0-2019-07-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-0-2019-07-22" class="anchor"></a>basejump 0.11.0 (2019-07-22)</h2>
<p>Start of new release series. Version bump reflects changes in dependency packages. See the acidtest, bioverbs, freerange, syntactic, and transformer release notes for more details.</p>
</div>
    <div id="basejump-0-11-1-2019-07-24" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-1-2019-07-24" class="anchor"></a>basejump 0.11.1 (2019-07-24)</h2>
<div id="minor-changes-19" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-19" class="anchor"></a>Minor changes</h3>
<ul>
<li>Added back deprecated assert checks that are required for bcbioRNASeq v0.2 release series.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-2-2019-07-29" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-2-2019-07-29" class="anchor"></a>basejump 0.11.2 (2019-07-29)</h2>
<div id="minor-changes-18" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-18" class="anchor"></a>Minor changes</h3>
<ul>
<li>Reexporting new functions in syntactic: <code>makeLabel</code>, <code>makeTitle</code>, <code>makeWords</code>.</li>
<li>Updated documentation to include modification timestamp.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-3-2019-07-30" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-3-2019-07-30" class="anchor"></a>basejump 0.11.3 (2019-07-30)</h2>
<p>100th release!</p>
<div id="minor-changes-17" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-17" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>mapCellsToSamples</code>: Relaxed grep matching on <code>cells</code> input to support legacy bcbioSingleCell objects. This change was needed to improve <code>updateObject</code> method in the upcoming bcbioSingleCell update.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-4-2019-08-07" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-4-2019-08-07" class="anchor"></a>basejump 0.11.4 (2019-08-07)</h2>
<div id="major-changes-5" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-5" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>makeSampleData</code>: Switched to S4 method that works on <code>data.frame</code> and <code>DataFrame</code> class objects. Enforcing lower camel case for column names.</li>
<li>
<code>makeSummarizedExperiment</code> and <code>makeSingleCellExperiment</code>: Switched to S4 method approach that currently requires <code>SimpleList</code> input for assays. Previously, we allowed <code>list</code> input for assays, but we’re tightening this up to simply the package code base and improve consistency.</li>
<li>
<code>aggregate</code> methods now consistently return the primary assay named as <code>counts</code>. This follows the recommended conventions defined in SingleCellExperiment. Aggregation functions will now intentionally fail for <code>SummarizedExperiment</code> objects that don’t contain an assay named <code>counts</code>.</li>
</ul>
</div>
<div id="minor-changes-16" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-16" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved documentation consistency by offloading <code>params.Rd</code> file to new acidroxygen package. This will be linked in the other Acid Genomics packages.</li>
<li>Updated unit tests to follow new package conventions (see above changes).</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-5-2019-08-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-5-2019-08-11" class="anchor"></a>basejump 0.11.5 (2019-08-11)</h2>
<div id="new-functions-4" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-4" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>calculateMetrics</code>: Migrated code here from bcbioSingleCell. Improved method to support <code>DelayedArray</code> class for large matrices.</li>
</ul>
</div>
<div id="minor-changes-15" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-15" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved code coverage and adjusted unit tests for breaking changes seen due to new covr update.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-6-2019-08-12" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-6-2019-08-12" class="anchor"></a>basejump 0.11.6 (2019-08-12)</h2>
<div id="minor-changes-14" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-14" class="anchor"></a>Minor changes</h3>
<ul>
<li>Updated goalie dependency.</li>
<li>Tightened up <code>appendToBody</code> and <code>methodFormals</code> calls to have better backward compatibility with R 3.5.</li>
<li>
<code>calculateMetrics</code>: Bug fix for getting <code>rowData</code> via <code>mcols</code> without <code>use.names = TRUE</code>. This improves Bioconductor backward compatibility. Also updated internal code to call <code>hasMetrics</code> from goalie.</li>
<li>Made <code>separatorBar</code> defunct. Use <code>separator</code> function instead.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-7-2019-08-13" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-7-2019-08-13" class="anchor"></a>basejump 0.11.7 (2019-08-13)</h2>
<div id="minor-changes-13" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-13" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>filterCells</code>: Improved downstream handling of <code>nCells</code> argument. Ensure double filtering is still allowed.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-8-2019-08-19" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-8-2019-08-19" class="anchor"></a>basejump 0.11.8 (2019-08-19)</h2>
<div id="new-functions-3" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-3" class="anchor"></a>New functions</h3>
<ul>
<li>Migrated <code>readSampleData</code> and <code>readTx2Gene</code> from bcbioBase.</li>
<li>Reworked <code>readSampleData</code> internal code, but still supporting bcbio pipeline conventions (i.e. “description” column for samples) as the default. I’ve reworked this approach so we can also call <code>readSampleData</code> inside the Chromium package (for 10X Genomics single-cell RNA-seq data) without having to depend on the bcbio R packages.</li>
<li>Reexporting new dplyr-like methods that support S4 <code>DataFrame</code>: <code>inner_join</code>, <code>left_join</code>, <code>right_join</code>, <code>full_join</code>, <code>anti_join</code>; <code>mutate_all</code>, <code>mutate_at</code>, <code>mutate_if</code>; <code>select_all</code>, <code>select_at</code>, <code>select_if</code>.</li>
</ul>
</div>
<div id="major-changes-4" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-4" class="anchor"></a>Major changes</h3>
<ul>
<li>Removed dplyr and magrittr dependencies in internal code, where applicable.</li>
<li>
<code>aggregateCols</code>: Sped up return by calling <code>SingleCellExperiment</code> rather than <code>makeSingleCellExperiment</code>. Note that this now doesn’t return session information in the object.</li>
<li>
<code>cell2sample</code>: Renamed return from <code>tibble</code> to <code>tbl_df</code>, for consistency.</li>
<li>Made some previous deprecated functions still used in bcbioRNASeq v0.2 release series (which has now been updated to v0.3) defunct. This applies primarily to <code>assert*</code> functions that have been reworked using a goalie approach.</li>
<li>
<code>estimateSizeFactors</code>: Removed option to calculate “deseq2-median-ratio” using DESeq2. We may revisit this idea in a future release.</li>
<li>
<code>makeSampleData</code>: Improved internal code for <code>DataFrame</code> method.</li>
</ul>
</div>
<div id="minor-changes-12" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-12" class="anchor"></a>Minor changes</h3>
<ul>
<li>Updated basejump dependencies, in preparation for bioconda release update supporting bcbioRNASeq.</li>
<li>Now importing AnnotationDbi, BiocParallel, Biostrings, and biomaRt.</li>
<li>Safe to import <code>select</code> from AnnotationDbi, now that we’re no longer depending on dplyr.</li>
<li>Switched to using <code>droplevels</code> instead of <code>relevel</code> on S4 objects internally, where applicable. This is better supported by S4Vectors package.</li>
<li>Simplified reexport documentation for S4 functions and methods.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-9-2019-08-21" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-9-2019-08-21" class="anchor"></a>basejump 0.11.9 (2019-08-21)</h2>
<div id="minor-changes-11" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-11" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>meltCounts</code>: Added initial method support for <code>SingleCellExperiment</code>. Currently requires deparsing of the count matrix to call <code><a href="https://rdrr.io/pkg/reshape2/man/melt.html">reshape2::melt</a></code>. Returns columns with S4 run-length encoding (Rle) applied.</li>
<li>
<code>mapGenes</code>: Converted warning to message when <code>strict = FALSE</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-10-2019-08-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-10-2019-08-22" class="anchor"></a>basejump 0.11.10 (2019-08-22)</h2>
<div id="minor-changes-10" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-10" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>makeSummarizedExperiment</code>: Now automatically handles non-Ensembl gene symbols present in assays that aren’t defined in <code>rowRanges</code>. This applies primarily to 10X Cell Ranger v3 output, which includes some non-Ensembl gene symbols: CD11b, CD4, CD8a, HLA-DR, IgG1, PD-1, etc. The function still intentionally errors on unannotated Ensembl features, which often indicates an accidental release version mismatch.</li>
<li>Updated dependency versions.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-11-2019-08-27" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-11-2019-08-27" class="anchor"></a>basejump 0.11.11 (2019-08-27)</h2>
<p>Updated R dependency to 3.6.</p>
<div id="new-functions-2" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-2" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>melt</code>: Added S4 methods for melting data into long format. Provides support for <code>matrix</code>, <code>Matrix</code>, <code>DataFrame</code>, <code>SummarizedExperiment</code>, and <code>SingleCellExperiment</code> currently.</li>
<li>
<code>nonzeroRowsAndCols</code>: Quickly remove non-zero rows and columns from a matrix or <code>SummarizedExperiment</code>.</li>
</ul>
</div>
<div id="major-changes-3" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-3" class="anchor"></a>Major changes</h3>
<ul>
<li>Migrated <code>EggNOG</code> and <code>PANTHER</code> S4 classes to separate packages.</li>
</ul>
</div>
<div id="minor-changes-9" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-9" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>calculateMetrics</code>: Now calls <code>nonzeroRowsAndCols</code> internally first when <code>prefilter = FALSE</code>, speeding up calculations significantly for very large <code>SingleCellExperiment</code> objects. This was added to improve loading of example unfiltered 10X Genomics Chromium data.</li>
</ul>
</div>
<div id="deprecations" class="section level3">
<h3 class="hasAnchor">
<a href="#deprecations" class="anchor"></a>Deprecations</h3>
<ul>
<li>Deprecated <code>readSampleData</code> and <code>readTx2Gene</code> in favor of <code>importSampleData</code> and <code>importTx2Gene</code>, respectively.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-12-2019-09-03" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-12-2019-09-03" class="anchor"></a>basejump 0.11.12 (2019-09-03)</h2>
<div id="minor-changes-8" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-8" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>melt</code>: Added method support for contingency <code>table</code> class.</li>
<li>Removed <code>set_*</code> reexports from magrittr packages.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-13-2019-09-06" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-13-2019-09-06" class="anchor"></a>basejump 0.11.13 (2019-09-06)</h2>
<div id="major-changes-2" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-2" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>autopadZeros</code>: Added improved support for detection and automatic handling of zeros in need of padding on the left side of a character vector. This addition is necessary for handling of Genewiz processed FASTQ file names.</li>
<li>
<code>makeTx2Gene</code> functions now support <code>ignoreTxVersion</code> argument, similar to conventions defined in tximport package.</li>
</ul>
</div>
<div id="minor-changes-7" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-7" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>importSampleData</code>: Updated to tentatively support a general pipeline via the “none” argument. In this case, only a <code>sampleID</code> column is required in metadata. This is been developed in conjunction with my new Genewiz to kallisto Nextflow processing pipeline being implemented at CPI.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-14-2019-09-09" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-14-2019-09-09" class="anchor"></a>basejump 0.11.14 (2019-09-09)</h2>
<div id="minor-changes-6" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-6" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>mcolnames</code>: Moved S4 methods previously defined in syntactic here.</li>
<li>Improved website documentation.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-15-2019-09-16" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-15-2019-09-16" class="anchor"></a>basejump 0.11.15 (2019-09-16)</h2>
<div id="major-changes-1" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-1" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>melt</code>: Updated <code>min</code> and <code>minMethod</code> defaults for <code>matrix</code> method. The <code>min</code> argument now defaults to <code>-Inf</code>, and the <code>minMethod</code> now defaults to <code>"absolute"</code>, instead of <code>"perRow"</code>, since this behavior is more intuitive to the user.</li>
</ul>
</div>
<div id="minor-changes-5" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-5" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>nonzeroRowsAndCols</code>: Added <code>assay</code> argument, switching from internal <code>counts</code> usage, to make the <code>SummarizedExperiment</code> method more flexible.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-16-2019-09-25" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-16-2019-09-25" class="anchor"></a>basejump 0.11.16 (2019-09-25)</h2>
<div id="minor-changes-4" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-4" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrated S4 methods from syntactic that work on Bioconductor classes here. This keeps the syntactic package very light weight and focused only on character string sanitization.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-17-2019-10-09" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-17-2019-10-09" class="anchor"></a>basejump 0.11.17 (2019-10-09)</h2>
<div id="minor-changes-3" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-3" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>makeSampleData</code>: Made function slightly more flexible. Now allowing automatic rowname coercion from columns (“sampleID”, “rowname”, “rn”), similar to approach employed by data.table and tibble packages.</li>
<li>Now exporting <code>stripGeneVersions</code> alias, which uses the same code internally as <code>stripTranscriptVersions</code>.</li>
</ul>
</div>
<div id="disabled-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#disabled-methods" class="anchor"></a>Disabled methods</h3>
<ul>
<li>Disabled DelayedArray class methods for <code>calculateMetrics</code>, <code>estimateSizeFactors</code>, and <code>nonzeroRowsAndCols</code> until <code>is_pristine</code> bug in DelayedArray v0.11.8 is fixed on Bioconductor Devel (3.10). This is causing unit tests to fail otherwise. See <a href="https://github.com/Bioconductor/DelayedArray/issues/55">related issue</a>
</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-18-2019-10-13" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-18-2019-10-13" class="anchor"></a>basejump 0.11.18 (2019-10-13)</h2>
<div id="minor-changes-2" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-2" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>importSampleData</code>: Pipeline now defaulting to “none” instead of “bcbio”, since this flag is now properly hard coded in bcbio R packages. Added new Constellation (CPI) pipeline option.</li>
<li>
<code>makeSampleData</code>: Now checks for all NA columns and rows, similar to approach in <code>importSampleData</code>. This helps improve return consistency. Automatic rowname setting has been tweaked a bit to no longer attempt to remove original ID column.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-19-2019-10-23" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-19-2019-10-23" class="anchor"></a>basejump 0.11.19 (2019-10-23)</h2>
<div id="new-functions-1" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-1" class="anchor"></a>New functions</h3>
<ul>
<li>Genome version detection: <code>ensemblVersion</code>, <code>gencodeVersion</code>, <code>refseqVersion</code>, <code>flybaseVersion</code>, <code>wormbaseVersion</code>. Similar shell variants are available in the <a href="https://koopa.acidgenomics.com/">koopa</a> package.</li>
</ul>
</div>
<div id="minor-changes-1" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-1" class="anchor"></a>Minor changes</h3>
<ul>
<li>Moved some low-level functions to new <a href="https://acidbase.acidgenomics.com/">acidbase</a> package. Updated NAMESPACE and reexports to reflect these changes.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-20-2019-10-24" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-20-2019-10-24" class="anchor"></a>basejump 0.11.20 (2019-10-24)</h2>
<div id="new-reexports" class="section level3">
<h3 class="hasAnchor">
<a href="#new-reexports" class="anchor"></a>New reexports</h3>
<ul>
<li>Reexporting <code>metadata2</code> and <code>metadata2&lt;-</code> functions from transformer. These will be used internally in the pending DESeqAnalysis update.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-21-2019-11-07" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-21-2019-11-07" class="anchor"></a>basejump 0.11.21 (2019-11-07)</h2>
<div id="major-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes" class="anchor"></a>Major changes</h3>
<ul>
<li>Updated Bioconductor dependencies to require new 3.10 release.</li>
</ul>
</div>
<div id="bug-fixes" class="section level3">
<h3 class="hasAnchor">
<a href="#bug-fixes" class="anchor"></a>Bug fixes</h3>
<ul>
<li>
<code>filterCells</code> requires an internal <code>decode</code> step to handle <code>Rle</code> evaluation, which worked previously in Bioconductor 3.9 release.</li>
<li>Updated unit tests to reflect <code>SingleCellExperiment</code> example object resave in acidtest 0.2.7 update, which changed the numbers.</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-22-2019-11-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-22-2019-11-11" class="anchor"></a>basejump 0.11.22 (2019-11-11)</h2>
<div id="new-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>correlation</code>: Added S4 method support that mimics base <code>cor</code> methods, but is more flexible, supporting additional arguments via <code>...</code> in generic. This way we can provide intelligent and quick correlation calculations for nested assays inside a <code>SummarizedExperiment</code> and for <code>DESeqResults</code> (see DESeqAnalysis package).</li>
</ul>
</div>
</div>
    <div id="basejump-0-11-23-2019-11-19" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-11-23-2019-11-19" class="anchor"></a>basejump 0.11.23 (2019-11-19)</h2>
<div id="minor-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes" class="anchor"></a>Minor changes</h3>
<ul>
<li>Updated package documentation to support roxygen2 7.0 update.</li>
<li>Reworked <code>formalsList</code> global slightly. Using the <code>synesthesia</code> color palette for <code>heatmap.color</code> argument doesn’t always perform well enough, so I’m switching to a blue/black/yellow palette defined by <code>blueYellow</code> in acidplots instead for <code>heatmap.color</code>. The <code>synesthesia</code> palette performs really well for correlation heatmaps, and is now recommended by default via the <code>acid.heatmap.correlation.color</code> global now instead.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#basejump-0-11-0-2019-07-22">0.11.0</a></li>
        <li><a href="#basejump-0-11-1-2019-07-24">0.11.1</a></li>
        <li><a href="#basejump-0-11-2-2019-07-29">0.11.2</a></li>
        <li><a href="#basejump-0-11-3-2019-07-30">0.11.3</a></li>
        <li><a href="#basejump-0-11-4-2019-08-07">0.11.4</a></li>
        <li><a href="#basejump-0-11-5-2019-08-11">0.11.5</a></li>
        <li><a href="#basejump-0-11-6-2019-08-12">0.11.6</a></li>
        <li><a href="#basejump-0-11-7-2019-08-13">0.11.7</a></li>
        <li><a href="#basejump-0-11-8-2019-08-19">0.11.8</a></li>
        <li><a href="#basejump-0-11-9-2019-08-21">0.11.9</a></li>
        <li><a href="#basejump-0-11-10-2019-08-22">0.11.10</a></li>
        <li><a href="#basejump-0-11-11-2019-08-27">0.11.11</a></li>
        <li><a href="#basejump-0-11-12-2019-09-03">0.11.12</a></li>
        <li><a href="#basejump-0-11-13-2019-09-06">0.11.13</a></li>
        <li><a href="#basejump-0-11-14-2019-09-09">0.11.14</a></li>
        <li><a href="#basejump-0-11-15-2019-09-16">0.11.15</a></li>
        <li><a href="#basejump-0-11-16-2019-09-25">0.11.16</a></li>
        <li><a href="#basejump-0-11-17-2019-10-09">0.11.17</a></li>
        <li><a href="#basejump-0-11-18-2019-10-13">0.11.18</a></li>
        <li><a href="#basejump-0-11-19-2019-10-23">0.11.19</a></li>
        <li><a href="#basejump-0-11-20-2019-10-24">0.11.20</a></li>
        <li><a href="#basejump-0-11-21-2019-11-07">0.11.21</a></li>
        <li><a href="#basejump-0-11-22-2019-11-11">0.11.22</a></li>
        <li><a href="#basejump-0-11-23-2019-11-19">0.11.23</a></li>
      </ul>
    </div>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


