<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.8 • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Version 0.8" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.10.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small>0.8</small></h1>
      
    </div>

    <div id="basejump-0-8-0-alpha-2018-11-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-0-alpha-2018-11-11" class="anchor"></a>basejump 0.8.0-alpha (2018-11-11)</h2>
<div id="major-changes-10" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-10" class="anchor"></a>Major changes</h3>
<p>This release is the beginning of an attempt to rework the basejump codebase a bit and make the package easier to unit test on <a href="https://travis-ci.com/">Travis CI</a>. bcbio R packages will be pinned to v0.7.2 during this development period.</p>
<p>Here we are working to split out the functionality of basejump into several, smaller sub-packages:</p>
<ul>
<li>basejump.annotations</li>
<li>basejump.assertions</li>
<li>basejump.classes</li>
<li>basejump.coercion</li>
<li>basejump.developer</li>
<li>basejump.experiment</li>
<li>basejump.generics</li>
<li>basejump.globals</li>
<li>basejump.io</li>
<li>basejump.markdown</li>
<li>basejump.plots</li>
<li>basejump.sanitization</li>
</ul>
</div>
</div>
    <div id="basejump-0-8-1-alpha-2018-11-15" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-1-alpha-2018-11-15" class="anchor"></a>basejump 0.8.1-alpha (2018-11-15)</h2>
<div id="major-changes-9" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-9" class="anchor"></a>Major changes</h3>
<p>The split-out sub-package approach isn’t working quite right, so the code base has been consolidated back into a single basejump package. Development toward splitting the package will continue, but a conceptual re-imagining of how to organize the functions is needed.</p>
</div>
</div>
    <div id="basejump-0-8-2-alpha-2018-11-19" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-2-alpha-2018-11-19" class="anchor"></a>basejump 0.8.2-alpha (2018-11-19)</h2>
<div id="major-changes-8" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-8" class="anchor"></a>Major changes</h3>
<p>Migrating some additional base code that can be dispatched on either <code>SummarizedExperiment</code> or <code>SingleCellExperiment</code> from the bcbio R packages. We’re going to split out some of the single-cell RNA-seq functionality into separate packages, since a lot of my work moving forward deals with the 10X Genomics Cell Ranger platform, rather than the bcbio-supported inDrops platform.</p>
<p>In particular:</p>
<ul>
<li>
<code>barcodeRanksPerSample</code>.</li>
<li>
<code>filterCells</code>.</li>
<li>
<code>plotBarcodeRanks</code>.</li>
<li>
<code>plotCellCounts</code>.</li>
<li>
<code>plotGenesPerCell</code>.</li>
<li>
<code>plotMitoRatio</code>.</li>
<li>
<code>plotMitoVsCoding</code>.</li>
<li>
<code>plotNovelty</code>.</li>
<li>
<code>plotReadsPerCell</code>.</li>
<li>
<code>plotUMIsPerCell</code>.</li>
<li>
<code>plotUMIsVsGenes</code>.</li>
</ul>
</div>
<div id="minor-changes-29" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-29" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrated some <code>plotHeatmap</code> and <code>plotPCA</code> code from bcbio R packages.</li>
<li>Added new <code>formalsList</code> global variable, which stashes <code>getOption</code> defaults used in some functions, namely the save/load functions and some plotting functions.</li>
<li>Switched to using <code>formalsList</code> along with <code>formals</code> declaration internally to make the parameters more consistent across functions.</li>
<li>Miscellaneous fixes to working examples and documentation.</li>
<li>Split out sanitization functions into separate R files (e.g. <code>removeNA</code>).</li>
</ul>
</div>
</div>
    <div id="basejump-0-8-3-alpha-2018-11-25" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-3-alpha-2018-11-25" class="anchor"></a>basejump 0.8.3-alpha (2018-11-25)</h2>
<div id="minor-changes-28" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-28" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved usage of <code><a href="https://www.rdocumentation.org/packages/assertthat/topics/validate_that">assertthat::validate_that</a></code> in S4 class validity checks. Removed former approach using internal <code>.valid</code> function.</li>
<li>Temporarily soft deprecated some functions that will be formally deprecated in a future release. See <code>deprecated.R</code> file.</li>
<li>
<code>interestingGroups</code> doesn’t attempt validity check using <code>validObject</code> by default, which can be enabled instead using <code>check = TRUE</code>.</li>
<li>
<code>sampleData</code>: Tightened up internal assert checks.</li>
<li>
<code>PANTHER</code>: Minor tweaks to internal variable names inside <code>.splitTerms</code>.</li>
<li>
<code>transmit</code>: Improved messages. Temporarily disabled working example, since it consistently fails on <a href="https://travis-ci.com/">Travis CI</a>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-8-4-alpha-2018-11-28" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-4-alpha-2018-11-28" class="anchor"></a>basejump 0.8.4-alpha (2018-11-28)</h2>
<div id="new-functions-7" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-7" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>autopadZeros</code>. Useful for padding zeros inside of a character vector.</li>
<li>
<code>basenameSansExt</code>. Quickly get the basename without the file extension for desired file path(s). Surprisingly, this isn’t defined in the tools package so I wrote my own.</li>
<li>
<code>cleanSystemLibrary</code>. I ran into some shared library configuration issues on the Azure infrastructure, so this utility function is useful for checking wheter an R installation has a clean library.</li>
<li>
<code>plotGenderMarkers</code>: Migrated the <code>SummarizedExperiment</code> method from bcbioRNASeq package.</li>
</ul>
</div>
<div id="minor-changes-27" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-27" class="anchor"></a>Minor changes</h3>
<ul>
<li>Split out call functions defined in <code>calls.R</code> into separate R files. Refer to <code>dots</code>, for example.</li>
<li>Split out <code>environment.R</code> into separate R files. See <code>detectHPC</code> for example.</li>
<li>Miscellaneous documentation improvements.</li>
</ul>
</div>
</div>
    <div id="basejump-0-8-5-alpha-2018-11-29" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-5-alpha-2018-11-29" class="anchor"></a>basejump 0.8.5-alpha (2018-11-29)</h2>
<div id="new-functions-6" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-6" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>relevelRowRanges</code>, <code>relevelColData</code>.</li>
</ul>
</div>
<div id="deprecations-3" class="section level3">
<h3 class="hasAnchor">
<a href="#deprecations-3" class="anchor"></a>Deprecations</h3>
<ul>
<li>
<code>markdownPlotlist</code>. Renamed to <code>markdownPlots</code>.</li>
</ul>
</div>
<div id="minor-changes-26" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-26" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>autopadZeros</code>: Improved internal code to keep track of names for <code>character</code> method. Also added method support for <code>SummarizedExperiment</code>, which works on the column names (e.g. sample names) only by default.</li>
<li>
<code>Ensembl2Entrez</code>: Simplify validity check to require <code>integer</code> in <code>entrezID</code> column. Also reworked and improved internal code that supports run-length encoding (Rle) using <code>decode</code>.</li>
<li>
<code>export</code>: Improved documentation for <code>name</code> argument.</li>
<li>
<code>makeGRangesFromEnsembl</code>: Improved messages to user.</li>
<li>
<code>plotCountsPerBiotype</code>: Improved error messages for when the biotype isn’t defined.</li>
<li>
<code>plotPCA</code>: Add support for unique sample detection with <code>areSamplesUnique</code>, similar to the approach used in the bcbioRNASeq quality control R Markdown.</li>
<li>
<code>sampleNames</code>: Simplified <code>sampleName</code> extraction, using <code>[[</code> internally.</li>
<li>Miscellaneous documentation fixes.</li>
<li>Updated unit test for <code>sanitizeRowData</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-8-6-alpha-2018-11-30" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-6-alpha-2018-11-30" class="anchor"></a>basejump 0.8.6-alpha (2018-11-30)</h2>
<div id="minor-changes-25" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-25" class="anchor"></a>Minor changes</h3>
<ul>
<li>Now importing <code>hasUniqueCols</code> from goalie. Switched from previous approach using <code>areSamplesUnique</code>.</li>
<li>Documentation improvement for <code>genomeBuild</code>, used in <code>makeGRanges</code> functions.</li>
<li>Miscellaneous documentation improvements to pass build checks.</li>
</ul>
</div>
</div>
    <div id="basejump-0-8-7-alpha-2018-12-01" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-7-alpha-2018-12-01" class="anchor"></a>basejump 0.8.7-alpha (2018-12-01)</h2>
<div id="new-functions-5" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-5" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>humanize</code>. New generic that enables easy conversion to human-friendly column and/or row names. Useful for CSV file export in particular.</li>
</ul>
</div>
<div id="minor-changes-24" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-24" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>export</code>: Added <code>humanize</code> argument support.</li>
<li>Improved grep pattern matching inside <code>makeNames</code> functions (e.g. <code>.sanitizeAcronyms</code>).</li>
<li>
<code>sampleData</code>: Improved blacklist pattern matching against <code>Seurat</code> objects for <code>SingleCellExperiment</code> method.</li>
</ul>
</div>
</div>
    <div id="basejump-0-8-8-alpha-2018-12-03" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-8-8-alpha-2018-12-03" class="anchor"></a>basejump 0.8.8-alpha (2018-12-03)</h2>
<div id="new-functions-4" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-4" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>deg</code>: Utility function to quickly obtain differentially expressed gene (DEG) identifiers as a character vector.</li>
</ul>
</div>
<div id="major-changes-7" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-7" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>plotHeatmap</code>: Now defining row scaling internally, rather than relying on the functionality defined in <code>pheatmap</code>.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#basejump-0-8-0-alpha-2018-11-11">0.8.0</a></li>
        <li><a href="#basejump-0-8-1-alpha-2018-11-15">0.8.1</a></li>
        <li><a href="#basejump-0-8-2-alpha-2018-11-19">0.8.2</a></li>
        <li><a href="#basejump-0-8-3-alpha-2018-11-25">0.8.3</a></li>
        <li><a href="#basejump-0-8-4-alpha-2018-11-28">0.8.4</a></li>
        <li><a href="#basejump-0-8-5-alpha-2018-11-29">0.8.5</a></li>
        <li><a href="#basejump-0-8-6-alpha-2018-11-30">0.8.6</a></li>
        <li><a href="#basejump-0-8-7-alpha-2018-12-01">0.8.7</a></li>
        <li><a href="#basejump-0-8-8-alpha-2018-12-03">0.8.8</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

