<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.12 • basejump</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Version 0.12" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.13.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://acidgenomics.com/">
    <span class="fas fa fas fa-vial fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/acidgenomics/basejump/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small>0.12</small></h1>
      
    </div>

    <div id="basejump-0120-2020-01-20" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0120-2020-01-20" class="anchor"></a>basejump 0.12.0 (2020-01-20)</h2>
<div id="major-changes-2" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-2" class="anchor"></a>Major changes</h3>
<ul>
<li>Migrated functions from brio (now renamed to pipette), freerange, and transformer, in preparation for Bioconductor submission and reviews.</li>
<li>Updated messages to utilize the cli package.</li>
</ul>
</div>
</div>
    <div id="basejump-0121-2020-01-30" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0121-2020-01-30" class="anchor"></a>basejump 0.12.1 (2020-01-30)</h2>
<div id="major-changes-1" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-1" class="anchor"></a>Major changes</h3>
<ul>
<li>Reworked <code>aggregate</code>, <code>aggregateCols</code>, <code>aggregateRows</code> support to reflect internal migration away from Matrix.utils dependency.</li>
</ul>
</div>
<div id="minor-changes-18" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-18" class="anchor"></a>Minor changes</h3>
<ul>
<li>NAMESPACE fix for unexpected removal of Matrix.utils from CRAN. Now defining the previously imported aggregate.Matrix function directly here in basejump.</li>
<li>
<code>aggregate</code>: Now defining matrix method.</li>
<li>
<code>aggregate*</code> generics now consistently use “x” instead of “object”.</li>
</ul>
</div>
</div>
    <div id="basejump-0122-2020-02-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0122-2020-02-18" class="anchor"></a>basejump 0.12.2 (2020-02-18)</h2>
<div id="minor-changes-17" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-17" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrating <code>pseudobulk</code> methods to pointillism package. Rethinking the approach used here to work better with per-cluster aggregation operations. Will be updated in the next pointillism release.</li>
</ul>
</div>
</div>
    <div id="basejump-0123-2020-02-24" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0123-2020-02-24" class="anchor"></a>basejump 0.12.3 (2020-02-24)</h2>
<div id="minor-changes-16" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-16" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>makeSingleCellExperiment</code>, <code>makeSummarizedExperiment</code>: Removed <code>spikeNames</code> support due to a breaking change in SingleCellExperiment, which has removed <code>isSpike</code> in favor of <code>altExps</code>. Refer to the SingleCellExperiment documentation for updated current best practice for handling spike-in transcripts, which now requires slotting a separate SummarizedExperiment object containing things like ERCCs inside the main SingleCellExperiment.</li>
</ul>
</div>
</div>
    <div id="basejump-0124-2020-03-15" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0124-2020-03-15" class="anchor"></a>basejump 0.12.4 (2020-03-15)</h2>
<div id="minor-changes-15" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-15" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>matchEnsemblReleaseToURL</code>, <code>matchHumanOrthologs</code>: update unit tests to reflect Ensembl server migration, which has rendered Ensembl archives inaccessible via biomaRt until March 24th. Unit tests now check against current release instead of a pinned archive release.</li>
</ul>
</div>
</div>
    <div id="basejump-0125-2020-05-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0125-2020-05-11" class="anchor"></a>basejump 0.12.5 (2020-05-11)</h2>
<div id="minor-changes-14" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-14" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>importSampleData</code>: The <code>pipeline = "cpi"</code> option is now defunct. Use <code>pipeline = "none', sheet = 2L</code> for CPI samples.</li>
<li>
<code>importSampleData</code>: Added optional <code>autopadZeros</code> argument for easy handling of sample identifiers, which are often input by users without correct padding. This helps easily convert sample_1, sample_2, … sample_10 to expected sample_01, sample_02, … sample_10 sorting. Currently disabled by default.</li>
<li>
<code>importTx2Gene</code>: Added <code>ignoreGeneVersion</code> option, now enabled by default. This helps process gene identifiers by default in a manner suitable for downstream tximport-DESeq2 workflow.</li>
<li>
<code>headtail</code>: Removed Unicode support in favor of simple ASCII return, to avoid build warnings in latest R 4.0 release version.</li>
<li>Miscellaneous unit test updates to reflect changes in <code>DataFrame</code> class and NCBI server updates.</li>
<li>
<code>camelCase</code>, <code>dottedCase</code>, <code>organism</code>, <code>snakeCase</code>, <code>upperCamelCase</code>: S4 methods for <code>DataFrame</code> are now defined directly against <code>DataFrame</code>, instead of attempting to inherit from <code>DataTable</code> virtual class. This will break otherwise on bioc-devel 3.12, which seems to have changed inheritance.</li>
</ul>
</div>
</div>
    <div id="basejump-0126-2020-05-17" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0126-2020-05-17" class="anchor"></a>basejump 0.12.6 (2020-05-17)</h2>
<div id="minor-changes-13" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-13" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>convertGenesToSymbols</code>: Added method support for <code>GRanges</code> class objects. Automatically sets names as unique gene symbols.</li>
<li>
<code>HGNC2Ensembl</code>: Enforce TSV handling internally in <code>import</code> call.</li>
<li>
<code>MGI2Ensembl</code>: Fix for column name handling.</li>
</ul>
</div>
</div>
    <div id="basejump-0127-2020-05-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0127-2020-05-22" class="anchor"></a>basejump 0.12.7 (2020-05-22)</h2>
<div id="minor-changes-12" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-12" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>aggregate</code>, <code>aggregateCols</code>, <code>aggregateRows</code>: Relaxed assert checks on validity of dimnames, so we can use internally in acidgsea package, which needs to handle gene symbols containing syntactically invalid hyphens.</li>
</ul>
</div>
</div>
    <div id="basejump-0128-2020-06-15" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0128-2020-06-15" class="anchor"></a>basejump 0.12.8 (2020-06-15)</h2>
<div id="minor-changes-11" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-11" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>autopadZeros</code>: Migrated character method support to syntactic package, since this is useful for low-level code run inside koopa.</li>
</ul>
</div>
</div>
    <div id="basejump-0129-2020-07-24" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0129-2020-07-24" class="anchor"></a>basejump 0.12.9 (2020-07-24)</h2>
<div id="minor-changes-10" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-10" class="anchor"></a>Minor changes</h3>
<ul>
<li>Maintenance release, updating minimum R dependency to 4.0.</li>
</ul>
</div>
</div>
    <div id="basejump-01210-2020-08-04" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-01210-2020-08-04" class="anchor"></a>basejump 0.12.10 (2020-08-04)</h2>
<div id="minor-changes-9" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-9" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrated <code>alphaThreshold</code> and <code>lfcThreshold</code> methods to DESeqAnalysis package. These are not used in other packages and may not be generally applicable to <code>SummarizedExperiment</code> class, so rethinking here.</li>
</ul>
</div>
</div>
    <div id="basejump-01211-2020-08-13" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-01211-2020-08-13" class="anchor"></a>basejump 0.12.11 (2020-08-13)</h2>
<div id="minor-changes-8" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-8" class="anchor"></a>Minor changes</h3>
<ul>
<li>Now reexporting <code>requireNamespaces</code> from acidbase package.</li>
</ul>
</div>
</div>
    <div id="basejump-01212-2020-08-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-01212-2020-08-18" class="anchor"></a>basejump 0.12.12 (2020-08-18)</h2>
<div id="new-methods" class="section level3">
<h3 class="hasAnchor">
<a href="#new-methods" class="anchor"></a>New methods</h3>
<ul>
<li>
<code>intersectAll</code>: Defined <code>list</code> method.</li>
<li>
<code>intersectionMatrix</code>: Defined <code>list</code> method.</li>
</ul>
</div>
</div>
    <div id="basejump-01213-2020-08-25" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-01213-2020-08-25" class="anchor"></a>basejump 0.12.13 (2020-08-25)</h2>
<div id="minor-changes-7" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-7" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrated some globals to acidbase package, for improved consistency.</li>
</ul>
</div>
</div>
    <div id="basejump-01214-2020-09-14" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-01214-2020-09-14" class="anchor"></a>basejump 0.12.14 (2020-09-14)</h2>
<div id="new-functions-1" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-1" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>splitByLevel</code>: Easily split a data frame into a list using a defined factor column (<code>f</code> argument). Can easily include the reference level with <code>ref = TRUE</code>, which is useful for statistical calculations on pairwise contrasts.</li>
</ul>
</div>
</div>
    <div id="basejump-01215-2020-09-25" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-01215-2020-09-25" class="anchor"></a>basejump 0.12.15 (2020-09-25)</h2>
<div id="new-classes" class="section level3">
<h3 class="hasAnchor">
<a href="#new-classes" class="anchor"></a>New classes</h3>
<ul>
<li>
<code>Protein2Gene</code>: <code>DataFrame</code> with <code>proteinID</code>, <code>geneID</code>, and <code>geneName</code> columns. Use corresponding <code>makeProtein2GeneFromEnsembl</code> to generate object simplify using Ensembl protein IDs as input.</li>
</ul>
</div>
<div id="new-functions" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>getEnsDb</code>: Now exporting function that was used internally to obtain <code>EnsDb</code> object from AnnotationHub.</li>
<li>
<code>makeProtein2GeneFromEnsembl</code>: New utility function that takes Ensembl protein identifiers as input, and returns corresponding gene identifiers and gene names (i.e. HUGO gene symbols).</li>
</ul>
</div>
<div id="minor-changes-6" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-6" class="anchor"></a>Minor changes</h3>
<ul>
<li>Reworked some internal code in <code>makeGRangesFromEnesmbl</code> to enable export of new <code>getEnsDb</code> function.</li>
<li>Reworked internal handling of AnnotationHub and EnsDb metadata.</li>
<li>
<code>makeGRangesFromEnsembl</code> / <code>getEnsDb</code>: Improved sorting of Ensembl releases so that current release greater than 99 returns as expected. Since Ensembl is now at 101, we need to convert to integers internally instead of attempting to sort on strings.</li>
</ul>
</div>
</div>
    <div id="basejump-01216-2020-09-29" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-01216-2020-09-29" class="anchor"></a>basejump 0.12.16 (2020-09-29)</h2>
<div id="minor-changes-5" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-5" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>makeProtein2GeneFromEnsembl</code>: Improved error message on match failure. Now returns the protein IDs that failed to match more clearly.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


