<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.5 • basejump</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Version 0.5" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.12.14</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small>0.5</small></h1>
      
    </div>

    <div id="basejump-050-2018-04-13" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-050-2018-04-13" class="anchor"></a>basejump 0.5.0 (2018-04-13)</h2>
<div id="new-functions-21" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-21" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>emptyRanges</code> enables easy creation of placeholder ranges for <code>GRanges</code> objects, where transgene and FASTA spike-ins are needed.</li>
<li>
<code>hgnc2gene</code> enables easy mapping of <a href="https://www.genenames.org/">HGNC</a> to <a href="https://www.ensembl.org/">Ensembl</a> gene identifiers.</li>
<li>
<code>mgi2gene</code> enables easy mapping of <a href="http://www.informatics.jax.org/">MGI</a> to <a href="https://www.ensembl.org/">Ensembl</a> gene identifiers.</li>
<li>
<code>panther</code> function enables easy querying of the <a href="http://pantherdb.org/">PANTHER</a> website. Human, mouse, nematode worm, and fruit fly are currently supported. The specific <a href="http://pantherdb.org/">PANTHER</a> release (e.g. 13) can be declared using the <code>release</code> argument. Otherwise, the function will return the most recent annotations from the <a href="http://pantherdb.org/">PANTHER</a> website.</li>
<li>Added <code>isURL</code> check function.</li>
<li>
<code>readJSON</code> adds support for JSON files. Like the other read functions, it supports both local files and remote URLs.</li>
</ul>
</div>
<div id="ggplot2-themes" class="section level3">
<h3 class="hasAnchor">
<a href="#ggplot2-themes" class="anchor"></a>ggplot2 themes</h3>
<ul>
<li>
<code>theme_midnight</code> and <code>theme_paperwhite</code> provide minimal, high contrast <a href="http://ggplot2.tidyverse.org/">ggplot2</a> themes with bold sans serif labels.</li>
</ul>
</div>
<div id="major-changes-26" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-26" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>loadData</code> now supports <code>.rda</code>, <code>.rds</code>, and <code>.RData</code> files. The function will error by design if multiple data extensions are detected inside the directory specified with the <code>dir</code> argument.</li>
</ul>
</div>
<div id="minor-changes-83" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-83" class="anchor"></a>Minor changes</h3>
<ul>
<li>Consolidated assert check function code.</li>
<li>Moved assertive imports to <code>basejump-package.R</code> file.</li>
<li>Consolidated globals inside package to <code>globals.R</code> file.</li>
<li>Removed internal <code>.biocLite</code> function. Now using <code>requireNamespace</code> instead, without attempting to install automatically.</li>
<li>Added internal support for safe loading RDS files.</li>
<li>Switched back to using <code>message</code>, <code>warning</code>, and <code>stop</code> instead of the <a href="http://rlang.tidyverse.org/">rlang</a> equivalents.</li>
<li>Improved internal method declaration using <code>getMethod</code> where applicable.</li>
<li>
<code>multiassignAsEnvir</code> is now recommended in place of <code>multiassignAsNewEnvir</code>.</li>
<li>
<code>readFileByExtension</code> will now attempt to use the <a href="https://cran.r-project.org/package=rio">rio</a> package for file extensions that are not natively supported.</li>
<li>
<code>writeCounts</code> now uses <code>mapply</code> internally.</li>
<li>Migrated <code>assertFormalAnnotationCol</code> to <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package.</li>
</ul>
</div>
</div>
    <div id="basejump-051-2018-04-16" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-051-2018-04-16" class="anchor"></a>basejump 0.5.1 (2018-04-16)</h2>
<div id="minor-changes-82" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-82" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>emptyRanges</code>: Now using <code>match.arg</code> internally to capture <code>seqname</code> argument.</li>
<li>Removed legacy <code>.assignCamelArgs</code> and <code>.assignCamelFormals</code> internal functions.</li>
<li>Improved internal handling of XLSX files in <code>localOrRemoteFile</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-052-2018-04-26" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-052-2018-04-26" class="anchor"></a>basejump 0.5.2 (2018-04-26)</h2>
<div id="minor-changes-81" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-81" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved documentation for assert check functions.</li>
<li>Deprecated <code>geomean</code> in favor of <code>geometricMean</code>.</li>
<li>Simplified internal code for <code>grepString</code>.</li>
<li>Added message during <code>hgnc2gene</code> call.</li>
<li>Miscellaneous documentation fixes.</li>
<li>Moved internal constructors into the S4 method definitions, where applicable.</li>
<li>Simplified default parameter definition for <code>panther(organism = "XXX")</code>.</li>
<li>Improved code coverage, using <code>nocov</code> where appropriate.</li>
</ul>
</div>
</div>
    <div id="basejump-053-2018-04-30" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-053-2018-04-30" class="anchor"></a>basejump 0.5.3 (2018-04-30)</h2>
<div id="minor-changes-80" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-80" class="anchor"></a>Minor changes</h3>
<ul>
<li>Improved internal S4 method code for <code>fixNA</code> and <code>removeNA</code>.</li>
<li>Tweaked gray accent colors in <code>theme_midnight</code> and <code>theme_paperwhite</code>. Now using British spelling internally for ggplot code.</li>
<li>Improved <code>strip.background</code> for <code>theme_paperwhite</code>, removing the black box around the labels when facet wrapping is enabled.</li>
</ul>
</div>
</div>
    <div id="basejump-054-2018-05-08" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-054-2018-05-08" class="anchor"></a>basejump 0.5.4 (2018-05-08)</h2>
<ul>
<li>Migrated <code>sanitizeSampleData</code> to <a href="https://bioinformatics.sph.harvard.edu/bcbioBase/">bcbioBase</a> package.</li>
<li>Updated <a href="https://bioconductor.org/">Bioconductor</a> install method for 3.7.</li>
</ul>
</div>
    <div id="basejump-055-2018-05-15" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-055-2018-05-15" class="anchor"></a>basejump 0.5.5 (2018-05-15)</h2>
<div id="major-changes-25" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-25" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>readGFF</code> now uses <code><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">rtracklayer::import</a></code> internally to return GFF file as a <code>GRanges</code> object instead of a <code>data.frame</code>.</li>
</ul>
</div>
<div id="minor-changes-79" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-79" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>assertIsGFF</code> and <code>parseGFFAttributes</code> functions are now defunct.</li>
<li>Simplified internal GFF handling code for <code>makeGRangesFromGFF</code>, <code>makeGene2symbolFromGFF</code>, and <code>makeTx2geneFromGFF</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-056-2018-05-19" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-056-2018-05-19" class="anchor"></a>basejump 0.5.6 (2018-05-19)</h2>
<div id="minor-changes-78" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-78" class="anchor"></a>Minor changes</h3>
<ul>
<li>Fixed NAMESPACE issue with <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html">GenomeInfoDb::seqnames</a></code>.</li>
<li>Improved <code>readGFF</code> working example to reflect switch to <code>GRanges</code> return.</li>
<li>Added <a href="https://www.apple.com/macos/">macOS</a> bioc-release image to <a href="https://travis-ci.com/">Travis CI</a> build checks.</li>
</ul>
</div>
</div>
    <div id="basejump-057-2018-05-24" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-057-2018-05-24" class="anchor"></a>basejump 0.5.7 (2018-05-24)</h2>
<div id="minor-changes-77" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-77" class="anchor"></a>Minor changes</h3>
<ul>
<li>Tweaked gray color accents for <code>theme_midnight</code> and <code>theme_paperwhite</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-058-2018-06-05" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-058-2018-06-05" class="anchor"></a>basejump 0.5.8 (2018-06-05)</h2>
<div id="minor-changes-76" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-76" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>makeGRangesFromEnsembl</code> now supports remapping of <a href="https://genome.ucsc.edu/">UCSC</a> genome build to Ensembl. However, this isn’t recommended, and will warn the user.</li>
<li>
<code>convertUCSCBuildToEnsembl</code> now returns <code>NULL</code> instead of erroring on genome build match failure.</li>
<li>
<code>stripTranscriptVersions</code> now matches “<code>.</code>”, “<code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code>”, and “<code>_</code>” version delimiters.</li>
<li>Made unused <code>dynamicPlotlist</code> function defunct.</li>
<li>Consider deprecating <code>assertIsCharacterOrNULL</code> and <code>assertIsDataFrameOrNULL</code> in a future release.</li>
</ul>
</div>
<div id="infrastructure-changes" class="section level3">
<h3 class="hasAnchor">
<a href="#infrastructure-changes" class="anchor"></a>Infrastructure changes</h3>
<ul>
<li>Reenable <a href="https://travis-ci.com/">Travis CI</a> blocklist, excluding develop branch from checks.</li>
<li>Reorganized documentation of deprecated and defunct functions.</li>
</ul>
</div>
</div>
    <div id="basejump-059-2018-06-12" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-059-2018-06-12" class="anchor"></a>basejump 0.5.9 (2018-06-12)</h2>
<div id="minor-changes-75" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-75" class="anchor"></a>Minor changes</h3>
<ul>
<li>Markdown function consistency improvements. Now all relevant <a href="https://daringfireball.net/projects/markdown/">Markdown</a> functions use <code>text</code> as the primary argument, instead of <code>object</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0510-2018-06-28" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0510-2018-06-28" class="anchor"></a>basejump 0.5.10 (2018-06-28)</h2>
<div id="minor-changes-74" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-74" class="anchor"></a>Minor changes</h3>
<ul>
<li>Removed <code>makeNames</code> argument from <code>readFileByExtension</code> function. Use the <code>makeNames</code> family of functions manually after data import instead. This helps avoid unwanted sanitization of data.</li>
<li>Simplified assert checks for internal load function.</li>
<li>Improved code coverage.</li>
<li>
<a href="https://ci.appveyor.com/">AppVeyor CI</a> updates to work with <a href="https://bioconductor.org/">Bioconductor</a> 3.8 devel.</li>
</ul>
</div>
</div>
    <div id="basejump-0511-2018-07-09" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0511-2018-07-09" class="anchor"></a>basejump 0.5.11 (2018-07-09)</h2>
<div id="minor-changes-73" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-73" class="anchor"></a>Minor changes</h3>
<ul>
<li>Now requiring <a href="http://ggplot2.tidyverse.org/">ggplot2</a> v3.0 internally.</li>
<li>
<code>theme_midnight</code> and <code>theme_paperwhite</code> now extend <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">ggplot2::theme_linedraw</a></code>, improving the consistency between these themes.</li>
<li>Example data is now consistenly formatted using snake case: <code>rnaseq_counts</code> and <code>single_cell_counts</code>, instead of the previous camel case conventions: <code>rnaseqCounts</code>, <code>singleCellCounts</code>.</li>
<li>Camel variants of the ggplot themes are now deprecated.</li>
<li>Updated internal gene synonyms data from <a href="https://www.nih.gov/">NIH</a>.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


