<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Version 0.9 • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Version 0.9" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.10.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1>Changelog <small>0.9</small></h1>
      
    </div>

    <div id="basejump-0-9-0-2018-12-12" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-0-2018-12-12" class="anchor"></a>basejump 0.9.0 (2018-12-12)</h2>
<p>I bumped the release series from v0.8 to v0.9 because this represents a significant change to the internal codebase, where I have now switched to using my new <a href="https://goalie.acidgenomics.com/">goalie</a> assert check engine from <a href="https://cran.r-project.org/package=assertive">assertive</a>.</p>
<div id="new-functions-2" class="section level3">
<h3 class="hasAnchor">
<a href="#new-functions-2" class="anchor"></a>New functions</h3>
<ul>
<li>
<code>decode</code>: Decode S4 run-length encoding (Rle).</li>
<li>
<code>encode</code>: Apply S4 run-length encoding (Rle).</li>
<li>
<code>geneNames</code>: Convenience function that returns gene names (symbols) mapped to the stable, but not human-friendly gene identifiers.</li>
<li>
<code>matchesGene2Symbol</code>, <code>matchesInterestingGroups</code>: New functions designed to match corresponding <code>Gene2Symbol</code> objects or <code>interestingGroups</code>.</li>
<li>
<code>pasteURL</code>: Convenience function that generates URL strings.</li>
<li>
<code>sanitizeColData</code>: rework of previous <code>sanitizeSampleData</code> approach.</li>
</ul>
</div>
<div id="major-changes-5" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-5" class="anchor"></a>Major changes</h3>
<ul>
<li>Now using <a href="https://goalie.acidgenomics.com/">goalie</a> package instead of <a href="https://cran.r-project.org/package=assertive">assertive</a> for internal assert checks. The new <code><a href="https://www.rdocumentation.org/packages/goalie/topics/assert">goalie::assert</a></code> function is more flexible in many cases. Similarly, <code><a href="https://www.rdocumentation.org/packages/goalie/topics/validate">goalie::validate</a></code> is now being used in place of <code><a href="https://www.rdocumentation.org/packages/assertthat/topics/validate_that">assertthat::validate_that</a></code>.</li>
<li>
<code>plotHeatmap</code> now calculates the z-score normalization internally, rather than relying upon the codebase inside <a href="https://cran.r-project.org/package=pheatmap">pheatmap</a>.</li>
<li>
<code>export</code>: Improved <code>SummarizedExperiment</code> method to also write <code>Gene2Symbol</code> and <code>Ensembl2Entrez</code> mappings to disk, when defined. The human-friendly output formal has been renamed from <code>human</code> to <code>humanize</code>, to reflect an action (verb). This corresponds better to our <code>humanize</code> generic function. Also reworked some internal code that handles output of colData and rowData to disk.</li>
<li>
<code>makeGRangesFromEnsembl</code>: Switched to using S4 run-length encoding (Rle) in our metadata column (mcols) return. This functionality matches the conventions used by GenomicRanges in the <code>GRanges</code> return, and reduces the memory footprint of very large annotation objects.</li>
</ul>
</div>
<div id="minor-changes-23" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-23" class="anchor"></a>Minor changes</h3>
<ul>
<li>Note that some internal instances of <code>has_length</code> aren’t quite strict enough. Switch to using <code>length(x) &gt; 0L</code> or improved <code>hasLength</code> assert check defined in the <a href="https://goalie.acidgenomics.com/">goalie</a> package.</li>
<li>
<code>geneSynonyms</code>: Switched to using new <code>pasteURL</code> function internally instead of using <code>paste</code> with <code>/</code> separator.</li>
<li>
<code>HGNC2Ensembl</code> generator: switch to using <code>pasteURL</code> internally.</li>
<li>
<code>loadData</code>, and other related load family functions: simplified internal code using our new <a href="https://goalie.acidgenomics.com/">goalie</a> asserts.</li>
</ul>
</div>
<div id="deprecations-2" class="section level3">
<h3 class="hasAnchor">
<a href="#deprecations-2" class="anchor"></a>Deprecations</h3>
<ul>
<li>Removed <code>assertFormalGene2symbol</code> from deprecations.</li>
</ul>
</div>
<div id="documentation-1" class="section level3">
<h3 class="hasAnchor">
<a href="#documentation-1" class="anchor"></a>Documentation</h3>
<ul>
<li>Improved documentation style throughout the package, switching from the usage of scalar types like <code>string</code> to <code><a href="https://www.rdocumentation.org/packages/base/topics/character">character(1)</a></code>, and <code>boolean</code> to <code><a href="https://www.rdocumentation.org/packages/base/topics/logical">logical(1)</a></code>. This better matches the actual data structure in R. Some other packages like <a href="https://cran.r-project.org/package=checkmate">checkmate</a> also use this convention, which I think is more readable than my previous approach.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-1-2018-12-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-1-2018-12-22" class="anchor"></a>basejump 0.9.1 (2018-12-22)</h2>
<p>This release defines the initial point where <a href="https://basejump.acidgenomics.com/">basejump</a> begins to import <a href="https://bioverbs.acidgenomics.com/">bioverbs</a>.</p>
<div id="major-changes-4" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-4" class="anchor"></a>Major changes</h3>
<ul>
<li>Now importing generics using our <a href="https://bioverbs.acidgenomics.com/">bioverbs</a> S4 generic package. All generics previously defined in <a href="https://basejump.acidgenomics.com/">basejump</a> will continue to be reexported, to maintain backward compatibility for reverse dependencies (revdeps).</li>
</ul>
</div>
<div id="minor-changes-22" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-22" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>aggregateCellsToSamples</code>: Split out S4 method to a separate file. Previously was defined in <code>aggregate-methods.R</code>.</li>
<li>Reorganized collapse family of functions. Refer to changes in <code>collapse-methods.R</code>, which is now split out to <code>collaseToString-methods.R</code>.</li>
<li>Split out the markdown family of functions back out into separate files.</li>
<li>
<code>matchesGene2Symbol</code>, <code>matchesInterestingGroups</code>: Reworked internal code and moved to separate files. No longer relies upon <code>makeTestFunction</code> from <a href="https://cran.r-project.org/package=checkmate">checkmate</a> package.</li>
</ul>
</div>
<div id="documentation" class="section level3">
<h3 class="hasAnchor">
<a href="#documentation" class="anchor"></a>Documentation</h3>
<ul>
<li>Switched documentation titles to sentence case from title case. It’s generally more readable.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-2-2019-01-07" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-2-2019-01-07" class="anchor"></a>basejump 0.9.2 (2019-01-07)</h2>
<p>This release defines the initial point where <a href="https://basejump.acidgenomics.com/">basejump</a> becomes even more modular, offloading some functions to new <a href="https://brio.acidgenomics.com/">brio</a>, <a href="https://syntactic.acidgenomics.com/">syntactic</a>, and S4Transformer packages.</p>
<p>Note that all offloaded functions will continue to be reexported in <a href="https://basejump.acidgenomics.com/">basejump</a>. If you notice a function that is missing and not correctly re-exported, please file an issue.</p>
<p><em>Note that S4Transformer has since been renamed to <a href="https://transformer.acidgenomics.com/">transformer</a>.</em></p>
<div id="offloaded-to-s4transformer" class="section level3">
<h3 class="hasAnchor">
<a href="#offloaded-to-s4transformer" class="anchor"></a>Offloaded to S4Transformer</h3>
<ul>
<li>
<code>as</code> coercion methods moved to S4Transformer package. This methods define our useful interconversions between Bioconductor and tidyverse data classes, including <code>DataFrame</code> and <code>tbl_df</code> (tibble).</li>
<li>
<code>coerceS4ToList</code> / <code>flatFiles</code>.</li>
</ul>
</div>
<div id="offloaded-to-bb8" class="section level3">
<h3 class="hasAnchor">
<a href="#offloaded-to-bb8" class="anchor"></a>Offloaded to bb8</h3>
<ul>
<li>
<code>cleanSystemLibrary</code>. This doesn’t scale well to all installations and is really only intended for personal use, so bb8 package is more appropriate.</li>
<li>Documentation functions, including <code>parseRd</code>, <code>RdTags</code>, <code>saveRdExamples</code>, and <code>tabular</code> are outside the scope of <a href="https://basejump.acidgenomics.com/">basejump</a>.</li>
</ul>
</div>
<div id="offloaded-to-brio-2" class="section level3">
<h3 class="hasAnchor">
<a href="#offloaded-to-brio-2" class="anchor"></a>Offloaded to brio</h3>
<ul>
<li>
<code>basenameSansExt</code>.</li>
<li>
<code>dots</code>.</li>
<li>
<code>export</code>.</li>
<li>
<code>import</code>.</li>
<li>
<code>initDir</code>.</li>
<li>
<code>loadData</code>.</li>
<li>
<code>localOrRemoteFile</code>.</li>
<li>
<code>pasteURL</code>.</li>
<li>
<code>realpath</code>.</li>
<li>
<code>sanitizeColData</code>.</li>
<li>
<code>sanitizeRowData</code>.</li>
<li>
<code>sanitizeSampleData</code>.</li>
<li>
<code>saveData</code>.</li>
<li>
<code>transmit</code>.</li>
<li>
<code>writeCounts</code>.</li>
</ul>
</div>
<div id="offloaded-to-goalie-2" class="section level3">
<h3 class="hasAnchor">
<a href="#offloaded-to-goalie-2" class="anchor"></a>Offloaded to goalie</h3>
<ul>
<li>
<code>matchArgsToDoCall</code>.</li>
<li>
<code>MethodDefinition</code>.</li>
<li>
<code>standardizeCall</code>.</li>
</ul>
</div>
<div id="minor-changes-21" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-21" class="anchor"></a>Minor changes</h3>
<ul>
<li>Added <code>nullOK</code> support to <a href="https://goalie.acidgenomics.com/">goalie</a> assert checks, where applicable.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-3-2019-01-08" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-3-2019-01-08" class="anchor"></a>basejump 0.9.3 (2019-01-08)</h2>
<div id="minor-changes-20" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-20" class="anchor"></a>Minor changes</h3>
<ul>
<li>Reorganized imports in <code>DESCRIPTION</code> file to make them more human readable. Note that <a href="https://basejump.acidgenomics.com/">basejump</a> sub-packages are imported first, then <a href="https://bioconductor.org/">Bioconductor</a> packages, followed by <a href="https://cran.r-project.org/">CRAN</a> packages, and required default packages.</li>
<li>Split out NAMESPACE imports into a separate <code>imports.R</code> file.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-4-2019-01-12" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-4-2019-01-12" class="anchor"></a>basejump 0.9.4 (2019-01-12)</h2>
<div id="minor-changes-19" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-19" class="anchor"></a>Minor changes</h3>
<ul>
<li>Consolidated reexports from <a href="https://basejump.acidgenomics.com/">basejump</a> sub-packages into <code>reexports.R</code> file.</li>
<li>Miscellaneous documentation updates, improving link appearance for functions exported in other packages.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-5-2019-01-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-5-2019-01-22" class="anchor"></a>basejump 0.9.5 (2019-01-22)</h2>
<div id="minor-changes-18" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-18" class="anchor"></a>Minor changes</h3>
<ul>
<li>Updated basejump subpackage dependencies.</li>
<li>Needed to add <code>decode</code> call internally for some plotting functions, to ensure that run-length encoded (Rle) rowData gets handled correctly.</li>
<li>Bug fix for internal <code>interestingGroups</code> handling in plot functions.</li>
</ul>
</div>
<div id="offloaded-to-brio-1" class="section level3">
<h3 class="hasAnchor">
<a href="#offloaded-to-brio-1" class="anchor"></a>Offloaded to brio</h3>
<ul>
<li>
<code>decode</code>, <code>encode</code>. These are useful for data sanitization. Still re-exported here in <a href="https://basejump.acidgenomics.com/">basejump</a>.</li>
</ul>
</div>
<div id="offloaded-to-goalie-1" class="section level3">
<h3 class="hasAnchor">
<a href="#offloaded-to-goalie-1" class="anchor"></a>Offloaded to goalie</h3>
<ul>
<li>
<code>printString</code>. This is a low-level function that is useful for setting the cause attribute in error messages. Still reexported here in basejump.</li>
</ul>
</div>
<div id="deprecations-1" class="section level3">
<h3 class="hasAnchor">
<a href="#deprecations-1" class="anchor"></a>Deprecations</h3>
<ul>
<li>
<code>sanitizeRowData</code> has been deprecated in favor of <code>atomize</code>.</li>
<li>
<code>sanitizeAnnotable</code> deprecation has been updated to point to <code>atomize</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-6-2019-01-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-6-2019-01-22" class="anchor"></a>basejump 0.9.6 (2019-01-22)</h2>
<div id="minor-changes-17" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-17" class="anchor"></a>Minor changes</h3>
<ul>
<li>Note that S4Transformer package import has been renamed to transformer.</li>
<li>Needed to add <code>decode</code> call internally for some plotting functions, to ensure that run-length encoded (Rle) rowData gets handled correctly.</li>
<li>Bug fix for internal <code>interestingGroups</code> handling in plot functions.</li>
</ul>
</div>
<div id="offloaded-to-brio" class="section level3">
<h3 class="hasAnchor">
<a href="#offloaded-to-brio" class="anchor"></a>Offloaded to brio</h3>
<ul>
<li>
<code>decode</code>, <code>encode</code>. These are useful for data sanitization. Still re-exported here in <a href="https://basejump.acidgenomics.com/">basejump</a>.</li>
</ul>
</div>
<div id="offloaded-to-goalie" class="section level3">
<h3 class="hasAnchor">
<a href="#offloaded-to-goalie" class="anchor"></a>Offloaded to goalie</h3>
<ul>
<li>
<code>printString</code>. This is a low-level function that is useful for setting the cause attribute in error messages. Still reexported here in basejump.</li>
</ul>
</div>
<div id="deprecations" class="section level3">
<h3 class="hasAnchor">
<a href="#deprecations" class="anchor"></a>Deprecations</h3>
<ul>
<li>
<code>sanitizeRowData</code> has been deprecated in favor of <code>atomize</code>.</li>
<li>
<code>sanitizeAnnotable</code> deprecation has been updated to point to <code>atomize</code>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-7-2019-01-23" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-7-2019-01-23" class="anchor"></a>basejump 0.9.7 (2019-01-23)</h2>
<div id="minor-changes-16" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-16" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>decode</code> and <code>encode</code> are properly reexported from <a href="https://brio.acidgenomics.com/">brio</a>.</li>
<li>Updated Travis CI and AppVeyor CI configurations.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-8-2019-02-08" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-8-2019-02-08" class="anchor"></a>basejump 0.9.8 (2019-02-08)</h2>
<div id="minor-changes-15" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-15" class="anchor"></a>Minor changes</h3>
<ul>
<li>Deprecating <code>plotGene</code> in favor of <code>plotCounts</code>. This change will be reflected in future updates of packages that depend on <a href="https://basejump.acidgenomics.com/">basejump</a>, including the bcbio R packages.</li>
<li>Split out subpackage reexports into separate files.</li>
<li>Reexporting <code>assert</code> from goalie package.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-9-2019-02-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-9-2019-02-11" class="anchor"></a>basejump 0.9.9 (2019-02-11)</h2>
<div id="minor-changes-14" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-14" class="anchor"></a>Minor changes</h3>
<ul>
<li>Code fixes to provide backward compatibility support for R 3.4. Tested using R 3.4.1 with Bioconductor 3.6 release.</li>
<li>Needed to add <code>unname</code> to some assert checks for expected <code><a href="https://www.rdocumentation.org/packages/base/topics/logical">logical(1)</a></code> return, which only happens in R 3.4 but not R 3.5.</li>
<li>
<code>uniteInterestingGroups</code>: Improved internal assert checks and name handling.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-10-2019-02-17" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-10-2019-02-17" class="anchor"></a>basejump 0.9.10 (2019-02-17)</h2>
<div id="minor-changes-13" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-13" class="anchor"></a>Minor changes</h3>
<ul>
<li>Working on making the current basejump code base completely backward compatible with bcbioBase v0.4.1 and bcbioRNASeq v0.2.8 release series.</li>
<li>Now reexporting <code><a href="https://www.rdocumentation.org/packages/goalie/topics/bapply">goalie::bapply</a></code> and additional useful pipes from <a href="https://magrittr.tidyverse.org/">magrittr</a> package.</li>
<li>Keeping the now deprecated <code>plotGene</code> generic reexported, while encouring users to update their code to use <code>plotCounts</code> instead.</li>
<li>
<code>interestingGroups</code>: Simplified internal assert checks, removing <code>matchesInterestingGroups</code>, which can become circular.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-11-2019-02-25" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-11-2019-02-25" class="anchor"></a>basejump 0.9.11 (2019-02-25)</h2>
<div id="minor-changes-12" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-12" class="anchor"></a>Minor changes</h3>
<ul>
<li>Updated <a href="https://transformer.acidgenomics.com/">transformer</a> package reexports to include new <a href="https://cran.r-project.org/package=data.table">data.table</a> coercion methods.</li>
<li>Added additional useful compression function reexports from <a href="https://brio.acidgenomics.com/">brio</a> package.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-12-2019-03-11" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-12-2019-03-11" class="anchor"></a>basejump 0.9.12 (2019-03-11)</h2>
<div id="major-changes-3" class="section level3">
<h3 class="hasAnchor">
<a href="#major-changes-3" class="anchor"></a>Major changes</h3>
<ul>
<li>
<code>makeGRangesFromGFF</code>: Reworked internal code, making it more modular. Added initial support for RefSeq GFF3 files. Also improved sanitization and special handling of files from FlyBase and WormBase.</li>
</ul>
</div>
<div id="minor-changes-11" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-11" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>plotHeatmap</code> family: Bug fix needed for internal <code>is.na</code> call on <code>annotationCol</code>, which should be wrapped with <code>any</code> to return boolean. This errors (as it should) on R 3.6, but I missed it on R 3.5.</li>
<li>Consolidating GRanges return code defined in <code>.makeGRanges</code>, which is run for both GFF file and ensembldb import. We’ve improved the Rle encoding steps here to work with complex GFF3 files (e.g. GENCODE).</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-13-2019-03-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-13-2019-03-18" class="anchor"></a>basejump 0.9.13 (2019-03-18)</h2>
<div id="minor-changes-10" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-10" class="anchor"></a>Minor changes</h3>
<ul>
<li>Updated dependencies, specificially brio and goalie.</li>
<li>Miscellaneous documentation improvements.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-14-2019-03-18" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-14-2019-03-18" class="anchor"></a>basejump 0.9.14 (2019-03-18)</h2>
<div id="minor-changes-9" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-9" class="anchor"></a>Minor changes</h3>
<ul>
<li>Additional bug fixes for <code>sampleData</code> and blacklisted metadata handling.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-15-2019-03-22" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-15-2019-03-22" class="anchor"></a>basejump 0.9.15 (2019-03-22)</h2>
<div id="minor-changes-8" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-8" class="anchor"></a>Minor changes</h3>
<ul>
<li>Migrated code to <a href="https://acidgenomics.com/">Acid Genomics</a>.</li>
</ul>
</div>
</div>
    <div id="basejump-0-9-16-2019-03-23" class="section level2">
<h2 class="hasAnchor">
<a href="#basejump-0-9-16-2019-03-23" class="anchor"></a>basejump 0.9.16 (2019-03-23)</h2>
<div id="minor-changes-7" class="section level3">
<h3 class="hasAnchor">
<a href="#minor-changes-7" class="anchor"></a>Minor changes</h3>
<ul>
<li>
<code>convertSampleIDsToNames</code>: Removed code to assign <code>sampleName</code> column to <code>NULL</code>. This step doesn’t work consistently for <code>DataFrame</code> across Bioconductor installations, and has been found to error on R 3.4 and the current bioc-devel on AppVeyor.</li>
<li>Miscellaneous documentation fixes, removing extra formatting in titles.</li>
<li>
<code>makeGRangesFromGFF</code>: Compressed Ensembl GTF file example was erroring out on AppVeyor CI, due to Windows’ poor handling of temp files on non-admin accounts. Switched to a non-gzipped example file to avoid this issue. Also removed tabular table from documentation, which currently doesn’t render correctly via pkgdown.</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#basejump-0-9-0-2018-12-12">0.9.0</a></li>
        <li><a href="#basejump-0-9-1-2018-12-22">0.9.1</a></li>
        <li><a href="#basejump-0-9-2-2019-01-07">0.9.2</a></li>
        <li><a href="#basejump-0-9-3-2019-01-08">0.9.3</a></li>
        <li><a href="#basejump-0-9-4-2019-01-12">0.9.4</a></li>
        <li><a href="#basejump-0-9-5-2019-01-22">0.9.5</a></li>
        <li><a href="#basejump-0-9-6-2019-01-22">0.9.6</a></li>
        <li><a href="#basejump-0-9-7-2019-01-23">0.9.7</a></li>
        <li><a href="#basejump-0-9-8-2019-02-08">0.9.8</a></li>
        <li><a href="#basejump-0-9-9-2019-02-11">0.9.9</a></li>
        <li><a href="#basejump-0-9-10-2019-02-17">0.9.10</a></li>
        <li><a href="#basejump-0-9-11-2019-02-25">0.9.11</a></li>
        <li><a href="#basejump-0-9-12-2019-03-11">0.9.12</a></li>
        <li><a href="#basejump-0-9-13-2019-03-18">0.9.13</a></li>
        <li><a href="#basejump-0-9-14-2019-03-18">0.9.14</a></li>
        <li><a href="#basejump-0-9-15-2019-03-22">0.9.15</a></li>
        <li><a href="#basejump-0-9-16-2019-03-23">0.9.16</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

