<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>basejump • basejump</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="basejump">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>basejump</h1>
                        <h4 class="author">Michael Steinbaugh</h4>
            
            <h4 class="date">6 August 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/acidgenomics/basejump/blob/master/vignettes/basejump.Rmd"><code>vignettes/basejump.Rmd</code></a></small>
      <div class="hidden name"><code>basejump.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>The basejump package is an infrastructure toolkit that extends the base functionality of Bioconductor <span class="citation">(Huber et al. 2015)</span>. The package leverages the S4 object system for object-oriented programming, and defines multiple additional generic functions for use in genomics research. basejump provides simple, user-friendly functions for the acquisition of genome annotations from multiple online databases, including native support for Ensembl <span class="citation">(T. Hubbard et al. 2002)</span> and websites supporting standard FASTA and GTF/GFF file formats. Consistent handling of sample metadata remains a challenge for many bioinformatic analyses, and basejump aims to address this by providing a suite of sanitization functions to help standardize these variable inputs. Additionally, interactive read/write operations in R can be cumbersome and non-trivial when working with multiple data objects; here we provide additional functions designed for interactive use that aim to reduce friction and provide consistent handling of multiple common file formats used in genomics research.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(basejump)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(SummarizedExperiment)
<span class="kw"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="dt">acid.test =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(RangedSummarizedExperiment, <span class="dt">package =</span> <span class="st">"acidtest"</span>)
rse &lt;-<span class="st"> </span>RangedSummarizedExperiment</code></pre>
<p>This vignette focuses on the most common user-facing functions that fall into the following categories:</p>
<ul>
<li>Annotation functions</li>
<li>Read/write functions</li>
<li>Data functions</li>
<li>Syntactic naming functions</li>
<li>Math and science functions</li>
</ul>
<p>There are additional function families whose functionality lie outside the scope of this vignette. Consult the <a href="https://basejump.acidgenomics.com/">basejump</a> website or package documentation for more information of these types of functions, which are more technical in nature and predominantly developer-facing:</p>
<ul>
<li>R Markdown functions</li>
<li>Atomic vector functions</li>
<li>Coercion methods</li>
<li>Developer functions</li>
<li>Assert check functions</li>
</ul>
</div>
<div id="the-s4-object-system" class="section level1">
<h1 class="hasAnchor">
<a href="#the-s4-object-system" class="anchor"></a>The S4 object system</h1>
<p>The <a href="https://basejump.acidgenomics.com/">basejump</a> package defines generics and methods for object-oriented programming with the S4 object system, which is used extensively by the <a href="https://bioconductor.org/">Bioconductor</a> project. These resources describe the S4 object system in detail:</p>
<ul>
<li>
<a href="https://www.bioconductor.org/help/course-materials/2017/Zurich/S4-classes-and-methods.html">S4 classes and methods</a>, Martin Morgan and Hervé Pagès, 2017.</li>
<li>
<a href="https://bioconductor.org/packages/devel/bioc/vignettes/S4Vectors/inst/doc/S4QuickOverview.pdf">A quick overview of the S4 class system</a>, Hervé Pagès, 2016.</li>
<li>
<a href="http://adv-r.had.co.nz/S4.html">The S4 object system</a>, Hadley Wickham.</li>
<li>
<a href="http://kasperdanielhansen.github.io/genbioconductor">Bioconductor for Genomic Data Science</a>, Kasper Hansen, 2016.</li>
<li>
<a href="https://www.bioconductor.org/help/course-materials/2010/AdvancedR/S4InBioconductor.pdf">S4 System Development in Bioconductor</a>, Patrick Aboyoun, 2010.</li>
</ul>
<p>In <a href="https://www.r-project.org/">R</a>, you can check whether a function is using the S4 object system with the <code><a href="https://rdrr.io/r/base/isS4.html">isS4()</a></code> function. Additionally, <code>showMethods()</code> and <code>getMethod()</code> are useful for exploring source code of S4 methods. In contrast, to obtain information on functions using the <a href="http://adv-r.had.co.nz/S3.html">S3</a> object system, use <code><a href="https://rdrr.io/r/utils/methods.html">methods()</a></code> instead.</p>
</div>
<div id="annotation-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#annotation-functions" class="anchor"></a>Annotation functions</h1>
<p>Obtaining versioned genome annotations quickly and reliably from online databases such as <a href="https://www.ensembl.org/">Ensembl</a> <span class="citation">(T. Hubbard et al. 2002)</span>, <a href="https://www.gencodegenes.org/">GENCODE</a> <span class="citation">(Harrow et al. 2012)</span>, <a href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a> <span class="citation">(Pruitt, Tatusova, and Maglott 2007)</span>, and the <a href="http://genome.ucsc.edu/">UCSC Genome Browser</a> <span class="citation">(Kent et al. 2002)</span> remains overly challenging. <a href="https://basejump.acidgenomics.com/">basejump</a> aims to help address this issue by providing native gene- and transcript-level annotation support by helping users interface with <a href="https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a> and <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a>, along with a robust set of tools to parse <a href="http://mblab.wustl.edu/GTF22.html" title="General Transfer Format">GTF</a> and <a href="https://useast.ensembl.org/info/website/upload/gff.html" title="General Feature Format">GFF</a> annotation files. Annotation file parsing is handled internally by the <a href="https://bioconductor.org/packages/rtracklayer/">rtracklayer</a> package <span class="citation">(Lawrence, Gentleman, and Carey 2009)</span>. Genome annotations are returned as <code>GRanges</code> class objects, defined in the <a href="https://bioconductor.org/packages/GenomicRanges/">GenomicRanges</a> package <span class="citation">(Lawrence et al. 2013)</span>, which allows for easy access to position, chromosome, strand, and additional metadata saved in the <code><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols()</a></code> slot of the object. <code>GRanges</code> can be coerced to a standard <code>data.frame</code> using the <code><a href="../reference/reexports-S4.html">as.data.frame()</a></code> function.</p>
<div id="ensembl-annotations" class="section level2">
<h2 class="hasAnchor">
<a href="#ensembl-annotations" class="anchor"></a>Ensembl annotations</h2>
<p><code><a href="https://rdrr.io/pkg/freerange/man/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl()</a></code> supports multiple genome builds, and versioned releases from <a href="https://www.ensembl.org/">Ensembl</a>.</p>
<div id="current-release" class="section level3">
<h3 class="hasAnchor">
<a href="#current-release" class="anchor"></a>Current release</h3>
<p>By default, the function will use the latest release version available from <a href="https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a> and <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a>.</p>
<pre class="sourceCode r"><code class="sourceCode r">grch38 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/freerange/man/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl</a></span>(<span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>)
<span class="kw"><a href="../reference/summary.html">summary</a></span>(grch38)</code></pre>
<pre><code>## [1] "GRanges object with 65774 ranges and 8 metadata columns"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(grch38))</code></pre>
<pre><code>## [1] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419" "ENSG00000000457"
## [5] "ENSG00000000460" "ENSG00000000938"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html">metadata</a></span>(grch38))</code></pre>
<pre><code>## [1] "package"        "version"        "date"           "organism"      
## [5] "genomeBuild"    "ensemblRelease" "ensembldb"      "level"         
## [9] "id"</code></pre>
<p>Data inside a <code>GRanges</code> object can be accessed with a number of functions defined in the <a href="https://bioconductor.org/packages/GenomeInfoDb/">GenomeInfoDb</a> and <a href="https://www.bioconductor.org/packages/IRanges/">IRanges</a> packages.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqnames</span>(grch38)</code></pre>
<pre><code>## factor-Rle of length 65774 with 49163 runs
##   Lengths:      2      1      4      3 ...      1      1      1      2
##   Values :      X     20      1      6 ...     10     15     22     11
## Levels(319): X 20 1 6 3 7 ... LRG_239 LRG_311 LRG_721 LRG_741 LRG_93</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqinfo</span>(grch38)</code></pre>
<pre><code>## Seqinfo object with 319 sequences from GRCh38 genome:
##   seqnames seqlengths isCircular genome
##   X         156040895      FALSE GRCh38
##   20         64444167      FALSE GRCh38
##   1         248956422      FALSE GRCh38
##   6         170805979      FALSE GRCh38
##   3         198295559      FALSE GRCh38
##   ...             ...        ...    ...
##   LRG_239      114904      FALSE GRCh38
##   LRG_311      115492      FALSE GRCh38
##   LRG_721       33396      FALSE GRCh38
##   LRG_741      231167      FALSE GRCh38
##   LRG_93        22459      FALSE GRCh38</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ranges</span>(grch38)</code></pre>
<pre><code>## IRanges object with 65774 ranges and 0 metadata columns:
##                       start       end     width
##                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   ENSG00000000003 100627109 100639991     12883
##   ENSG00000000005 100584802 100599885     15084
##   ENSG00000000419  50934867  50958555     23689
##   ENSG00000000457 169849631 169894267     44637
##   ENSG00000000460 169662007 169854080    192074
##               ...       ...       ...       ...
##            LRG_94  70597348  70602775      5428
##            LRG_96  55203594  55289803     86210
##            LRG_97  37225270  37244265     18996
##            LRG_98  36568013  36579762     11750
##            LRG_99  36591943  36598262      6320</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">strand</span>(grch38)</code></pre>
<pre><code>## factor-Rle of length 65774 with 32627 runs
##   Lengths: 1 1 2 1 1 1 2 1 1 1 1 4 3 1 1 ... 1 1 2 1 1 1 7 3 6 1 1 1 3 1 1
##   Values : - + - + - + - + - + - + - + - ... - + - + - + - + - + - + - + -
## Levels(3): + - *</code></pre>
<p>The <code>GRanges</code> class is a powerful container for genomic coordinates and metadata. For example, we can easily the gene counts per chromosome.</p>
<pre class="sourceCode r"><code class="sourceCode r">grch38 %&gt;%
<span class="st">    </span><span class="kw">seqnames</span>() %&gt;%
<span class="st">    </span><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>() %&gt;%
<span class="st">    </span><span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="dt">decreasing =</span> <span class="ot">TRUE</span>) %&gt;%
<span class="st">    </span><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="dt">n =</span> <span class="dv">24</span>) %&gt;%
<span class="st">    </span>.[<span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(.))]</code></pre>
<pre><code>## .
##    1   10   11   12   13   14   15   16   17   18   19    2   20   21   22 
## 5457 2329 3432 3072 1392 2306 2262 2623 3148 1216 3025 4193 1449  893 1399 
##    3    4    5    6    7    8    9    X    Y 
## 3189 2643 3016 3021 3000 2455 2375 2507  544</code></pre>
<p>Gene metadata is contained in <code><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols()</a></code>, using a <code>DataFrame</code> internally. Here’a the current list of gene-level annotation columns returned from <code><a href="https://rdrr.io/pkg/freerange/man/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl()</a></code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span>(grch38) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>()</code></pre>
<pre><code>## [1] "broadClass"     "description"    "entrezID"       "geneBiotype"   
## [5] "geneID"         "geneIDVersion"  "geneName"       "seqCoordSystem"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">mcols</a></span>(grch38) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="dt">X =</span> ., <span class="dt">FUN =</span> head)</code></pre>
<pre><code>## $broadClass
## factor-Rle of length 6 with 1 run
##   Lengths:      6
##   Values : coding
## Levels(8): coding ig mito noncoding other pseudo small tcr
## 
## $description
## factor-Rle of length 6 with 6 runs
##   Lengths:                                                              ...
##   Values :                                                          tetr...
## Levels(35744): 1-acylglycerol-3-phosphate O-acyltransferase 1 [Source:HG...
## 
## $entrezID
## $entrezID$ENSG00000000003
## [1] 7105
## 
## $entrezID$ENSG00000000005
## [1] 64102
## 
## $entrezID$ENSG00000000419
## [1] 8813
## 
## $entrezID$ENSG00000000457
## [1] 57147
## 
## $entrezID$ENSG00000000460
## [1] 55732
## 
## $entrezID$ENSG00000000938
## [1] 2268
## 
## 
## $geneBiotype
## factor-Rle of length 6 with 1 run
##   Lengths:              6
##   Values : protein_coding
## Levels(45): 3prime_overlapping_ncrna antisense ... vaultRNA
## 
## $geneID
## character-Rle of length 6 with 6 runs
##   Lengths:                 1                 1 ...                 1
##   Values : "ENSG00000000003" "ENSG00000000005" ... "ENSG00000000938"
## 
## $geneIDVersion
## character-Rle of length 6 with 6 runs
##   Lengths:                    1 ...                    1
##   Values : "ENSG00000000003.13" ... "ENSG00000000938.11"
## 
## $geneName
## factor-Rle of length 6 with 6 runs
##   Lengths:        1        1        1        1        1        1
##   Values :   TSPAN6     TNMD     DPM1    SCYL3 C1orf112      FGR
## Levels(59074): 5_8S_rRNA 5S_rRNA 7SK A1BG ... ZYG11B ZYX ZZEF1 ZZZ3
## 
## $seqCoordSystem
## factor-Rle of length 6 with 1 run
##   Lengths:          6
##   Values : chromosome
## Levels(3): chromosome lrg scaffold</code></pre>
<p>Transcript-level annotations are also supported, using the <code>level = "transcripts"</code> argument.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/freerange/man/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl</a></span>(<span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>, <span class="dt">level =</span> <span class="st">"transcripts"</span>)</code></pre>
<pre><code>## GRanges object with 214285 ranges and 14 metadata columns:
##                   seqnames              ranges strand | broadClass
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;Rle&gt;
##   ENST00000000233        7 127588345-127591705      + |     coding
##   ENST00000000412       12     8940365-8949955      - |     coding
##   ENST00000000442       11   64305578-64316738      + |     coding
##   ENST00000001008       12     2794953-2805423      + |     coding
##   ENST00000001146        2   72129238-72148038      - |     coding
##               ...      ...                 ...    ... .        ...
##          LRG_94t1       10   70597348-70602775      - |      other
##          LRG_96t1       15   55203594-55289803      - |      other
##          LRG_97t1       22   37225270-37244265      - |      other
##          LRG_98t1       11   36568013-36579762      + |      other
##          LRG_99t1       11   36591943-36598262      - |      other
##                                                                                                                                  description
##                                                                                                                                        &lt;Rle&gt;
##   ENST00000000233                                                                ADP-ribosylation factor 5 [Source:HGNC Symbol;Acc:HGNC:658]
##   ENST00000000412                                         mannose-6-phosphate receptor (cation dependent) [Source:HGNC Symbol;Acc:HGNC:6752]
##   ENST00000000442                                                         estrogen-related receptor alpha [Source:HGNC Symbol;Acc:HGNC:3471]
##   ENST00000001008                                                          FK506 binding protein 4, 59kDa [Source:HGNC Symbol;Acc:HGNC:3720]
##   ENST00000001146                                 cytochrome P450, family 26, subfamily B, polypeptide 1 [Source:HGNC Symbol;Acc:HGNC:20581]
##               ...                                                                                                                        ...
##          LRG_94t1                                                       perforin 1 (pore forming protein) [Source:HGNC Symbol;Acc:HGNC:9360]
##          LRG_96t1                                                      RAB27A, member RAS oncogene family [Source:HGNC Symbol;Acc:HGNC:9766]
##          LRG_97t1 ras-related C3 botulinum toxin substrate 2 (rho family, small GTP binding protein Rac2) [Source:HGNC Symbol;Acc:HGNC:9802]
##          LRG_98t1                                                         recombination activating gene 1 [Source:HGNC Symbol;Acc:HGNC:9831]
##          LRG_99t1                                                         recombination activating gene 2 [Source:HGNC Symbol;Acc:HGNC:9832]
##                   entrezID    geneBiotype          geneID
##                     &lt;list&gt;          &lt;Rle&gt;           &lt;Rle&gt;
##   ENST00000000233      381 protein_coding ENSG00000004059
##   ENST00000000412     4074 protein_coding ENSG00000003056
##   ENST00000000442     2101 protein_coding ENSG00000173153
##   ENST00000001008     2288 protein_coding ENSG00000004478
##   ENST00000001146    56603 protein_coding ENSG00000003137
##               ...      ...            ...             ...
##          LRG_94t1     5551       LRG_gene          LRG_94
##          LRG_96t1     5873       LRG_gene          LRG_96
##          LRG_97t1     5880       LRG_gene          LRG_97
##          LRG_98t1     5896       LRG_gene          LRG_98
##          LRG_99t1     5897       LRG_gene          LRG_99
##                        geneIDVersion geneName seqCoordSystem
##                                &lt;Rle&gt;    &lt;Rle&gt;          &lt;Rle&gt;
##   ENST00000000233  ENSG00000004059.9     ARF5     chromosome
##   ENST00000000412  ENSG00000003056.6     M6PR     chromosome
##   ENST00000000442 ENSG00000173153.12    ESRRA     chromosome
##   ENST00000001008  ENSG00000004478.7    FKBP4     chromosome
##   ENST00000001146  ENSG00000003137.7  CYP26B1     chromosome
##               ...                ...      ...            ...
##          LRG_94t1           LRG_94.1     PRF1     chromosome
##          LRG_96t1           LRG_96.1   RAB27A     chromosome
##          LRG_97t1           LRG_97.1     RAC2     chromosome
##          LRG_98t1           LRG_98.1     RAG1     chromosome
##          LRG_99t1           LRG_99.1     RAG2     chromosome
##                   transcriptBiotype transcriptCdsSeqEnd
##                               &lt;Rle&gt;               &lt;Rle&gt;
##   ENST00000000233    protein_coding           127591299
##   ENST00000000412    protein_coding             8946404
##   ENST00000000442    protein_coding            64315966
##   ENST00000001008    protein_coding             2803258
##   ENST00000001146    protein_coding            72147834
##               ...               ...                 ...
##          LRG_94t1          LRG_gene            70600902
##          LRG_96t1          LRG_gene            55234934
##          LRG_97t1          LRG_gene            37244148
##          LRG_98t1          LRG_gene            36576436
##          LRG_99t1          LRG_gene            36594168
##                   transcriptCdsSeqStart    transcriptID
##                                   &lt;Rle&gt;           &lt;Rle&gt;
##   ENST00000000233             127588499 ENST00000000233
##   ENST00000000412               8941818 ENST00000000412
##   ENST00000000442              64307180 ENST00000000442
##   ENST00000001008               2795140 ENST00000001008
##   ENST00000001146              72132227 ENST00000001146
##               ...                   ...             ...
##          LRG_94t1              70598053        LRG_94t1
##          LRG_96t1              55205507        LRG_96t1
##          LRG_97t1              37226673        LRG_97t1
##          LRG_98t1              36573305        LRG_98t1
##          LRG_99t1              36592585        LRG_99t1
##                   transcriptIDVersion  transcriptName
##                                 &lt;Rle&gt;           &lt;Rle&gt;
##   ENST00000000233   ENST00000000233.8 ENST00000000233
##   ENST00000000412   ENST00000000412.6 ENST00000000412
##   ENST00000000442   ENST00000000442.9 ENST00000000442
##   ENST00000001008   ENST00000001008.5 ENST00000001008
##   ENST00000001146   ENST00000001146.5 ENST00000001146
##               ...                 ...             ...
##          LRG_94t1          LRG_94t1.1        LRG_94t1
##          LRG_96t1          LRG_96t1.1        LRG_96t1
##          LRG_97t1          LRG_97t1.1        LRG_97t1
##          LRG_98t1          LRG_98t1.1        LRG_98t1
##          LRG_99t1          LRG_99t1.1        LRG_99t1
##   -------
##   seqinfo: 319 sequences from GRCh38 genome</code></pre>
</div>
<div id="versioned-releases" class="section level3">
<h3 class="hasAnchor">
<a href="#versioned-releases" class="anchor"></a>Versioned releases</h3>
<p>Versioned releases are supported by the “<code>release</code>” argument, which currently supports back to <a href="https://www.ensembl.org/">Ensembl</a> 87. If an older <a href="https://www.ensembl.org/">Ensembl</a> release is required, resort to using a <a href="http://mblab.wustl.edu/GTF22.html" title="General Transfer Format">GTF</a>/<a href="https://useast.ensembl.org/info/website/upload/gff.html" title="General Feature Format">GFF</a> file or browse the <a href="https://bioconductor.org/">Bioconductor</a> website to see if a suitable annotation database package is available.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/freerange/man/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl</a></span>(
    <span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>,
    <span class="dt">genomeBuild =</span> <span class="st">"GRCh38"</span>,
    <span class="dt">release =</span> <span class="dv">90</span>
)</code></pre>
<pre><code>## GRanges object with 64661 ranges and 8 metadata columns:
##                   seqnames              ranges strand | broadClass
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;Rle&gt;
##   ENSG00000000003        X 100627109-100639991      - |     coding
##   ENSG00000000005        X 100584802-100599885      + |     coding
##   ENSG00000000419       20   50934867-50958555      - |     coding
##   ENSG00000000457        1 169849631-169894267      - |     coding
##   ENSG00000000460        1 169662007-169854080      + |     coding
##               ...      ...                 ...    ... .        ...
##           LRG_995        1   77948402-77979205      - |      other
##           LRG_996       12   56080025-56103507      + |      other
##           LRG_997        6 117288367-117425855      - |      other
##           LRG_998        6   41934933-42048894      - |      other
##           LRG_999       19   42268537-42295796      + |      other
##                                                                                                      description
##                                                                                                            &lt;Rle&gt;
##   ENSG00000000003                                              tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:11858]
##   ENSG00000000005                                                tenomodulin [Source:HGNC Symbol;Acc:HGNC:17757]
##   ENSG00000000419 dolichyl-phosphate mannosyltransferase subunit 1, catalytic [Source:HGNC Symbol;Acc:HGNC:3005]
##   ENSG00000000457                                   SCY1 like pseudokinase 3 [Source:HGNC Symbol;Acc:HGNC:19285]
##   ENSG00000000460                        chromosome 1 open reading frame 112 [Source:HGNC Symbol;Acc:HGNC:25565]
##               ...                                                                                            ...
##           LRG_995                      far upstream element binding protein 1 [Source:HGNC Symbol;Acc:HGNC:4004]
##           LRG_996                           erb-b2 receptor tyrosine kinase 3 [Source:HGNC Symbol;Acc:HGNC:3431]
##           LRG_997             ROS proto-oncogene 1, receptor tyrosine kinase [Source:HGNC Symbol;Acc:HGNC:10261]
##           LRG_998                                                   cyclin D3 [Source:HGNC Symbol;Acc:HGNC:1585]
##           LRG_999                          capicua transcriptional repressor [Source:HGNC Symbol;Acc:HGNC:14214]
##                   entrezID    geneBiotype          geneID
##                     &lt;list&gt;          &lt;Rle&gt;           &lt;Rle&gt;
##   ENSG00000000003     7105 protein_coding ENSG00000000003
##   ENSG00000000005    64102 protein_coding ENSG00000000005
##   ENSG00000000419     8813 protein_coding ENSG00000000419
##   ENSG00000000457    57147 protein_coding ENSG00000000457
##   ENSG00000000460    55732 protein_coding ENSG00000000460
##               ...      ...            ...             ...
##           LRG_995     8880       LRG_gene         LRG_995
##           LRG_996     2065       LRG_gene         LRG_996
##           LRG_997     6098       LRG_gene         LRG_997
##           LRG_998      896       LRG_gene         LRG_998
##           LRG_999    23152       LRG_gene         LRG_999
##                        geneIDVersion geneName seqCoordSystem
##                                &lt;Rle&gt;    &lt;Rle&gt;          &lt;Rle&gt;
##   ENSG00000000003 ENSG00000000003.14   TSPAN6     chromosome
##   ENSG00000000005  ENSG00000000005.5     TNMD     chromosome
##   ENSG00000000419 ENSG00000000419.12     DPM1     chromosome
##   ENSG00000000457 ENSG00000000457.13    SCYL3     chromosome
##   ENSG00000000460 ENSG00000000460.16 C1orf112     chromosome
##               ...                ...      ...            ...
##           LRG_995          LRG_995.1    FUBP1     chromosome
##           LRG_996          LRG_996.1    ERBB3     chromosome
##           LRG_997          LRG_997.1     ROS1     chromosome
##           LRG_998          LRG_998.1    CCND3     chromosome
##           LRG_999          LRG_999.1      CIC     chromosome
##   -------
##   seqinfo: 388 sequences from GRCh38 genome</code></pre>
</div>
<div id="grch37" class="section level3">
<h3 class="hasAnchor">
<a href="#grch37" class="anchor"></a>GRCh37</h3>
<p>The legacy <a href="https://grch37.ensembl.org/">Ensembl GRCh37 genome build</a> (release 75; also known as <a href="https://genome.ucsc.edu/cgi-bin/hgGateway?db=hg19">UCSC hg19</a>) is also supported, but is no longer recommended for use in new analyses. Internally, support for <a href="https://grch37.ensembl.org/">GRCh37</a> is provided by the <a href="https://bioconductor.org/packages/EnsDb.Hsapiens.v75/">EnsDb.Hsapiens.v75</a> annotation database package.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/freerange/man/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl</a></span>(
    <span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>,
    <span class="dt">genomeBuild =</span> <span class="st">"GRCh37"</span>
)</code></pre>
<pre><code>## GRanges object with 64102 ranges and 6 metadata columns:
##                   seqnames              ranges strand | broadClass
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;Rle&gt;
##   ENSG00000000003        X   99883667-99894988      - |     coding
##   ENSG00000000005        X   99839799-99854882      + |     coding
##   ENSG00000000419       20   49551404-49575092      - |     coding
##   ENSG00000000457        1 169818772-169863408      - |     coding
##   ENSG00000000460        1 169631245-169823221      + |     coding
##               ...      ...                 ...    ... .        ...
##            LRG_94       10   72357104-72362531      - |      other
##            LRG_96       15   55495792-55582001      - |      other
##            LRG_97       22   37621310-37640305      - |      other
##            LRG_98       11   36589563-36601312      + |      other
##            LRG_99       11   36613493-36619812      - |      other
##                   entrezID    geneBiotype          geneID geneName
##                     &lt;list&gt;          &lt;Rle&gt;           &lt;Rle&gt;    &lt;Rle&gt;
##   ENSG00000000003     7105 protein_coding ENSG00000000003   TSPAN6
##   ENSG00000000005    64102 protein_coding ENSG00000000005     TNMD
##   ENSG00000000419     8813 protein_coding ENSG00000000419     DPM1
##   ENSG00000000457    57147 protein_coding ENSG00000000457    SCYL3
##   ENSG00000000460    55732 protein_coding ENSG00000000460 C1orf112
##               ...      ...            ...             ...      ...
##            LRG_94     5551       LRG_gene          LRG_94   LRG_94
##            LRG_96     5873       LRG_gene          LRG_96   LRG_96
##            LRG_97     5880       LRG_gene          LRG_97   LRG_97
##            LRG_98     5896       LRG_gene          LRG_98   LRG_98
##            LRG_99     5897       LRG_gene          LRG_99   LRG_99
##                   seqCoordSystem
##                            &lt;Rle&gt;
##   ENSG00000000003     chromosome
##   ENSG00000000005     chromosome
##   ENSG00000000419     chromosome
##   ENSG00000000457     chromosome
##   ENSG00000000460     chromosome
##               ...            ...
##            LRG_94     chromosome
##            LRG_96     chromosome
##            LRG_97     chromosome
##            LRG_98     chromosome
##            LRG_99     chromosome
##   -------
##   seqinfo: 273 sequences from GRCh37 genome</code></pre>
</div>
</div>
<div id="gene-mapping-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-mapping-functions" class="anchor"></a>Gene mapping functions</h2>
<p>Gene-to-symbol and transcript-to-gene mappings can be easily acquired with the <code><a href="../reference/defunct.html">gene2symbol()</a></code> and <code><a href="../reference/defunct.html">tx2gene()</a></code> functions. Both of these functions return a <code>data.frame</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/makeGene2Symbol.html">makeGene2SymbolFromEnsembl</a></span>(<span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>)</code></pre>
<pre><code>## Gene2Symbol with 65774 rows and 2 columns
##                          geneID    geneName
##                     &lt;character&gt; &lt;character&gt;
## ENSG00000000003 ENSG00000000003      TSPAN6
## ENSG00000000005 ENSG00000000005        TNMD
## ENSG00000000419 ENSG00000000419        DPM1
## ENSG00000000457 ENSG00000000457       SCYL3
## ENSG00000000460 ENSG00000000460    C1orf112
## ...                         ...         ...
## LRG_94                   LRG_94      PRF1.1
## LRG_96                   LRG_96    RAB27A.1
## LRG_97                   LRG_97      RAC2.1
## LRG_98                   LRG_98      RAG1.1
## LRG_99                   LRG_99      RAG2.1</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/makeTx2Gene.html">makeTx2GeneFromEnsembl</a></span>(<span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>)</code></pre>
<pre><code>## Tx2Gene with 214285 rows and 2 columns
##                    transcriptID          geneID
##                     &lt;character&gt;     &lt;character&gt;
## ENST00000000233 ENST00000000233 ENSG00000004059
## ENST00000000412 ENST00000000412 ENSG00000003056
## ENST00000000442 ENST00000000442 ENSG00000173153
## ENST00000001008 ENST00000001008 ENSG00000004478
## ENST00000001146 ENST00000001146 ENSG00000003137
## ...                         ...             ...
## LRG_94t1               LRG_94t1          LRG_94
## LRG_96t1               LRG_96t1          LRG_96
## LRG_97t1               LRG_97t1          LRG_97
## LRG_98t1               LRG_98t1          LRG_98
## LRG_99t1               LRG_99t1          LRG_99</code></pre>
</div>
<div id="parse-gtfgff-files" class="section level2">
<h2 class="hasAnchor">
<a href="#parse-gtfgff-files" class="anchor"></a>Parse GTF/GFF files</h2>
<p>For organisms that are not well supported on <a href="https://www.ensembl.org/">Ensembl</a>, we recommend using <a href="http://mblab.wustl.edu/GTF22.html" title="General Transfer Format">GTF</a> or <a href="https://useast.ensembl.org/info/website/upload/gff.html" title="General Feature Format">GFF</a> files for gene annotations. In general, <a href="http://mblab.wustl.edu/GTF22.html" title="General Transfer Format">GTF</a> files are more consistently standardized and recommended over GFF3, if possible. <code><a href="../reference/defunct.html">makeTx2geneFromGFF()</a></code> and <code><a href="../reference/defunct.html">makeGene2symbolFromGFF()</a></code> are utility functions that can return transcript-to-gene and gene-to-symbol mappings from the genomic ranges returned by <code><a href="https://rdrr.io/pkg/freerange/man/makeGRangesFromGFF.html">makeGRangesFromGFF()</a></code>.</p>
<div id="gtf" class="section level3">
<h3 class="hasAnchor">
<a href="#gtf" class="anchor"></a>GTF</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/freerange/man/makeGRangesFromGFF.html">makeGRangesFromGFF</a></span>(
    <span class="dt">file =</span> <span class="kw"><a href="https://rdrr.io/pkg/brio/man/pasteURL.html">pasteURL</a></span>(
        <span class="st">"ftp.ensembl.org"</span>,
        <span class="st">"pub"</span>,
        <span class="st">"release-92"</span>,
        <span class="st">"gtf"</span>,
        <span class="st">"homo_sapiens"</span>,
        <span class="st">"Homo_sapiens.GRCh38.92.gtf.gz"</span>,
        <span class="dt">protocol =</span> <span class="st">"ftp"</span>
    )
)</code></pre>
</div>
<div id="gff3" class="section level3">
<h3 class="hasAnchor">
<a href="#gff3" class="anchor"></a>GFF3</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/freerange/man/makeGRangesFromGFF.html">makeGRangesFromGFF</a></span>(
    <span class="dt">file =</span> <span class="kw"><a href="https://rdrr.io/pkg/brio/man/pasteURL.html">pasteURL</a></span>(
        <span class="st">"ftp.ensembl.org"</span>,
        <span class="st">"pub"</span>,
        <span class="st">"release-92"</span>,
        <span class="st">"gff3"</span>,
        <span class="st">"homo_sapiens"</span>,
        <span class="st">"Homo_sapiens.GRCh38.92.gff3.gz"</span>,
        <span class="dt">protocol =</span> <span class="st">"ftp"</span>
    )
)</code></pre>
</div>
</div>
<div id="gene-interconversion-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-interconversion-functions" class="anchor"></a>Gene interconversion functions</h2>
<p>Dealing with gene identifier to gene name (symbol) mapping in a non-destructive manner remains challenging for the bioinformatics community. <a href="https://basejump.acidgenomics.com/">basejump</a> provides <code><a href="../reference/convertGenesToSymbols.html">convertGenesToSymbols()</a></code> and <code><a href="../reference/convertGenesToSymbols.html">convertSymbolsToGenes()</a></code> generics that define methods for <code>SummarizedExperiment</code>, making interconversion between gene IDs and gene names easier. Internally, these mappings are handled by the <code>geneID</code> and <code>geneName</code> <code>mcols</code> columns inside of the <code>rowRanges</code> slot.</p>
<pre class="sourceCode r"><code class="sourceCode r">rse_symbols &lt;-<span class="st"> </span><span class="kw"><a href="../reference/convertGenesToSymbols.html">convertGenesToSymbols</a></span>(rse)
<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(rse_symbols))</code></pre>
<pre><code>## [1] "TSPAN6"   "TNMD"     "DPM1"     "SCYL3"    "C1orf112" "FGR"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">rse_genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/convertGenesToSymbols.html">convertSymbolsToGenes</a></span>(rse_symbols)
<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(rse_genes))</code></pre>
<pre><code>## [1] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419" "ENSG00000000457"
## [5] "ENSG00000000460" "ENSG00000000938"</code></pre>
<p>To convert from transcript- to gene-level easily, the <code><a href="../reference/convertTranscriptsToGenes.html">convertTranscriptsToGenes()</a></code> and <code><a href="../reference/stripTranscriptVersions.html">stripTranscriptVersions()</a></code> functions are also provided (not shown).</p>
</div>
<div id="query-external-annotation-databases" class="section level2">
<h2 class="hasAnchor">
<a href="#query-external-annotation-databases" class="anchor"></a>Query external annotation databases</h2>
<p>In addition to interfacing with <a href="https://www.ensembl.org/">Ensembl</a>, <a href="https://basejump.acidgenomics.com/">basejump</a> exports a number of utility functions to interface with other genomic databases, including:</p>
<ul>
<li>
<a href="#eggnog">EggNOG</a> <span class="citation">(Huerta-Cepas et al. 2016)</span>.</li>
<li>
<a href="#hgnc">HGNC</a> <span class="citation">(Yates et al. 2017)</span>.</li>
<li>
<a href="#mgi">MGI</a> <span class="citation">(C. L. Smith et al. 2018)</span>.</li>
<li>
<a href="#panther">PANTHER</a> <span class="citation">(Mi et al. 2017)</span>.</li>
</ul>
<div id="hgnc" class="section level3">
<h3 class="hasAnchor">
<a href="#hgnc" class="anchor"></a>HGNC</h3>
<pre class="sourceCode r"><code class="sourceCode r">hgnc2gene &lt;-<span class="st"> </span><span class="kw"><a href="../reference/HGNC2Ensembl-class.html">HGNC2Ensembl</a></span>()
<span class="kw"><a href="../reference/summary.html">summary</a></span>(hgnc2gene)</code></pre>
<pre><code>##       Length        Class         Mode 
##            2 HGNC2Ensembl           S4</code></pre>
</div>
<div id="mgi" class="section level3">
<h3 class="hasAnchor">
<a href="#mgi" class="anchor"></a>MGI</h3>
<pre class="sourceCode r"><code class="sourceCode r">mgi2ensembl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/MGI2Ensembl-class.html">MGI2Ensembl</a></span>()
<span class="kw"><a href="../reference/summary.html">summary</a></span>(mgi2ensembl)</code></pre>
<pre><code>##      Length       Class        Mode 
##           2 MGI2Ensembl          S4</code></pre>
</div>
<div id="eggnog" class="section level3">
<h3 class="hasAnchor">
<a href="#eggnog" class="anchor"></a>EggNOG</h3>
<pre class="sourceCode r"><code class="sourceCode r">eggnog &lt;-<span class="st"> </span><span class="kw"><a href="../reference/EggNOG-class.html">EggNOG</a></span>()
<span class="kw"><a href="../reference/summary.html">summary</a></span>(eggnog)</code></pre>
<pre><code>## Length  Class   Mode 
##      2 EggNOG     S4</code></pre>
</div>
<div id="panther" class="section level3">
<h3 class="hasAnchor">
<a href="#panther" class="anchor"></a>PANTHER</h3>
<pre class="sourceCode r"><code class="sourceCode r">panther &lt;-<span class="st"> </span><span class="kw"><a href="../reference/PANTHER-class.html">PANTHER</a></span>(<span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>)
<span class="kw"><a href="../reference/summary.html">summary</a></span>(panther)</code></pre>
<pre><code>##  Length   Class    Mode 
##       9 PANTHER      S4</code></pre>
</div>
<div id="wormbase" class="section level3">
<h3 class="hasAnchor">
<a href="#wormbase" class="anchor"></a>WormBase</h3>
<p>Note that for <em>Caenorhabditis elegans</em> genome annotations, we recommend additional use of our specialized <a href="https://wormbase.acidgenomics.com/">wormbase package</a>, which queries the <a href="https://wormbase.org/">WormBase</a> database <span class="citation">(Stein et al. 2001)</span> directly, and provides support of versioned releases (e.g. WS265) that have additional metadata not currently available on <a href="https://www.ensembl.org/">Ensembl</a>.</p>
</div>
</div>
</div>
<div id="importexport-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#importexport-functions" class="anchor"></a>Import/export functions</h1>
<p>FIXME This section is still under development.</p>
</div>
<div id="data-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#data-functions" class="anchor"></a>Data functions</h1>
<p>FIXME This section is still under development.</p>
</div>
<div id="sanitization-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#sanitization-functions" class="anchor"></a>Sanitization functions</h1>
<p>The <code>makeNames</code> suite of function, consisting primarily of <code><a href="https://rdrr.io/pkg/syntactic/man/camelCase.html">camel()</a></code>, <code><a href="https://rdrr.io/pkg/syntactic/man/dottedCase.html">dotted()</a></code>, <code><a href="https://rdrr.io/pkg/syntactic/man/snakeCase.html">snake()</a></code>, and <code><a href="https://rdrr.io/pkg/syntactic/man/upperCamelCase.html">upperCamel()</a></code>, provide S4 method support for sanitizing common data structures used in genomics research and on <a href="https://bioconductor.org/">Bioconductor</a>.</p>
<p>To see how these functions work, let’s load up an example <code>list</code> named <code>mn</code> (short for <code>makeNames</code>).</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/brio/man/loadRemoteData.html">loadRemoteData</a></span>(<span class="dt">url =</span> <span class="kw"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span>(basejumpTestsURL, <span class="st">"mn.rda"</span>))
<span class="kw"><a href="https://rdrr.io/r/base/class.html">class</a></span>(mn)</code></pre>
<pre><code>## [1] "list"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(mn)</code></pre>
<pre><code>## [1] "character"      "namedCharacter" "factor"         "dataFrame"     
## [5] "matrix"         "tibble"         "list"</code></pre>
<div id="atomic-vectors" class="section level2">
<h2 class="hasAnchor">
<a href="#atomic-vectors" class="anchor"></a>Atomic vectors</h2>
<div id="character" class="section level3">
<h3 class="hasAnchor">
<a href="#character" class="anchor"></a>Character</h3>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>mn$character
<span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(x)</code></pre>
<pre><code>##  [1] "hello world"     "HELLO WORLD"     "RNAi clones"    
##  [4] "nCount"          "tx2gene"         "TX2GeneID"      
##  [7] "G2M.Score"       "worfdbHTMLRemap" "Mazda RX4"      
## [10] "%GC"             "5prime"          "5'-3' bias"     
## [13] "123"             NA</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/camelCase.html">camel</a></span>(x)</code></pre>
<pre><code>##  [1] "helloWorld"      "helloWORLD"      "rnaiClones"     
##  [4] "nCount"          "tx2gene"         "tx2GeneID"      
##  [7] "g2mScore"        "worfdbHTMLRemap" "mazdaRX4"       
## [10] "percentGC"       "x5prime"         "x5x3Bias"       
## [13] "x123"            NA</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/dottedCase.html">dotted</a></span>(x)</code></pre>
<pre><code>##  [1] "hello.world"       "HELLO.WORLD"       "RNAI.clones"      
##  [4] "n.Count"           "tx2gene"           "TX2.Gene.ID"      
##  [7] "G2M.Score"         "worfdb.HTML.Remap" "Mazda.RX4"        
## [10] "percent.GC"        "X5prime"           "X5.3.bias"        
## [13] "X123"              NA</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/snakeCase.html">snake</a></span>(x)</code></pre>
<pre><code>##  [1] "hello_world"       "hello_world"       "rnai_clones"      
##  [4] "n_count"           "tx2gene"           "tx2_gene_id"      
##  [7] "g2m_score"         "worfdb_html_remap" "mazda_rx4"        
## [10] "percent_gc"        "x5prime"           "x5_3_bias"        
## [13] "x123"              NA</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/upperCamelCase.html">upperCamel</a></span>(x)</code></pre>
<pre><code>##  [1] "HelloWorld"      "HELLOWORLD"      "RNAIClones"     
##  [4] "NCount"          "Tx2gene"         "TX2GeneID"      
##  [7] "G2MScore"        "WorfdbHTMLRemap" "MazdaRX4"       
## [10] "PercentGC"       "X5prime"         "X5X3Bias"       
## [13] "X123"            NA</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/makeNames.html">makeNames</a></span>(x)</code></pre>
<pre><code>##  [1] "hello_world"     "HELLO_WORLD"     "RNAi_clones"    
##  [4] "nCount"          "tx2gene"         "TX2GeneID"      
##  [7] "G2M_Score"       "worfdbHTMLRemap" "Mazda_RX4"      
## [10] "X_GC"            "X5prime"         "X5__3__bias"    
## [13] "X123"            "NA_"</code></pre>
</div>
<div id="named-character" class="section level3">
<h3 class="hasAnchor">
<a href="#named-character" class="anchor"></a>Named character</h3>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>mn$namedCharacter
<span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(x)</code></pre>
<pre><code>##        Item.A        Item.B 
## "hello world" "HELLO WORLD"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/camelCase.html">camel</a></span>(x)</code></pre>
<pre><code>##        itemA        itemB 
## "helloWorld" "helloWORLD"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/dottedCase.html">dotted</a></span>(x)</code></pre>
<pre><code>##        Item.A        Item.B 
## "hello.world" "HELLO.WORLD"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/snakeCase.html">snake</a></span>(x)</code></pre>
<pre><code>##        item_a        item_b 
## "hello_world" "hello_world"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/upperCamelCase.html">upperCamel</a></span>(x)</code></pre>
<pre><code>##        ItemA        ItemB 
## "HelloWorld" "HELLOWORLD"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/makeNames.html">makeNames</a></span>(x)</code></pre>
<pre><code>## [1] "hello_world" "HELLO_WORLD"</code></pre>
</div>
<div id="factor" class="section level3">
<h3 class="hasAnchor">
<a href="#factor" class="anchor"></a>Factor</h3>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>mn$factor
<span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(x)</code></pre>
<pre><code>## sample 1 sample 2 sample 3 sample 4 
##  group 1  group 1  group 2  group 2 
## Levels: group 1 group 2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/camelCase.html">camel</a></span>(x)</code></pre>
<pre><code>## sample1 sample2 sample3 sample4 
##  group1  group1  group2  group2 
## Levels: group1 group2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/dottedCase.html">dotted</a></span>(x)</code></pre>
<pre><code>## sample.1 sample.2 sample.3 sample.4 
##  group.1  group.1  group.2  group.2 
## Levels: group.1 group.2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/snakeCase.html">snake</a></span>(x)</code></pre>
<pre><code>## sample_1 sample_2 sample_3 sample_4 
##  group_1  group_1  group_2  group_2 
## Levels: group_1 group_2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/upperCamelCase.html">upperCamel</a></span>(x)</code></pre>
<pre><code>## Sample1 Sample2 Sample3 Sample4 
##  Group1  Group1  Group2  Group2 
## Levels: Group1 Group2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/makeNames.html">makeNames</a></span>(x)</code></pre>
<pre><code>## [1] "group_1"   "group_1_1" "group_2"   "group_2_1"</code></pre>
</div>
</div>
<div id="data-frames" class="section level2">
<h2 class="hasAnchor">
<a href="#data-frames" class="anchor"></a>Data frames</h2>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>datasets::USArrests
<span class="kw"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span>(x)</code></pre>
<pre><code>## [[1]]
##  [1] "Alabama"        "Alaska"         "Arizona"        "Arkansas"      
##  [5] "California"     "Colorado"       "Connecticut"    "Delaware"      
##  [9] "Florida"        "Georgia"        "Hawaii"         "Idaho"         
## [13] "Illinois"       "Indiana"        "Iowa"           "Kansas"        
## [17] "Kentucky"       "Louisiana"      "Maine"          "Maryland"      
## [21] "Massachusetts"  "Michigan"       "Minnesota"      "Mississippi"   
## [25] "Missouri"       "Montana"        "Nebraska"       "Nevada"        
## [29] "New Hampshire"  "New Jersey"     "New Mexico"     "New York"      
## [33] "North Carolina" "North Dakota"   "Ohio"           "Oklahoma"      
## [37] "Oregon"         "Pennsylvania"   "Rhode Island"   "South Carolina"
## [41] "South Dakota"   "Tennessee"      "Texas"          "Utah"          
## [45] "Vermont"        "Virginia"       "Washington"     "West Virginia" 
## [49] "Wisconsin"      "Wyoming"       
## 
## [[2]]
## [1] "Murder"   "Assault"  "UrbanPop" "Rape"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/camelCase.html">camel</a></span>(x, <span class="dt">rownames =</span> <span class="ot">TRUE</span>, <span class="dt">colnames =</span> <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span>()</code></pre>
<pre><code>## [[1]]
##  [1] "alabama"       "alaska"        "arizona"       "arkansas"     
##  [5] "california"    "colorado"      "connecticut"   "delaware"     
##  [9] "florida"       "georgia"       "hawaii"        "idaho"        
## [13] "illinois"      "indiana"       "iowa"          "kansas"       
## [17] "kentucky"      "louisiana"     "maine"         "maryland"     
## [21] "massachusetts" "michigan"      "minnesota"     "mississippi"  
## [25] "missouri"      "montana"       "nebraska"      "nevada"       
## [29] "newHampshire"  "newJersey"     "newMexico"     "newYork"      
## [33] "northCarolina" "northDakota"   "ohio"          "oklahoma"     
## [37] "oregon"        "pennsylvania"  "rhodeIsland"   "southCarolina"
## [41] "southDakota"   "tennessee"     "texas"         "utah"         
## [45] "vermont"       "virginia"      "washington"    "westVirginia" 
## [49] "wisconsin"     "wyoming"      
## 
## [[2]]
## [1] "murder"   "assault"  "urbanPop" "rape"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/dottedCase.html">dotted</a></span>(x, <span class="dt">rownames =</span> <span class="ot">TRUE</span>, <span class="dt">colnames =</span> <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span>()</code></pre>
<pre><code>## [[1]]
##  [1] "Alabama"        "Alaska"         "Arizona"        "Arkansas"      
##  [5] "California"     "Colorado"       "Connecticut"    "Delaware"      
##  [9] "Florida"        "Georgia"        "Hawaii"         "Idaho"         
## [13] "Illinois"       "Indiana"        "Iowa"           "Kansas"        
## [17] "Kentucky"       "Louisiana"      "Maine"          "Maryland"      
## [21] "Massachusetts"  "Michigan"       "Minnesota"      "Mississippi"   
## [25] "Missouri"       "Montana"        "Nebraska"       "Nevada"        
## [29] "New.Hampshire"  "New.Jersey"     "New.Mexico"     "New.York"      
## [33] "North.Carolina" "North.Dakota"   "Ohio"           "Oklahoma"      
## [37] "Oregon"         "Pennsylvania"   "Rhode.Island"   "South.Carolina"
## [41] "South.Dakota"   "Tennessee"      "Texas"          "Utah"          
## [45] "Vermont"        "Virginia"       "Washington"     "West.Virginia" 
## [49] "Wisconsin"      "Wyoming"       
## 
## [[2]]
## [1] "Murder"    "Assault"   "Urban.Pop" "Rape"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/snakeCase.html">snake</a></span>(x, <span class="dt">rownames =</span> <span class="ot">TRUE</span>, <span class="dt">colnames =</span> <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span>()</code></pre>
<pre><code>## [[1]]
##  [1] "alabama"        "alaska"         "arizona"        "arkansas"      
##  [5] "california"     "colorado"       "connecticut"    "delaware"      
##  [9] "florida"        "georgia"        "hawaii"         "idaho"         
## [13] "illinois"       "indiana"        "iowa"           "kansas"        
## [17] "kentucky"       "louisiana"      "maine"          "maryland"      
## [21] "massachusetts"  "michigan"       "minnesota"      "mississippi"   
## [25] "missouri"       "montana"        "nebraska"       "nevada"        
## [29] "new_hampshire"  "new_jersey"     "new_mexico"     "new_york"      
## [33] "north_carolina" "north_dakota"   "ohio"           "oklahoma"      
## [37] "oregon"         "pennsylvania"   "rhode_island"   "south_carolina"
## [41] "south_dakota"   "tennessee"      "texas"          "utah"          
## [45] "vermont"        "virginia"       "washington"     "west_virginia" 
## [49] "wisconsin"      "wyoming"       
## 
## [[2]]
## [1] "murder"    "assault"   "urban_pop" "rape"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/upperCamelCase.html">upperCamel</a></span>(x, <span class="dt">rownames =</span> <span class="ot">TRUE</span>, <span class="dt">colnames =</span> <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span>()</code></pre>
<pre><code>## [[1]]
##  [1] "Alabama"       "Alaska"        "Arizona"       "Arkansas"     
##  [5] "California"    "Colorado"      "Connecticut"   "Delaware"     
##  [9] "Florida"       "Georgia"       "Hawaii"        "Idaho"        
## [13] "Illinois"      "Indiana"       "Iowa"          "Kansas"       
## [17] "Kentucky"      "Louisiana"     "Maine"         "Maryland"     
## [21] "Massachusetts" "Michigan"      "Minnesota"     "Mississippi"  
## [25] "Missouri"      "Montana"       "Nebraska"      "Nevada"       
## [29] "NewHampshire"  "NewJersey"     "NewMexico"     "NewYork"      
## [33] "NorthCarolina" "NorthDakota"   "Ohio"          "Oklahoma"     
## [37] "Oregon"        "Pennsylvania"  "RhodeIsland"   "SouthCarolina"
## [41] "SouthDakota"   "Tennessee"     "Texas"         "Utah"         
## [45] "Vermont"       "Virginia"      "Washington"    "WestVirginia" 
## [49] "Wisconsin"     "Wyoming"      
## 
## [[2]]
## [1] "Murder"   "Assault"  "UrbanPop" "Rape"</code></pre>
</div>
<div id="lists" class="section level2">
<h2 class="hasAnchor">
<a href="#lists" class="anchor"></a>Lists</h2>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>mn$list
<span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(x)</code></pre>
<pre><code>## $Item.A
## [1] 1 2
## 
## $Item.B
## [1] 3 4</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/camelCase.html">camel</a></span>(x) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>()</code></pre>
<pre><code>## [1] "itemA" "itemB"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/dottedCase.html">dotted</a></span>(x) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>()</code></pre>
<pre><code>## [1] "Item.A" "Item.B"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/snakeCase.html">snake</a></span>(x) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>()</code></pre>
<pre><code>## [1] "item_a" "item_b"</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/pkg/syntactic/man/upperCamelCase.html">upperCamel</a></span>(x) %&gt;%<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>()</code></pre>
<pre><code>## [1] "ItemA" "ItemB"</code></pre>
</div>
</div>
<div id="r-session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#r-session-information" class="anchor"></a>R session information</h1>
<pre class="sourceCode r"><code class="sourceCode r">utils::<span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</code></pre>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-redhat-linux-gnu (64-bit)
## Running under: Red Hat Enterprise Linux
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib64/libopenblas-r0.3.3.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] SummarizedExperiment_1.14.1 DelayedArray_0.10.0        
##  [3] BiocParallel_1.18.0         matrixStats_0.54.0         
##  [5] Biobase_2.44.0              GenomicRanges_1.36.0       
##  [7] GenomeInfoDb_1.20.0         IRanges_2.18.1             
##  [9] S4Vectors_0.22.0            BiocGenerics_0.30.0        
## [11] basejump_0.11.4             BiocStyle_2.12.0           
## 
## loaded via a namespace (and not attached):
##   [1] ProtGenerics_1.16.0           bitops_1.0-6                 
##   [3] fs_1.3.1                      bit64_0.9-7                  
##   [5] progress_1.2.2                httr_1.4.1                   
##   [7] rprojroot_1.3-2               syntactic_0.2.3              
##   [9] tools_3.6.1                   backports_1.1.4              
##  [11] R6_2.4.0                      lazyeval_0.2.2               
##  [13] DBI_1.0.0                     withr_2.1.2                  
##  [15] prettyunits_1.0.2             tidyselect_0.2.5             
##  [17] bit_1.1-14                    curl_4.0                     
##  [19] compiler_3.6.1                cli_1.1.0                    
##  [21] xml2_1.2.1                    desc_1.2.0                   
##  [23] rtracklayer_1.44.2            bookdown_0.12                
##  [25] readr_1.3.1                   rappdirs_0.3.1               
##  [27] pkgdown_1.3.0.9100            commonmark_1.7               
##  [29] stringr_1.4.0                 digest_0.6.20                
##  [31] Rsamtools_2.0.0               rmarkdown_1.14               
##  [33] R.utils_2.9.0                 XVector_0.24.0               
##  [35] pkgconfig_2.0.2               htmltools_0.3.6              
##  [37] sessioninfo_1.1.1             ensembldb_2.8.0              
##  [39] dbplyr_1.4.2                  rlang_0.4.0                  
##  [41] RSQLite_2.1.2                 shiny_1.3.2                  
##  [43] DelayedMatrixStats_1.6.0      dplyr_0.8.3                  
##  [45] R.oo_1.22.0                   RCurl_1.95-4.12              
##  [47] magrittr_1.5                  GenomeInfoDbData_1.2.1       
##  [49] Matrix_1.2-17                 Rcpp_1.0.2                   
##  [51] R.methodsS3_1.7.1             stringi_1.4.3                
##  [53] yaml_2.2.0                    MASS_7.3-51.4                
##  [55] zlibbioc_1.30.0               brio_0.3.2                   
##  [57] plyr_1.8.4                    goalie_0.3.1                 
##  [59] BiocFileCache_1.8.0           AnnotationHub_2.16.0         
##  [61] grid_3.6.1                    blob_1.2.0                   
##  [63] transformer_0.2.1             promises_1.0.1               
##  [65] crayon_1.3.4                  lattice_0.20-38              
##  [67] Biostrings_2.52.0             bioverbs_0.2.1               
##  [69] GenomicFeatures_1.36.4        hms_0.5.0                    
##  [71] zeallot_0.1.0                 knitr_1.23                   
##  [73] pillar_1.4.2                  reshape2_1.4.3               
##  [75] biomaRt_2.40.3                XML_3.98-1.20                
##  [77] glue_1.3.1                    evaluate_0.14                
##  [79] freerange_0.2.0               data.table_1.12.2            
##  [81] BiocManager_1.30.4            vctrs_0.2.0                  
##  [83] httpuv_1.5.1                  tidyr_0.8.3                  
##  [85] grr_0.9.5                     purrr_0.3.2                  
##  [87] assertthat_0.2.1              xfun_0.8                     
##  [89] mime_0.7                      xtable_1.8-4                 
##  [91] AnnotationFilter_1.8.0        roxygen2_6.1.1               
##  [93] later_0.8.0                   SingleCellExperiment_1.6.0   
##  [95] tibble_2.1.3                  GenomicAlignments_1.20.1     
##  [97] Matrix.utils_0.9.7            AnnotationDbi_1.46.0         
##  [99] memoise_1.1.0                 interactiveDisplayBase_1.22.0</code></pre>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/agxufd">shared library</a> on <a href="https://paperpile.com/">Paperpile</a>.</p>
<div class="references">
<p>Harrow, Jennifer, Adam Frankish, Jose M Gonzalez, Electra Tapanari, Mark Diekhans, Felix Kokocinski, Bronwen L Aken, et al. 2012. “GENCODE: the Reference Human Genome Annotation for the ENCODE Project.” <em>Genome Res.</em> 22 (9) (September): 1760–1774. doi:<a href="http://dx.doi.org/10.1101/gr.135350.111">10.1101/gr.135350.111</a>. <a href="http://dx.doi.org/10.1101/gr.135350.111">http://dx.doi.org/10.1101/gr.135350.111</a>.</p>
<p>Hubbard, T, D Barker, E Birney, G Cameron, Y Chen, L Clark, T Cox, et al. 2002. “The Ensembl Genome Database Project.” <em>Nucleic Acids Res.</em> 30 (1) (January): 38–41. <a href="https://www.ncbi.nlm.nih.gov/pubmed/11752248">https://www.ncbi.nlm.nih.gov/pubmed/11752248</a>.</p>
<p>Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” <em>Nat. Methods</em> 12 (2) (February): 115–121. doi:<a href="http://dx.doi.org/10.1038/nmeth.3252">10.1038/nmeth.3252</a>. <a href="http://dx.doi.org/10.1038/nmeth.3252">http://dx.doi.org/10.1038/nmeth.3252</a>.</p>
<p>Huerta-Cepas, Jaime, Damian Szklarczyk, Kristoffer Forslund, Helen Cook, Davide Heller, Mathias C Walter, Thomas Rattei, et al. 2016. “eggNOG 4.5: a Hierarchical Orthology Framework with Improved Functional Annotations for Eukaryotic, Prokaryotic and Viral Sequences.” <em>Nucleic Acids Res.</em> 44 (D1) (January): D286–93. doi:<a href="http://dx.doi.org/10.1093/nar/gkv1248">10.1093/nar/gkv1248</a>. <a href="http://dx.doi.org/10.1093/nar/gkv1248">http://dx.doi.org/10.1093/nar/gkv1248</a>.</p>
<p>Kent, W James, Charles W Sugnet, Terrence S Furey, Krishna M Roskin, Tom H Pringle, Alan M Zahler, and David Haussler. 2002. “The Human Genome Browser at UCSC.” <em>Genome Res.</em> 12 (6) (June): 996–1006. doi:<a href="http://dx.doi.org/10.1101/gr.229102">10.1101/gr.229102</a>. <a href="http://dx.doi.org/10.1101/gr.229102">http://dx.doi.org/10.1101/gr.229102</a>.</p>
<p>Lawrence, Michael, Robert Gentleman, and Vincent Carey. 2009. “rtracklayer: an R Package for Interfacing with Genome Browsers.” <em>Bioinformatics</em> 25 (14) (July): 1841–1842. doi:<a href="http://dx.doi.org/10.1093/bioinformatics/btp328">10.1093/bioinformatics/btp328</a>. <a href="http://dx.doi.org/10.1093/bioinformatics/btp328">http://dx.doi.org/10.1093/bioinformatics/btp328</a>.</p>
<p>Lawrence, Michael, Wolfgang Huber, Herv<span>é</span> Pag<span>è</span>s, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9 (8) (August): e1003118. doi:<a href="http://dx.doi.org/10.1371/journal.pcbi.1003118">10.1371/journal.pcbi.1003118</a>. <a href="http://dx.doi.org/10.1371/journal.pcbi.1003118">http://dx.doi.org/10.1371/journal.pcbi.1003118</a>.</p>
<p>Mi, Huaiyu, Xiaosong Huang, Anushya Muruganujan, Haiming Tang, Caitlin Mills, Diane Kang, and Paul D Thomas. 2017. “PANTHER Version 11: expanded Annotation Data from Gene Ontology and Reactome Pathways, and Data Analysis Tool Enhancements.” <em>Nucleic Acids Res.</em> 45 (D1) (January): D183–D189. doi:<a href="http://dx.doi.org/10.1093/nar/gkw1138">10.1093/nar/gkw1138</a>. <a href="http://dx.doi.org/10.1093/nar/gkw1138">http://dx.doi.org/10.1093/nar/gkw1138</a>.</p>
<p>Pruitt, Kim D, Tatiana Tatusova, and Donna R Maglott. 2007. “NCBI Reference Sequences (RefSeq): a Curated Non-Redundant Sequence Database of Genomes, Transcripts and Proteins.” <em>Nucleic Acids Res.</em> 35 (Database issue) (January): D61–5. doi:<a href="http://dx.doi.org/10.1093/nar/gkl842">10.1093/nar/gkl842</a>. <a href="http://dx.doi.org/10.1093/nar/gkl842">http://dx.doi.org/10.1093/nar/gkl842</a>.</p>
<p>Smith, Cynthia L, Judith A Blake, James A Kadin, Joel E Richardson, Carol J Bult, and Mouse Genome Database Group. 2018. “Mouse Genome Database (MGD)-2018: knowledgebase for the Laboratory Mouse.” <em>Nucleic Acids Res.</em> 46 (D1) (January): D836–D842. doi:<a href="http://dx.doi.org/10.1093/nar/gkx1006">10.1093/nar/gkx1006</a>. <a href="http://dx.doi.org/10.1093/nar/gkx1006">http://dx.doi.org/10.1093/nar/gkx1006</a>.</p>
<p>Stein, L, P Sternberg, R Durbin, J Thierry-Mieg, and J Spieth. 2001. “WormBase: network Access to the Genome and Biology of Caenorhabditis Elegans.” <em>Nucleic Acids Res.</em> 29 (1) (January): 82–86. doi:<a href="http://dx.doi.org/10.1093/nar/29.1.82">10.1093/nar/29.1.82</a>. <a href="https://www.ncbi.nlm.nih.gov/pubmed/11125056">https://www.ncbi.nlm.nih.gov/pubmed/11125056</a>.</p>
<p>Yates, Bethan, Bryony Braschi, Kristian A Gray, Ruth L Seal, Susan Tweedie, and Elspeth A Bruford. 2017. “Genenames.Org: the HGNC and VGNC Resources in 2017.” <em>Nucleic Acids Res.</em> 45 (D1) (January): D619–D625. doi:<a href="http://dx.doi.org/10.1093/nar/gkw1033">10.1093/nar/gkw1033</a>. <a href="http://dx.doi.org/10.1093/nar/gkw1033">http://dx.doi.org/10.1093/nar/gkw1033</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#the-s4-object-system">The S4 object system</a></li>
      <li>
<a href="#annotation-functions">Annotation functions</a><ul class="nav nav-pills nav-stacked">
<li><a href="#ensembl-annotations">Ensembl annotations</a></li>
      <li><a href="#gene-mapping-functions">Gene mapping functions</a></li>
      <li><a href="#parse-gtfgff-files">Parse GTF/GFF files</a></li>
      <li><a href="#gene-interconversion-functions">Gene interconversion functions</a></li>
      <li><a href="#query-external-annotation-databases">Query external annotation databases</a></li>
      </ul>
</li>
      <li><a href="#importexport-functions">Import/export functions</a></li>
      <li><a href="#data-functions">Data functions</a></li>
      <li>
<a href="#sanitization-functions">Sanitization functions</a><ul class="nav nav-pills nav-stacked">
<li><a href="#atomic-vectors">Atomic vectors</a></li>
      <li><a href="#data-frames">Data frames</a></li>
      <li><a href="#lists">Lists</a></li>
      </ul>
</li>
      <li><a href="#r-session-information">R session information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://steinbaugh.com/">Michael Steinbaugh</a>, <a href="https://acidgenomics.com/">Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
