<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>basejump • basejump</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="basejump">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.7.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>basejump</h1>
                        <h4 class="author">Michael Steinbaugh</h4>
            
            <h4 class="date">26 August 2018</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/steinbaugh/basejump/blob/master/vignettes/basejump.Rmd"><code>vignettes/basejump.Rmd</code></a></small>
      <div class="hidden name"><code>basejump.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>The basejump package is an infrastructure toolkit that extends the base functionality of Bioconductor <span class="citation">(Huber et al. 2015)</span>. The package leverages the S4 object system for object-oriented programming, and defines multiple additional generic functions for use in genomics research. basejump provides simple, user-friendly functions for the acquisition of genome annotations from multiple online databases, including native support for Ensembl <span class="citation">(T. Hubbard et al. 2002)</span> and websites supporting standard FASTA and GFF file formats. Consistent handling of sample metadata remains a challenge for many bioinformatic analyses, and basejump aims to address this by providing a suite of sanitization functions to help standardize these variable inputs. Additionally, interactive read/write operations in R can be cumbersome and non-trivial when working with multiple data objects; here we provide additional functions designed for interactive use that aim to reduce friction and provide consistent handling of multiple common file formats used in genomics research.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This vignette focuses on the most common user-facing functions that fall into the following categories:</p>
<ul>
<li>Annotation functions</li>
<li>Read/write functions</li>
<li>Data functions</li>
<li>Plot functions</li>
<li>Sanitization functions</li>
<li>Math and science functions</li>
</ul>
<p>There are additional function families whose functionality lie outside the scope of this vignette. Consult the <a href="https://steinbaugh.com/basejump">basejump</a> website or package documentation for more information of these types of functions, which are more technical in nature and predominantly developer-facing:</p>
<ul>
<li>R Markdown functions</li>
<li>Atomic vector functions</li>
<li>Coercion methods</li>
<li>Developer functions</li>
<li>Assert check functions</li>
</ul>
</div>
<div id="the-s4-object-system" class="section level1">
<h1 class="hasAnchor">
<a href="#the-s4-object-system" class="anchor"></a>The S4 object system</h1>
<p>The <a href="https://steinbaugh.com/basejump">basejump</a> package defines generics and methods for object-oriented programming with the S4 object system, which is used extensively by the <a href="https://bioconductor.org">Bioconductor</a> project. These resources describe the S4 object system in detail:</p>
<ul>
<li>
<a href="https://www.bioconductor.org/help/course-materials/2017/Zurich/S4-classes-and-methods.html">S4 classes and methods</a>, Martin Morgan and Hervé Pagès, 2017.</li>
<li>
<a href="https://bioconductor.org/packages/devel/bioc/vignettes/S4Vectors/inst/doc/S4QuickOverview.pdf">A quick overview of the S4 class system</a>, Hervé Pagès, 2016.</li>
<li>
<a href="http://adv-r.had.co.nz/S4.html">The S4 object system</a>, Hadley Wickham.</li>
<li>
<a href="http://kasperdanielhansen.github.io/genbioconductor">Bioconductor for Genomic Data Science</a>, Kasper Hansen, 2016.</li>
<li>
<a href="https://www.bioconductor.org/help/course-materials/2010/AdvancedR/S4InBioconductor.pdf">S4 System Development in Bioconductor</a>, Patrick Aboyoun, 2010.</li>
</ul>
<p>Here is a current list of S4 generics and/or methods exported by the package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exports &lt;-<span class="st"> </span><span class="kw">getNamespaceExports</span>(<span class="st">"basejump"</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">sort</span>() <span class="op">%&gt;%</span>
<span class="st">    </span>.[<span class="op">!</span><span class="kw">grepl</span>(<span class="st">"^</span><span class="ch">\\</span><span class="st">."</span>, .)]
isS4 &lt;-<span class="st"> </span><span class="kw">vapply</span>(
    <span class="dt">X =</span> exports,
    <span class="dt">FUN =</span> <span class="cf">function</span>(x) {
        <span class="kw">isS4</span>(<span class="kw">get</span>(x, <span class="dt">envir =</span> <span class="kw">asNamespace</span>(<span class="st">"basejump"</span>)))
    },
    <span class="dt">FUN.VALUE =</span> <span class="kw">logical</span>(1L)
)
exports[isS4]</code></pre></div>
<pre><code>##  [1] "aggregateFeatures"         "aggregateReplicates"      
##  [3] "broadClass"                "camel"                    
##  [5] "coerce"                    "collapseToString"         
##  [7] "convertGenesToSymbols"     "convertSymbolsToGenes"    
##  [9] "convertTranscriptsToGenes" "counts"                   
## [11] "detectOrganism"            "dotted"                   
## [13] "fixNA"                     "flatFiles"                
## [15] "foldChangeToLogRatio"      "gene2symbol"              
## [17] "geometricMean"             "interestingGroups"        
## [19] "interestingGroups&lt;-"       "logRatioToFoldChange"     
## [21] "mcols"                     "metrics"                  
## [23] "plotCorrelationHeatmap"    "plotGene"                 
## [25] "plotHeatmap"               "plotQC"                   
## [27] "plotQuantileHeatmap"       "removeNA"                 
## [29] "sampleData"                "sampleData&lt;-"             
## [31] "sampleNames"               "selectSamples"            
## [33] "seqnames"                  "snake"                    
## [35] "stripTranscriptVersions"   "SummarizedExperiment"     
## [37] "uniteInterestingGroups"    "upperCamel"</code></pre>
<p>In <a href="https://www.r-project.org">R</a>, you can check whether a function is using the S4 object system with the <code><a href="http://www.rdocumentation.org/packages/base/topics/isS4">base::isS4()</a></code> function. Additionally, <code><a href="http://www.rdocumentation.org/packages/methods/topics/showMethods">methods::showMethods()</a></code> and <code><a href="http://www.rdocumentation.org/packages/methods/topics/getMethod">methods::getMethod()</a></code>. In contrast, to obtain information on functions using the <a href="http://adv-r.had.co.nz/S3.html">S3</a> object system, use <code><a href="http://www.rdocumentation.org/packages/utils/topics/methods">utils::methods()</a></code> instead.</p>
</div>
<div id="annotation-functions" class="section level1">
<h1 class="hasAnchor">
<a href="#annotation-functions" class="anchor"></a>Annotation functions</h1>
<p>Obtaining versioned genome annotations quickly and reliably from online databases such as <a href="https://www.ensembl.org">Ensembl</a> <span class="citation">(T. Hubbard et al. 2002)</span>, <a href="https://www.gencodegenes.org">GENCODE</a> <span class="citation">(Harrow et al. 2012)</span>, <a href="https://www.ncbi.nlm.nih.gov/refseq">RefSeq</a> <span class="citation">(Pruitt, Tatusova, and Maglott 2007)</span>, and the <a href="http://genome.ucsc.edu">UCSC Genome Browser</a> <span class="citation">(Kent et al. 2002)</span> remains overly challenging. <a href="https://steinbaugh.com/basejump">basejump</a> aims to help address this issue by providing native gene- and transcript-level annotation support by helping users interface with <a href="https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a> and <a href="https://doi.org/doi:10.18129/B9.bioc.ensembldb">ensembldb</a>, along with a robust set of tools to parse general feature format (GFF) and general transfer format (GTF; a.k.a. GFFv2) annotation files. GFF file parsing is handled internally by the <a href="https://bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a> <span class="citation">(Lawrence, Gentleman, and Carey 2009)</span> package. Genome annotations are returned as <code>GRanges</code> objects, defined in the <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a> <span class="citation">(Lawrence et al. 2013)</span> package, which allows for easy access to position, chromosome, strand, and additional metadata saved in the <code>mcols</code> slot of the object. <code>GRanges</code> can be coerced to a standard <code>data.frame</code> using the <code><a href="http://www.rdocumentation.org/packages/BiocGenerics/topics/as.data.frame">BiocGenerics::as.data.frame()</a></code> function.</p>
<div id="ensembl-annotations" class="section level2">
<h2 class="hasAnchor">
<a href="#ensembl-annotations" class="anchor"></a>Ensembl annotations</h2>
<p><code><a href="../reference/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl()</a></code> supports multiple genome builds, and versioned releases from <a href="https://www.ensembl.org">Ensembl</a>. By default, the function will use the latest release version available from <a href="https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a> and <a href="https://doi.org/doi:10.18129/B9.bioc.ensembldb">ensembldb</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grch38 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl</a></span>(<span class="st">"Homo sapiens"</span>)</code></pre></div>
<pre><code>## Making GRanges from Ensembl with AnnotationHub 2.12.0 (2018-04-30)</code></pre>
<pre><code>## Ensembl 92 EnsDb for Homo Sapiens</code></pre>
<pre><code>## id: "AH60977"
## organism: "Homo sapiens"
## build: "GRCh38"
## release: "92"
## format: "genes"</code></pre>
<pre><code>## Standardizing the metadata columns</code></pre>
<pre><code>## Defining broad class using: geneName, geneBiotype, seqnames</code></pre>
<pre><code>## Sorting ranges by geneID</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(grch38)</code></pre></div>
<pre><code>## [1] "GRanges object with 65256 ranges and 7 metadata columns"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>(grch38)</code></pre></div>
<pre><code>## Formal class 'GRanges' [package "GenomicRanges"] with 7 slots
##   ..@ seqnames       :Formal class 'Rle' [package "S4Vectors"] with 4 slots
##   ..@ ranges         :Formal class 'IRanges' [package "IRanges"] with 6 slots
##   ..@ strand         :Formal class 'Rle' [package "S4Vectors"] with 4 slots
##   ..@ seqinfo        :Formal class 'Seqinfo' [package "GenomeInfoDb"] with 4 slots
##   ..@ elementMetadata:Formal class 'DataFrame' [package "S4Vectors"] with 6 slots
##   ..@ elementType    : chr "ANY"
##   ..@ metadata       : list()</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">names</span>(grch38))</code></pre></div>
<pre><code>## [1] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419" "ENSG00000000457"
## [5] "ENSG00000000460" "ENSG00000000938"</code></pre>
<p>Data inside a <code>GRanges</code> object can be accessed with a number of functions defined in the <a href="https://bioconductor.org/packages/release/bioc/html/GenomeInfoDb.html">GenomeInfoDb</a> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">GenomeInfoDb<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomeInfoDb/topics/seqinfo">seqnames</a></span>(grch38)</code></pre></div>
<pre><code>## factor-Rle of length 65256 with 48641 runs
##   Lengths:      2      1      4      3 ...      1      1      2      1
##   Values :      X     20      1      6 ...      1     12      6     19
## Levels(423): X 20 1 6 3 7 ... LRG_311 LRG_721 LRG_741 LRG_763 LRG_93</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">GenomeInfoDb<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomeInfoDb/topics/seqinfo">seqinfo</a></span>(grch38)</code></pre></div>
<pre><code>## Seqinfo object with 423 sequences from GRCh38 genome:
##   seqnames seqlengths isCircular genome
##   X         156040895      FALSE GRCh38
##   20         64444167      FALSE GRCh38
##   1         248956422      FALSE GRCh38
##   6         170805979      FALSE GRCh38
##   3         198295559      FALSE GRCh38
##   ...             ...        ...    ...
##   LRG_311      115492      FALSE GRCh38
##   LRG_721       33396      FALSE GRCh38
##   LRG_741      231167      FALSE GRCh38
##   LRG_763      176286      FALSE GRCh38
##   LRG_93        22459      FALSE GRCh38</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">IRanges<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/IRanges/topics/range-squeezers">ranges</a></span>(grch38)</code></pre></div>
<pre><code>## IRanges object with 65256 ranges and 0 metadata columns:
##                       start       end     width
##                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   ENSG00000000003 100627109 100639991     12883
##   ENSG00000000005 100584802 100599885     15084
##   ENSG00000000419  50934867  50958555     23689
##   ENSG00000000457 169849631 169894267     44637
##   ENSG00000000460 169662007 169854080    192074
##               ...       ...       ...       ...
##           LRG_995  77948402  77979205     30804
##           LRG_996  56080025  56103507     23483
##           LRG_997 117288367 117425855    137489
##           LRG_998  41934933  42048894    113962
##           LRG_999  42268537  42295796     27260</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">BiocGenerics<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/BiocGenerics/topics/strand">strand</a></span>(grch38)</code></pre></div>
<pre><code>## factor-Rle of length 65256 with 32450 runs
##   Lengths: 1 1 2 1 1 1 2 1 1 1 1 4 3 1 1 ... 1 4 1 1 1 3 1 1 1 1 1 3 1 2 1
##   Values : - + - + - + - + - + - + - + - ... + - + - + - + - + - + - + - +
## Levels(3): + - *</code></pre>
<p>The <code>GRanges</code> class is a powerful container for genomic coordinates and metadata. For example, we can easily the gene counts per chromosome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grch38 <span class="op">%&gt;%</span>
<span class="st">    </span>GenomeInfoDb<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomeInfoDb/topics/seqinfo">seqnames</a></span>(.) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">table</span>() <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">sort</span>(<span class="dt">decreasing =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span>
<span class="st">    </span><span class="kw">head</span>(<span class="dt">n =</span> <span class="dv">24</span>) <span class="op">%&gt;%</span>
<span class="st">    </span>.[<span class="kw">sort</span>(<span class="kw">names</span>(.))]</code></pre></div>
<pre><code>## .
##    1   10   11   12   13   14   15   16   17   18   19    2   20   21   22 
## 5349 2236 3314 2990 1327 2237 2183 2529 3058 1185 2996 4048 1414  845 1367 
##    3    4    5    6    7    8    9    X    Y 
## 3064 2521 2887 2899 2926 2390 2292 2410  518</code></pre>
<p>Gene metadata is contained in the <code>mcols</code> slot of the object, using a <code>DataFrame</code> internally. Here’a the current list of gene-level annotation columns returned from <code><a href="../reference/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl()</a></code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">S4Vectors<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class">mcols</a></span>(grch38) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">colnames</span>()</code></pre></div>
<pre><code>## [1] "broadClass"     "description"    "entrezID"       "geneBiotype"   
## [5] "geneID"         "geneName"       "seqCoordSystem"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">S4Vectors<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class">mcols</a></span>(grch38) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>()</code></pre></div>
<pre><code>## Formal class 'DataFrame' [package "S4Vectors"] with 6 slots
##   ..@ rownames       : NULL
##   ..@ nrows          : int 65256
##   ..@ listData       :List of 7
##   .. ..$ broadClass    : Factor w/ 8 levels "coding","ig",..: 1 1 1 1 1 1 1 1 1 1 ...
##   .. .. ..- attr(*, "names")= chr [1:65256] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419" "ENSG00000000457" ...
##   .. ..$ description   : Factor w/ 37499 levels "1-acylglycerol-3-phosphate O-acyltransferase 1 [Source:HGNC Symbol;Acc:HGNC:324]",..: 33719 33552 6386 30004 3887 8291 4876 943 9310 20214 ...
##   .. ..$ entrezID      :List of 65256
##   .. ..$ geneBiotype   : Factor w/ 47 levels "3prime_overlapping_ncRNA",..: 23 23 23 23 23 23 23 23 23 23 ...
##   .. ..$ geneID        : chr [1:65256] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419" "ENSG00000000457" ...
##   .. ..$ geneName      : Factor w/ 58334 levels "A1BG","A1BG-AS1",..: 55560 54793 26480 50775 22680 28411 24138 28838 29147 40197 ...
##   .. ..$ seqCoordSystem: Factor w/ 3 levels "chromosome","lrg",..: 1 1 1 1 1 1 1 1 1 1 ...
##   ..@ elementType    : chr "ANY"
##   ..@ elementMetadata: NULL
##   ..@ metadata       : list()</code></pre>
<p>Transcript-level annotations are also supported, using the <code>format = "transcripts"</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">transcripts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl</a></span>(
    <span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>,
    <span class="dt">format =</span> <span class="st">"transcripts"</span>
)</code></pre></div>
<pre><code>## Making GRanges from Ensembl with AnnotationHub 2.12.0 (2018-04-30)</code></pre>
<pre><code>## Ensembl 92 EnsDb for Homo Sapiens</code></pre>
<pre><code>## id: "AH60977"
## organism: "Homo sapiens"
## build: "GRCh38"
## release: "92"
## format: "transcripts"</code></pre>
<pre><code>## Standardizing the metadata columns</code></pre>
<pre><code>## Defining broad class using: geneName, transcriptBiotype, seqnames</code></pre>
<pre><code>## Sorting ranges by transcriptID</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(transcripts)</code></pre></div>
<pre><code>## [1] "GRanges object with 225358 ranges and 13 metadata columns"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">S4Vectors<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/S4Vectors/topics/Vector-class">mcols</a></span>(transcripts) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>()</code></pre></div>
<pre><code>## Formal class 'DataFrame' [package "S4Vectors"] with 6 slots
##   ..@ rownames       : NULL
##   ..@ nrows          : int 225358
##   ..@ listData       :List of 13
##   .. ..$ broadClass            : Factor w/ 9 levels "coding","decaying",..: 1 1 1 1 1 1 1 1 1 1 ...
##   .. .. ..- attr(*, "names")= chr [1:225358] "ENST00000000233" "ENST00000000412" "ENST00000000442" "ENST00000001008" ...
##   .. ..$ description           : Factor w/ 37499 levels "1-acylglycerol-3-phosphate O-acyltransferase 1 [Source:HGNC Symbol;Acc:HGNC:324]",..: 660 15606 7212 8412 5519 19360 943 6655 10155 30158 ...
##   .. ..$ entrezID              :List of 225358
##   .. ..$ geneBiotype           : Factor w/ 47 levels "3prime_overlapping_ncRNA",..: 23 23 23 23 23 23 23 23 23 23 ...
##   .. ..$ geneID                : Factor w/ 65256 levels "ENSG00000000003",..: 46 33 13473 52 35 42 8 37 25 15 ...
##   .. ..$ geneName              : Factor w/ 58334 levels "A1BG","A1BG-AS1",..: 21098 35853 27506 28456 25628 39996 28838 25796 31016 50913 ...
##   .. ..$ seqCoordSystem        : Factor w/ 3 levels "chromosome","lrg",..: 1 1 1 1 1 1 1 1 1 1 ...
##   .. ..$ transcriptBiotype     : Factor w/ 50 levels "3prime_overlapping_ncRNA",..: 25 25 25 25 25 25 25 25 25 25 ...
##   .. ..$ transcriptCdsSeqEnd   : int [1:225358] 127591299 8946404 64315966 2803258 72147834 37248350 143511634 90019341 11400005 50188115 ...
##   .. ..$ transcriptCdsSeqStart : int [1:225358] 127588499 8941818 64307180 2795140 72132227 37231706 143495707 90006335 11399082 50159623 ...
##   .. ..$ transcriptID          : chr [1:225358] "ENST00000000233" "ENST00000000412" "ENST00000000442" "ENST00000001008" ...
##   .. ..$ transcriptName        : chr [1:225358] "ENST00000000233" "ENST00000000412" "ENST00000000442" "ENST00000001008" ...
##   .. ..$ transcriptSupportLevel: int [1:225358] 1 1 1 1 1 1 1 2 1 1 ...
##   ..@ elementType    : chr "ANY"
##   ..@ elementMetadata: NULL
##   ..@ metadata       : list()</code></pre>
<div id="versioned-releases" class="section level3">
<h3 class="hasAnchor">
<a href="#versioned-releases" class="anchor"></a>Versioned releases</h3>
<p>Versioned releases are supported by the “<code>release</code>” argument, which currently supports back to <a href="https://www.ensembl.org">Ensembl</a> 87. If an older <a href="https://www.ensembl.org">Ensembl</a> release is required, resort to using a GFF file or browse the <a href="https://bioconductor.org">Bioconductor</a> website to see if a suitable annotation database package is available.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grch38_<span class="dv">90</span> &lt;-<span class="st"> </span><span class="kw"><a href="../reference/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl</a></span>(
    <span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>,
    <span class="dt">release =</span> <span class="dv">90</span>
)</code></pre></div>
<pre><code>## Making GRanges from Ensembl with AnnotationHub 2.12.0 (2018-04-30)</code></pre>
<pre><code>## Ensembl 90 EnsDb for Homo Sapiens</code></pre>
<pre><code>## id: "AH57757"
## organism: "Homo sapiens"
## build: "GRCh38"
## release: "90"
## format: "genes"</code></pre>
<pre><code>## Standardizing the metadata columns</code></pre>
<pre><code>## Defining broad class using: geneName, geneBiotype, seqnames</code></pre>
<pre><code>## Sorting ranges by geneID</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(grch38_<span class="dv">90</span>)</code></pre></div>
<pre><code>## [1] "GRanges object with 64661 ranges and 7 metadata columns"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">GenomeInfoDb<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomeInfoDb/topics/seqinfo">seqinfo</a></span>(grch38_<span class="dv">90</span>)</code></pre></div>
<pre><code>## Seqinfo object with 388 sequences from GRCh38 genome:
##   seqnames seqlengths isCircular genome
##   X         156040895      FALSE GRCh38
##   20         64444167      FALSE GRCh38
##   1         248956422      FALSE GRCh38
##   6         170805979      FALSE GRCh38
##   3         198295559      FALSE GRCh38
##   ...             ...        ...    ...
##   LRG_239      114904      FALSE GRCh38
##   LRG_311      115492      FALSE GRCh38
##   LRG_721       33396      FALSE GRCh38
##   LRG_741      231167      FALSE GRCh38
##   LRG_93        22459      FALSE GRCh38</code></pre>
</div>
<div id="grch37" class="section level3">
<h3 class="hasAnchor">
<a href="#grch37" class="anchor"></a>GRCh37</h3>
<p>The legacy <a href="https://grch37.ensembl.org">Ensembl GRCh37 genome build</a> (release 75; also known as <a href="https://genome.ucsc.edu/cgi-bin/hgGateway?db=hg19">UCSC hg19</a>) is also supported, but is no longer recommended for use in new analyses. Internally, support for <a href="https://grch37.ensembl.org">GRCh37</a> is provided by the <a href="https://bioconductor.org/packages/release/data/annotation/html/EnsDb.Hsapiens.v75.html">EnsDb.Hsapiens.v75</a> annotation database package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grch37 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl</a></span>(
    <span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>,
    <span class="dt">build =</span> <span class="st">"GRCh37"</span>
)</code></pre></div>
<pre><code>## id: "EnsDb.Hsapiens.v75"
## organism: "Homo sapiens"
## build: "GRCh37"
## release: "75"
## format: "genes"</code></pre>
<pre><code>## Standardizing the metadata columns</code></pre>
<pre><code>## Defining broad class using: geneName, geneBiotype, seqnames</code></pre>
<pre><code>## Sorting ranges by geneID</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(grch37)</code></pre></div>
<pre><code>## [1] "GRanges object with 64102 ranges and 6 metadata columns"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">GenomeInfoDb<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/GenomeInfoDb/topics/seqinfo">seqinfo</a></span>(grch37)</code></pre></div>
<pre><code>## Seqinfo object with 273 sequences from GRCh37 genome:
##   seqnames seqlengths isCircular genome
##   X         155270560      FALSE GRCh37
##   20         63025520      FALSE GRCh37
##   1         249250621      FALSE GRCh37
##   6         171115067      FALSE GRCh37
##   3         198022430      FALSE GRCh37
##   ...             ...        ...    ...
##   LRG_187       13555      FALSE GRCh37
##   LRG_239      114904      FALSE GRCh37
##   LRG_311      115492      FALSE GRCh37
##   LRG_415       44382      FALSE GRCh37
##   LRG_93        22459      FALSE GRCh37</code></pre>
</div>
<div id="data-frame-return" class="section level3">
<h3 class="hasAnchor">
<a href="#data-frame-return" class="anchor"></a>Data frame return</h3>
<p>For convenience, if you prefer to return annotations as a <code>data.frame</code> instead of <code>GRanges</code>, we provide the <code><a href="../reference/annotable.html">annotable()</a></code> function. Internally, this calls the <code><a href="../reference/makeGRangesFromEnsembl.html">makeGRangesFromEnsembl()</a></code> function but automatically coerces the return to a standard <code>data.frame</code>, with the identifiers as the row names. For quick access to current <a href="https://www.ensembl.org">Ensembl</a> annotations in <code>tibble</code> format, we also recommend the <a href="https://github.com/stephenturner/annotables">annotables</a> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/annotable.html">annotable</a></span>(<span class="st">"Homo sapiens"</span>)
<span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>(genes)</code></pre></div>
<pre><code>## Observations: 65,256
## Variables: 12
## $ seqnames       &lt;fct&gt; X, X, 20, 1, 1, 1, 1, 6, 6, 6, 1, 1, X, 6, 3, 7...
## $ start          &lt;int&gt; 100627109, 100584802, 50934867, 169849631, 1696...
## $ end            &lt;int&gt; 100639991, 100599885, 50958555, 169894267, 1698...
## $ width          &lt;int&gt; 12883, 15084, 23689, 44637, 192074, 23214, 9562...
## $ strand         &lt;fct&gt; -, +, -, -, +, -, +, -, -, +, -, +, -, +, +, +,...
## $ broadClass     &lt;fct&gt; coding, coding, coding, coding, coding, coding,...
## $ description    &lt;fct&gt; tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:1185...
## $ entrezID       &lt;I(list)&gt;  7105, 64102,  8813, 57147, 55732,  2268,  ...
## $ geneBiotype    &lt;fct&gt; protein_coding, protein_coding, protein_coding,...
## $ geneID         &lt;chr&gt; "ENSG00000000003", "ENSG00000000005", "ENSG0000...
## $ geneName       &lt;fct&gt; TSPAN6, TNMD, DPM1, SCYL3, C1orf112, FGR, CFH, ...
## $ seqCoordSystem &lt;fct&gt; chromosome, chromosome, chromosome, chromosome,...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">transcripts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/annotable.html">annotable</a></span>(<span class="st">"Homo sapiens"</span>)
<span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>(transcripts)</code></pre></div>
<pre><code>## Observations: 65,256
## Variables: 12
## $ seqnames       &lt;fct&gt; X, X, 20, 1, 1, 1, 1, 6, 6, 6, 1, 1, X, 6, 3, 7...
## $ start          &lt;int&gt; 100627109, 100584802, 50934867, 169849631, 1696...
## $ end            &lt;int&gt; 100639991, 100599885, 50958555, 169894267, 1698...
## $ width          &lt;int&gt; 12883, 15084, 23689, 44637, 192074, 23214, 9562...
## $ strand         &lt;fct&gt; -, +, -, -, +, -, +, -, -, +, -, +, -, +, +, +,...
## $ broadClass     &lt;fct&gt; coding, coding, coding, coding, coding, coding,...
## $ description    &lt;fct&gt; tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:1185...
## $ entrezID       &lt;I(list)&gt;  7105, 64102,  8813, 57147, 55732,  2268,  ...
## $ geneBiotype    &lt;fct&gt; protein_coding, protein_coding, protein_coding,...
## $ geneID         &lt;chr&gt; "ENSG00000000003", "ENSG00000000005", "ENSG0000...
## $ geneName       &lt;fct&gt; TSPAN6, TNMD, DPM1, SCYL3, C1orf112, FGR, CFH, ...
## $ seqCoordSystem &lt;fct&gt; chromosome, chromosome, chromosome, chromosome,...</code></pre>
<p>Note that <code>GRanges</code> objects can also be coerced to a <code>data.frame</code> using the <code>as.data.frame()</code> coercion method.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">grch38 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>()</code></pre></div>
<pre><code>## Observations: 65,256
## Variables: 12
## $ seqnames       &lt;fct&gt; X, X, 20, 1, 1, 1, 1, 6, 6, 6, 1, 1, X, 6, 3, 7...
## $ start          &lt;int&gt; 100627109, 100584802, 50934867, 169849631, 1696...
## $ end            &lt;int&gt; 100639991, 100599885, 50958555, 169894267, 1698...
## $ width          &lt;int&gt; 12883, 15084, 23689, 44637, 192074, 23214, 9562...
## $ strand         &lt;fct&gt; -, +, -, -, +, -, +, -, -, +, -, +, -, +, +, +,...
## $ broadClass     &lt;fct&gt; coding, coding, coding, coding, coding, coding,...
## $ description    &lt;fct&gt; tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:1185...
## $ entrezID       &lt;I(list)&gt;  7105, 64102,  8813, 57147, 55732,  2268,  ...
## $ geneBiotype    &lt;fct&gt; protein_coding, protein_coding, protein_coding,...
## $ geneID         &lt;chr&gt; "ENSG00000000003", "ENSG00000000005", "ENSG0000...
## $ geneName       &lt;fct&gt; TSPAN6, TNMD, DPM1, SCYL3, C1orf112, FGR, CFH, ...
## $ seqCoordSystem &lt;fct&gt; chromosome, chromosome, chromosome, chromosome,...</code></pre>
</div>
</div>
<div id="gene-mapping-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#gene-mapping-functions" class="anchor"></a>Gene mapping functions</h2>
<p>Gene-to-symbol and transcript-to-gene mappings can be easily acquired with the <code><a href="../reference/gene2symbol.html">gene2symbol()</a></code> and <code><a href="../reference/defunct.html">tx2gene()</a></code> functions. Both of these functions return a <code>data.frame</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene2symbol &lt;-<span class="st"> </span><span class="kw"><a href="../reference/makeGene2symbolFromEnsembl.html">makeGene2symbolFromEnsembl</a></span>(<span class="st">"Homo sapiens"</span>)
<span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>(gene2symbol)</code></pre></div>
<pre><code>## Observations: 65,256
## Variables: 2
## $ geneID   &lt;chr&gt; "ENSG00000000003", "ENSG00000000005", "ENSG0000000041...
## $ geneName &lt;chr&gt; "TSPAN6", "TNMD", "DPM1", "SCYL3", "C1orf112", "FGR",...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tx2gene &lt;-<span class="st"> </span><span class="kw"><a href="../reference/makeTx2geneFromEnsembl.html">makeTx2geneFromEnsembl</a></span>(<span class="st">"Homo sapiens"</span>)
<span class="kw"><a href="http://www.rdocumentation.org/packages/tibble/topics/glimpse">glimpse</a></span>(tx2gene)</code></pre></div>
<pre><code>## Observations: 225,358
## Variables: 2
## $ transcriptID &lt;chr&gt; "ENST00000000233", "ENST00000000412", "ENST000000...
## $ geneID       &lt;chr&gt; "ENSG00000004059", "ENSG00000003056", "ENSG000001...</code></pre>
</div>
</div>
<div id="r-session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#r-session-information" class="anchor"></a>R session information</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sessioninfo<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/sessioninfo/topics/session_info">session_info</a></span>()</code></pre></div>
<pre><code>## ─ Session info ──────────────────────────────────────────────────────────
##  setting  value                       
##  version  R version 3.5.1 (2018-07-02)
##  os       macOS High Sierra 10.13.6   
##  system   x86_64, darwin15.6.0        
##  ui       X11                         
##  language (EN)                        
##  collate  en_US.UTF-8                 
##  tz       America/New_York            
##  date     2018-08-26                  
## 
## ─ Packages ──────────────────────────────────────────────────────────────
##  package                * version   date       source         
##  AnnotationDbi            1.42.1    2018-05-08 Bioconductor   
##  AnnotationFilter         1.4.0     2018-05-01 Bioconductor   
##  AnnotationHub            2.12.0    2018-05-01 Bioconductor   
##  assertive.base           0.0-7     2016-12-30 CRAN (R 3.5.0) 
##  assertive.code           0.0-1     2015-10-06 CRAN (R 3.5.0) 
##  assertive.data           0.0-1     2015-10-06 CRAN (R 3.5.0) 
##  assertive.files          0.0-2     2016-05-10 CRAN (R 3.5.0) 
##  assertive.numbers        0.0-2     2016-05-09 CRAN (R 3.5.0) 
##  assertive.properties     0.0-4     2016-12-30 CRAN (R 3.5.0) 
##  assertive.sets           0.0-3     2016-12-30 CRAN (R 3.5.0) 
##  assertive.strings        0.0-3     2016-05-10 CRAN (R 3.5.0) 
##  assertive.types          0.0-3     2016-12-30 CRAN (R 3.5.0) 
##  assertthat               0.2.0     2017-04-11 CRAN (R 3.5.0) 
##  backports                1.1.2     2017-12-13 CRAN (R 3.5.0) 
##  basejump               * 0.7.3     2018-08-26 Bioconductor   
##  bindr                    0.1.1     2018-03-13 CRAN (R 3.5.0) 
##  bindrcpp               * 0.2.2     2018-03-29 CRAN (R 3.5.0) 
##  Biobase                * 2.40.0    2018-05-01 Bioconductor   
##  BiocGenerics           * 0.26.0    2018-05-01 Bioconductor   
##  BiocInstaller            1.30.0    2018-05-01 Bioconductor   
##  BiocParallel           * 1.14.2    2018-07-08 Bioconductor   
##  BiocStyle              * 2.8.2     2018-05-30 Bioconductor   
##  biomaRt                  2.36.1    2018-05-24 Bioconductor   
##  Biostrings               2.48.0    2018-05-01 Bioconductor   
##  bit                      1.1-14    2018-05-29 CRAN (R 3.5.0) 
##  bit64                    0.9-7     2017-05-08 CRAN (R 3.5.0) 
##  bitops                   1.0-6     2013-08-17 CRAN (R 3.5.0) 
##  blob                     1.1.1     2018-03-25 CRAN (R 3.5.0) 
##  bookdown                 0.7       2018-02-18 CRAN (R 3.5.0) 
##  clisymbols               1.2.0     2017-05-21 CRAN (R 3.5.0) 
##  codetools                0.2-15    2016-10-05 CRAN (R 3.5.1) 
##  colorspace               1.3-2     2016-12-14 CRAN (R 3.5.0) 
##  commonmark               1.5       2018-04-28 CRAN (R 3.5.0) 
##  cowplot                  0.9.3     2018-07-15 CRAN (R 3.5.1) 
##  crayon                   1.3.4     2017-09-16 CRAN (R 3.5.0) 
##  curl                     3.2       2018-03-28 CRAN (R 3.5.0) 
##  DBI                      1.0.0     2018-05-02 CRAN (R 3.5.0) 
##  DelayedArray           * 0.6.5     2018-08-15 Bioconductor   
##  desc                     1.2.0     2018-05-01 CRAN (R 3.5.0) 
##  digest                   0.6.15    2018-01-28 CRAN (R 3.5.0) 
##  dplyr                    0.7.6     2018-06-29 CRAN (R 3.5.0) 
##  EnsDb.Hsapiens.v75       2.99.0    2018-04-25 Bioconductor   
##  ensembldb                2.4.1     2018-05-07 Bioconductor   
##  evaluate                 0.11      2018-07-17 CRAN (R 3.5.1) 
##  fs                       1.2.5     2018-07-30 CRAN (R 3.5.1) 
##  GenomeInfoDb           * 1.16.0    2018-05-01 Bioconductor   
##  GenomeInfoDbData         1.1.0     2018-04-25 Bioconductor   
##  GenomicAlignments        1.16.0    2018-05-01 Bioconductor   
##  GenomicFeatures          1.32.2    2018-08-13 Bioconductor   
##  GenomicRanges          * 1.32.6    2018-07-20 Bioconductor   
##  ggplot2                  3.0.0     2018-07-03 CRAN (R 3.5.0) 
##  glue                     1.3.0     2018-07-17 CRAN (R 3.5.1) 
##  grr                      0.9.5     2016-08-26 CRAN (R 3.5.0) 
##  gtable                   0.2.0     2016-02-26 CRAN (R 3.5.0) 
##  hms                      0.4.2     2018-03-10 CRAN (R 3.5.0) 
##  htmltools                0.3.6     2017-04-28 CRAN (R 3.5.0) 
##  httpuv                   1.4.5     2018-07-19 CRAN (R 3.5.0) 
##  httr                     1.3.1     2017-08-20 CRAN (R 3.5.0) 
##  interactiveDisplayBase   1.18.0    2018-05-01 Bioconductor   
##  IRanges                * 2.14.10   2018-05-16 Bioconductor   
##  jsonlite                 1.5       2017-06-01 CRAN (R 3.5.0) 
##  knitr                    1.20      2018-02-20 CRAN (R 3.5.0) 
##  later                    0.7.3     2018-06-08 CRAN (R 3.5.0) 
##  lattice                  0.20-35   2017-03-25 CRAN (R 3.5.1) 
##  lazyeval                 0.2.1     2017-10-29 CRAN (R 3.5.0) 
##  magrittr                 1.5       2014-11-22 CRAN (R 3.5.0) 
##  MASS                     7.3-50    2018-04-30 CRAN (R 3.5.1) 
##  Matrix                   1.2-14    2018-04-13 CRAN (R 3.5.1) 
##  Matrix.utils             0.9.7     2018-04-19 CRAN (R 3.5.0) 
##  matrixStats            * 0.54.0    2018-07-23 CRAN (R 3.5.0) 
##  memoise                  1.1.0     2017-04-21 CRAN (R 3.5.0) 
##  mime                     0.5       2016-07-07 CRAN (R 3.5.0) 
##  munsell                  0.5.0     2018-06-12 CRAN (R 3.5.0) 
##  pbapply                  1.3-4     2018-01-10 CRAN (R 3.5.0) 
##  pheatmap                 1.0.10    2018-05-19 CRAN (R 3.5.0) 
##  pillar                   1.3.0     2018-07-14 CRAN (R 3.5.0) 
##  pkgconfig                2.0.2     2018-08-16 CRAN (R 3.5.0) 
##  pkgdown                  1.1.0     2018-06-02 CRAN (R 3.5.0) 
##  plyr                     1.8.4     2016-06-08 CRAN (R 3.5.0) 
##  prettyunits              1.0.2     2015-07-13 CRAN (R 3.5.0) 
##  progress                 1.2.0     2018-06-14 CRAN (R 3.5.0) 
##  promises                 1.0.1     2018-04-13 CRAN (R 3.5.0) 
##  ProtGenerics             1.12.0    2018-05-01 Bioconductor   
##  purrr                    0.2.5     2018-05-29 CRAN (R 3.5.0) 
##  R.methodsS3              1.7.1     2016-02-16 CRAN (R 3.5.0) 
##  R.oo                     1.22.0    2018-04-22 CRAN (R 3.5.0) 
##  R.utils                  2.6.0     2017-11-05 CRAN (R 3.5.0) 
##  R6                       2.2.2     2017-06-17 CRAN (R 3.5.0) 
##  RColorBrewer             1.1-2     2014-12-07 CRAN (R 3.5.0) 
##  Rcpp                     0.12.18   2018-07-23 CRAN (R 3.5.0) 
##  RCurl                    1.95-4.11 2018-07-15 cran (@1.95-4.)
##  readr                    1.1.1     2017-05-16 CRAN (R 3.5.0) 
##  rlang                    0.2.2     2018-08-16 CRAN (R 3.5.0) 
##  rmarkdown                1.10      2018-06-11 CRAN (R 3.5.0) 
##  roxygen2                 6.1.0     2018-07-27 CRAN (R 3.5.1) 
##  rprojroot                1.3-2     2018-01-03 CRAN (R 3.5.0) 
##  Rsamtools                1.32.2    2018-07-03 Bioconductor   
##  RSQLite                  2.1.1     2018-05-06 CRAN (R 3.5.0) 
##  rstudioapi               0.7       2017-09-07 CRAN (R 3.5.0) 
##  rtracklayer              1.40.5    2018-08-20 Bioconductor   
##  S4Vectors              * 0.18.3    2018-06-08 Bioconductor   
##  scales                   1.0.0     2018-08-09 CRAN (R 3.5.0) 
##  sessioninfo              1.0.0     2017-06-21 CRAN (R 3.5.0) 
##  shiny                    1.1.0     2018-05-17 CRAN (R 3.5.0) 
##  stringi                  1.2.4     2018-07-20 CRAN (R 3.5.0) 
##  stringr                  1.3.1     2018-05-10 CRAN (R 3.5.0) 
##  SummarizedExperiment   * 1.10.1    2018-05-11 Bioconductor   
##  tibble                   1.4.2     2018-01-22 CRAN (R 3.5.0) 
##  tidyr                    0.8.1     2018-05-18 CRAN (R 3.5.0) 
##  tidyselect               0.2.4     2018-02-26 CRAN (R 3.5.0) 
##  withr                    2.1.2     2018-03-15 CRAN (R 3.5.0) 
##  xfun                     0.3       2018-07-06 CRAN (R 3.5.1) 
##  XML                      3.98-1.16 2018-08-19 CRAN (R 3.5.1) 
##  xml2                     1.2.0     2018-01-24 CRAN (R 3.5.0) 
##  xtable                   1.8-2     2016-02-05 CRAN (R 3.5.0) 
##  XVector                  0.20.0    2018-05-01 Bioconductor   
##  yaml                     2.2.0     2018-07-25 CRAN (R 3.5.0) 
##  zlibbioc                 1.26.0    2018-05-01 Bioconductor</code></pre>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/agxufd">shared library</a> on <a href="https://paperpile.com">Paperpile</a>.</p>
<div id="refs" class="references">
<div id="ref-Harrow2012-hx">
<p>Harrow, Jennifer, Adam Frankish, Jose M Gonzalez, Electra Tapanari, Mark Diekhans, Felix Kokocinski, Bronwen L Aken, et al. 2012. “GENCODE: The Reference Human Genome Annotation for the ENCODE Project.” <em>Genome Res.</em> 22 (9): 1760–74. doi:<a href="https://doi.org/10.1101/gr.135350.111">10.1101/gr.135350.111</a>.</p>
</div>
<div id="ref-Hubbard2002-ah">
<p>Hubbard, T, D Barker, E Birney, G Cameron, Y Chen, L Clark, T Cox, et al. 2002. “The Ensembl Genome Database Project.” <em>Nucleic Acids Res.</em> 30 (1): 38–41. <a href="https://www.ncbi.nlm.nih.gov/pubmed/11752248" class="uri">https://www.ncbi.nlm.nih.gov/pubmed/11752248</a>.</p>
</div>
<div id="ref-Huber2015-nw">
<p>Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” <em>Nat. Methods</em> 12 (2): 115–21. doi:<a href="https://doi.org/10.1038/nmeth.3252">10.1038/nmeth.3252</a>.</p>
</div>
<div id="ref-Kent2002-fo">
<p>Kent, W James, Charles W Sugnet, Terrence S Furey, Krishna M Roskin, Tom H Pringle, Alan M Zahler, and David Haussler. 2002. “The Human Genome Browser at UCSC.” <em>Genome Res.</em> 12 (6): 996–1006. doi:<a href="https://doi.org/10.1101/gr.229102">10.1101/gr.229102</a>.</p>
</div>
<div id="ref-Lawrence2009-zx">
<p>Lawrence, Michael, Robert Gentleman, and Vincent Carey. 2009. “Rtracklayer: An R Package for Interfacing with Genome Browsers.” <em>Bioinformatics</em> 25 (14): 1841–2. doi:<a href="https://doi.org/10.1093/bioinformatics/btp328">10.1093/bioinformatics/btp328</a>.</p>
</div>
<div id="ref-Lawrence2013-vz">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9 (8): e1003118. doi:<a href="https://doi.org/10.1371/journal.pcbi.1003118">10.1371/journal.pcbi.1003118</a>.</p>
</div>
<div id="ref-Pruitt2007-vu">
<p>Pruitt, Kim D, Tatiana Tatusova, and Donna R Maglott. 2007. “NCBI Reference Sequences (RefSeq): A Curated Non-Redundant Sequence Database of Genomes, Transcripts and Proteins.” <em>Nucleic Acids Res.</em> 35 (Database issue): D61–5. doi:<a href="https://doi.org/10.1093/nar/gkl842">10.1093/nar/gkl842</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#the-s4-object-system">The S4 object system</a></li>
      <li>
<a href="#annotation-functions">Annotation functions</a><ul class="nav nav-pills nav-stacked">
<li><a href="#ensembl-annotations">Ensembl annotations</a></li>
      <li><a href="#gene-mapping-functions">Gene mapping functions</a></li>
      </ul>
</li>
      <li><a href="#r-session-information">R session information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="/">Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
