<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>basejump • basejump</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="basejump">
<meta property="og:description" content="basejump">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.12.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/acidgenomics/basejump/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="basejump_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>basejump</h1>
                        <h4 class="author">Michael Steinbaugh</h4>
            
            <h4 class="date">18 August 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/acidgenomics/basejump/blob/master/vignettes/basejump.Rmd"><code>vignettes/basejump.Rmd</code></a></small>
      <div class="hidden name"><code>basejump.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>The basejump package is an infrastructure toolkit that extends the base functionality of Bioconductor <span class="citation">(Huber et al. 2015)</span>. The package leverages the S4 object system for object-oriented programming, and defines multiple additional generic functions for use in genomics research. basejump provides simple, user-friendly functions for the acquisition of genome annotations from multiple online databases, including native support for Ensembl <span class="citation">(Hubbard et al. 2002)</span> and websites supporting standard FASTA and GTF/GFF file formats. Consistent handling of sample metadata remains a challenge for many bioinformatic analyses, and basejump aims to address this by providing a suite of sanitization functions to help standardize these variable inputs. Additionally, interactive read/write operations in R can be cumbersome and non-trivial when working with multiple data objects; here we provide additional functions designed for interactive use that aim to reduce friction and provide consistent handling of multiple common file formats used in genomics research.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">basejump</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">SummarizedExperiment</span>)
<span class="fu"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="kw">acid.test</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">RangedSummarizedExperiment</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"acidtest"</span>)
<span class="no">rse</span> <span class="kw">&lt;-</span> <span class="no">RangedSummarizedExperiment</span></pre></body></html></div>
</div>
<div id="the-s4-object-system" class="section level1">
<h1 class="hasAnchor">
<a href="#the-s4-object-system" class="anchor"></a>The S4 object system</h1>
<p>The <a href="https://basejump.acidgenomics.com/">basejump</a> package defines generics and methods for object-oriented programming with the S4 object system, which is used extensively by the <a href="https://bioconductor.org/">Bioconductor</a> project. These resources describe the S4 object system in detail:</p>
<ul>
<li>
<a href="https://www.bioconductor.org/help/course-materials/2017/Zurich/S4-classes-and-methods.html">S4 classes and methods</a>, Martin Morgan and Hervé Pagès, 2017.</li>
<li>
<a href="https://bioconductor.org/packages/devel/bioc/vignettes/S4Vectors/inst/doc/S4QuickOverview.pdf">A quick overview of the S4 class system</a>, Hervé Pagès, 2016.</li>
<li>
<a href="http://adv-r.had.co.nz/S4.html">The S4 object system</a>, Hadley Wickham.</li>
<li>
<a href="http://kasperdanielhansen.github.io/genbioconductor">Bioconductor for Genomic Data Science</a>, Kasper Hansen, 2016.</li>
<li>
<a href="https://www.bioconductor.org/help/course-materials/2010/AdvancedR/S4InBioconductor.pdf">S4 System Development in Bioconductor</a>, Patrick Aboyoun, 2010.</li>
</ul>
<p>In <a href="https://www.r-project.org/">R</a>, you can check whether a function is using the S4 object system with the <code><a href="https://rdrr.io/r/base/isS4.html">isS4()</a></code> function. Additionally, <code>showMethods()</code> and <code>getMethod()</code> are useful for exploring source code of S4 methods. In contrast, to obtain information on functions using the <a href="http://adv-r.had.co.nz/S3.html">S3</a> object system, use <code><a href="https://rdrr.io/r/utils/methods.html">methods()</a></code> instead.</p>
</div>
<div id="r-session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#r-session-information" class="anchor"></a>R session information</h1>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="kw pkg">utils</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</pre></body></html></div>
<pre><code>## R version 4.0.2 (2020-06-22)
## Platform: x86_64-apple-darwin17.0 (64-bit)
## Running under: macOS Catalina 10.15.6
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] SummarizedExperiment_1.18.2 DelayedArray_0.14.1        
##  [3] matrixStats_0.56.0          Biobase_2.48.0             
##  [5] GenomicRanges_1.40.0        GenomeInfoDb_1.24.2        
##  [7] IRanges_2.22.2              S4Vectors_0.26.1           
##  [9] BiocGenerics_0.34.0         basejump_0.12.12           
## [11] BiocStyle_2.16.0           
## 
## loaded via a namespace (and not attached):
##   [1] ProtGenerics_1.20.0           bitops_1.0-6                 
##   [3] fs_1.5.0                      bit64_4.0.2                  
##   [5] progress_1.2.2                httr_1.4.2                   
##   [7] rprojroot_1.3-2               syntactic_0.4.2              
##   [9] tools_4.0.2                   backports_1.1.8              
##  [11] R6_2.4.1                      colorspace_1.4-1             
##  [13] DBI_1.1.0                     lazyeval_0.2.2               
##  [15] withr_2.2.0                   tidyselect_1.1.0             
##  [17] prettyunits_1.1.1             bit_4.0.4                    
##  [19] curl_4.3                      compiler_4.0.2               
##  [21] acidgenerics_0.3.10           cli_2.0.2                    
##  [23] desc_1.2.0                    rtracklayer_1.48.0           
##  [25] bookdown_0.20                 scales_1.1.1                 
##  [27] acidbase_0.1.13               readr_1.3.1                  
##  [29] askpass_1.1                   rappdirs_0.3.1               
##  [31] pkgdown_1.5.1                 stringr_1.4.0                
##  [33] digest_0.6.25                 Rsamtools_2.4.0              
##  [35] rmarkdown_2.3                 XVector_0.28.0               
##  [37] pkgconfig_2.0.3               htmltools_0.5.0              
##  [39] sessioninfo_1.1.1             dbplyr_1.4.4                 
##  [41] fastmap_1.0.1                 ensembldb_2.12.1             
##  [43] rlang_0.4.7                   rstudioapi_0.11              
##  [45] RSQLite_2.2.0                 shiny_1.5.0                  
##  [47] generics_0.0.2                vroom_1.2.1                  
##  [49] BiocParallel_1.22.0           dplyr_1.0.1                  
##  [51] RCurl_1.98-1.2                magrittr_1.5                 
##  [53] GenomeInfoDbData_1.2.3        Matrix_1.2-18                
##  [55] munsell_0.5.0                 Rcpp_1.0.5                   
##  [57] fansi_0.4.1                   lifecycle_0.2.0              
##  [59] stringi_1.4.6                 yaml_2.2.1                   
##  [61] MASS_7.3-51.6                 zlibbioc_1.34.0              
##  [63] goalie_0.4.8                  BiocFileCache_1.12.1         
##  [65] AnnotationHub_2.20.1          grid_4.0.2                   
##  [67] blob_1.2.1                    promises_1.1.1               
##  [69] crayon_1.3.4                  lattice_0.20-41              
##  [71] Biostrings_2.56.0             GenomicFeatures_1.40.1       
##  [73] hms_0.5.3                     pipette_0.4.11               
##  [75] knitr_1.29                    pillar_1.4.6                 
##  [77] biomaRt_2.44.1                XML_3.99-0.5                 
##  [79] glue_1.4.1                    BiocVersion_3.11.1           
##  [81] evaluate_0.14                 data.table_1.13.0            
##  [83] BiocManager_1.30.10           vctrs_0.3.2                  
##  [85] httpuv_1.5.4                  openssl_1.4.2                
##  [87] purrr_0.3.4                   assertthat_0.2.1             
##  [89] xfun_0.16                     mime_0.9                     
##  [91] xtable_1.8-4                  AnnotationFilter_1.12.0      
##  [93] later_1.1.0.1                 SingleCellExperiment_1.10.1  
##  [95] tibble_3.0.3                  GenomicAlignments_1.24.0     
##  [97] AnnotationDbi_1.50.3          memoise_1.1.0                
##  [99] ellipsis_0.3.1                interactiveDisplayBase_1.26.3</code></pre>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/agxufd">shared library</a> on <a href="https://paperpile.com/">Paperpile</a>.</p>
<div id="refs" class="references">
<div id="ref-Hubbard2002-ah">
<p>Hubbard, T, D Barker, E Birney, G Cameron, Y Chen, L Clark, T Cox, et al. 2002. “The Ensembl Genome Database Project.” <em>Nucleic Acids Res.</em> 30 (1): 38–41. <a href="https://www.ncbi.nlm.nih.gov/pubmed/11752248">https://www.ncbi.nlm.nih.gov/pubmed/11752248</a>.</p>
</div>
<div id="ref-Huber2015-nw">
<p>Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” <em>Nat. Methods</em> 12 (2): 115–21. <a href="https://doi.org/10.1038/nmeth.3252">https://doi.org/10.1038/nmeth.3252</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://steinbaugh.com/">Michael Steinbaugh</a>, <a href="https://acidgenomics.com/">Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
