<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>basejump • basejump</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="basejump">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>basejump</h1>
                        <h4 class="author">Michael Steinbaugh</h4>
            
            <h4 class="date">25 September 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/acidgenomics/basejump/blob/master/vignettes/basejump.Rmd"><code>vignettes/basejump.Rmd</code></a></small>
      <div class="hidden name"><code>basejump.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>The basejump package is an infrastructure toolkit that extends the base functionality of Bioconductor <span class="citation">(Huber et al. 2015)</span>. The package leverages the S4 object system for object-oriented programming, and defines multiple additional generic functions for use in genomics research. basejump provides simple, user-friendly functions for the acquisition of genome annotations from multiple online databases, including native support for Ensembl <span class="citation">(Hubbard et al. 2002)</span> and websites supporting standard FASTA and GTF/GFF file formats. Consistent handling of sample metadata remains a challenge for many bioinformatic analyses, and basejump aims to address this by providing a suite of sanitization functions to help standardize these variable inputs. Additionally, interactive read/write operations in R can be cumbersome and non-trivial when working with multiple data objects; here we provide additional functions designed for interactive use that aim to reduce friction and provide consistent handling of multiple common file formats used in genomics research.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(basejump)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(SummarizedExperiment)
<span class="kw"><a href="https://rdrr.io/r/base/options.html">options</a></span>(<span class="dt">acid.test =</span> <span class="ot">TRUE</span>)
<span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(RangedSummarizedExperiment, <span class="dt">package =</span> <span class="st">"acidtest"</span>)
rse &lt;-<span class="st"> </span>RangedSummarizedExperiment</code></pre>
</div>
<div id="the-s4-object-system" class="section level1">
<h1 class="hasAnchor">
<a href="#the-s4-object-system" class="anchor"></a>The S4 object system</h1>
<p>The <a href="https://basejump.acidgenomics.com/">basejump</a> package defines generics and methods for object-oriented programming with the S4 object system, which is used extensively by the <a href="https://bioconductor.org/">Bioconductor</a> project. These resources describe the S4 object system in detail:</p>
<ul>
<li>
<a href="https://www.bioconductor.org/help/course-materials/2017/Zurich/S4-classes-and-methods.html">S4 classes and methods</a>, Martin Morgan and Hervé Pagès, 2017.</li>
<li>
<a href="https://bioconductor.org/packages/devel/bioc/vignettes/S4Vectors/inst/doc/S4QuickOverview.pdf">A quick overview of the S4 class system</a>, Hervé Pagès, 2016.</li>
<li>
<a href="http://adv-r.had.co.nz/S4.html">The S4 object system</a>, Hadley Wickham.</li>
<li>
<a href="http://kasperdanielhansen.github.io/genbioconductor">Bioconductor for Genomic Data Science</a>, Kasper Hansen, 2016.</li>
<li>
<a href="https://www.bioconductor.org/help/course-materials/2010/AdvancedR/S4InBioconductor.pdf">S4 System Development in Bioconductor</a>, Patrick Aboyoun, 2010.</li>
</ul>
<p>In <a href="https://www.r-project.org/">R</a>, you can check whether a function is using the S4 object system with the <code><a href="https://rdrr.io/r/base/isS4.html">isS4()</a></code> function. Additionally, <code>showMethods()</code> and <code>getMethod()</code> are useful for exploring source code of S4 methods. In contrast, to obtain information on functions using the <a href="http://adv-r.had.co.nz/S3.html">S3</a> object system, use <code><a href="https://rdrr.io/r/utils/methods.html">methods()</a></code> instead.</p>
</div>
<div id="r-session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#r-session-information" class="anchor"></a>R session information</h1>
<pre class="sourceCode r"><code class="sourceCode r">utils::<span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</code></pre>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-redhat-linux-gnu (64-bit)
## Running under: Red Hat Enterprise Linux Server 7.7 (Maipo)
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib64/libopenblas-r0.3.3.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] SummarizedExperiment_1.14.1 DelayedArray_0.10.0        
##  [3] BiocParallel_1.18.1         matrixStats_0.55.0         
##  [5] Biobase_2.44.0              GenomicRanges_1.36.1       
##  [7] GenomeInfoDb_1.20.0         IRanges_2.18.2             
##  [9] S4Vectors_0.22.1            BiocGenerics_0.30.0        
## [11] basejump_0.11.16            BiocStyle_2.12.0           
## 
## loaded via a namespace (and not attached):
##  [1] ProtGenerics_1.16.0           bitops_1.0-6                 
##  [3] fs_1.3.1                      bit64_0.9-7                  
##  [5] progress_1.2.2                httr_1.4.1                   
##  [7] rprojroot_1.3-2               syntactic_0.3.0              
##  [9] tools_3.6.1                   backports_1.1.4              
## [11] R6_2.4.0                      colorspace_1.4-1             
## [13] lazyeval_0.2.2                DBI_1.0.0                    
## [15] withr_2.1.2                   tidyselect_0.2.5             
## [17] prettyunits_1.0.2             bit_1.1-14                   
## [19] curl_4.0                      compiler_3.6.1               
## [21] cli_1.1.0                     desc_1.2.0                   
## [23] rtracklayer_1.44.4            bookdown_0.13                
## [25] scales_1.0.0                  rappdirs_0.3.1               
## [27] pkgdown_1.4.1                 stringr_1.4.0                
## [29] digest_0.6.20                 Rsamtools_2.0.0              
## [31] rmarkdown_1.15                R.utils_2.9.0                
## [33] XVector_0.24.0                pkgconfig_2.0.2              
## [35] htmltools_0.3.6               sessioninfo_1.1.1            
## [37] ensembldb_2.8.0               dbplyr_1.4.2                 
## [39] rlang_0.4.0                   RSQLite_2.1.2                
## [41] shiny_1.3.2                   DelayedMatrixStats_1.6.1     
## [43] dplyr_0.8.3                   R.oo_1.22.0                  
## [45] RCurl_1.95-4.12               magrittr_1.5                 
## [47] GenomeInfoDbData_1.2.1        Matrix_1.2-17                
## [49] munsell_0.5.0                 Rcpp_1.0.2                   
## [51] R.methodsS3_1.7.1             stringi_1.4.3                
## [53] yaml_2.2.0                    MASS_7.3-51.4                
## [55] zlibbioc_1.30.0               brio_0.3.8                   
## [57] goalie_0.3.8                  BiocFileCache_1.8.0          
## [59] AnnotationHub_2.16.1          grid_3.6.1                   
## [61] blob_1.2.0                    transformer_0.2.7            
## [63] promises_1.0.1                crayon_1.3.4                 
## [65] lattice_0.20-38               Biostrings_2.52.0            
## [67] bioverbs_0.2.10               GenomicFeatures_1.36.4       
## [69] hms_0.5.1                     zeallot_0.1.0                
## [71] knitr_1.25                    pillar_1.4.2                 
## [73] biomaRt_2.40.4                XML_3.98-1.20                
## [75] glue_1.3.1                    evaluate_0.14                
## [77] freerange_0.2.6               data.table_1.12.2            
## [79] BiocManager_1.30.4            vctrs_0.2.0                  
## [81] httpuv_1.5.2                  grr_0.9.5                    
## [83] purrr_0.3.2                   assertthat_0.2.1             
## [85] xfun_0.9                      mime_0.7                     
## [87] xtable_1.8-4                  AnnotationFilter_1.8.0       
## [89] later_0.8.0                   SingleCellExperiment_1.6.0   
## [91] tibble_2.1.3                  GenomicAlignments_1.20.1     
## [93] Matrix.utils_0.9.7            AnnotationDbi_1.46.1         
## [95] memoise_1.1.0                 interactiveDisplayBase_1.22.0</code></pre>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/agxufd">shared library</a> on <a href="https://paperpile.com/">Paperpile</a>.</p>
<div class="references">
<p>Hubbard, T, D Barker, E Birney, G Cameron, Y Chen, L Clark, T Cox, et al. 2002. “The Ensembl Genome Database Project.” <em>Nucleic Acids Res.</em> 30 (1) (January): 38–41. <a href="https://www.ncbi.nlm.nih.gov/pubmed/11752248">https://www.ncbi.nlm.nih.gov/pubmed/11752248</a>.</p>
<p>Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” <em>Nat. Methods</em> 12 (2) (February): 115–121. doi:<a href="http://dx.doi.org/10.1038/nmeth.3252">10.1038/nmeth.3252</a>. <a href="http://dx.doi.org/10.1038/nmeth.3252">http://dx.doi.org/10.1038/nmeth.3252</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#the-s4-object-system">The S4 object system</a></li>
      <li><a href="#r-session-information">R session information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://steinbaugh.com/">Michael Steinbaugh</a>, <a href="https://acidgenomics.com/">Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
