<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>basejump • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="basejump" />

<meta property="og:description" content="" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-article">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.10.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>


<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>basejump</h1>
                        <h4 class="author">Michael Steinbaugh</h4>
            
            <h4 class="date">17 April 2019</h4>
      
      <small class="dont-index">Source: <a href='https://github.com/acidgenomics/basejump/blob/master/vignettes/basejump.Rmd'><code>vignettes/basejump.Rmd</code></a></small>
      <div class="hidden name"><code>basejump.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>The basejump package is an infrastructure toolkit that extends the base functionality of Bioconductor <span class="citation">(<span class="citeproc-not-found" data-reference-id="Huber2015-nw"><strong>???</strong></span>)</span>. The package leverages the S4 object system for object-oriented programming, and defines multiple additional generic functions for use in genomics research. basejump provides simple, user-friendly functions for the acquisition of genome annotations from multiple online databases, including native support for Ensembl <span class="citation">(<span class="citeproc-not-found" data-reference-id="Hubbard2002-ah"><strong>???</strong></span>)</span> and websites supporting standard FASTA and GTF/GFF file formats. Consistent handling of sample metadata remains a challenge for many bioinformatic analyses, and basejump aims to address this by providing a suite of sanitization functions to help standardize these variable inputs. Additionally, interactive read/write operations in R can be cumbersome and non-trivial when working with multiple data objects; here we provide additional functions designed for interactive use that aim to reduce friction and provide consistent handling of multiple common file formats used in genomics research.</p>
    </div>
    
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>This vignette focuses on the most common user-facing functions that fall into the following categories:</p>
<ul>
<li>Annotation functions</li>
<li>Read/write functions</li>
<li>Data functions</li>
<li>Syntactic naming functions</li>
<li>Math and science functions</li>
</ul>
<p>There are additional function families whose functionality lie outside the scope of this vignette. Consult the <a href="https://basejump.acidgenomics.com/">basejump</a> website or package documentation for more information of these types of functions, which are more technical in nature and predominantly developer-facing:</p>
<ul>
<li>R Markdown functions</li>
<li>Atomic vector functions</li>
<li>Coercion methods</li>
<li>Developer functions</li>
<li>Assert check functions</li>
</ul>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(basejump)
<span class="kw">library</span>(SummarizedExperiment)
<span class="kw">options</span>(<span class="dt">acid.test =</span> <span class="ot">TRUE</span>)
<span class="kw">data</span>(rse, <span class="dt">package =</span> <span class="st">&quot;acidtest&quot;</span>)</code></pre>
</div>
<div id="the-s4-object-system" class="section level1">
<h1>The S4 object system</h1>
<p>The <a href="https://basejump.acidgenomics.com/">basejump</a> package defines generics and methods for object-oriented programming with the S4 object system, which is used extensively by the <a href="https://bioconductor.org/">Bioconductor</a> project. These resources describe the S4 object system in detail:</p>
<ul>
<li><a href="https://www.bioconductor.org/help/course-materials/2017/Zurich/S4-classes-and-methods.html">S4 classes and methods</a>, Martin Morgan and Hervé Pagès, 2017.</li>
<li><a href="https://bioconductor.org/packages/devel/bioc/vignettes/S4Vectors/inst/doc/S4QuickOverview.pdf">A quick overview of the S4 class system</a>, Hervé Pagès, 2016.</li>
<li><a href="http://adv-r.had.co.nz/S4.html">The S4 object system</a>, Hadley Wickham.</li>
<li><a href="http://kasperdanielhansen.github.io/genbioconductor">Bioconductor for Genomic Data Science</a>, Kasper Hansen, 2016.</li>
<li><a href="https://www.bioconductor.org/help/course-materials/2010/AdvancedR/S4InBioconductor.pdf">S4 System Development in Bioconductor</a>, Patrick Aboyoun, 2010.</li>
</ul>
<p>In <a href="https://www.r-project.org/">R</a>, you can check whether a function is using the S4 object system with the [<code>isS4()</code>][base::isS4] function. Additionally, [<code>showMethods()</code>][methods::showMethods] and [<code>getMethod()</code>][methods::getMethod] are useful for exploring source code of S4 methods. In contrast, to obtain information on functions using the <a href="http://adv-r.had.co.nz/S3.html">S3</a> object system, use [<code>methods()</code>][utils::methods] instead.</p>
</div>
<div id="annotation-functions" class="section level1">
<h1>Annotation functions</h1>
<p>Obtaining versioned genome annotations quickly and reliably from online databases such as <a href="https://www.ensembl.org/">Ensembl</a> <span class="citation">(<span class="citeproc-not-found" data-reference-id="Hubbard2002-ah"><strong>???</strong></span>)</span>, <a href="https://www.gencodegenes.org/">GENCODE</a> <span class="citation">(<span class="citeproc-not-found" data-reference-id="Harrow2012-hx"><strong>???</strong></span>)</span>, <a href="https://www.ncbi.nlm.nih.gov/refseq/">RefSeq</a> <span class="citation">(<span class="citeproc-not-found" data-reference-id="Pruitt2007-vu"><strong>???</strong></span>)</span>, and the <a href="http://genome.ucsc.edu/">UCSC Genome Browser</a> <span class="citation">(<span class="citeproc-not-found" data-reference-id="Kent2002-fo"><strong>???</strong></span>)</span> remains overly challenging. <a href="https://basejump.acidgenomics.com/">basejump</a> aims to help address this issue by providing native gene- and transcript-level annotation support by helping users interface with <a href="https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a> and <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a>, along with a robust set of tools to parse <a href="http://mblab.wustl.edu/GTF22.html" title="General Transfer Format">GTF</a> and <a href="https://useast.ensembl.org/info/website/upload/gff.html" title="General Feature Format">GFF</a> annotation files. Annotation file parsing is handled internally by the <a href="https://bioconductor.org/packages/rtracklayer/">rtracklayer</a> package <span class="citation">(<span class="citeproc-not-found" data-reference-id="Lawrence2009-zx"><strong>???</strong></span>)</span>. Genome annotations are returned as [<code>GRanges</code>][GenomicRanges::GRanges] objects, defined in the <a href="https://bioconductor.org/packages/GenomicRanges/">GenomicRanges</a> package <span class="citation">(<span class="citeproc-not-found" data-reference-id="Lawrence2013-vz"><strong>???</strong></span>)</span>, which allows for easy access to position, chromosome, strand, and additional metadata saved in the [<code>mcols()</code>][S4Vectors::mcols] slot of the object. <code>GRanges</code> can be coerced to a standard <code>data.frame</code> using the [<code>as.data.frame()</code>][BiocGenerics::as.data.frame] function.</p>
<div id="ensembl-annotations" class="section level2">
<h2>Ensembl annotations</h2>
<p><code>makeGRangesFromEnsembl()</code> supports multiple genome builds, and versioned releases from <a href="https://www.ensembl.org/">Ensembl</a>.</p>
<div id="current-release" class="section level3">
<h3>Current release</h3>
<p>By default, the function will use the latest release version available from <a href="https://bioconductor.org/packages/release/bioc/html/AnnotationHub.html">AnnotationHub</a> and <a href="https://bioconductor.org/packages/ensembldb/">ensembldb</a>.</p>
<pre class="sourceCode r"><code class="sourceCode r">grch38 &lt;-<span class="st"> </span><span class="kw">makeGRangesFromEnsembl</span>(<span class="dt">organism =</span> <span class="st">&quot;Homo sapiens&quot;</span>)
<span class="kw">summary</span>(grch38)</code></pre>
<pre><code>## [1] &quot;GRanges object with 65687 ranges and 8 metadata columns&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">names</span>(grch38))</code></pre>
<pre><code>## [1] &quot;ENSG00000000003&quot; &quot;ENSG00000000005&quot; &quot;ENSG00000000419&quot; &quot;ENSG00000000457&quot;
## [5] &quot;ENSG00000000460&quot; &quot;ENSG00000000938&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(<span class="kw">metadata</span>(grch38))</code></pre>
<pre><code>## [1] &quot;package&quot;        &quot;version&quot;        &quot;date&quot;           &quot;organism&quot;      
## [5] &quot;genomeBuild&quot;    &quot;ensemblRelease&quot; &quot;ensembldb&quot;      &quot;level&quot;         
## [9] &quot;id&quot;</code></pre>
<p>Data inside a <code>GRanges</code> object can be accessed with a number of functions defined in the <a href="https://bioconductor.org/packages/GenomeInfoDb/">GenomeInfoDb</a> and <a href="https://www.bioconductor.org/packages/IRanges/">IRanges</a> packages.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqnames</span>(grch38)</code></pre>
<pre><code>## factor-Rle of length 65687 with 49080 runs
##   Lengths:      2      1      4      3 ...      1      1      2      1
##   Values :      X     20      1      6 ...      1     12      6     19
## Levels(423): X 20 1 6 3 7 ... LRG_311 LRG_721 LRG_741 LRG_763 LRG_93</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">seqinfo</span>(grch38)</code></pre>
<pre><code>## Seqinfo object with 423 sequences from GRCh38 genome:
##   seqnames seqlengths isCircular genome
##   X         156040895      FALSE GRCh38
##   20         64444167      FALSE GRCh38
##   1         248956422      FALSE GRCh38
##   6         170805979      FALSE GRCh38
##   3         198295559      FALSE GRCh38
##   ...             ...        ...    ...
##   LRG_311      115492      FALSE GRCh38
##   LRG_721       33396      FALSE GRCh38
##   LRG_741      231167      FALSE GRCh38
##   LRG_763      176286      FALSE GRCh38
##   LRG_93        22459      FALSE GRCh38</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ranges</span>(grch38)</code></pre>
<pre><code>## IRanges object with 65687 ranges and 0 metadata columns:
##                       start       end     width
##                   &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   ENSG00000000003 100627109 100639991     12883
##   ENSG00000000005 100584802 100599885     15084
##   ENSG00000000419  50934867  50958555     23689
##   ENSG00000000457 169849631 169894267     44637
##   ENSG00000000460 169662007 169854080    192074
##               ...       ...       ...       ...
##           LRG_995  77948402  77979205     30804
##           LRG_996  56080025  56103507     23483
##           LRG_997 117288367 117425855    137489
##           LRG_998  41934933  42048894    113962
##           LRG_999  42268537  42295796     27260</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">strand</span>(grch38)</code></pre>
<pre><code>## factor-Rle of length 65687 with 32670 runs
##   Lengths: 1 1 2 1 1 1 2 1 1 1 1 4 3 1 1 ... 1 4 1 1 1 3 1 1 1 1 1 3 1 2 1
##   Values : - + - + - + - + - + - + - + - ... + - + - + - + - + - + - + - +
## Levels(3): + - *</code></pre>
<p>The <code>GRanges</code> class is a powerful container for genomic coordinates and metadata. For example, we can easily the gene counts per chromosome.</p>
<pre class="sourceCode r"><code class="sourceCode r">grch38 %&gt;%
<span class="st">    </span><span class="kw">seqnames</span>() %&gt;%
<span class="st">    </span><span class="kw">table</span>() %&gt;%
<span class="st">    </span><span class="kw">sort</span>(<span class="dt">decreasing =</span> <span class="ot">TRUE</span>) %&gt;%
<span class="st">    </span><span class="kw">head</span>(<span class="dt">n =</span> <span class="dv">24</span>) %&gt;%
<span class="st">    </span>.[<span class="kw">sort</span>(<span class="kw">names</span>(.))]</code></pre>
<pre><code>## .
##    1   10   11   12   13   14   15   16   17   18   19    2   20   21   22 
## 5405 2271 3342 3001 1348 2252 2206 2541 3074 1200 3004 4067 1417  846 1375 
##    3    4    5    6    7    8    9    X    Y 
## 3090 2526 2903 2940 2946 2405 2308 2418  518</code></pre>
<p>Gene metadata is contained in the [<code>mcols()</code>][S4Vectors::mcols] slot of the object, using a <code>DataFrame</code> internally. Here’a the current list of gene-level annotation columns returned from <code>makeGRangesFromEnsembl()</code>:</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mcols</span>(grch38) %&gt;%<span class="st"> </span><span class="kw">colnames</span>()</code></pre>
<pre><code>## [1] &quot;broadClass&quot;     &quot;description&quot;    &quot;entrezID&quot;       &quot;geneBiotype&quot;   
## [5] &quot;geneID&quot;         &quot;geneIDVersion&quot;  &quot;geneName&quot;       &quot;seqCoordSystem&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mcols</span>(grch38) %&gt;%<span class="st"> </span><span class="kw">lapply</span>(<span class="dt">X =</span> ., <span class="dt">FUN =</span> head)</code></pre>
<pre><code>## $broadClass
## factor-Rle of length 6 with 1 run
##   Lengths:      6
##   Values : coding
## Levels(8): coding ig mito noncoding other pseudo small tcr
## 
## $description
## factor-Rle of length 6 with 6 runs
##   Lengths:                                                              ...
##   Values :                                              tetraspanin 6 [S...
## Levels(45585): 1-acylglycerol-3-phosphate O-acyltransferase 1 [Source:HG...
## 
## $entrezID
## $entrezID$ENSG00000000003
## [1] 7105
## 
## $entrezID$ENSG00000000005
## [1] 64102
## 
## $entrezID$ENSG00000000419
## [1] 8813
## 
## $entrezID$ENSG00000000457
## [1] 57147
## 
## $entrezID$ENSG00000000460
## [1] 55732
## 
## $entrezID$ENSG00000000938
## [1] 2268
## 
## 
## $geneBiotype
## factor-Rle of length 6 with 1 run
##   Lengths:              6
##   Values : protein_coding
## Levels(48): 3prime_overlapping_ncRNA antisense ... vaultRNA
## 
## $geneID
## character-Rle of length 6 with 6 runs
##   Lengths:                 1                 1 ...                 1
##   Values : &quot;ENSG00000000003&quot; &quot;ENSG00000000005&quot; ... &quot;ENSG00000000938&quot;
## 
## $geneIDVersion
## character-Rle of length 6 with 6 runs
##   Lengths:                    1 ...                    1
##   Values : &quot;ENSG00000000003.14&quot; ... &quot;ENSG00000000938.12&quot;
## 
## $geneName
## factor-Rle of length 6 with 6 runs
##   Lengths:        1        1        1        1        1        1
##   Values :   TSPAN6     TNMD     DPM1    SCYL3 C1orf112      FGR
## Levels(58698): A1BG A1BG-AS1 A1CF A2M A2M-AS1 ... ZYG11B ZYX ZYXP1 ZZEF1
## 
## $seqCoordSystem
## factor-Rle of length 6 with 1 run
##   Lengths:          6
##   Values : chromosome
## Levels(3): chromosome lrg scaffold</code></pre>
<p>Transcript-level annotations are also supported, using the <code>level = &quot;transcripts&quot;</code> argument.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">makeGRangesFromEnsembl</span>(<span class="dt">organism =</span> <span class="st">&quot;Homo sapiens&quot;</span>, <span class="dt">level =</span> <span class="st">&quot;transcripts&quot;</span>)</code></pre>
<pre><code>## GRanges object with 228432 ranges and 15 metadata columns:
##                   seqnames              ranges strand | broadClass
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;Rle&gt;
##   ENST00000000233        7 127588345-127591705      + |     coding
##   ENST00000000412       12     8940365-8949955      - |     coding
##   ENST00000000442       11   64305578-64316738      + |     coding
##   ENST00000001008       12     2794953-2805423      + |     coding
##   ENST00000001146        2   72129238-72148038      - |     coding
##               ...      ...                 ...    ... .        ...
##         LRG_998t1        6   41934933-41941848      - |      other
##         LRG_998t2        6   41934933-42048894      - |      other
##         LRG_999t1       19   42284665-42295796      + |      other
##         LRG_999t2       19   42268537-42295796      + |      other
##          LRG_99t1       11   36591943-36598262      - |      other
##                                                                                          description
##                                                                                                &lt;Rle&gt;
##   ENST00000000233                        ADP ribosylation factor 5 [Source:HGNC Symbol;Acc:HGNC:658]
##   ENST00000000412  mannose-6-phosphate receptor, cation dependent [Source:HGNC Symbol;Acc:HGNC:6752]
##   ENST00000000442                 estrogen related receptor alpha [Source:HGNC Symbol;Acc:HGNC:3471]
##   ENST00000001008                         FK506 binding protein 4 [Source:HGNC Symbol;Acc:HGNC:3720]
##   ENST00000001146 cytochrome P450 family 26 subfamily B member 1 [Source:HGNC Symbol;Acc:HGNC:20581]
##               ...                                                                                ...
##         LRG_998t1                                       cyclin D3 [Source:HGNC Symbol;Acc:HGNC:1585]
##         LRG_998t2                                       cyclin D3 [Source:HGNC Symbol;Acc:HGNC:1585]
##         LRG_999t1              capicua transcriptional repressor [Source:HGNC Symbol;Acc:HGNC:14214]
##         LRG_999t2              capicua transcriptional repressor [Source:HGNC Symbol;Acc:HGNC:14214]
##          LRG_99t1                      recombination activating 2 [Source:HGNC Symbol;Acc:HGNC:9832]
##                   entrezID    geneBiotype          geneID
##                     &lt;list&gt;          &lt;Rle&gt;           &lt;Rle&gt;
##   ENST00000000233      381 protein_coding ENSG00000004059
##   ENST00000000412     4074 protein_coding ENSG00000003056
##   ENST00000000442     2101 protein_coding ENSG00000173153
##   ENST00000001008     2288 protein_coding ENSG00000004478
##   ENST00000001146    56603 protein_coding ENSG00000003137
##               ...      ...            ...             ...
##         LRG_998t1      896       LRG_gene         LRG_998
##         LRG_998t2      896       LRG_gene         LRG_998
##         LRG_999t1    23152       LRG_gene         LRG_999
##         LRG_999t2    23152       LRG_gene         LRG_999
##          LRG_99t1     5897       LRG_gene          LRG_99
##                        geneIDVersion geneName seqCoordSystem
##                                &lt;Rle&gt;    &lt;Rle&gt;          &lt;Rle&gt;
##   ENST00000000233 ENSG00000004059.10     ARF5     chromosome
##   ENST00000000412  ENSG00000003056.7     M6PR     chromosome
##   ENST00000000442 ENSG00000173153.13    ESRRA     chromosome
##   ENST00000001008  ENSG00000004478.7    FKBP4     chromosome
##   ENST00000001146  ENSG00000003137.8  CYP26B1     chromosome
##               ...                ...      ...            ...
##         LRG_998t1          LRG_998.1    CCND3     chromosome
##         LRG_998t2          LRG_998.1    CCND3     chromosome
##         LRG_999t1          LRG_999.1      CIC     chromosome
##         LRG_999t2          LRG_999.1      CIC     chromosome
##          LRG_99t1           LRG_99.1     RAG2     chromosome
##                   transcriptBiotype transcriptCdsSeqEnd
##                               &lt;Rle&gt;               &lt;Rle&gt;
##   ENST00000000233    protein_coding           127591299
##   ENST00000000412    protein_coding             8946404
##   ENST00000000442    protein_coding            64315966
##   ENST00000001008    protein_coding             2803258
##   ENST00000001146    protein_coding            72147834
##               ...               ...                 ...
##         LRG_998t1          LRG_gene            41941649
##         LRG_998t2          LRG_gene            41940540
##         LRG_999t1          LRG_gene            42295191
##         LRG_999t2          LRG_gene            42295191
##          LRG_99t1          LRG_gene            36594168
##                   transcriptCdsSeqStart    transcriptID
##                                   &lt;Rle&gt;           &lt;Rle&gt;
##   ENST00000000233             127588499 ENST00000000233
##   ENST00000000412               8941818 ENST00000000412
##   ENST00000000442              64307180 ENST00000000442
##   ENST00000001008               2795140 ENST00000001008
##   ENST00000001146              72132227 ENST00000001146
##               ...                   ...             ...
##         LRG_998t1              41935940       LRG_998t1
##         LRG_998t2              41935940       LRG_998t2
##         LRG_999t1              42284705       LRG_999t1
##         LRG_999t2              42271784       LRG_999t2
##          LRG_99t1              36592585        LRG_99t1
##                   transcriptIDVersion  transcriptName
##                                 &lt;Rle&gt;           &lt;Rle&gt;
##   ENST00000000233   ENST00000000233.9 ENST00000000233
##   ENST00000000412   ENST00000000412.7 ENST00000000412
##   ENST00000000442  ENST00000000442.10 ENST00000000442
##   ENST00000001008   ENST00000001008.5 ENST00000001008
##   ENST00000001146   ENST00000001146.6 ENST00000001146
##               ...                 ...             ...
##         LRG_998t1         LRG_998t1.1       LRG_998t1
##         LRG_998t2         LRG_998t2.1       LRG_998t2
##         LRG_999t1         LRG_999t1.1       LRG_999t1
##         LRG_999t2         LRG_999t2.1       LRG_999t2
##          LRG_99t1          LRG_99t1.1        LRG_99t1
##                   transcriptSupportLevel
##                                    &lt;Rle&gt;
##   ENST00000000233                      1
##   ENST00000000412                      1
##   ENST00000000442                      1
##   ENST00000001008                      1
##   ENST00000001146                      1
##               ...                    ...
##         LRG_998t1                   &lt;NA&gt;
##         LRG_998t2                   &lt;NA&gt;
##         LRG_999t1                   &lt;NA&gt;
##         LRG_999t2                   &lt;NA&gt;
##          LRG_99t1                   &lt;NA&gt;
##   -------
##   seqinfo: 423 sequences from GRCh38 genome</code></pre>
</div>
<div id="versioned-releases" class="section level3">
<h3>Versioned releases</h3>
<p>Versioned releases are supported by the “<code>release</code>” argument, which currently supports back to <a href="https://www.ensembl.org/">Ensembl</a> 87. If an older <a href="https://www.ensembl.org/">Ensembl</a> release is required, resort to using a <a href="http://mblab.wustl.edu/GTF22.html" title="General Transfer Format">GTF</a>/<a href="https://useast.ensembl.org/info/website/upload/gff.html" title="General Feature Format">GFF</a> file or browse the <a href="https://bioconductor.org/">Bioconductor</a> website to see if a suitable annotation database package is available.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">makeGRangesFromEnsembl</span>(
    <span class="dt">organism =</span> <span class="st">&quot;Homo sapiens&quot;</span>,
    <span class="dt">genomeBuild =</span> <span class="st">&quot;GRCh38&quot;</span>,
    <span class="dt">release =</span> <span class="dv">90</span>
)</code></pre>
<pre><code>## GRanges object with 64661 ranges and 8 metadata columns:
##                   seqnames              ranges strand | broadClass
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;Rle&gt;
##   ENSG00000000003        X 100627109-100639991      - |     coding
##   ENSG00000000005        X 100584802-100599885      + |     coding
##   ENSG00000000419       20   50934867-50958555      - |     coding
##   ENSG00000000457        1 169849631-169894267      - |     coding
##   ENSG00000000460        1 169662007-169854080      + |     coding
##               ...      ...                 ...    ... .        ...
##           LRG_995        1   77948402-77979205      - |      other
##           LRG_996       12   56080025-56103507      + |      other
##           LRG_997        6 117288367-117425855      - |      other
##           LRG_998        6   41934933-42048894      - |      other
##           LRG_999       19   42268537-42295796      + |      other
##                                                                                                      description
##                                                                                                            &lt;Rle&gt;
##   ENSG00000000003                                              tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:11858]
##   ENSG00000000005                                                tenomodulin [Source:HGNC Symbol;Acc:HGNC:17757]
##   ENSG00000000419 dolichyl-phosphate mannosyltransferase subunit 1, catalytic [Source:HGNC Symbol;Acc:HGNC:3005]
##   ENSG00000000457                                   SCY1 like pseudokinase 3 [Source:HGNC Symbol;Acc:HGNC:19285]
##   ENSG00000000460                        chromosome 1 open reading frame 112 [Source:HGNC Symbol;Acc:HGNC:25565]
##               ...                                                                                            ...
##           LRG_995                      far upstream element binding protein 1 [Source:HGNC Symbol;Acc:HGNC:4004]
##           LRG_996                           erb-b2 receptor tyrosine kinase 3 [Source:HGNC Symbol;Acc:HGNC:3431]
##           LRG_997             ROS proto-oncogene 1, receptor tyrosine kinase [Source:HGNC Symbol;Acc:HGNC:10261]
##           LRG_998                                                   cyclin D3 [Source:HGNC Symbol;Acc:HGNC:1585]
##           LRG_999                          capicua transcriptional repressor [Source:HGNC Symbol;Acc:HGNC:14214]
##                   entrezID    geneBiotype          geneID
##                     &lt;list&gt;          &lt;Rle&gt;           &lt;Rle&gt;
##   ENSG00000000003     7105 protein_coding ENSG00000000003
##   ENSG00000000005    64102 protein_coding ENSG00000000005
##   ENSG00000000419     8813 protein_coding ENSG00000000419
##   ENSG00000000457    57147 protein_coding ENSG00000000457
##   ENSG00000000460    55732 protein_coding ENSG00000000460
##               ...      ...            ...             ...
##           LRG_995     8880       LRG_gene         LRG_995
##           LRG_996     2065       LRG_gene         LRG_996
##           LRG_997     6098       LRG_gene         LRG_997
##           LRG_998      896       LRG_gene         LRG_998
##           LRG_999    23152       LRG_gene         LRG_999
##                        geneIDVersion geneName seqCoordSystem
##                                &lt;Rle&gt;    &lt;Rle&gt;          &lt;Rle&gt;
##   ENSG00000000003 ENSG00000000003.14   TSPAN6     chromosome
##   ENSG00000000005  ENSG00000000005.5     TNMD     chromosome
##   ENSG00000000419 ENSG00000000419.12     DPM1     chromosome
##   ENSG00000000457 ENSG00000000457.13    SCYL3     chromosome
##   ENSG00000000460 ENSG00000000460.16 C1orf112     chromosome
##               ...                ...      ...            ...
##           LRG_995          LRG_995.1    FUBP1     chromosome
##           LRG_996          LRG_996.1    ERBB3     chromosome
##           LRG_997          LRG_997.1     ROS1     chromosome
##           LRG_998          LRG_998.1    CCND3     chromosome
##           LRG_999          LRG_999.1      CIC     chromosome
##   -------
##   seqinfo: 388 sequences from GRCh38 genome</code></pre>
</div>
<div id="grch37" class="section level3">
<h3>GRCh37</h3>
<p>The legacy <a href="https://grch37.ensembl.org/">Ensembl GRCh37 genome build</a> (release 75; also known as <a href="https://genome.ucsc.edu/cgi-bin/hgGateway?db=hg19">UCSC hg19</a>) is also supported, but is no longer recommended for use in new analyses. Internally, support for <a href="https://grch37.ensembl.org/">GRCh37</a> is provided by the <a href="https://bioconductor.org/packages/EnsDb.Hsapiens.v75/">EnsDb.Hsapiens.v75</a> annotation database package.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">makeGRangesFromEnsembl</span>(
    <span class="dt">organism =</span> <span class="st">&quot;Homo sapiens&quot;</span>,
    <span class="dt">genomeBuild =</span> <span class="st">&quot;GRCh37&quot;</span>
)</code></pre>
<pre><code>## GRanges object with 64102 ranges and 6 metadata columns:
##                   seqnames              ranges strand | broadClass
##                      &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |      &lt;Rle&gt;
##   ENSG00000000003        X   99883667-99894988      - |     coding
##   ENSG00000000005        X   99839799-99854882      + |     coding
##   ENSG00000000419       20   49551404-49575092      - |     coding
##   ENSG00000000457        1 169818772-169863408      - |     coding
##   ENSG00000000460        1 169631245-169823221      + |     coding
##               ...      ...                 ...    ... .        ...
##            LRG_94       10   72357104-72362531      - |      other
##            LRG_96       15   55495792-55582001      - |      other
##            LRG_97       22   37621310-37640305      - |      other
##            LRG_98       11   36589563-36601312      + |      other
##            LRG_99       11   36613493-36619812      - |      other
##                   entrezID    geneBiotype          geneID geneName
##                     &lt;list&gt;          &lt;Rle&gt;           &lt;Rle&gt;    &lt;Rle&gt;
##   ENSG00000000003     7105 protein_coding ENSG00000000003   TSPAN6
##   ENSG00000000005    64102 protein_coding ENSG00000000005     TNMD
##   ENSG00000000419     8813 protein_coding ENSG00000000419     DPM1
##   ENSG00000000457    57147 protein_coding ENSG00000000457    SCYL3
##   ENSG00000000460    55732 protein_coding ENSG00000000460 C1orf112
##               ...      ...            ...             ...      ...
##            LRG_94     5551       LRG_gene          LRG_94   LRG_94
##            LRG_96     5873       LRG_gene          LRG_96   LRG_96
##            LRG_97     5880       LRG_gene          LRG_97   LRG_97
##            LRG_98     5896       LRG_gene          LRG_98   LRG_98
##            LRG_99     5897       LRG_gene          LRG_99   LRG_99
##                   seqCoordSystem
##                            &lt;Rle&gt;
##   ENSG00000000003     chromosome
##   ENSG00000000005     chromosome
##   ENSG00000000419     chromosome
##   ENSG00000000457     chromosome
##   ENSG00000000460     chromosome
##               ...            ...
##            LRG_94     chromosome
##            LRG_96     chromosome
##            LRG_97     chromosome
##            LRG_98     chromosome
##            LRG_99     chromosome
##   -------
##   seqinfo: 273 sequences from GRCh37 genome</code></pre>
</div>
</div>
<div id="gene-mapping-functions" class="section level2">
<h2>Gene mapping functions</h2>
<p>Gene-to-symbol and transcript-to-gene mappings can be easily acquired with the <code>gene2symbol()</code> and <code>tx2gene()</code> functions. Both of these functions return a <code>data.frame</code>.</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">makeGene2SymbolFromEnsembl</span>(<span class="dt">organism =</span> <span class="st">&quot;Homo sapiens&quot;</span>)</code></pre>
<pre><code>## Gene2Symbol with 65687 rows and 2 columns
##                          geneID    geneName
##                     &lt;character&gt; &lt;character&gt;
## ENSG00000000003 ENSG00000000003      TSPAN6
## ENSG00000000005 ENSG00000000005        TNMD
## ENSG00000000419 ENSG00000000419        DPM1
## ENSG00000000457 ENSG00000000457       SCYL3
## ENSG00000000460 ENSG00000000460    C1orf112
## ...                         ...         ...
## LRG_995                 LRG_995     FUBP1.1
## LRG_996                 LRG_996     ERBB3.1
## LRG_997                 LRG_997      ROS1.1
## LRG_998                 LRG_998     CCND3.1
## LRG_999                 LRG_999       CIC.1</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">makeTx2GeneFromEnsembl</span>(<span class="dt">organism =</span> <span class="st">&quot;Homo sapiens&quot;</span>)</code></pre>
<pre><code>## Tx2Gene with 228432 rows and 2 columns
##                    transcriptID          geneID
##                     &lt;character&gt;     &lt;character&gt;
## ENST00000000233 ENST00000000233 ENSG00000004059
## ENST00000000412 ENST00000000412 ENSG00000003056
## ENST00000000442 ENST00000000442 ENSG00000173153
## ENST00000001008 ENST00000001008 ENSG00000004478
## ENST00000001146 ENST00000001146 ENSG00000003137
## ...                         ...             ...
## LRG_998t1             LRG_998t1         LRG_998
## LRG_998t2             LRG_998t2         LRG_998
## LRG_999t1             LRG_999t1         LRG_999
## LRG_999t2             LRG_999t2         LRG_999
## LRG_99t1               LRG_99t1          LRG_99</code></pre>
</div>
<div id="parse-gtfgff-files" class="section level2">
<h2>Parse GTF/GFF files</h2>
<p>For organisms that are not well supported on <a href="https://www.ensembl.org/">Ensembl</a>, we recommend using <a href="http://mblab.wustl.edu/GTF22.html" title="General Transfer Format">GTF</a> or <a href="https://useast.ensembl.org/info/website/upload/gff.html" title="General Feature Format">GFF</a> files for gene annotations. In general, <a href="http://mblab.wustl.edu/GTF22.html" title="General Transfer Format">GTF</a> files are more consistently standardized and recommended over GFF3, if possible. <code>makeTx2geneFromGFF()</code> and <code>makeGene2symbolFromGFF()</code> are utility functions that can return transcript-to-gene and gene-to-symbol mappings from the genomic ranges returned by <code>makeGRangesFromGFF()</code>.</p>
<div id="gtf" class="section level3">
<h3>GTF</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">makeGRangesFromGFF</span>(
    <span class="dt">file =</span> <span class="kw">pasteURL</span>(
        <span class="st">&quot;ftp.ensembl.org&quot;</span>,
        <span class="st">&quot;pub&quot;</span>,
        <span class="st">&quot;release-92&quot;</span>,
        <span class="st">&quot;gtf&quot;</span>,
        <span class="st">&quot;homo_sapiens&quot;</span>,
        <span class="st">&quot;Homo_sapiens.GRCh38.92.gtf.gz&quot;</span>,
        <span class="dt">protocol =</span> <span class="st">&quot;ftp&quot;</span>
    )
)</code></pre>
</div>
<div id="gff3" class="section level3">
<h3>GFF3</h3>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">makeGRangesFromGFF</span>(
    <span class="dt">file =</span> <span class="kw">pasteURL</span>(
        <span class="st">&quot;ftp.ensembl.org&quot;</span>,
        <span class="st">&quot;pub&quot;</span>,
        <span class="st">&quot;release-92&quot;</span>,
        <span class="st">&quot;gff3&quot;</span>,
        <span class="st">&quot;homo_sapiens&quot;</span>,
        <span class="st">&quot;Homo_sapiens.GRCh38.92.gff3.gz&quot;</span>,
        <span class="dt">protocol =</span> <span class="st">&quot;ftp&quot;</span>
    )
)</code></pre>
</div>
</div>
<div id="gene-interconversion-functions" class="section level2">
<h2>Gene interconversion functions</h2>
<p>Dealing with gene identifier to gene name (symbol) mapping in a non-destructive manner remains challenging for the bioinformatics community. <a href="https://basejump.acidgenomics.com/">basejump</a> provides <code>convertGenesToSymbols()</code> and <code>convertSymbolsToGenes()</code> generics that define methods for <code>SummarizedExperiment</code>, making interconversion between gene IDs and gene names easier. Internally, these mappings are handled by the <code>geneID</code> and <code>geneName</code> <code>mcols</code> columns inside of the <code>rowRanges</code> slot.</p>
<pre class="sourceCode r"><code class="sourceCode r">rse_symbols &lt;-<span class="st"> </span><span class="kw">convertGenesToSymbols</span>(rse)
<span class="kw">head</span>(<span class="kw">rownames</span>(rse_symbols))</code></pre>
<pre><code>## [1] &quot;TSPAN6&quot;   &quot;TNMD&quot;     &quot;DPM1&quot;     &quot;SCYL3&quot;    &quot;C1orf112&quot; &quot;FGR&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r">rse_genes &lt;-<span class="st"> </span><span class="kw">convertSymbolsToGenes</span>(rse_symbols)
<span class="kw">head</span>(<span class="kw">rownames</span>(rse_genes))</code></pre>
<pre><code>## [1] &quot;ENSG00000000003&quot; &quot;ENSG00000000005&quot; &quot;ENSG00000000419&quot; &quot;ENSG00000000457&quot;
## [5] &quot;ENSG00000000460&quot; &quot;ENSG00000000938&quot;</code></pre>
<p>To convert from transcript- to gene-level easily, the <code>convertTranscriptsToGenes()</code> and <code>stripTranscriptVersions()</code> functions are also provided (not shown).</p>
</div>
<div id="query-external-annotation-databases" class="section level2">
<h2>Query external annotation databases</h2>
<p>In addition to interfacing with <a href="https://www.ensembl.org/">Ensembl</a>, <a href="https://basejump.acidgenomics.com/">basejump</a> exports a number of utility functions to interface with other genomic databases, including:</p>
<ul>
<li><a href="#eggnog">EggNOG</a> <span class="citation">(<span class="citeproc-not-found" data-reference-id="Huerta-Cepas2016-ql"><strong>???</strong></span>)</span>.</li>
<li><a href="#hgnc">HGNC</a> <span class="citation">(<span class="citeproc-not-found" data-reference-id="Yates2017-fq"><strong>???</strong></span>)</span>.</li>
<li><a href="#mgi">MGI</a> <span class="citation">(<span class="citeproc-not-found" data-reference-id="Smith2018-ds"><strong>???</strong></span>)</span>.</li>
<li><a href="#panther">PANTHER</a> <span class="citation">(<span class="citeproc-not-found" data-reference-id="Mi2017-jl"><strong>???</strong></span>)</span>.</li>
</ul>
<div id="hgnc" class="section level3">
<h3>HGNC</h3>
<pre class="sourceCode r"><code class="sourceCode r">hgnc2gene &lt;-<span class="st"> </span><span class="kw">HGNC2Ensembl</span>()
<span class="kw">summary</span>(hgnc2gene)</code></pre>
<pre><code>##       Length        Class         Mode 
##            2 HGNC2Ensembl           S4</code></pre>
</div>
<div id="mgi" class="section level3">
<h3>MGI</h3>
<pre class="sourceCode r"><code class="sourceCode r">mgi2ensembl &lt;-<span class="st"> </span><span class="kw">MGI2Ensembl</span>()
<span class="kw">summary</span>(mgi2ensembl)</code></pre>
<pre><code>##      Length       Class        Mode 
##           2 MGI2Ensembl          S4</code></pre>
</div>
<div id="eggnog" class="section level3">
<h3>EggNOG</h3>
<pre class="sourceCode r"><code class="sourceCode r">eggnog &lt;-<span class="st"> </span><span class="kw">EggNOG</span>()
<span class="kw">summary</span>(eggnog)</code></pre>
<pre><code>## Length  Class   Mode 
##      2 EggNOG     S4</code></pre>
</div>
<div id="panther" class="section level3">
<h3>PANTHER</h3>
<pre class="sourceCode r"><code class="sourceCode r">panther &lt;-<span class="st"> </span><span class="kw">PANTHER</span>(<span class="dt">organism =</span> <span class="st">&quot;Homo sapiens&quot;</span>)
<span class="kw">summary</span>(panther)</code></pre>
<pre><code>##  Length   Class    Mode 
##       9 PANTHER      S4</code></pre>
</div>
<div id="wormbase" class="section level3">
<h3>WormBase</h3>
<p>Note that for <em>Caenorhabditis elegans</em> genome annotations, we recommend additional use of our specialized <a href="https://wormbase.acidgenomics.com/">wormbase package</a>, which queries the <a href="https://wormbase.org/">WormBase</a> database <span class="citation">(<span class="citeproc-not-found" data-reference-id="Stein2001-mk"><strong>???</strong></span>)</span> directly, and provides support of versioned releases (e.g. WS265) that have additional metadata not currently available on <a href="https://www.ensembl.org/">Ensembl</a>.</p>
</div>
</div>
</div>
<div id="importexport-functions" class="section level1">
<h1>Import/export functions</h1>
<p>FIXME This section is still under development.</p>
</div>
<div id="data-functions" class="section level1">
<h1>Data functions</h1>
<p>FIXME This section is still under development.</p>
</div>
<div id="sanitization-functions" class="section level1">
<h1>Sanitization functions</h1>
<p>The <code>makeNames</code> suite of function, consisting primarily of <code>camel()</code>, <code>dotted()</code>, <code>snake()</code>, and <code>upperCamel()</code>, provide S4 method support for sanitizing common data structures used in genomics research and on <a href="https://bioconductor.org/">Bioconductor</a>.</p>
<p>To see how these functions work, let’s load up an example <code>list</code> named <code>mn</code> (short for <code>makeNames</code>).</p>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">loadRemoteData</span>(<span class="dt">url =</span> <span class="kw">file.path</span>(basejumpTestsURL, <span class="st">&quot;mn.rda&quot;</span>))
<span class="kw">class</span>(mn)</code></pre>
<pre><code>## [1] &quot;list&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(mn)</code></pre>
<pre><code>## [1] &quot;character&quot;      &quot;namedCharacter&quot; &quot;factor&quot;         &quot;dataFrame&quot;     
## [5] &quot;matrix&quot;         &quot;tibble&quot;         &quot;list&quot;</code></pre>
<div id="atomic-vectors" class="section level2">
<h2>Atomic vectors</h2>
<div id="character" class="section level3">
<h3>Character</h3>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>mn$character
<span class="kw">print</span>(x)</code></pre>
<pre><code>##  [1] &quot;hello world&quot;     &quot;HELLO WORLD&quot;     &quot;RNAi clones&quot;    
##  [4] &quot;nCount&quot;          &quot;tx2gene&quot;         &quot;TX2GeneID&quot;      
##  [7] &quot;G2M.Score&quot;       &quot;worfdbHTMLRemap&quot; &quot;Mazda RX4&quot;      
## [10] &quot;%GC&quot;             &quot;5prime&quot;          &quot;5&#39;-3&#39; bias&quot;     
## [13] &quot;123&quot;             NA</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">camel</span>(x)</code></pre>
<pre><code>##  [1] &quot;helloWorld&quot;      &quot;helloWORLD&quot;      &quot;rnaiClones&quot;     
##  [4] &quot;nCount&quot;          &quot;tx2gene&quot;         &quot;tx2GeneID&quot;      
##  [7] &quot;g2mScore&quot;        &quot;worfdbHTMLRemap&quot; &quot;mazdaRX4&quot;       
## [10] &quot;percentGC&quot;       &quot;x5prime&quot;         &quot;x5x3Bias&quot;       
## [13] &quot;x123&quot;            NA</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dotted</span>(x)</code></pre>
<pre><code>##  [1] &quot;hello.world&quot;       &quot;HELLO.WORLD&quot;       &quot;RNAI.clones&quot;      
##  [4] &quot;n.Count&quot;           &quot;tx2gene&quot;           &quot;TX2.Gene.ID&quot;      
##  [7] &quot;G2M.Score&quot;         &quot;worfdb.HTML.Remap&quot; &quot;Mazda.RX4&quot;        
## [10] &quot;percent.GC&quot;        &quot;X5prime&quot;           &quot;X5.3.bias&quot;        
## [13] &quot;X123&quot;              NA</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">snake</span>(x)</code></pre>
<pre><code>##  [1] &quot;hello_world&quot;       &quot;hello_world&quot;       &quot;rnai_clones&quot;      
##  [4] &quot;n_count&quot;           &quot;tx2gene&quot;           &quot;tx2_gene_id&quot;      
##  [7] &quot;g2m_score&quot;         &quot;worfdb_html_remap&quot; &quot;mazda_rx4&quot;        
## [10] &quot;percent_gc&quot;        &quot;x5prime&quot;           &quot;x5_3_bias&quot;        
## [13] &quot;x123&quot;              NA</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upperCamel</span>(x)</code></pre>
<pre><code>##  [1] &quot;HelloWorld&quot;      &quot;HELLOWORLD&quot;      &quot;RNAIClones&quot;     
##  [4] &quot;NCount&quot;          &quot;Tx2gene&quot;         &quot;TX2GeneID&quot;      
##  [7] &quot;G2MScore&quot;        &quot;WorfdbHTMLRemap&quot; &quot;MazdaRX4&quot;       
## [10] &quot;PercentGC&quot;       &quot;X5prime&quot;         &quot;X5X3Bias&quot;       
## [13] &quot;X123&quot;            NA</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">makeNames</span>(x)</code></pre>
<pre><code>##  [1] &quot;hello_world&quot;     &quot;HELLO_WORLD&quot;     &quot;RNAi_clones&quot;    
##  [4] &quot;nCount&quot;          &quot;tx2gene&quot;         &quot;TX2GeneID&quot;      
##  [7] &quot;G2M_Score&quot;       &quot;worfdbHTMLRemap&quot; &quot;Mazda_RX4&quot;      
## [10] &quot;X_GC&quot;            &quot;X5prime&quot;         &quot;X5__3__bias&quot;    
## [13] &quot;X123&quot;            &quot;NA_&quot;</code></pre>
</div>
<div id="named-character" class="section level3">
<h3>Named character</h3>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>mn$namedCharacter
<span class="kw">print</span>(x)</code></pre>
<pre><code>##        Item.A        Item.B 
## &quot;hello world&quot; &quot;HELLO WORLD&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">camel</span>(x)</code></pre>
<pre><code>##        itemA        itemB 
## &quot;helloWorld&quot; &quot;helloWORLD&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dotted</span>(x)</code></pre>
<pre><code>##        Item.A        Item.B 
## &quot;hello.world&quot; &quot;HELLO.WORLD&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">snake</span>(x)</code></pre>
<pre><code>##        item_a        item_b 
## &quot;hello_world&quot; &quot;hello_world&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upperCamel</span>(x)</code></pre>
<pre><code>##        ItemA        ItemB 
## &quot;HelloWorld&quot; &quot;HELLOWORLD&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">makeNames</span>(x)</code></pre>
<pre><code>## [1] &quot;hello_world&quot; &quot;HELLO_WORLD&quot;</code></pre>
</div>
<div id="factor" class="section level3">
<h3>Factor</h3>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>mn$factor
<span class="kw">print</span>(x)</code></pre>
<pre><code>## sample 1 sample 2 sample 3 sample 4 
##  group 1  group 1  group 2  group 2 
## Levels: group 1 group 2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">camel</span>(x)</code></pre>
<pre><code>## sample1 sample2 sample3 sample4 
##  group1  group1  group2  group2 
## Levels: group1 group2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dotted</span>(x)</code></pre>
<pre><code>## sample.1 sample.2 sample.3 sample.4 
##  group.1  group.1  group.2  group.2 
## Levels: group.1 group.2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">snake</span>(x)</code></pre>
<pre><code>## sample_1 sample_2 sample_3 sample_4 
##  group_1  group_1  group_2  group_2 
## Levels: group_1 group_2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upperCamel</span>(x)</code></pre>
<pre><code>## Sample1 Sample2 Sample3 Sample4 
##  Group1  Group1  Group2  Group2 
## Levels: Group1 Group2</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">makeNames</span>(x)</code></pre>
<pre><code>## [1] &quot;group_1&quot;   &quot;group_1_1&quot; &quot;group_2&quot;   &quot;group_2_1&quot;</code></pre>
</div>
</div>
<div id="data-frames" class="section level2">
<h2>Data frames</h2>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>datasets::USArrests
<span class="kw">dimnames</span>(x)</code></pre>
<pre><code>## [[1]]
##  [1] &quot;Alabama&quot;        &quot;Alaska&quot;         &quot;Arizona&quot;        &quot;Arkansas&quot;      
##  [5] &quot;California&quot;     &quot;Colorado&quot;       &quot;Connecticut&quot;    &quot;Delaware&quot;      
##  [9] &quot;Florida&quot;        &quot;Georgia&quot;        &quot;Hawaii&quot;         &quot;Idaho&quot;         
## [13] &quot;Illinois&quot;       &quot;Indiana&quot;        &quot;Iowa&quot;           &quot;Kansas&quot;        
## [17] &quot;Kentucky&quot;       &quot;Louisiana&quot;      &quot;Maine&quot;          &quot;Maryland&quot;      
## [21] &quot;Massachusetts&quot;  &quot;Michigan&quot;       &quot;Minnesota&quot;      &quot;Mississippi&quot;   
## [25] &quot;Missouri&quot;       &quot;Montana&quot;        &quot;Nebraska&quot;       &quot;Nevada&quot;        
## [29] &quot;New Hampshire&quot;  &quot;New Jersey&quot;     &quot;New Mexico&quot;     &quot;New York&quot;      
## [33] &quot;North Carolina&quot; &quot;North Dakota&quot;   &quot;Ohio&quot;           &quot;Oklahoma&quot;      
## [37] &quot;Oregon&quot;         &quot;Pennsylvania&quot;   &quot;Rhode Island&quot;   &quot;South Carolina&quot;
## [41] &quot;South Dakota&quot;   &quot;Tennessee&quot;      &quot;Texas&quot;          &quot;Utah&quot;          
## [45] &quot;Vermont&quot;        &quot;Virginia&quot;       &quot;Washington&quot;     &quot;West Virginia&quot; 
## [49] &quot;Wisconsin&quot;      &quot;Wyoming&quot;       
## 
## [[2]]
## [1] &quot;Murder&quot;   &quot;Assault&quot;  &quot;UrbanPop&quot; &quot;Rape&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">camel</span>(x, <span class="dt">rownames =</span> <span class="ot">TRUE</span>, <span class="dt">colnames =</span> <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span><span class="kw">dimnames</span>()</code></pre>
<pre><code>## [[1]]
##  [1] &quot;alabama&quot;       &quot;alaska&quot;        &quot;arizona&quot;       &quot;arkansas&quot;     
##  [5] &quot;california&quot;    &quot;colorado&quot;      &quot;connecticut&quot;   &quot;delaware&quot;     
##  [9] &quot;florida&quot;       &quot;georgia&quot;       &quot;hawaii&quot;        &quot;idaho&quot;        
## [13] &quot;illinois&quot;      &quot;indiana&quot;       &quot;iowa&quot;          &quot;kansas&quot;       
## [17] &quot;kentucky&quot;      &quot;louisiana&quot;     &quot;maine&quot;         &quot;maryland&quot;     
## [21] &quot;massachusetts&quot; &quot;michigan&quot;      &quot;minnesota&quot;     &quot;mississippi&quot;  
## [25] &quot;missouri&quot;      &quot;montana&quot;       &quot;nebraska&quot;      &quot;nevada&quot;       
## [29] &quot;newHampshire&quot;  &quot;newJersey&quot;     &quot;newMexico&quot;     &quot;newYork&quot;      
## [33] &quot;northCarolina&quot; &quot;northDakota&quot;   &quot;ohio&quot;          &quot;oklahoma&quot;     
## [37] &quot;oregon&quot;        &quot;pennsylvania&quot;  &quot;rhodeIsland&quot;   &quot;southCarolina&quot;
## [41] &quot;southDakota&quot;   &quot;tennessee&quot;     &quot;texas&quot;         &quot;utah&quot;         
## [45] &quot;vermont&quot;       &quot;virginia&quot;      &quot;washington&quot;    &quot;westVirginia&quot; 
## [49] &quot;wisconsin&quot;     &quot;wyoming&quot;      
## 
## [[2]]
## [1] &quot;murder&quot;   &quot;assault&quot;  &quot;urbanPop&quot; &quot;rape&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dotted</span>(x, <span class="dt">rownames =</span> <span class="ot">TRUE</span>, <span class="dt">colnames =</span> <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span><span class="kw">dimnames</span>()</code></pre>
<pre><code>## [[1]]
##  [1] &quot;Alabama&quot;        &quot;Alaska&quot;         &quot;Arizona&quot;        &quot;Arkansas&quot;      
##  [5] &quot;California&quot;     &quot;Colorado&quot;       &quot;Connecticut&quot;    &quot;Delaware&quot;      
##  [9] &quot;Florida&quot;        &quot;Georgia&quot;        &quot;Hawaii&quot;         &quot;Idaho&quot;         
## [13] &quot;Illinois&quot;       &quot;Indiana&quot;        &quot;Iowa&quot;           &quot;Kansas&quot;        
## [17] &quot;Kentucky&quot;       &quot;Louisiana&quot;      &quot;Maine&quot;          &quot;Maryland&quot;      
## [21] &quot;Massachusetts&quot;  &quot;Michigan&quot;       &quot;Minnesota&quot;      &quot;Mississippi&quot;   
## [25] &quot;Missouri&quot;       &quot;Montana&quot;        &quot;Nebraska&quot;       &quot;Nevada&quot;        
## [29] &quot;New.Hampshire&quot;  &quot;New.Jersey&quot;     &quot;New.Mexico&quot;     &quot;New.York&quot;      
## [33] &quot;North.Carolina&quot; &quot;North.Dakota&quot;   &quot;Ohio&quot;           &quot;Oklahoma&quot;      
## [37] &quot;Oregon&quot;         &quot;Pennsylvania&quot;   &quot;Rhode.Island&quot;   &quot;South.Carolina&quot;
## [41] &quot;South.Dakota&quot;   &quot;Tennessee&quot;      &quot;Texas&quot;          &quot;Utah&quot;          
## [45] &quot;Vermont&quot;        &quot;Virginia&quot;       &quot;Washington&quot;     &quot;West.Virginia&quot; 
## [49] &quot;Wisconsin&quot;      &quot;Wyoming&quot;       
## 
## [[2]]
## [1] &quot;Murder&quot;    &quot;Assault&quot;   &quot;Urban.Pop&quot; &quot;Rape&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">snake</span>(x, <span class="dt">rownames =</span> <span class="ot">TRUE</span>, <span class="dt">colnames =</span> <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span><span class="kw">dimnames</span>()</code></pre>
<pre><code>## [[1]]
##  [1] &quot;alabama&quot;        &quot;alaska&quot;         &quot;arizona&quot;        &quot;arkansas&quot;      
##  [5] &quot;california&quot;     &quot;colorado&quot;       &quot;connecticut&quot;    &quot;delaware&quot;      
##  [9] &quot;florida&quot;        &quot;georgia&quot;        &quot;hawaii&quot;         &quot;idaho&quot;         
## [13] &quot;illinois&quot;       &quot;indiana&quot;        &quot;iowa&quot;           &quot;kansas&quot;        
## [17] &quot;kentucky&quot;       &quot;louisiana&quot;      &quot;maine&quot;          &quot;maryland&quot;      
## [21] &quot;massachusetts&quot;  &quot;michigan&quot;       &quot;minnesota&quot;      &quot;mississippi&quot;   
## [25] &quot;missouri&quot;       &quot;montana&quot;        &quot;nebraska&quot;       &quot;nevada&quot;        
## [29] &quot;new_hampshire&quot;  &quot;new_jersey&quot;     &quot;new_mexico&quot;     &quot;new_york&quot;      
## [33] &quot;north_carolina&quot; &quot;north_dakota&quot;   &quot;ohio&quot;           &quot;oklahoma&quot;      
## [37] &quot;oregon&quot;         &quot;pennsylvania&quot;   &quot;rhode_island&quot;   &quot;south_carolina&quot;
## [41] &quot;south_dakota&quot;   &quot;tennessee&quot;      &quot;texas&quot;          &quot;utah&quot;          
## [45] &quot;vermont&quot;        &quot;virginia&quot;       &quot;washington&quot;     &quot;west_virginia&quot; 
## [49] &quot;wisconsin&quot;      &quot;wyoming&quot;       
## 
## [[2]]
## [1] &quot;murder&quot;    &quot;assault&quot;   &quot;urban_pop&quot; &quot;rape&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upperCamel</span>(x, <span class="dt">rownames =</span> <span class="ot">TRUE</span>, <span class="dt">colnames =</span> <span class="ot">TRUE</span>) %&gt;%<span class="st"> </span><span class="kw">dimnames</span>()</code></pre>
<pre><code>## [[1]]
##  [1] &quot;Alabama&quot;       &quot;Alaska&quot;        &quot;Arizona&quot;       &quot;Arkansas&quot;     
##  [5] &quot;California&quot;    &quot;Colorado&quot;      &quot;Connecticut&quot;   &quot;Delaware&quot;     
##  [9] &quot;Florida&quot;       &quot;Georgia&quot;       &quot;Hawaii&quot;        &quot;Idaho&quot;        
## [13] &quot;Illinois&quot;      &quot;Indiana&quot;       &quot;Iowa&quot;          &quot;Kansas&quot;       
## [17] &quot;Kentucky&quot;      &quot;Louisiana&quot;     &quot;Maine&quot;         &quot;Maryland&quot;     
## [21] &quot;Massachusetts&quot; &quot;Michigan&quot;      &quot;Minnesota&quot;     &quot;Mississippi&quot;  
## [25] &quot;Missouri&quot;      &quot;Montana&quot;       &quot;Nebraska&quot;      &quot;Nevada&quot;       
## [29] &quot;NewHampshire&quot;  &quot;NewJersey&quot;     &quot;NewMexico&quot;     &quot;NewYork&quot;      
## [33] &quot;NorthCarolina&quot; &quot;NorthDakota&quot;   &quot;Ohio&quot;          &quot;Oklahoma&quot;     
## [37] &quot;Oregon&quot;        &quot;Pennsylvania&quot;  &quot;RhodeIsland&quot;   &quot;SouthCarolina&quot;
## [41] &quot;SouthDakota&quot;   &quot;Tennessee&quot;     &quot;Texas&quot;         &quot;Utah&quot;         
## [45] &quot;Vermont&quot;       &quot;Virginia&quot;      &quot;Washington&quot;    &quot;WestVirginia&quot; 
## [49] &quot;Wisconsin&quot;     &quot;Wyoming&quot;      
## 
## [[2]]
## [1] &quot;Murder&quot;   &quot;Assault&quot;  &quot;UrbanPop&quot; &quot;Rape&quot;</code></pre>
</div>
<div id="lists" class="section level2">
<h2>Lists</h2>
<pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span>mn$list
<span class="kw">print</span>(x)</code></pre>
<pre><code>## $Item.A
## [1] 1 2
## 
## $Item.B
## [1] 3 4</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">camel</span>(x) %&gt;%<span class="st"> </span><span class="kw">names</span>()</code></pre>
<pre><code>## [1] &quot;itemA&quot; &quot;itemB&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dotted</span>(x) %&gt;%<span class="st"> </span><span class="kw">names</span>()</code></pre>
<pre><code>## [1] &quot;Item.A&quot; &quot;Item.B&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">snake</span>(x) %&gt;%<span class="st"> </span><span class="kw">names</span>()</code></pre>
<pre><code>## [1] &quot;item_a&quot; &quot;item_b&quot;</code></pre>
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">upperCamel</span>(x) %&gt;%<span class="st"> </span><span class="kw">names</span>()</code></pre>
<pre><code>## [1] &quot;ItemA&quot; &quot;ItemB&quot;</code></pre>
</div>
</div>
<div id="r-session-information" class="section level1">
<h1>R session information</h1>
<pre class="sourceCode r"><code class="sourceCode r">utils::<span class="kw">sessionInfo</span>()</code></pre>
<pre><code>## R version 3.5.2 (2018-12-20)
## Platform: x86_64-redhat-linux-gnu (64-bit)
## Running under: Red Hat Enterprise Linux
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    datasets  utils     grDevices graphics  stats    
## [8] methods   base     
## 
## other attached packages:
##  [1] SummarizedExperiment_1.12.0 DelayedArray_0.8.0         
##  [3] BiocParallel_1.16.6         matrixStats_0.54.0         
##  [5] Biobase_2.42.0              GenomicRanges_1.34.0       
##  [7] GenomeInfoDb_1.18.2         IRanges_2.16.0             
##  [9] S4Vectors_0.20.1            BiocGenerics_0.28.0        
## [11] basejump_0.10.3             BiocStyle_2.10.0           
## 
## loaded via a namespace (and not attached):
##  [1] ProtGenerics_1.14.0           bitops_1.0-6                 
##  [3] fs_1.2.7                      bit64_0.9-7                  
##  [5] progress_1.2.0                httr_1.4.0                   
##  [7] rprojroot_1.3-2               syntactic_0.1.8              
##  [9] tools_3.5.2                   backports_1.1.4              
## [11] R6_2.4.0                      lazyeval_0.2.2               
## [13] DBI_1.0.0                     withr_2.1.2                  
## [15] tidyselect_0.2.5              prettyunits_1.0.2            
## [17] bit_1.1-14                    curl_3.3                     
## [19] compiler_3.5.2                cli_1.1.0                    
## [21] xml2_1.2.0                    desc_1.2.0                   
## [23] rtracklayer_1.42.2            bookdown_0.9                 
## [25] readr_1.3.1                   pkgdown_1.3.0                
## [27] commonmark_1.7                stringr_1.4.0                
## [29] digest_0.6.18                 Rsamtools_1.34.1             
## [31] rmarkdown_1.12                R.utils_2.8.0                
## [33] XVector_0.22.0                pkgconfig_2.0.2              
## [35] htmltools_0.3.6               sessioninfo_1.1.1            
## [37] ensembldb_2.6.8               rlang_0.3.4                  
## [39] RSQLite_2.1.1                 shiny_1.3.1                  
## [41] dplyr_0.8.0.1                 R.oo_1.22.0                  
## [43] RCurl_1.95-4.12               magrittr_1.5                 
## [45] GenomeInfoDbData_1.2.0        Matrix_1.2-17                
## [47] Rcpp_1.0.1                    R.methodsS3_1.7.1            
## [49] stringi_1.4.3                 yaml_2.2.0                   
## [51] MASS_7.3-51.3                 zlibbioc_1.28.0              
## [53] brio_0.1.18                   plyr_1.8.4                   
## [55] goalie_0.2.14                 AnnotationHub_2.14.5         
## [57] grid_3.5.2                    blob_1.1.1                   
## [59] transformer_0.1.8             promises_1.0.1               
## [61] crayon_1.3.4                  lattice_0.20-38              
## [63] Biostrings_2.50.2             bioverbs_0.1.12              
## [65] GenomicFeatures_1.34.8        hms_0.4.2                    
## [67] knitr_1.22                    pillar_1.3.1                 
## [69] reshape2_1.4.3                biomaRt_2.38.0               
## [71] XML_3.98-1.19                 glue_1.3.1                   
## [73] evaluate_0.13                 freerange_0.1.2              
## [75] data.table_1.12.2             BiocManager_1.30.4           
## [77] httpuv_1.5.1                  tidyr_0.8.3                  
## [79] grr_0.9.5                     purrr_0.3.2                  
## [81] assertthat_0.2.1              xfun_0.6                     
## [83] mime_0.6                      xtable_1.8-3                 
## [85] AnnotationFilter_1.6.0        roxygen2_6.1.1               
## [87] later_0.8.0                   SingleCellExperiment_1.4.1   
## [89] tibble_2.1.1                  GenomicAlignments_1.18.1     
## [91] Matrix.utils_0.9.7            AnnotationDbi_1.44.0         
## [93] memoise_1.1.0                 interactiveDisplayBase_1.20.0</code></pre>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/agxufd">shared library</a> on <a href="https://paperpile.com/">Paperpile</a>.</p>
<div class="references">

</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2>Contents</h2>
      <ul>
      <li><a href="#introduction">Introduction</a></li>
      <li><a href="#the-s4-object-system">The S4 object system</a></li>
      <li><a href="#annotation-functions">Annotation functions</a><ul>
      <li><a href="#ensembl-annotations">Ensembl annotations</a></li>
      <li><a href="#gene-mapping-functions">Gene mapping functions</a></li>
      <li><a href="#parse-gtfgff-files">Parse GTF/GFF files</a></li>
      <li><a href="#gene-interconversion-functions">Gene interconversion functions</a></li>
      <li><a href="#query-external-annotation-databases">Query external annotation databases</a></li>
      </ul></li>
      <li><a href="#importexport-functions">Import/export functions</a></li>
      <li><a href="#data-functions">Data functions</a></li>
      <li><a href="#sanitization-functions">Sanitization functions</a><ul>
      <li><a href="#atomic-vectors">Atomic vectors</a></li>
      <li><a href="#data-frames">Data frames</a></li>
      <li><a href="#lists">Lists</a></li>
      </ul></li>
      <li><a href="#r-session-information">R session information</a></li>
      <li><a href="#references">References</a></li>
      </ul>
    </div>
      </div>

</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>
