<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Ensembl Annotations — annotable • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Ensembl Annotations — annotable" />

<meta property="og:description" content="Ensembl Annotations" />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Ensembl Annotations</h1>
    <small>Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/annotable.R'><code>R/annotable.R</code></a></small>
    </div>

    
    <p>Ensembl Annotations</p>
    

    <pre class="usage"><span class='fu'>annotable</span>(<span class='no'>organism</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>), <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>organism</th>
      <td><p>Full latin organism name (e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p>Fetch ranges as "<code>genes</code>" or "<code>transcripts</code>".</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><em>Optional.</em> Genome assembly name (e.g. "<code>GRCh38</code>"). If
<code>NULL</code>, defaults to the most recent build available.</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p><em>Optional.</em> Release version (e.g. <code>90</code>). If <code>NULL</code>, defaults
to the most recent release available.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>data.frame</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Simply specify the desired organism, using the full latin name. For example,
we can obtain human annotations with <code>Homo sapiens</code>. Optionally, specific
Ensembl genome builds (e.g. <code>GRCh38</code>) and release versions (e.g. <code>87</code>) are
supported.</p>
<p>Under the hood, this function fetches annotations from AnnotationHub using
the ensembldb package. AnnotationHub supports versioned Ensembl releases,
back to version 87.</p>
    
    <h2 class="hasAnchor" id="broad-class-definitions"><a class="anchor" href="#broad-class-definitions"></a>Broad Class Definitions</h2>

    
    <p>For gene and transcript tables, when <code>broadClass = TRUE</code>, a <code>broadClass</code>
column is added, which generalizes the gene types into a smaller number of
semantically-meaningful groups:</p><ul>
<li><p><code>coding</code></p></li>
<li><p><code>noncoding</code></p></li>
<li><p><code>pseudo</code></p></li>
<li><p><code>small</code></p></li>
<li><p><code>decaying</code></p></li>
<li><p><code>ig</code> (immunoglobulin)</p></li>
<li><p><code>tcr</code> (T cell receptor)</p></li>
<li><p><code>other</code></p></li>
</ul>
    
    <h2 class="hasAnchor" id="grch-legacy-annotations"><a class="anchor" href="#grch-legacy-annotations"></a>GRCh37 Legacy Annotations</h2>

    
    <p>The <em>Homo sapiens</em> GRCh37 (release 75) build is supported by internally
querying the
<a href='https://doi.org/doi:10.18129/B9.bioc.EnsDb.Hsapiens.v75'>EnsDb.Hsapiens.v75</a>
package.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Gene Functions: <code>convertGenesToSymbols</code>,
  <code>convertTranscriptsToGenes</code>,
  <code>detectOrganism</code>,
  <code>stripTranscriptVersions</code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Genes ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>annotable</span>(<span class='st'>"Homo sapiens"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl with AnnotationHub 2.10.1 (2017-10-27)</span></div><div class='output co'>#&gt; <span class='message'>id: "AH57757"</span>
#&gt; <span class='message'>organism: "Homo sapiens"</span>
#&gt; <span class='message'>build: "GRCh38"</span>
#&gt; <span class='message'>release: "90"</span>
#&gt; <span class='message'>format: "genes"</span></div><div class='output co'>#&gt; <span class='message'>Using geneName instead of symbol</span></div><div class='output co'>#&gt; <span class='message'>Setting geneBiotype as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting seqCoordSystem as factor</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, geneBiotype, seqnames</span></div><div class='input'><span class='fu'><a href='http://www.rdocumentation.org/packages/tibble/topics/glimpse'>glimpse</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Observations: 64,661
#&gt; Variables: 12
#&gt; $ seqnames       &lt;fct&gt; X, X, 20, 1, 1, 1, 1, 6, 6, 6, 1, 1, X, 6, 3, 7, 7, ...
#&gt; $ start          &lt;int&gt; 100627109, 100584802, 50934867, 169849631, 169662007...
#&gt; $ end            &lt;int&gt; 100639991, 100599885, 50958555, 169894267, 169854080...
#&gt; $ width          &lt;int&gt; 12883, 15084, 23689, 44637, 192074, 23214, 95627, 16...
#&gt; $ strand         &lt;fct&gt; -, +, -, -, +, -, +, -, -, +, -, +, -, +, +, +, +, -...
#&gt; $ geneID         &lt;chr&gt; "ENSG00000000003", "ENSG00000000005", "ENSG000000004...
#&gt; $ geneName       &lt;chr&gt; "TSPAN6", "TNMD", "DPM1", "SCYL3", "C1orf112", "FGR"...
#&gt; $ geneBiotype    &lt;fct&gt; protein_coding, protein_coding, protein_coding, prot...
#&gt; $ description    &lt;chr&gt; "tetraspanin 6 [Source:HGNC Symbol;Acc:HGNC:11858]",...
#&gt; $ seqCoordSystem &lt;fct&gt; chromosome, chromosome, chromosome, chromosome, chro...
#&gt; $ entrezID       &lt;I(list)&gt;  7105, 64102,  8813, 57147, 55732,  2268,  3075,...
#&gt; $ broadClass     &lt;fct&gt; coding, coding, coding, coding, coding, coding, codi...</div><div class='input'>
<span class='co'># Transcripts ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>annotable</span>(<span class='st'>"Homo sapiens"</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"transcripts"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl with AnnotationHub 2.10.1 (2017-10-27)</span></div><div class='output co'>#&gt; <span class='message'>id: "AH57757"</span>
#&gt; <span class='message'>organism: "Homo sapiens"</span>
#&gt; <span class='message'>build: "GRCh38"</span>
#&gt; <span class='message'>release: "90"</span>
#&gt; <span class='message'>format: "transcripts"</span></div><div class='output co'>#&gt; <span class='message'>Using geneName instead of symbol</span></div><div class='output co'>#&gt; <span class='message'>Setting geneBiotype as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting seqCoordSystem as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting txBiotype as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting txSupportLevel as factor</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, txBiotype, seqnames</span></div><div class='input'><span class='fu'><a href='http://www.rdocumentation.org/packages/tibble/topics/glimpse'>glimpse</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Observations: 220,144
#&gt; Variables: 18
#&gt; $ seqnames       &lt;fct&gt; 7, 12, 11, 12, 2, 2, 6, 16, 4, 3, 7, 7, 11, 1, 1, 7,...
#&gt; $ start          &lt;int&gt; 127588345, 8940365, 64305578, 2794953, 72129238, 372...
#&gt; $ end            &lt;int&gt; 127591705, 8949955, 64316738, 2805423, 72148038, 372...
#&gt; $ width          &lt;int&gt; 3361, 9591, 11161, 10471, 18801, 17495, 16880, 14609...
#&gt; $ strand         &lt;fct&gt; +, -, +, +, -, +, -, -, -, +, +, -, -, -, +, +, +, -...
#&gt; $ txID           &lt;chr&gt; "ENST00000000233", "ENST00000000412", "ENST000000004...
#&gt; $ txName         &lt;chr&gt; "ENST00000000233", "ENST00000000412", "ENST000000004...
#&gt; $ txBiotype      &lt;fct&gt; protein_coding, protein_coding, protein_coding, prot...
#&gt; $ geneID         &lt;chr&gt; "ENSG00000004059", "ENSG00000003056", "ENSG000001731...
#&gt; $ geneName       &lt;chr&gt; "ARF5", "M6PR", "ESRRA", "FKBP4", "CYP26B1", "NDUFAF...
#&gt; $ geneBiotype    &lt;fct&gt; protein_coding, protein_coding, protein_coding, prot...
#&gt; $ description    &lt;chr&gt; "ADP ribosylation factor 5 [Source:HGNC Symbol;Acc:H...
#&gt; $ txCdsSeqStart  &lt;int&gt; 127588499, 8941818, 64307180, 2795140, 72132227, 372...
#&gt; $ txCdsSeqEnd    &lt;int&gt; 127591299, 8946404, 64315966, 2803258, 72147834, 372...
#&gt; $ txSupportLevel &lt;fct&gt; 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 5, 1, 1, 1, 1...
#&gt; $ seqCoordSystem &lt;fct&gt; chromosome, chromosome, chromosome, chromosome, chro...
#&gt; $ entrezID       &lt;I(list)&gt;    381,   4074,   2101,   2288,  56603,  55471, ...
#&gt; $ broadClass     &lt;fct&gt; coding, coding, coding, coding, coding, coding, codi...</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#broad-class-definitions">Broad Class Definitions</a></li>

      <li><a href="#grch-legacy-annotations">GRCh37 Legacy Annotations</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

