<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Ensembl Annotations — annotable • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">basejump</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Ensembl Annotations</h1>
    </div>

    
    <p>Quickly obtain gene annotations from <a href='http://www.ensembl.org/'>Ensembl</a>.</p>
    

    <pre class="usage"><span class='fu'>annotable</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for character</span>
<span class='fu'>annotable</span>(<span class='no'>object</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"gene"</span>, <span class='no'>release</span>,
  <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='co'># S4 method for data.frame</span>
<span class='fu'>annotable</span>(<span class='no'>object</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object. Default recommended usage is to provide full latin
organism name as a string. Also supports input of pre-built tibbles from
the annotables data package.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p><em>Additional arguments (for the S4 generic definition).</em></p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p>Desired table format, either <code>gene</code>, <code>tx2gene</code>, or
<code>gene2symbol</code>.</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p><em>Optional</em>. Ensembl release version. Defaults to the most
current release available on AnnotationHub.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>If <code>TRUE</code>, suppress any status messages and/or progress bars.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>data.frame with unique rows per gene or transcript.</p>
    
    <h2 class="hasAnchor" id="using-annotationhub"><a class="anchor" href="#using-annotationhub"></a>Using AnnotationHub</h2>

    
    <p>The default recommended approach is to simply specify the desired organism,
using the full latin name. For example, we can obtain human annotations with
<code>Homo sapiens</code> and mouse annotations with <code>Mus musculus</code>. Under the hood, the
function fetches these annotations from AnnotationHub using the ensembldb
package. AnnotationHub supports versioned Ensembl releases, back to version
87.</p>
    
    <h2 class="hasAnchor" id="using-pre-compiled-annotables"><a class="anchor" href="#using-pre-compiled-annotables"></a>Using Pre-Compiled Annotables</h2>

    
    <p>Generally we' don't recommend this approach. However, pass-in of a tibble
generated using <a href='https://doi.org/doi:10.18129/B9.bioc.biomaRt'>biomaRt</a> in
the <a href='https://github.com/stephenturner/annotables'>annotables</a> data package is
useful for adding support for the outdated (but still frequently used)
GRCh37/hg19 human genome.</p>
    
    <h2 class="hasAnchor" id="broad-class-definitions"><a class="anchor" href="#broad-class-definitions"></a>Broad Class Definitions</h2>

    
    <p>A <code>broadClass</code> column is added, which generalizes the gene types into a
smaller number of semantically-meaningful groups:</p><ul>
<li><p><code>coding</code></p></li>
<li><p><code>noncoding</code></p></li>
<li><p><code>pseudo</code></p></li>
<li><p><code>small</code></p></li>
<li><p><code>decaying</code></p></li>
<li><p><code>ig</code> (immunoglobulin)</p></li>
<li><p><code>tcr</code> (T cell receptor)</p></li>
<li><p><code>other</code></p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <ul>
<li><p><a href='https://doi.org/doi:10.18129/B9.bioc.AnnotationHub'>AnnotationHub</a>.</p></li>
<li><p><a href='https://doi.org/doi:10.18129/B9.bioc.ensembldb'>ensembldb</a>.</p></li>
</ul>
    <p>Other Gene Annotation Utilities: <code><a href='detectOrganism.html'>detectOrganism</a></code>,
  <code><a href='gene2symbolFromGFF.html'>gene2symbolFromGFF</a></code>,
  <code><a href='gene2symbol.html'>gene2symbol</a></code>, <code><a href='symbol2gene.html'>symbol2gene</a></code>,
  <code><a href='tx2geneFromGFF.html'>tx2geneFromGFF</a></code>, <code><a href='tx2gene.html'>tx2gene</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>annotable</span>(<span class='st'>"Mus musculus"</span>) <span class='kw'>%&gt;%</span> <span class='fu'>str</span>()</div><div class='output co'>#&gt; <span class='message'>Loading Ensembl annotations from AnnotationHub</span>
#&gt; <span class='message'>2017-10-27</span></div><div class='output co'>#&gt; <span class='message'>EnsDB AH57770: Mus musculus Ensembl 90</span></div><div class='output co'>#&gt; 'data.frame':	53110 obs. of  10 variables:
#&gt;  $ ensgene       : chr  "ENSMUSG00000000001" "ENSMUSG00000000003" "ENSMUSG00000000028" "ENSMUSG00000000031" ...
#&gt;  $ symbol        : chr  "Gnai3" "Pbsn" "Cdc45" "H19" ...
#&gt;  $ description   : chr  "guanine nucleotide binding protein (G protein), alpha inhibiting 3 [Source:MGI Symbol;Acc:MGI:95773]" "probasin [Source:MGI Symbol;Acc:MGI:1860484]" "cell division cycle 45 [Source:MGI Symbol;Acc:MGI:1338073]" "H19, imprinted maternally expressed transcript [Source:MGI Symbol;Acc:MGI:95891]" ...
#&gt;  $ biotype       : chr  "protein_coding" "protein_coding" "protein_coding" "lincRNA" ...
#&gt;  $ broadClass    : chr  "coding" "coding" "coding" "noncoding" ...
#&gt;  $ geneSeqStart  : int  108107280 77837901 18780447 142575529 161117193 108343354 121237253 17281185 5861482 120405281 ...
#&gt;  $ geneSeqEnd    : int  108146146 77853623 18811987 142578143 161258213 108414396 121255856 17289115 5870394 120530186 ...
#&gt;  $ seqName       : chr  "3" "X" "16" "7" ...
#&gt;  $ seqStrand     : int  -1 -1 -1 -1 1 1 1 1 1 1 ...
#&gt;  $ seqCoordSystem: chr  "chromosome" "chromosome" "chromosome" "chromosome" ...</div><div class='input'>
# Alternate approach, with pre-compiled tibble
</div><span class='co'># NOT RUN {</span>
<span class='fu'>annotable</span>(<span class='kw pkg'>annotables</span><span class='kw ns'>::</span><span class='no'><a href='http://www.rdocumentation.org/packages/annotables/topics/grch37'>grch37</a></span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#using-annotationhub">Using AnnotationHub</a></li>

      <li><a href="#using-pre-compiled-annotables">Using Pre-Compiled Annotables</a></li>

      <li><a href="#broad-class-definitions">Broad Class Definitions</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
