<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Convert Ensembl Identifiers to Gene Symbols — convertGenesToSymbols • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Convert Ensembl Identifiers to Gene Symbols — convertGenesToSymbols" />

<meta property="og:description" content="Convert Ensembl Identifiers to Gene Symbols" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Convert Ensembl Identifiers to Gene Symbols</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/steinbaugh/basejump/blob/master/R/convertGenesToSymbols-methods.R'><code>R/convertGenesToSymbols-methods.R</code></a></small>
    <div class="hidden name"><code>convertGenesToSymbols.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Convert Ensembl Identifiers to Gene Symbols</p>
    
    </div>

    <pre class="usage"><span class='fu'>convertGenesToSymbols</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>convertSymbolsToGenes</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for character</span>
<span class='fu'>convertGenesToSymbols</span>(<span class='no'>object</span>, <span class='kw'>gene2symbol</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>build</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># S4 method for matrix</span>
<span class='fu'>convertGenesToSymbols</span>(<span class='no'>object</span>, <span class='kw'>gene2symbol</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>build</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># S4 method for sparseMatrix</span>
<span class='fu'>convertGenesToSymbols</span>(<span class='no'>object</span>,
  <span class='kw'>gene2symbol</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>build</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>convertGenesToSymbols</span>(<span class='no'>object</span>)

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>convertSymbolsToGenes</span>(<span class='no'>object</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>gene2symbol</th>
      <td><p><code>Gene2Symbol</code> or <code>NULL</code>. Gene-to-symbol mappings. If set
<code>NULL</code>, the function will attempt to download the mappings from Ensembl
automatically.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code>string</code>. Full latin organism name (e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>build</th>
      <td><p><code>string</code> or <code>NULL</code>. Genome build assembly name (e.g.
"<code>GRCh38</code>"). If set <code>NULL</code>, defaults to the most recent build available.</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p><code>scalar integer</code> or <code>NULL</code>. Release version (e.g. <code>90</code>). If
set <code>NULL</code>, defaults to the most recent release available.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Object with gene IDs converted to names (symbols).</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Identifier Mapping Functions: <code><a href='Ensembl2Entrez.html'>Ensembl2Entrez</a></code>,
  <code><a href='Gene2Symbol.html'>Gene2Symbol</a></code>, <code><a href='HGNC2Ensembl.html'>HGNC2Ensembl</a></code>,
  <code><a href='MGI2Ensembl.html'>MGI2Ensembl</a></code>, <code><a href='Tx2Gene.html'>Tx2Gene</a></code>,
  <code><a href='convertTranscriptsToGenes.html'>convertTranscriptsToGenes</a></code>,
  <code><a href='convertUCSCBuildToEnsembl.html'>convertUCSCBuildToEnsembl</a></code>,
  <code><a href='mapGenes.html'>mapGenes</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'>data</span>(<span class='no'>rse_small</span>)
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='no'>rse_small</span>

<span class='no'>gene2symbol</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='Gene2Symbol.html'>gene2symbol</a></span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: 'gene2symbol' is deprecated.</span>
#&gt; <span class='warning'>Use 'Gene2Symbol' instead.</span>
#&gt; <span class='warning'>See help("Deprecated")</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>gene2symbol</span>)</div><div class='output co'>#&gt; Gene2Symbol 0.99.0
#&gt; https://steinbaugh.com/basejump
#&gt; citation("basejump")
#&gt; DataFrame with 50 rows and 2 columns
#&gt;                 geneID    geneName
#&gt;            &lt;character&gt; &lt;character&gt;
#&gt; gene1  ENSG00000000003      TSPAN6
#&gt; gene2  ENSG00000000005        TNMD
#&gt; gene3  ENSG00000000419        DPM1
#&gt; gene4  ENSG00000000457       SCYL3
#&gt; gene5  ENSG00000000460    C1orf112
#&gt; ...                ...         ...
#&gt; gene46 ENSG00000004059        ARF5
#&gt; gene47 ENSG00000004139       SARM1
#&gt; gene48 ENSG00000004142     POLDIP2
#&gt; gene49 ENSG00000004399      PLXND1
#&gt; gene50 ENSG00000004455         AK2
#&gt; 50 genes; 50 symbols</div><div class='input'><span class='no'>genes</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='no'>gene2symbol</span><span class='kw'>[[</span><span class='st'>"geneID"</span>]])

<span class='co'>## character ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>convertGenesToSymbols</span>(<span class='no'>genes</span>, <span class='kw'>gene2symbol</span> <span class='kw'>=</span> <span class='no'>gene2symbol</span>)
<span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; ENSG00000000003 ENSG00000000005 ENSG00000000419 ENSG00000000457 ENSG00000000460 
#&gt;        "TSPAN6"          "TNMD"          "DPM1"         "SCYL3"      "C1orf112" 
#&gt; ENSG00000000938 
#&gt;           "FGR" </div><div class='input'>
<span class='co'>## matrix ====</span>
<span class='no'>samples</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='fu'>colnames</span>(<span class='no'>object</span>))
<span class='no'>counts</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(
    <span class='kw'>data</span> <span class='kw'>=</span> <span class='fu'>seq_len</span>(<span class='fu'>length</span>(<span class='no'>genes</span>) * <span class='fu'>length</span>(<span class='no'>samples</span>)),
    <span class='kw'>byrow</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
    <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='fu'>length</span>(<span class='no'>genes</span>),
    <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fu'>length</span>(<span class='no'>samples</span>),
    <span class='kw'>dimnames</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='no'>genes</span>, <span class='no'>samples</span>)
)
<span class='fu'>print</span>(<span class='no'>counts</span>)</div><div class='output co'>#&gt;                 sample1 sample2 sample3 sample4
#&gt; ENSG00000000003       1       2       3       4
#&gt; ENSG00000000005       5       6       7       8
#&gt; ENSG00000000419       9      10      11      12
#&gt; ENSG00000000457      13      14      15      16
#&gt; ENSG00000000460      17      18      19      20
#&gt; ENSG00000000938      21      22      23      24</div><div class='input'><span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>convertGenesToSymbols</span>(<span class='no'>counts</span>, <span class='kw'>gene2symbol</span> <span class='kw'>=</span> <span class='no'>gene2symbol</span>)
<span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt;          sample1 sample2 sample3 sample4
#&gt; TSPAN6         1       2       3       4
#&gt; TNMD           5       6       7       8
#&gt; DPM1           9      10      11      12
#&gt; SCYL3         13      14      15      16
#&gt; C1orf112      17      18      19      20
#&gt; FGR           21      22      23      24</div><div class='input'>
<span class='co'>## SummarizedExperiment ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>convertGenesToSymbols</span>(<span class='no'>rse_small</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: 'gene2symbol' is deprecated.</span>
#&gt; <span class='warning'>Use 'Gene2Symbol' instead.</span>
#&gt; <span class='warning'>See help("Deprecated")</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 50 4 
#&gt; metadata(3): version date interestingGroups
#&gt; assays(1): counts
#&gt; rownames(50): TSPAN6 TNMD ... PLXND1 AK2
#&gt; rowData names(5): geneID geneName geneBiotype broadClass entrezID
#&gt; colnames(4): sample1 sample2 sample3 sample4
#&gt; colData names(2): genotype treatment</div><div class='input'><span class='co'>## Interconvert back to gene IDs.</span>
<span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'>convertSymbolsToGenes</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: 'gene2symbol' is deprecated.</span>
#&gt; <span class='warning'>Use 'Gene2Symbol' instead.</span>
#&gt; <span class='warning'>See help("Deprecated")</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>y</span>)</div><div class='output co'>#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 50 4 
#&gt; metadata(3): version date interestingGroups
#&gt; assays(1): counts
#&gt; rownames(50): ENSG00000000003 ENSG00000000005 ... ENSG00000004399
#&gt;   ENSG00000004455
#&gt; rowData names(5): geneID geneName geneBiotype broadClass entrezID
#&gt; colnames(4): sample1 sample2 sample3 sample4
#&gt; colData names(2): genotype treatment</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

