<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Convert Ensembl Transcripts to Genes — convertTranscriptsToGenes • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Convert Ensembl Transcripts to Genes — convertTranscriptsToGenes" />

<meta property="og:description" content="Convert Ensembl Transcripts to Genes" />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.5.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Convert Ensembl Transcripts to Genes</h1>
    <small>Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/steinbaugh/basejump/blob/master/R/methods-convertTranscriptsToGenes.R'><code>R/methods-convertTranscriptsToGenes.R</code></a></small>
    <div class="hidden name"><code>convertTranscriptsToGenes.Rd</code></div>
    </div>

    
    <p>Convert Ensembl Transcripts to Genes</p>
    

    <pre class="usage"><span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for character</span>
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>, <span class='kw'>tx2gene</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># S4 method for matrix</span>
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for data.frame</span>
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for DataFrame</span>
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for dgCMatrix</span>
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for dgTMatrix</span>
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>object</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>tx2gene</th>
      <td><p><em>Optional.</em> Transcript-to-gene mappings. If <code>NULL</code>, will
attempt to download from Ensembl using the desired <code>organism</code>,
<code>genomeBuild</code>, and <code>release</code> arguments.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><em>Optional.</em> Organism name. Normally this argument is
unnecessary and can be left unset. If a count matrix starts with a
FASTA spike-in (e.g. "EGFP"), then automatic genome detection based on the
first gene identifier will fail. In this case, the desired organism must be
manually declared.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><em>Optional.</em> Genome build assembly name (e.g. "<code>GRCh38</code>").
If <code>NULL</code>, defaults to the most recent build available.</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p><em>Optional.</em> Release version (e.g. <code>90</code>). If <code>NULL</code>, defaults
to the most recent release available.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Same class as original object.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Gene Annotation Functions: <code><a href='annotable.html'>annotable</a></code>,
  <code>broadClass</code>,
  <code>convertGenesToSymbols</code>,
  <code><a href='convertUCSCBuildToEnsembl.html'>convertUCSCBuildToEnsembl</a></code>,
  <code>detectOrganism</code>, <code><a href='emptyRanges.html'>emptyRanges</a></code>,
  <code><a href='hgnc2gene.html'>hgnc2gene</a></code>,
  <code><a href='makeGRangesFromEnsembl.html'>makeGRangesFromEnsembl</a></code>,
  <code><a href='makeGRangesFromGFF.html'>makeGRangesFromGFF</a></code>,
  <code><a href='makeGene2symbolFromEnsembl.html'>makeGene2symbolFromEnsembl</a></code>,
  <code><a href='makeGene2symbolFromGFF.html'>makeGene2symbolFromGFF</a></code>,
  <code><a href='makeTx2geneFromEnsembl.html'>makeTx2geneFromEnsembl</a></code>,
  <code><a href='makeTx2geneFromGFF.html'>makeTx2geneFromGFF</a></code>, <code><a href='mgi2gene.html'>mgi2gene</a></code>,
  <code><a href='panther.html'>panther</a></code>, <code><a href='parseGFFAttributes.html'>parseGFFAttributes</a></code>,
  <code>stripTranscriptVersions</code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># character ====</span>
<span class='no'>transcripts</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(
    <span class='st'>"ENSMUST00000000001"</span>,
    <span class='st'>"ENSMUST00000000003"</span>,
    <span class='st'>"ENSMUST00000114041"</span>
)
<span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>transcripts</span>)</div><div class='output co'>#&gt; <span class='message'>Obtaining transcript-to-gene mappings from Ensembl</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl with AnnotationHub 2.10.1 (2017-10-27)</span></div><div class='output co'>#&gt; <span class='message'>id: "AH57770"</span>
#&gt; <span class='message'>organism: "Mus musculus"</span>
#&gt; <span class='message'>build: "GRCm38"</span>
#&gt; <span class='message'>release: "90"</span>
#&gt; <span class='message'>format: "transcripts"</span></div><div class='output co'>#&gt; <span class='message'>Standardizing the metadata columns</span></div><div class='output co'>#&gt; <span class='message'>Using geneName instead of symbol</span></div><div class='output co'>#&gt; <span class='message'>Setting geneBiotype as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting seqCoordSystem as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting txBiotype as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting txSupportLevel as factor</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, txBiotype, seqnames</span></div><div class='output co'>#&gt;   ENSMUST00000000001   ENSMUST00000000003   ENSMUST00000114041 
#&gt; "ENSMUSG00000000001" "ENSMUSG00000000003" "ENSMUSG00000000003" </div><div class='input'>
<span class='co'># matrix ====</span>
<span class='no'>mat</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(
    <span class='kw'>data</span> <span class='kw'>=</span> <span class='fu'>seq</span>(<span class='fl'>1L</span>:<span class='fl'>6L</span>),
    <span class='kw'>byrow</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
    <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='fl'>3L</span>,
    <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>2L</span>,
    <span class='kw'>dimnames</span> <span class='kw'>=</span> <span class='fu'>list</span>(
        <span class='fu'>c</span>(<span class='st'>"ENSMUST00000000001"</span>, <span class='st'>"ENSMUST00000000003"</span>, <span class='st'>"ENSMUST00000114041"</span>),
        <span class='fu'>c</span>(<span class='st'>"sample_1"</span>, <span class='st'>"sample_2"</span>)
    )
)
<span class='fu'>print</span>(<span class='no'>mat</span>)</div><div class='output co'>#&gt;                    sample_1 sample_2
#&gt; ENSMUST00000000001        1        2
#&gt; ENSMUST00000000003        3        4
#&gt; ENSMUST00000114041        5        6</div><div class='input'><span class='no'>mat</span> <span class='kw'>&lt;-</span> <span class='fu'>convertTranscriptsToGenes</span>(<span class='no'>mat</span>)</div><div class='output co'>#&gt; <span class='message'>Obtaining transcript-to-gene mappings from Ensembl</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl with AnnotationHub 2.10.1 (2017-10-27)</span></div><div class='output co'>#&gt; <span class='message'>id: "AH57770"</span>
#&gt; <span class='message'>organism: "Mus musculus"</span>
#&gt; <span class='message'>build: "GRCm38"</span>
#&gt; <span class='message'>release: "90"</span>
#&gt; <span class='message'>format: "transcripts"</span></div><div class='output co'>#&gt; <span class='message'>Standardizing the metadata columns</span></div><div class='output co'>#&gt; <span class='message'>Using geneName instead of symbol</span></div><div class='output co'>#&gt; <span class='message'>Setting geneBiotype as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting seqCoordSystem as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting txBiotype as factor</span></div><div class='output co'>#&gt; <span class='message'>Setting txSupportLevel as factor</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, txBiotype, seqnames</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>mat</span>)</div><div class='output co'>#&gt;                    sample_1 sample_2
#&gt; ENSMUSG00000000001        1        2
#&gt; ENSMUSG00000000003        3        4
#&gt; ENSMUSG00000000003        5        6</div><div class='input'><span class='fu'>rownames</span>(<span class='no'>mat</span>)</div><div class='output co'>#&gt;   ENSMUST00000000001   ENSMUST00000000003   ENSMUST00000114041 
#&gt; "ENSMUSG00000000001" "ENSMUSG00000000003" "ENSMUSG00000000003" </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

