<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Export — export • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Export — export" />

<meta property="og:description" content="Export data out of R and write to disk." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Export</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/steinbaugh/basejump/blob/master/R/export-methods.R'><code>R/export-methods.R</code></a></small>
    <div class="hidden name"><code>export.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Export data out of R and write to disk.</p>
    
    </div>

    <pre class="usage"><span class='fu'>export</span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='co'># S4 method for ANY</span>
<span class='fu'>export</span>(<span class='no'>x</span>, <span class='no'>file</span>, <span class='no'>format</span>, <span class='no'>...</span>)

<span class='co'># S4 method for sparseMatrix</span>
<span class='fu'>export</span>(<span class='no'>x</span>, <span class='no'>file</span>, <span class='no'>format</span>)

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>export</span>(<span class='no'>x</span>, <span class='kw'>name</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>dir</span> <span class='kw'>=</span> <span class='st'>"."</span>,
  <span class='kw'>compress</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>humanize</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>slotNames</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"assays"</span>,
  <span class='st'>"colData"</span>, <span class='st'>"rowData"</span>, <span class='st'>"Gene2Symbol"</span>, <span class='st'>"Ensembl2Entrez"</span>))

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>export</span>(<span class='no'>x</span>, <span class='kw'>name</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>dir</span> <span class='kw'>=</span> <span class='st'>"."</span>,
  <span class='kw'>compress</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>humanize</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>slotNames</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"assays"</span>,
  <span class='st'>"colData"</span>, <span class='st'>"rowData"</span>, <span class='st'>"Gene2Symbol"</span>, <span class='st'>"Ensembl2Entrez"</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x</th>
      <td><p>An object supporting <code>dim</code> to be written to disk.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>. File path. Specify <code>file</code> or <code>format</code> but not
both.</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>. An optional character string containing the
file format, which can be used to override the format inferred from <code>file</code>,
or in lieu of specifying <code>file</code>.</p></td>
    </tr>
    <tr>
      <th>name</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>. Name to use on disk. If left <code>NULL</code>, will use the
name of the object instead.</p></td>
    </tr>
    <tr>
      <th>dir</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>. Directory path.</p></td>
    </tr>
    <tr>
      <th>compress</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>. Apply gzip compression to all files.</p></td>
    </tr>
    <tr>
      <th>humanize</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>. Automatically convert gene IDs to gene symbols
in the <code>rownames</code> and sample IDs to sample names in the <code>colnames</code>.</p></td>
    </tr>
    <tr>
      <th>slotNames</th>
      <td><p><code>character</code>. Names of slots to include when writing to disk.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Invisible <code>character</code>. File path(s).</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This is a wrapper for <code><a href='https://www.rdocumentation.org/packages/rio/topics/export'>rio::export</a></code> that adds support for additional S4
classes in Bioconductor.</p>
    
    <h2 class="hasAnchor" id="row-names"><a class="anchor" href="#row-names"></a>Row names</h2>

    
    <p>The standard <code><a href='https://www.rdocumentation.org/packages/rio/topics/export'>rio::export</a></code> call will drop rownames when exporting to CSV. The
readr family of functions also never write rownames. This is a <em>really poor</em>
default setting for handling genomic data, which often contain gene
identifiers in the rownames. Here we're performing any internal tibble
coercion step to ensure rownames are always moved to a "<code>rowname</code>" column in
the CSV export.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='https://www.rdocumentation.org/packages/rio/topics/export'>rio::export</a></code>.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>library</a></span>(<span class='no'>SummarizedExperiment</span>)</div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomicRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: stats4</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: S4Vectors</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'S4Vectors'</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:base':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    expand.grid</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: IRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: GenomeInfoDb</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: Biobase</span></div><div class='output co'>#&gt; <span class='message'>Welcome to Bioconductor</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    Vignettes contain introductory material; view with</span>
#&gt; <span class='message'>    'browseVignettes()'. To cite Bioconductor, see</span>
#&gt; <span class='message'>    'citation("Biobase")', and for packages 'citation("pkgname")'.</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: DelayedArray</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: matrixStats</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'matrixStats'</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:Biobase':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    anyMissing, rowMedians</span></div><div class='output co'>#&gt; <span class='message'>The following object is masked from 'package:basejump':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    rowRanges</span></div><div class='output co'>#&gt; <span class='message'>Loading required package: BiocParallel</span></div><div class='output co'>#&gt; <span class='message'></span>
#&gt; <span class='message'>Attaching package: 'DelayedArray'</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:matrixStats':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</span></div><div class='output co'>#&gt; <span class='message'>The following objects are masked from 'package:base':</span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>    aperm, apply</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>library</a></span>(<span class='no'>SingleCellExperiment</span>)
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>rse</span>, <span class='no'>sce</span>)

<span class='co'>## matrix ====</span>
<span class='no'>rnaseq_counts</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>SummarizedExperiment</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class'>assay</a></span>(<span class='no'>rse</span>)
<span class='fu'>export</span>(<span class='no'>rnaseq_counts</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"csv"</span>)</div><div class='output co'>#&gt; <span class='message'>Exported rnaseq_counts.csv.</span></div><div class='input'>
<span class='co'>## sparseMatrix ====</span>
<span class='no'>single_cell_counts</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>SummarizedExperiment</span><span class='kw ns'>::</span><span class='fu'><a href='https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class'>assay</a></span>(<span class='no'>sce</span>)
<span class='fu'>export</span>(<span class='no'>single_cell_counts</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"mtx"</span>)</div><div class='output co'>#&gt; <span class='message'>Exported single_cell_counts.mtx, single_cell_counts.mtx.colnames, single_cell_counts.mtx.rownames.</span></div><div class='input'>
<span class='co'>## Clean up.</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/files'>file.remove</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(
    <span class='st'>"rnaseq_counts.csv"</span>,
    <span class='st'>"single_cell_counts.mtx"</span>,
    <span class='st'>"single_cell_counts.mtx.colnames"</span>,
    <span class='st'>"single_cell_counts.mtx.rownames"</span>
))</div><div class='output co'>#&gt; [1] TRUE TRUE TRUE TRUE</div><div class='input'>
<span class='co'>## SummarizedExperiment ====</span>
<span class='fu'>export</span>(<span class='no'>rse</span>, <span class='kw'>dir</span> <span class='kw'>=</span> <span class='st'>"example"</span>)</div><div class='output co'>#&gt; <span class='message'>Exporting assays: counts</span></div><div class='output co'>#&gt; <span class='message'>Exported counts.csv.</span></div><div class='output co'>#&gt; <span class='message'>Exported colData.csv.</span></div><div class='output co'>#&gt; <span class='message'>Exported rowData.csv.</span></div><div class='output co'>#&gt; <span class='message'>Exported Gene2Symbol.csv.</span></div><div class='output co'>#&gt; <span class='message'>4 genes don't map to Entrez.</span></div><div class='output co'>#&gt; <span class='message'>2 genes map to multiple Entrez IDs.</span></div><div class='output co'>#&gt; <span class='message'>Returning with 1:1 mappings using oldest Entrez ID per gene.</span></div><div class='output co'>#&gt; <span class='message'>Exported Ensembl2Entrez.csv.</span></div><div class='output co'>#&gt; <span class='message'>Exported rse to /data01/git/basejump/docs/reference/example/rse.</span></div><div class='input'>
<span class='co'>## SingleCellExperiment ====</span>
<span class='fu'>export</span>(<span class='no'>sce</span>, <span class='kw'>dir</span> <span class='kw'>=</span> <span class='st'>"example"</span>)</div><div class='output co'>#&gt; <span class='message'>Exporting assays: counts</span></div><div class='output co'>#&gt; <span class='message'>Exported counts.mtx, counts.mtx.colnames, counts.mtx.rownames.</span></div><div class='output co'>#&gt; <span class='message'>Exported colData.csv.</span></div><div class='output co'>#&gt; <span class='message'>Exported rowData.csv.</span></div><div class='output co'>#&gt; <span class='message'>Exported Gene2Symbol.csv.</span></div><div class='output co'>#&gt; <span class='message'>4 genes don't map to Entrez.</span></div><div class='output co'>#&gt; <span class='message'>2 genes map to multiple Entrez IDs.</span></div><div class='output co'>#&gt; <span class='message'>Returning with 1:1 mappings using oldest Entrez ID per gene.</span></div><div class='output co'>#&gt; <span class='message'>Exported Ensembl2Entrez.csv.</span></div><div class='output co'>#&gt; <span class='message'>Exported sce to /data01/git/basejump/docs/reference/example/sce.</span></div><div class='output co'>#&gt; $assays
#&gt; $assays$counts
#&gt;                                                                       matrix 
#&gt;          "/data01/git/basejump/docs/reference/example/sce/assays/counts.mtx" 
#&gt;                                                                     barcodes 
#&gt; "/data01/git/basejump/docs/reference/example/sce/assays/counts.mtx.colnames" 
#&gt;                                                                        genes 
#&gt; "/data01/git/basejump/docs/reference/example/sce/assays/counts.mtx.rownames" 
#&gt; 
#&gt; 
#&gt; $colData
#&gt; [1] "/data01/git/basejump/docs/reference/example/sce/colData.csv"
#&gt; 
#&gt; $rowData
#&gt; [1] "/data01/git/basejump/docs/reference/example/sce/rowData.csv"
#&gt; 
#&gt; $Gene2Symbol
#&gt; [1] "/data01/git/basejump/docs/reference/example/sce/Gene2Symbol.csv"
#&gt; 
#&gt; $Ensembl2Entrez
#&gt; [1] "/data01/git/basejump/docs/reference/example/sce/Ensembl2Entrez.csv"
#&gt; </div><div class='output co'>#&gt; <span class='message'>Exporting reducedDims: </span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#row-names">Row names</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

