<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Filter cells — filterCells • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Filter cells — filterCells" />

<meta property="og:description" content="Filter cells" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Filter cells</h1>
    <small class="dont-index">Source: <a href='https://github.com/acidgenomics/basejump/blob/master/R/filterCells-methods.R'><code>R/filterCells-methods.R</code></a></small>
    <div class="hidden name"><code>filterCells.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Filter cells</p>
    
    </div>

    <pre class="usage"><span class='fu'>filterCells</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>filterCells</span>(<span class='no'>object</span>, <span class='kw'>minCounts</span> <span class='kw'>=</span> <span class='fl'>1L</span>,
  <span class='kw'>maxCounts</span> <span class='kw'>=</span> <span class='fl'>Inf</span>, <span class='kw'>minFeatures</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='kw'>maxFeatures</span> <span class='kw'>=</span> <span class='fl'>Inf</span>,
  <span class='kw'>minNovelty</span> <span class='kw'>=</span> <span class='fl'>0L</span>, <span class='kw'>maxMitoRatio</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='kw'>minCellsPerFeature</span> <span class='kw'>=</span> <span class='fl'>1L</span>,
  <span class='kw'>nCells</span> <span class='kw'>=</span> <span class='fl'>Inf</span>, <span class='kw'>countsCol</span> <span class='kw'>=</span> <span class='st'>"nCount"</span>, <span class='kw'>featuresCol</span> <span class='kw'>=</span> <span class='st'>"nFeature"</span>,
  <span class='kw'>noveltyCol</span> <span class='kw'>=</span> <span class='st'>"log10FeaturesPerCount"</span>, <span class='kw'>mitoRatioCol</span> <span class='kw'>=</span> <span class='st'>"mitoRatio"</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>minCounts, maxCounts</th>
      <td><p><code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code>.
Minimum/maximum number of counts per cell.
Applies to UMI disambiguated counts for droplet scRNA-seq.
Matches <code>nUMI</code> then <code>nCount</code> column in
<code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>colData()</a></code> internally.
Previously named <code>minUMIs</code>/<code>maxUMIs</code> in bcbioSingleCell.</p></td>
    </tr>
    <tr>
      <th>minFeatures, maxFeatures</th>
      <td><p><code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code>.
Minimum/maximum number of features (i.e. genes) detected.
Matches <code>nFeature</code>in <code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>colData()</a></code> internally.
Previously named <code>minGenes</code>/<code>maxGenes</code> in bcbioSingleCell.</p></td>
    </tr>
    <tr>
      <th>minNovelty</th>
      <td><p><code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code> (<code>0</code>-<code>1</code>).
Minimum novelty score (log10 features per UMI).
Matches <code>log10FeaturesPerCount</code> then <code>log10FeaturesPerUMI</code> (legacy)
<code><a href='https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html'>colData()</a></code> internally.</p></td>
    </tr>
    <tr>
      <th>maxMitoRatio</th>
      <td><p><code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code> (<code>0</code>-<code>1</code>).
Maximum relative mitochondrial abundance.</p></td>
    </tr>
    <tr>
      <th>minCellsPerFeature</th>
      <td><p><code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code>.
Include genes with non-zero expression in at least this many cells.
Previously named <code>minCellsPerGene</code> in bcbioSingleCell.</p></td>
    </tr>
    <tr>
      <th>nCells</th>
      <td><p><code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code>.
Expected number of cells per sample.
Don't set this by default, unless you're confident of your capture.</p></td>
    </tr>
    <tr>
      <th>countsCol, featuresCol, noveltyCol, mitoRatioCol</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Column mapping name.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>SingleCellExperiment</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Apply feature (i.e. gene/transcript) detection, novelty score, and
mitochondrial abundance cutoffs to cellular barcodes. By default we recommend
applying the same filtering cutoff to all samples. The filtering parameters
now support per-sample cutoffs, defined using a named <code>numeric</code> vector. When
matching per sample, be sure to use the <code><a href='sampleNames.html'>sampleNames()</a></code> return values (i.e.
the <code>sampleName</code> column in <code><a href='sampleData.html'>sampleData()</a></code>).</p>
<p>Filtering information gets slotted into <code><a href='https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html'>metadata()</a></code>
as <code>filterCells</code> metadata.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Updated 2019-08-13.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>SingleCellExperiment</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"acidtest"</span>)

<span class='co'>## SingleCellExperiment ====</span>
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='no'>SingleCellExperiment</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>filterCells</span>(<span class='no'>object</span>)</div><div class='output co'>#&gt; <span class='message'>Calculating 100 cellular barcode metrics.</span></div><div class='output co'>#&gt; <span class='message'>497 coding features detected.</span></div><div class='output co'>#&gt; <span class='message'>0 mitochondrial features detected.</span></div><div class='output co'>#&gt; <span class='message'>Pre-filter:</span>
#&gt; <span class='message'>  - 100 cells</span>
#&gt; <span class='message'>  - 500 features</span></div><div class='output co'>#&gt; <span class='message'>Post-filter:</span>
#&gt; <span class='message'>  - 100 cells (100%)</span>
#&gt; <span class='message'>  - 478 features (95.6%)</span></div><div class='output co'>#&gt; <span class='message'>Per argument:</span>
#&gt; <span class='message'>  minCounts   maxCounts minFeatures maxFeatures  minNovelty </span>
#&gt; <span class='message'>        100         100         100         100         100 </span></div><div class='output co'>#&gt; <span class='message'>Per sample, per argument:</span>
#&gt; <span class='message'>$sample1</span>
#&gt; <span class='message'>  minCounts   maxCounts minFeatures maxFeatures  minNovelty </span>
#&gt; <span class='message'>         64          64          64          64          64 </span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>$sample2</span>
#&gt; <span class='message'>  minCounts   maxCounts minFeatures maxFeatures  minNovelty </span>
#&gt; <span class='message'>         36          36          36          36          36 </span></div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: SingleCellExperiment 
#&gt; dim: 478 100 
#&gt; metadata(3): date filterCells subset
#&gt; assays(1): counts
#&gt; rownames(478): gene001 gene002 ... gene499 gene500
#&gt; rowData names(8): broadClass description ... geneName seqCoordSystem
#&gt; colnames(100): cell001 cell002 ... cell099 cell100
#&gt; colData names(7): sampleID nCount ... log10FeaturesPerCount mitoRatio
#&gt; reducedDimNames(0):
#&gt; spikeNames(0):</div><div class='input'>
<span class='co'>## Per sample cutoffs.</span>
<span class='fu'><a href='sampleNames.html'>sampleNames</a></span>(<span class='no'>object</span>)</div><div class='output co'>#&gt;   sample1   sample2 
#&gt; "sample1" "sample2" </div><div class='input'><span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>filterCells</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>,
    <span class='kw'>minCounts</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>sample1</span> <span class='kw'>=</span> <span class='fl'>100L</span>)
)</div><div class='output co'>#&gt; <span class='message'>Calculating 100 cellular barcode metrics.</span></div><div class='output co'>#&gt; <span class='message'>497 coding features detected.</span></div><div class='output co'>#&gt; <span class='message'>0 mitochondrial features detected.</span></div><div class='output co'>#&gt; <span class='message'>Pre-filter:</span>
#&gt; <span class='message'>  - 100 cells</span>
#&gt; <span class='message'>  - 500 features</span></div><div class='output co'>#&gt; <span class='message'>Post-filter:</span>
#&gt; <span class='message'>  - 100 cells (100%)</span>
#&gt; <span class='message'>  - 478 features (95.6%)</span></div><div class='output co'>#&gt; <span class='message'>Per argument:</span>
#&gt; <span class='message'>  minCounts   maxCounts minFeatures maxFeatures  minNovelty </span>
#&gt; <span class='message'>        100         100         100         100         100 </span></div><div class='output co'>#&gt; <span class='message'>Per sample, per argument:</span>
#&gt; <span class='message'>$sample1</span>
#&gt; <span class='message'>  minCounts   maxCounts minFeatures maxFeatures  minNovelty </span>
#&gt; <span class='message'>         64          64          64          64          64 </span>
#&gt; <span class='message'></span>
#&gt; <span class='message'>$sample2</span>
#&gt; <span class='message'>  minCounts   maxCounts minFeatures maxFeatures  minNovelty </span>
#&gt; <span class='message'>         36          36          36          36          36 </span></div><div class='input'><span class='fu'><a href='show.html'>show</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: SingleCellExperiment 
#&gt; dim: 478 100 
#&gt; metadata(3): date filterCells subset
#&gt; assays(1): counts
#&gt; rownames(478): gene001 gene002 ... gene499 gene500
#&gt; rowData names(8): broadClass description ... geneName seqCoordSystem
#&gt; colnames(100): cell001 cell002 ... cell099 cell100
#&gt; colData names(7): sampleID nCount ... log10FeaturesPerCount mitoRatio
#&gt; reducedDimNames(0):
#&gt; spikeNames(0):</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


