<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Import — import • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Import — import" />

<meta property="og:description" content="Read file by extension into R." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/import.R'><code>R/import.R</code></a></small>
    <div class="hidden name"><code>import.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Read file by extension into R.</p>
    
    </div>

    <pre class="usage"><span class='fu'>import</span>(<span class='no'>file</span>, <span class='kw'>dataFrame</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/options'>getOption</a></span>(<span class='st'>"basejump.data.frame"</span>, <span class='st'>"data.frame"</span>),
  <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>file</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>. File path.</p></td>
    </tr>
    <tr>
      <th>dataFrame</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>. Data frame class to return. Can be set
globally using the <code>basejump.data.frame</code> option.</p>
<p>Current recommendations (by priority):</p><ul>
<li><p><code>data.frame</code>: Base R default. Generally recommended.</p><ul>
<li><p>S3 class.</p></li>
<li><p>Allows rownames, but they're required and can't be set <code>NULL</code>.</p></li>
<li><p>See <code><a href='https://www.rdocumentation.org/packages/utils/topics/help'>help(topic = "data.frame", package = "base")</a></code> for details.</p></li>
</ul></li>
<li><p><code>DataFrame</code>: Recommended when working with Bioconductor packages.</p><ul>
<li><p>S4 class; inherits <code>DataTable</code> virtual class.</p></li>
<li><p>Allows rownames, but they're optional and can be set <code>NULL</code>.</p></li>
<li><p>See <code><a href='https://www.rdocumentation.org/packages/utils/topics/help'>help(topic = "DataFrame", package = "S4Vectors")</a></code> for details.</p></li>
</ul></li>
<li><p><code>tbl_df</code> (<code>tibble</code>): Recommended when working with tidyverse packages.</p><ul>
<li><p>S3 class; inherits <code>data.frame</code>.</p></li>
<li><p>Does not allow rownames.</p></li>
<li><p>See <code><a href='https://www.rdocumentation.org/packages/utils/topics/help'>help(topic = "tibble", package = "tibble")</a></code> for details.</p></li>
</ul></li>
<li><p><code>data.table</code>: Recommended when working with the data.table package.</p><ul>
<li><p>S3 class; inherits <code>data.frame</code>.</p></li>
<li><p>Does not allow rownames.</p></li>
<li><p>See <code><a href='https://www.rdocumentation.org/packages/utils/topics/help'>help(topic = "data.table", package = "data.table")</a></code> for details.</p></li>
</ul></li>
</ul></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Varies, depending on the file extension:</p><ul>
<li><p>Default: <code>DataFrame</code>.</p></li>
<li><p><code>MTX</code>: <code>sparseMatrix</code>.</p></li>
<li><p><code>GTF</code>/<code>GFF</code>: <code>GRanges</code>.</p></li>
<li><p><code>JSON</code>/<code>YAML</code>: <code>list</code>.</p></li>
<li><p>Source code or log: <code>character</code>.</p></li>
</ul>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Remote URLs and compressed files are supported. All extensions are case
insensitive.</p>
<p>This function supports automatic loading of common file types:</p><ul>
<li><p><code>CSV</code>: Comma Separated Values.<br />
Imported by <code><a href='https://www.rdocumentation.org/packages/data.table/topics/fread'>data.table::fread</a></code>.</p></li>
<li><p><code>TSV</code> Tab Separated Values.<br />
Imported by <code><a href='https://www.rdocumentation.org/packages/data.table/topics/fread'>data.table::fread</a></code>.</p></li>
<li><p><code>TXT</code>: Text file. <em>Ambiguous, not recommended</em>.<br />
Imported by <code><a href='https://www.rdocumentation.org/packages/data.table/topics/fread'>data.table::fread</a></code>.</p></li>
<li><p><code>XLSX</code>/<code>XLS</code>: Excel workbook.<br />
Imported by <code><a href='https://readxl.tidyverse.org/reference/read_excel.html'>readxl::read_excel</a></code>.</p></li>
<li><p><code>MTX</code>: MatrixMarket sparse matrix.<br />
Imported by <code><a href='https://www.rdocumentation.org/packages/Matrix/topics/externalFormats'>Matrix::readMM</a></code>.</p></li>
<li><p><code>GTF</code>/<code>GFF</code>/<code>GFF3</code>: General Feature Format.<br />
Imported by <code><a href='https://www.rdocumentation.org/packages/rtracklayer/topics/export'>rtracklayer::import</a></code>.</p></li>
<li><p><code>JSON</code>: JSON.
Imported by <code><a href='https://www.rdocumentation.org/packages/jsonlite/topics/read_json'>jsonlite::read_json</a></code>.</p></li>
<li><p><code>YAML</code>/<code>YML</code>: YAML.
Imported by <code><a href='https://www.rdocumentation.org/packages/yaml/topics/yaml.load'>yaml::yaml.load_file</a></code></p></li>
<li><p><code>RDA</code>/<code>RDATA</code>: R Data.</p><ul>
<li><p>Imported by <code>load</code>.</p></li>
<li><p>Must contain a single object.</p></li>
<li><p>Doesn't require internal object name to match, unlike <code>loadData</code>.</p></li>
</ul></li>
<li><p><code>RDS</code>: R Data Serialized.<br />
Imported by <code>readRDS</code>.</p></li>
</ul>
    <p>These file formats will be imported as source code lines by
<code><a href='https://readr.tidyverse.org/reference/read_lines.html'>readr::read_lines</a></code>: <code>LOG</code>, <code>MD</code>, <code>PY</code>, <code>R</code>, <code>RMD</code>, <code>SH</code>.</p>
<p>These file formats are blacklisted, and intentionally not supported:
<code>DOC</code>, <code>DOCX</code>, <code>PDF</code>, <code>PPT</code>, <code>PPTX</code>.</p>
<p>If a file format isn't supported natively (or blacklisted), the
<a href='https://cran.r-project.org/web/packages/rio/index.html'>rio</a> package will
be used as a fallback attempt. See <code><a href='https://www.rdocumentation.org/packages/rio/topics/import'>rio::import</a></code> for details.</p>
    
    <h2 class="hasAnchor" id="delimited-files-csv-tsv-"><a class="anchor" href="#delimited-files-csv-tsv-"></a>Delimited Files (CSV/TSV)</h2>

    
    <p><code>import</code> uses the <code>fread</code> function of the  <a href='https://cran.r-project.org/package=data.table'>data.table</a> package to
import standard CSV and TSV files. This should work automatically for most
files without issue.</p>
<p>Here are some notable exceptions:</p><ul>
<li><p>Columns headers should be <code>character</code>. In the event that a column name
is <code>numeric</code>, set <code>header = TRUE</code> here to force the column name.</p></li>
</ul>
    <p>See <code><a href='https://www.rdocumentation.org/packages/utils/topics/help'>help(topic = "fread", package = "data.table")</a></code> for details.</p>
<p>The <code>read_csv</code> and <code>read_tsv</code> functions of the <a href='https://readr.tidyverse.org'>readr</a> package
are good alternatives, which return <code>tibble</code> data frames (<code>tbl_df</code>).</p>
    
    <h2 class="hasAnchor" id="matrix-market-exchange-mtx-mex-"><a class="anchor" href="#matrix-market-exchange-mtx-mex-"></a>Matrix Market Exchange (MTX/MEX)</h2>

    
    <p>Reading a Matrix Market Exchange (<code>MTX</code>) file now requires <code>COLNAMES</code> and
<code>ROWNAMES</code> sidecar files containing the <code>colnames</code> and <code>rownames</code> of
the sparse matrix. Legacy support for manual loading of these sidecar files
is provided.</p>
    
    <h2 class="hasAnchor" id="general-feature-format-gff-gtf-"><a class="anchor" href="#general-feature-format-gff-gtf-"></a>General Feature Format (GFF/GTF)</h2>

    
    <p>The GFF (General Feature Format) format consists of one line per feature,
each containing 9 columns of data, plus optional track definition lines. The
GTF (General Transfer Format) is identical to GFF version 2.</p>
<p>Column names follow the <a href='https://bit.ly/2K6EBla'>Ensembl conventions</a>.
Additional information:</p><ul>
<li><p><a href='http://www.ensembl.org/info/website/upload/gff.html'>Ensembl</a></p></li>
<li><p><a href='http://www.gencodegenes.org/gencodeformat.html'>Gencode</a></p></li>
</ul>
    
    <h2 class="hasAnchor" id="bcbio-files"><a class="anchor" href="#bcbio-files"></a>bcbio files</h2>

    
    <p>Also supports some additional extensions commonly used with the
<a href='https://bcbio-nextgen.readthedocs.io'>bcbio</a> pipeline:</p><ul>
<li><p><code>COUNTS</code>: Counts table (e.g. RNA-seq aligned counts).</p></li>
<li><p><code>COLNAMES</code>: Sidecar file containing column names.</p></li>
<li><p><code>ROWNAMES</code>: Sidecar file containing row names.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><a href='http://readr.tidyverse.org'>readr</a>.</p></li>
<li><p><a href='http://readxl.tidyverse.org'>readxl</a>.</p></li>
<li><p><a href='https://cran.r-project.org/web/packages/Matrix/index.html'>Matrix</a>.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>url</span> <span class='kw'>&lt;-</span> <span class='no'>basejumpCacheURL</span>

<span class='co'>## Comma Separated Values</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/file.path'>file.path</a></span>(<span class='no'>url</span>, <span class='st'>"example.csv"</span>))</div><div class='output co'>#&gt; <span class='message'>Importing example.csv using data.table::fread().</span></div><div class='output co'>#&gt; <span class='message'>Setting rownames from `rowname` column.</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt;         genotype treatment
#&gt; sample1 wildtype   control
#&gt; sample2 knockout   control
#&gt; sample3 wildtype   treated
#&gt; sample4 knockout   treated</div><div class='input'>
<span class='co'>## Tab Separated Values</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/file.path'>file.path</a></span>(<span class='no'>url</span>, <span class='st'>"example.tsv"</span>))</div><div class='output co'>#&gt; <span class='message'>Importing example.tsv using data.table::fread().</span></div><div class='output co'>#&gt; <span class='message'>Setting rownames from `rowname` column.</span></div><div class='input'>
<span class='co'>## Microsoft Excel Worksheet</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/file.path'>file.path</a></span>(<span class='no'>url</span>, <span class='st'>"example.xlsx"</span>))</div><div class='output co'>#&gt; <span class='message'>Importing example.xlsx using readxl::read_excel().</span></div><div class='output co'>#&gt; <span class='message'>Setting rownames from `rowname` column.</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt;         genotype treatment
#&gt; sample1 wildtype   control
#&gt; sample2 knockout   control
#&gt; sample3 wildtype   treated
#&gt; sample4 knockout   treated</div><div class='input'>
<span class='co'>## R Data</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/file.path'>file.path</a></span>(<span class='no'>url</span>, <span class='st'>"rnaseq_counts.rda"</span>))</div><div class='output co'>#&gt; <span class='message'>Importing rnaseq_counts.rda using base::load().</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/class'>class</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "matrix"</div><div class='input'>
## GTF/GFF
# x <- import(file = file.path(url, "example.gtf"))
# summary(x)

## JSON
# x <- import(file = file.path(url, "example.json"))
# names(x)

## YAML
# x <- import(file = file.path(url, "example.yml"))
# names(x)

## Counts Table (i.e. aligned counts from bcbio)
# x <- import(file = file.path(url, "example.counts"))
# colSums(x)
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#delimited-files-csv-tsv-">Delimited Files (CSV/TSV)</a></li>

      <li><a href="#matrix-market-exchange-mtx-mex-">Matrix Market Exchange (MTX/MEX)</a></li>

      <li><a href="#general-feature-format-gff-gtf-">General Feature Format (GFF/GTF)</a></li>

      <li><a href="#bcbio-files">bcbio files</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

