<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Import — import • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Import — import" />

<meta property="og:description" content="Read file by extension into R." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/import.R'><code>R/import.R</code></a></small>
    <div class="hidden name"><code>import.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Read file by extension into R.</p>
    
    </div>

    <pre class="usage"><span class='fu'>import</span>(<span class='no'>file</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>file</th>
      <td><p><code>string</code>. File path.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Varies, depending on the file extension:</p><ul>
<li><p>Default: <code>DataFrame</code>.</p></li>
<li><p><code>MTX</code>: <code>sparseMatrix</code>.</p></li>
<li><p><code>GTF</code>/<code>GFF</code>: <code>GRanges</code>.</p></li>
<li><p><code>JSON</code>/<code>YAML</code>: <code>list</code>.</p></li>
<li><p>Source code or log: <code>character</code>.</p></li>
</ul>

<p>Note that data frames are returned as S4 <code>DataFrame</code> class instead of
<code>data.frame</code>, when applicable.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This is a wrapper for <code><a href='http://www.rdocumentation.org/packages/rio/topics/import'>rio::import()</a></code> that adds support for additional common
genomic data files. Remote URLs and compressed files are supported. All
extensions are case insensitive.</p>
<p>This function supports automatic loading of common file types:</p><ul>
<li><p><code>CSV</code>: Comma Separated Values.</p></li>
<li><p><code>TSV</code> Tab Separated Values.</p></li>
<li><p><code>XLSX</code>: Excel workbook.</p></li>
<li><p><code>MTV</code>: MatrixMarket sparse matrix.</p></li>
<li><p><code>GTF</code>/<code>GFF</code>/<code>GFF3</code>: General Feature Format.</p></li>
<li><p><code>RDA</code>/<code>RDATA</code>: R Data.</p><ul>
<li><p>Must contain a single object.</p></li>
<li><p>Doesn't require internal object name to match, unlike <code><a href='loadData.html'>loadData()</a></code>.</p></li>
</ul></li>
<li><p><code>RDS</code>: R Data Serialized.</p></li>
<li><p><code>JSON</code>: JSON.</p></li>
<li><p><code>YAML</code>/<code>YML</code>: YAML.</p></li>
</ul>
    <p>These file formats will be read as (source code) lines:
<code>LOG</code>, <code>MD</code>, <code>PY</code>, <code>R</code>, <code>RMD</code>, <code>SH</code>.</p>
<p>These file formats are blacklisted, and intentionally not supported:
<code>DOC</code>, <code>DOCX</code>, <code>PDF</code>, <code>PPT</code>, <code>PPTX</code>, <code>TXT</code>.</p>
<p>If the file format isn't supported natively (or blacklisted), the
<a href='https://cran.r-project.org/web/packages/rio/index.html'>rio</a> package will
be used as a fallback attempt.</p>
    
    <h2 class="hasAnchor" id="matrix-market-exchange-mtx-mex-"><a class="anchor" href="#matrix-market-exchange-mtx-mex-"></a>Matrix Market Exchange (MTX/MEX)</h2>

    
    <p>Reading a Matrix Market Exchange (<code>MTX</code>) file now requires <code>COLNAMES</code> and
<code>ROWNAMES</code> sidecar files containing the <code>colnames()</code> and <code>rownames()</code> of
the sparse matrix. Legacy support for manual loading of these sidecar files
is provided.</p>
    
    <h2 class="hasAnchor" id="general-feature-format-gff-gtf-"><a class="anchor" href="#general-feature-format-gff-gtf-"></a>General Feature Format (GFF/GTF)</h2>

    
    <p>The GFF (General Feature Format) format consists of one line per feature,
each containing 9 columns of data, plus optional track definition lines. The
GTF (General Transfer Format) is identical to GFF version 2.</p>
<p>Column names follow the <a href='https://bit.ly/2K6EBla'>Ensembl conventions</a>.
Additional information:</p><ul>
<li><p><a href='http://www.ensembl.org/info/website/upload/gff.html'>Ensembl</a></p></li>
<li><p><a href='http://www.gencodegenes.org/gencodeformat.html'>Gencode</a></p></li>
</ul>
    
    <h2 class="hasAnchor" id="bcbio-files"><a class="anchor" href="#bcbio-files"></a>bcbio files</h2>

    
    <p>Also supports some additional extensions commonly used with the
<a href='https://bcbio-nextgen.readthedocs.io'>bcbio</a> pipeline:</p><ul>
<li><p><code>COUNTS</code>: Counts table (e.g. RNA-seq aligned counts).</p></li>
<li><p><code>COLNAMES</code>: Sidecar file containing column names.</p></li>
<li><p><code>ROWNAMES</code>: Sidecar file containing row names.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><a href='http://readr.tidyverse.org'>readr</a>.</p></li>
<li><p><a href='http://readxl.tidyverse.org'>readxl</a>.</p></li>
<li><p><a href='https://cran.r-project.org/web/packages/Matrix/index.html'>Matrix</a>.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## R Data</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"rnaseq_counts.rda"</span>))</div><div class='output co'>#&gt; <span class='message'>Reading rnaseq_counts.rda.</span></div><div class='input'><span class='fu'>class</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "matrix"</div><div class='input'>
<span class='co'>## Comma Separated Values</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"example.csv"</span>))</div><div class='output co'>#&gt; <span class='message'>Reading example.csv.</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; DataFrame with 4 rows and 2 columns
#&gt;            genotype   treatment
#&gt;         &lt;character&gt; &lt;character&gt;
#&gt; sample1    wildtype     control
#&gt; sample2    knockout     control
#&gt; sample3    wildtype     treated
#&gt; sample4    knockout     treated</div><div class='input'>
<span class='co'>## Microsoft Excel Worksheet</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"example.xlsx"</span>))</div><div class='output co'>#&gt; <span class='message'>Reading example.xlsx.</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; DataFrame with 4 rows and 2 columns
#&gt;            genotype   treatment
#&gt;         &lt;character&gt; &lt;character&gt;
#&gt; sample1    wildtype     control
#&gt; sample2    knockout     control
#&gt; sample3    wildtype     treated
#&gt; sample4    knockout     treated</div><div class='input'>
<span class='co'>## GTF/GFF</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"example.gtf"</span>))</div><div class='output co'>#&gt; <span class='message'>Reading example.gtf.</span></div><div class='input'><span class='fu'>summary</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "GRanges object with 95 ranges and 22 metadata columns"</div><div class='input'>
<span class='co'>## JSON</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"example.json"</span>))</div><div class='output co'>#&gt; <span class='message'>Reading example.json.</span></div><div class='input'><span class='fu'>names</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "fasta_hash"        "genomes"           "gtf_hash"         
#&gt; [4] "input_fasta_files" "input_gtf_files"   "mem_gb"           
#&gt; [7] "mkref_version"     "threads"           "version"          </div><div class='input'>
<span class='co'>## YAML</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"example.yml"</span>))</div><div class='output co'>#&gt; <span class='message'>Reading example.yml.</span></div><div class='input'><span class='fu'>names</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "date"         "upload"       "bcbio_system" "samples"     </div><div class='input'>
<span class='co'>## Counts Table (i.e. aligned counts from bcbio)</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>import</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='fu'>file.path</span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"example.counts"</span>))</div><div class='output co'>#&gt; <span class='message'>Reading example.counts.</span></div><div class='output co'>#&gt; <span class='message'>Parsed with column specification:</span>
#&gt; <span class='message'>cols(</span>
#&gt; <span class='message'>  id = col_character(),</span>
#&gt; <span class='message'>  sample01 = col_integer(),</span>
#&gt; <span class='message'>  sample02 = col_integer(),</span>
#&gt; <span class='message'>  sample03 = col_integer(),</span>
#&gt; <span class='message'>  sample04 = col_integer(),</span>
#&gt; <span class='message'>  sample05 = col_integer(),</span>
#&gt; <span class='message'>  sample06 = col_integer(),</span>
#&gt; <span class='message'>  sample07 = col_integer(),</span>
#&gt; <span class='message'>  sample08 = col_integer(),</span>
#&gt; <span class='message'>  sample09 = col_integer(),</span>
#&gt; <span class='message'>  sample10 = col_integer(),</span>
#&gt; <span class='message'>  sample11 = col_integer(),</span>
#&gt; <span class='message'>  sample12 = col_integer()</span>
#&gt; <span class='message'>)</span></div><div class='input'><span class='fu'>colSums</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; sample01 sample02 sample03 sample04 sample05 sample06 sample07 sample08 
#&gt;      262      502      299      318      123      620      242      303 
#&gt; sample09 sample10 sample11 sample12 
#&gt;      199      258      258      157 </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#matrix-market-exchange-mtx-mex-">Matrix Market Exchange (MTX/MEX)</a></li>

      <li><a href="#general-feature-format-gff-gtf-">General Feature Format (GFF/GTF)</a></li>

      <li><a href="#bcbio-files">bcbio files</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

