<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Function reference" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.8.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-s-classes" class="hasAnchor"><a href="#section-s-classes" class="anchor"></a>S4 classes</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="EggNOG-class.html">EggNOG-class</a></code> </p>
        </td>
        <td><p>EggNOG Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="EggNOG.html">EggNOG()</a></code> </p>
        </td>
        <td><p>EggNOG Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Ensembl2Entrez-class.html">Ensembl2Entrez-class</a></code> </p>
        </td>
        <td><p>Ensembl-to-Entrez Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Ensembl2Entrez.html">Ensembl2Entrez()</a></code> </p>
        </td>
        <td><p>Ensembl-to-Entrez Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Gene2Symbol-class.html">Gene2Symbol-class</a></code> </p>
        </td>
        <td><p>Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Gene2Symbol.html">Gene2Symbol()</a></code> </p>
        </td>
        <td><p>Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="HGNC2Ensembl-class.html">HGNC2Ensembl-class</a></code> </p>
        </td>
        <td><p>HGNC-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="HGNC2Ensembl.html">HGNC2Ensembl()</a></code> </p>
        </td>
        <td><p>HGNC-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MGI2Ensembl-class.html">MGI2Ensembl-class</a></code> </p>
        </td>
        <td><p>MGI-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MGI2Ensembl.html">MGI2Ensembl()</a></code> </p>
        </td>
        <td><p>MGI-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PANTHER-class.html">PANTHER-class</a></code> </p>
        </td>
        <td><p>PANTHER Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PANTHER.html">PANTHER()</a></code> </p>
        </td>
        <td><p>PANTHER Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Tx2Gene-class.html">Tx2Gene-class</a></code> </p>
        </td>
        <td><p>Transcript-to-Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Tx2Gene.html">Tx2Gene()</a></code> </p>
        </td>
        <td><p>Transcript-to-Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="show.html">show(<i>&lt;EggNOG&gt;</i>)</a></code> <code><a href="show.html">show(<i>&lt;PANTHER&gt;</i>)</a></code> </p>
        </td>
        <td><p>Show an Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summary.html">summary(<i>&lt;Gene2Symbol&gt;</i>)</a></code> <code><a href="summary.html">summary(<i>&lt;Tx2Gene&gt;</i>)</a></code> </p>
        </td>
        <td><p>Object Summaries</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-import-export" class="hasAnchor"><a href="#section-import-export" class="anchor"></a>Import/export</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="assignAndSaveData.html">assignAndSaveData()</a></code> </p>
        </td>
        <td><p>Assign and Save Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="export.html">export()</a></code> </p>
        </td>
        <td><p>Export</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="import.html">import()</a></code> </p>
        </td>
        <td><p>Import</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadData.html">loadData()</a></code> </p>
        </td>
        <td><p>Load Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadDataAsName.html">loadDataAsName()</a></code> </p>
        </td>
        <td><p>Load Data as Name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadRemoteData.html">loadRemoteData()</a></code> </p>
        </td>
        <td><p>Load Remote Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="localOrRemoteFile.html">localOrRemoteFile()</a></code> </p>
        </td>
        <td><p>Dynamically Handle a Local or Remote File Path</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="saveData.html">saveData()</a></code> </p>
        </td>
        <td><p>Save Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="saveRdExamples.html">saveRdExamples()</a></code> </p>
        </td>
        <td><p>Save R Documentation Examples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="transmit.html">transmit()</a></code> </p>
        </td>
        <td><p>Transmit Files from a Remote FTP Server</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="writeCounts.html">writeCounts()</a></code> </p>
        </td>
        <td><p>Write Counts</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-genome-annotations" class="hasAnchor"><a href="#section-genome-annotations" class="anchor"></a>Genome annotations</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="emptyRanges.html">emptyRanges()</a></code> </p>
        </td>
        <td><p>Generate Empty Genomic Ranges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneSynonyms.html">geneSynonyms()</a></code> </p>
        </td>
        <td><p>Gene Synonyms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGRanges.html">makeGRangesFromEnsembl()</a></code> <code><a href="makeGRanges.html">makeGRangesFromEnsDb()</a></code> <code><a href="makeGRanges.html">makeGRangesFromGFF()</a></code> <code><a href="makeGRanges.html">annotable()</a></code> </p>
        </td>
        <td><p>Make Genomic Ranges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGene2Symbol.html">makeGene2SymbolFromEnsembl()</a></code> <code><a href="makeGene2Symbol.html">makeGene2SymbolFromEnsDb()</a></code> <code><a href="makeGene2Symbol.html">makeGene2SymbolFromGFF()</a></code> </p>
        </td>
        <td><p>Make Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeTx2Gene.html">makeTx2GeneFromEnsembl()</a></code> <code><a href="makeTx2Gene.html">makeTx2GeneFromEnsDb()</a></code> <code><a href="makeTx2Gene.html">makeTx2GeneFromGFF()</a></code> </p>
        </td>
        <td><p>Make Transcript-to-Gene Mappings</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-gene-identifier-mapping" class="hasAnchor"><a href="#section-gene-identifier-mapping" class="anchor"></a>Gene identifier mapping</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="convertGenesToSymbols.html">convertGenesToSymbols()</a></code> <code><a href="convertGenesToSymbols.html">convertSymbolsToGenes()</a></code> </p>
        </td>
        <td><p>Convert Genes to Symbols</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertSampleIDsToNames.html">convertSampleIDsToNames()</a></code> </p>
        </td>
        <td><p>Convert Sample Identifiers to Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertTranscriptsToGenes.html">convertTranscriptsToGenes()</a></code> </p>
        </td>
        <td><p>Convert Transcripts to Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertUCSCBuildToEnsembl.html">convertUCSCBuildToEnsembl()</a></code> </p>
        </td>
        <td><p>Convert UCSC Build to Ensembl</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapGenes.html">mapGenesToRownames()</a></code> <code><a href="mapGenes.html">mapGenesToIDs()</a></code> <code><a href="mapGenes.html">mapGenesToSymbols()</a></code> </p>
        </td>
        <td><p>Map Genes</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-counts" class="hasAnchor"><a href="#section-counts" class="anchor"></a>Counts</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="counts.html">counts(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="counts.html">`counts<-`(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meltCounts.html">meltCounts()</a></code> </p>
        </td>
        <td><p>Melt Count Matrix to Long Format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tpm.html">tpm(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Transcripts per Million</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-sample-level" class="hasAnchor"><a href="#section-sample-level" class="anchor"></a>Sample-level</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="makeSampleData.html">makeSampleData()</a></code> </p>
        </td>
        <td><p>Make Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeSummarizedExperiment.html">makeSummarizedExperiment()</a></code> </p>
        </td>
        <td><p>Make Summarized Experiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="metrics.html">metrics()</a></code> <code><a href="metrics.html">metricsPerSample()</a></code> </p>
        </td>
        <td><p>Metrics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="minimalSampleData.html">minimalSampleData()</a></code> </p>
        </td>
        <td><p>Minimal Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleData.html">sampleData()</a></code> <code><a href="sampleData.html">`sampleData<-`()</a></code> <code><a href="sampleData.html">sampleNames(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleNames.html">`sampleNames<-`(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;character&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sample Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="selectSamples.html">selectSamples()</a></code> </p>
        </td>
        <td><p>Select Samples</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-cell-level" class="hasAnchor"><a href="#section-cell-level" class="anchor"></a>Cell-level</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="aggregateCellsToSamples.html">aggregateCellsToSamples()</a></code> </p>
        </td>
        <td><p>Aggregate Cells to Samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cell2sample.html">cell2sample()</a></code> </p>
        </td>
        <td><p>Cell-to-Sample Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeSingleCellExperiment.html">makeSingleCellExperiment()</a></code> </p>
        </td>
        <td><p>Make Single-Cell Experiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapCellsToSamples.html">mapCellsToSamples()</a></code> </p>
        </td>
        <td><p>Map Cells to Samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetPerSample.html">subsetPerSample()</a></code> </p>
        </td>
        <td><p>Subset Per Sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="topCellsPerSample.html">topCellsPerSample()</a></code> </p>
        </td>
        <td><p>Top Cells per Sample</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-metadata" class="hasAnchor"><a href="#section-metadata" class="anchor"></a>Metadata</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="interestingGroups.html">interestingGroups()</a></code> <code><a href="interestingGroups.html">`interestingGroups<-`()</a></code> </p>
        </td>
        <td><p>Interesting Groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organism.html">organism(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;DataFrame&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;GRanges&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Organism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="uniteInterestingGroups.html">uniteInterestingGroups()</a></code> </p>
        </td>
        <td><p>Unite Interesting Groups</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-sanitization" class="hasAnchor"><a href="#section-sanitization" class="anchor"></a>Sanitization</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="makeNames.html">camel()</a></code> <code><a href="makeNames.html">dotted()</a></code> <code><a href="makeNames.html">snake()</a></code> <code><a href="makeNames.html">upperCamel()</a></code> <code><a href="makeNames.html">makeNames()</a></code> <code><a href="makeNames.html">makeDimnames()</a></code> </p>
        </td>
        <td><p>Make Syntactically Valid Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collapseToString.html">collapseToString()</a></code> </p>
        </td>
        <td><p>Collapse to String</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeNA.html">removeNA()</a></code> </p>
        </td>
        <td><p>Remove Rows and Columns Containing Only NA Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeNA.html">sanitizeNA()</a></code> </p>
        </td>
        <td><p>Sanitize NA Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizePercent.html">sanitizePercent()</a></code> </p>
        </td>
        <td><p>Sanitize Percent</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeRowData.html">sanitizeRowData()</a></code> </p>
        </td>
        <td><p>Sanitize Row Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeSampleData.html">sanitizeSampleData()</a></code> </p>
        </td>
        <td><p>Sanitize Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sortUnique.html">sortUnique()</a></code> </p>
        </td>
        <td><p>Sort and Make Unique</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stripTranscriptVersions.html">stripTranscriptVersions()</a></code> </p>
        </td>
        <td><p>Strip Transcript Versions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toStringUnique.html">toStringUnique()</a></code> </p>
        </td>
        <td><p>Convert to a Unique Character String</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-aggregation" class="hasAnchor"><a href="#section-aggregation" class="anchor"></a>Aggregation</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="aggregate.html">aggregateRows()</a></code> <code><a href="aggregate.html">aggregateCols()</a></code> </p>
        </td>
        <td><p>Aggregate Rows or Columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aggregateCellsToSamples.html">aggregateCellsToSamples()</a></code> </p>
        </td>
        <td><p>Aggregate Cells to Samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="combine.html">combine(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="combine.html">combine(<i>&lt;SingleCellExperiment&gt;</i>,<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Combining or merging different Bioconductor data structures</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-math" class="hasAnchor"><a href="#section-math" class="anchor"></a>Math</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="geometricMean.html">geometricMean()</a></code> </p>
        </td>
        <td><p>Geometric Mean</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="logRatio.html">foldChangeToLogRatio()</a></code> <code><a href="logRatio.html">logRatioToFoldChange()</a></code> </p>
        </td>
        <td><p>Interconvert Log Ratio and Fold Change Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="zerosVsDepth.html">zerosVsDepth()</a></code> </p>
        </td>
        <td><p>Percentage of Zeros vs. Library Depth</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-plots" class="hasAnchor"><a href="#section-plots" class="anchor"></a>Plots</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="plot5Prime3PrimeBias.html">plot5Prime3PrimeBias()</a></code> </p>
        </td>
        <td><p>Plot 5'-&gt;3' Bias</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotBarcodeRanks.html">plotBarcodeRanks()</a></code> </p>
        </td>
        <td><p>Plot Barcode Ranks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCellCounts.html">plotCellCounts()</a></code> </p>
        </td>
        <td><p>Plot Cell Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCountsPerBiotype.html">plotCountsPerBiotype()</a></code> <code><a href="plotCountsPerBiotype.html">plotCountsPerBroadClass()</a></code> </p>
        </td>
        <td><p>Plot Counts per Biotype</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCountsPerGene.html">plotCountsPerGene()</a></code> </p>
        </td>
        <td><p>Plot Counts Per Gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGHeatmap.html">plotDEGHeatmap()</a></code> </p>
        </td>
        <td><p>Differentially Expressed Gene Heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGPCA.html">plotDEGPCA()</a></code> </p>
        </td>
        <td><p>Plot DEG PCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotExonicMappingRate.html">plotExonicMappingRate()</a></code> </p>
        </td>
        <td><p>Plot Exonic Mapping Rate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGenderMarkers.html">plotGenderMarkers()</a></code> </p>
        </td>
        <td><p>Plot Sexually Dimorphic Gender Marker Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGene.html">plotGene()</a></code> </p>
        </td>
        <td><p>Plot Gene Expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGeneSaturation.html">plotGeneSaturation()</a></code> </p>
        </td>
        <td><p>Plot Gene Detection Saturation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGenesDetected.html">plotGenesDetected()</a></code> </p>
        </td>
        <td><p>Plot Genes Detected</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGenesPerCell.html">plotGenesPerCell()</a></code> </p>
        </td>
        <td><p>Plot Genes per Cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotHeatmap.html">plotHeatmap()</a></code> <code><a href="plotHeatmap.html">plotCorrelationHeatmap()</a></code> <code><a href="plotHeatmap.html">plotQuantileHeatmap()</a></code> </p>
        </td>
        <td><p>Plot Heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotIntronicMappingRate.html">plotIntronicMappingRate()</a></code> </p>
        </td>
        <td><p>Plot Intronic Mapping Rate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMappedReads.html">plotMappedReads()</a></code> </p>
        </td>
        <td><p>Plot Mapped Reads</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMappingRate.html">plotMappingRate()</a></code> </p>
        </td>
        <td><p>Plot Mapping Rate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMeanSD.html">plotMeanSD()</a></code> </p>
        </td>
        <td><p>Plot Row Standard Deviations vs. Row Means</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMitoRatio.html">plotMitoRatio()</a></code> </p>
        </td>
        <td><p>Plot Mitochondrial Transcript Abundance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMitoVsCoding.html">plotMitoVsCoding()</a></code> </p>
        </td>
        <td><p>Plot Mitochondrial vs. Coding Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotNovelty.html">plotNovelty()</a></code> </p>
        </td>
        <td><p>Plot Novelty Score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotQC.html">plotQC()</a></code> </p>
        </td>
        <td><p>Plot Quality Control</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotRRNAMappingRate.html">plotRRNAMappingRate()</a></code> </p>
        </td>
        <td><p>Plot Ribosomal RNA (rRNA) Mapping Rate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotReadsPerCell.html">plotReadsPerCell()</a></code> </p>
        </td>
        <td><p>Plot Read Counts per Cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTotalCounts.html">plotTotalCounts()</a></code> </p>
        </td>
        <td><p>Plot Total Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTotalReads.html">plotTotalReads()</a></code> </p>
        </td>
        <td><p>Plot Total Reads</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotUMIsPerCell.html">plotUMIsPerCell()</a></code> </p>
        </td>
        <td><p>Plot UMIs per Cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotUMIsVsGenes.html">plotUMIsVsGenes()</a></code> </p>
        </td>
        <td><p>Plot UMI and Gene Correlation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotVolcano.html">plotVolcano()</a></code> </p>
        </td>
        <td><p>Plot Volcano</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotZerosVsDepth.html">plotZerosVsDepth()</a></code> </p>
        </td>
        <td><p>Plot Percentage of Zeros vs. Library Depth</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-ggplot-" class="hasAnchor"><a href="#section-ggplot-" class="anchor"></a>ggplot2</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="ggplot2-geoms.html">basejump_geom_abline()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label_average()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label_repel()</a></code> </p>
        </td>
        <td><p>ggplot2 Geometric Objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ggplot2-themes.html">theme_paperwhite()</a></code> <code><a href="ggplot2-themes.html">theme_midnight()</a></code> </p>
        </td>
        <td><p>ggplot2 Themes</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-markdown" class="hasAnchor"><a href="#section-markdown" class="anchor"></a>Markdown</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="markdown.html">markdown()</a></code> </p>
        </td>
        <td><p>Markdown</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownHeader.html">markdownHeader()</a></code> <code><a href="markdownHeader.html">mdHeader()</a></code> </p>
        </td>
        <td><p>Markdown Header</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownLink.html">markdownLink()</a></code> <code><a href="markdownLink.html">mdLink()</a></code> </p>
        </td>
        <td><p>Markdown Hyperlink</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownList.html">markdownList()</a></code> <code><a href="markdownList.html">mdList()</a></code> </p>
        </td>
        <td><p>Markdown List</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownPlotlist.html">markdownPlotlist()</a></code> <code><a href="markdownPlotlist.html">mdPlotlist()</a></code> </p>
        </td>
        <td><p>Markdown Plotlist</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownTables.html">markdownTables()</a></code> </p>
        </td>
        <td><p>Create Multiple Tables in LaTeX, HTML, Markdown and reStructuredText</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepareTemplate.html">prepareTemplate()</a></code> </p>
        </td>
        <td><p>Prepare R Markdown Template File</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-coercion" class="hasAnchor"><a href="#section-coercion" class="anchor"></a>Coercion</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="EggNOG-class.html">EggNOG-class</a></code> </p>
        </td>
        <td><p>EggNOG Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="EggNOG.html">EggNOG()</a></code> </p>
        </td>
        <td><p>EggNOG Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Ensembl2Entrez-class.html">Ensembl2Entrez-class</a></code> </p>
        </td>
        <td><p>Ensembl-to-Entrez Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Ensembl2Entrez.html">Ensembl2Entrez()</a></code> </p>
        </td>
        <td><p>Ensembl-to-Entrez Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Gene2Symbol-class.html">Gene2Symbol-class</a></code> </p>
        </td>
        <td><p>Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Gene2Symbol.html">Gene2Symbol()</a></code> </p>
        </td>
        <td><p>Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="HGNC2Ensembl-class.html">HGNC2Ensembl-class</a></code> </p>
        </td>
        <td><p>HGNC-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="HGNC2Ensembl.html">HGNC2Ensembl()</a></code> </p>
        </td>
        <td><p>HGNC-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MGI2Ensembl-class.html">MGI2Ensembl-class</a></code> </p>
        </td>
        <td><p>MGI-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MGI2Ensembl.html">MGI2Ensembl()</a></code> </p>
        </td>
        <td><p>MGI-to-Ensembl Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MethodDefinition.html">methodFunction()</a></code> <code><a href="MethodDefinition.html">methodFormals()</a></code> <code><a href="MethodDefinition.html">hasLocal()</a></code> <code><a href="MethodDefinition.html">extractLocal()</a></code> </p>
        </td>
        <td><p>Access to the Data Inside an S4 Method Definition</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PANTHER-class.html">PANTHER-class</a></code> </p>
        </td>
        <td><p>PANTHER Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PANTHER.html">PANTHER()</a></code> </p>
        </td>
        <td><p>PANTHER Database Annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="RdTags.html">RdTags()</a></code> </p>
        </td>
        <td><p>R Documentation Tags</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Tx2Gene-class.html">Tx2Gene-class</a></code> </p>
        </td>
        <td><p>Transcript-to-Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Tx2Gene.html">Tx2Gene()</a></code> </p>
        </td>
        <td><p>Transcript-to-Gene Identifier Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aggregate.html">aggregateRows()</a></code> <code><a href="aggregate.html">aggregateCols()</a></code> </p>
        </td>
        <td><p>Aggregate Rows or Columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aggregateCellsToSamples.html">aggregateCellsToSamples()</a></code> </p>
        </td>
        <td><p>Aggregate Cells to Samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="alphaSummary.html">alphaSummary()</a></code> </p>
        </td>
        <td><p>Alpha Level Cutoff Summary Statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertFormalGene2Symbol.html">assertFormalGene2Symbol()</a></code> </p>
        </td>
        <td><p>Check Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertFormalInterestingGroups.html">assertFormalInterestingGroups()</a></code> </p>
        </td>
        <td><p>Check Interesting Groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assignAndSaveData.html">assignAndSaveData()</a></code> </p>
        </td>
        <td><p>Assign and Save Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="barcodePattern.html">barcodePattern</a></code> </p>
        </td>
        <td><p>Single-Cell Barcode Pattern
Trailing number is to match cellranger output.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="barcodeRanksPerSample.html">barcodeRanksPerSample()</a></code> </p>
        </td>
        <td><p>Barcode Ranks per Sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cell2sample.html">cell2sample()</a></code> </p>
        </td>
        <td><p>Cell-to-Sample Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cleanSystemLibrary.html">cleanSystemLibrary()</a></code> </p>
        </td>
        <td><p>Check for Clean System Library</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="coerceS4ToList.html">flatFiles()</a></code> <code><a href="coerceS4ToList.html">coerceS4ToList()</a></code> </p>
        </td>
        <td><p>Coerce Any S4 Object to List</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collapseToString.html">collapseToString()</a></code> </p>
        </td>
        <td><p>Collapse to String</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="combine.html">combine(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="combine.html">combine(<i>&lt;SingleCellExperiment&gt;</i>,<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Combining or merging different Bioconductor data structures</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="contrastName.html">contrastName()</a></code> </p>
        </td>
        <td><p>Contrast Name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertGenesToSymbols.html">convertGenesToSymbols()</a></code> <code><a href="convertGenesToSymbols.html">convertSymbolsToGenes()</a></code> </p>
        </td>
        <td><p>Convert Genes to Symbols</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertSampleIDsToNames.html">convertSampleIDsToNames()</a></code> </p>
        </td>
        <td><p>Convert Sample Identifiers to Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertTranscriptsToGenes.html">convertTranscriptsToGenes()</a></code> </p>
        </td>
        <td><p>Convert Transcripts to Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertUCSCBuildToEnsembl.html">convertUCSCBuildToEnsembl()</a></code> </p>
        </td>
        <td><p>Convert UCSC Build to Ensembl</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="counts.html">counts(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="counts.html">`counts<-`(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="decode.html">decode(<i>&lt;DataFrame&gt;</i>)</a></code> </p>
        </td>
        <td><p>Rle objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectHPC.html">detectHPC()</a></code> </p>
        </td>
        <td><p>Detect HPC Environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectLanes.html">detectLanes()</a></code> </p>
        </td>
        <td><p>Detect Sequencing Lanes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectOrganism.html">detectOrganism()</a></code> </p>
        </td>
        <td><p>Detect Organism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="df.html">df</a></code> </p>
        </td>
        <td><p>Example S4 DataFrame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dots.html">dots()</a></code> </p>
        </td>
        <td><p>Extract Dots from Function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="emptyRanges.html">emptyRanges()</a></code> </p>
        </td>
        <td><p>Generate Empty Genomic Ranges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="encode.html">encode()</a></code> </p>
        </td>
        <td><p>Encode Column Data Using Run-Length Encoding</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="export.html">export()</a></code> </p>
        </td>
        <td><p>Export</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="extPattern.html">extPattern</a></code> </p>
        </td>
        <td><p>File Extension Pattern
Note optional matching of gzip.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterCells.html">filterCells()</a></code> </p>
        </td>
        <td><p>Filter Cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gender_markers.html">gender_markers</a></code> </p>
        </td>
        <td><p>Sexually Dimorphic Gender Marker Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneSynonyms.html">geneSynonyms()</a></code> </p>
        </td>
        <td><p>Gene Synonyms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genomeMetadataNames.html">genomeMetadataNames</a></code> </p>
        </td>
        <td><p>Slot Names in Metadata Containing Genome Information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geometricMean.html">geometricMean()</a></code> </p>
        </td>
        <td><p>Geometric Mean</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ggplot2-geoms.html">basejump_geom_abline()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label_average()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label_repel()</a></code> </p>
        </td>
        <td><p>ggplot2 Geometric Objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ggplot2-themes.html">theme_paperwhite()</a></code> <code><a href="ggplot2-themes.html">theme_midnight()</a></code> </p>
        </td>
        <td><p>ggplot2 Themes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="headtail.html">headtail()</a></code> </p>
        </td>
        <td><p>Return the First and Last Part of an Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="import.html">import()</a></code> </p>
        </td>
        <td><p>Import</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="initDir.html">initDir()</a></code> </p>
        </td>
        <td><p>Initialize Directory</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="interestingGroups.html">interestingGroups()</a></code> <code><a href="interestingGroups.html">`interestingGroups<-`()</a></code> </p>
        </td>
        <td><p>Interesting Groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lanePattern.html">lanePattern</a></code> </p>
        </td>
        <td><p>Sequencing Lane Grep Pattern</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadData.html">loadData()</a></code> </p>
        </td>
        <td><p>Load Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadDataAsName.html">loadDataAsName()</a></code> </p>
        </td>
        <td><p>Load Data as Name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="loadRemoteData.html">loadRemoteData()</a></code> </p>
        </td>
        <td><p>Load Remote Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="localOrRemoteFile.html">localOrRemoteFile()</a></code> </p>
        </td>
        <td><p>Dynamically Handle a Local or Remote File Path</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="logRatio.html">foldChangeToLogRatio()</a></code> <code><a href="logRatio.html">logRatioToFoldChange()</a></code> </p>
        </td>
        <td><p>Interconvert Log Ratio and Fold Change Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGRanges.html">makeGRangesFromEnsembl()</a></code> <code><a href="makeGRanges.html">makeGRangesFromEnsDb()</a></code> <code><a href="makeGRanges.html">makeGRangesFromGFF()</a></code> <code><a href="makeGRanges.html">annotable()</a></code> </p>
        </td>
        <td><p>Make Genomic Ranges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGene2Symbol.html">makeGene2SymbolFromEnsembl()</a></code> <code><a href="makeGene2Symbol.html">makeGene2SymbolFromEnsDb()</a></code> <code><a href="makeGene2Symbol.html">makeGene2SymbolFromGFF()</a></code> </p>
        </td>
        <td><p>Make Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeNames.html">camel()</a></code> <code><a href="makeNames.html">dotted()</a></code> <code><a href="makeNames.html">snake()</a></code> <code><a href="makeNames.html">upperCamel()</a></code> <code><a href="makeNames.html">makeNames()</a></code> <code><a href="makeNames.html">makeDimnames()</a></code> </p>
        </td>
        <td><p>Make Syntactically Valid Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeSampleData.html">makeSampleData()</a></code> </p>
        </td>
        <td><p>Make Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeSingleCellExperiment.html">makeSingleCellExperiment()</a></code> </p>
        </td>
        <td><p>Make Single-Cell Experiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeSummarizedExperiment.html">makeSummarizedExperiment()</a></code> </p>
        </td>
        <td><p>Make Summarized Experiment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeTx2Gene.html">makeTx2GeneFromEnsembl()</a></code> <code><a href="makeTx2Gene.html">makeTx2GeneFromEnsDb()</a></code> <code><a href="makeTx2Gene.html">makeTx2GeneFromGFF()</a></code> </p>
        </td>
        <td><p>Make Transcript-to-Gene Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapCellsToSamples.html">mapCellsToSamples()</a></code> </p>
        </td>
        <td><p>Map Cells to Samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapGenes.html">mapGenesToRownames()</a></code> <code><a href="mapGenes.html">mapGenesToIDs()</a></code> <code><a href="mapGenes.html">mapGenesToSymbols()</a></code> </p>
        </td>
        <td><p>Map Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdown.html">markdown()</a></code> </p>
        </td>
        <td><p>Markdown</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownHeader.html">markdownHeader()</a></code> <code><a href="markdownHeader.html">mdHeader()</a></code> </p>
        </td>
        <td><p>Markdown Header</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownLink.html">markdownLink()</a></code> <code><a href="markdownLink.html">mdLink()</a></code> </p>
        </td>
        <td><p>Markdown Hyperlink</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownList.html">markdownList()</a></code> <code><a href="markdownList.html">mdList()</a></code> </p>
        </td>
        <td><p>Markdown List</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownPlotlist.html">markdownPlotlist()</a></code> <code><a href="markdownPlotlist.html">mdPlotlist()</a></code> </p>
        </td>
        <td><p>Markdown Plotlist</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownTables.html">markdownTables()</a></code> </p>
        </td>
        <td><p>Create Multiple Tables in LaTeX, HTML, Markdown and reStructuredText</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mat.html">mat</a></code> </p>
        </td>
        <td><p>Example Dense Matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matchArgsToDoCall.html">matchArgsToDoCall()</a></code> </p>
        </td>
        <td><p>Match Arguments to Do Call</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matchInterestingGroups.html">matchInterestingGroups()</a></code> </p>
        </td>
        <td><p>Match Interesting Groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meltCounts.html">meltCounts()</a></code> </p>
        </td>
        <td><p>Melt Count Matrix to Long Format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="metrics.html">metrics()</a></code> <code><a href="metrics.html">metricsPerSample()</a></code> </p>
        </td>
        <td><p>Metrics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="microplate.html">microplate()</a></code> </p>
        </td>
        <td><p>Microtiter Plate Well Identifiers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="minimalSampleData.html">minimalSampleData()</a></code> </p>
        </td>
        <td><p>Minimal Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multiassignAsEnvir.html">multiassignAsEnvir()</a></code> </p>
        </td>
        <td><p>Assign Multiple Objects as an Environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="naStrings.html">naStrings</a></code> </p>
        </td>
        <td><p>NA Strings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organism.html">organism(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;DataFrame&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;GRanges&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Organism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organism_mappings.html">organism_mappings</a></code> </p>
        </td>
        <td><p>Organism Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parseRd.html">parseRd()</a></code> </p>
        </td>
        <td><p>Parse R Documentation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot5Prime3PrimeBias.html">plot5Prime3PrimeBias()</a></code> </p>
        </td>
        <td><p>Plot 5'-&gt;3' Bias</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotBarcodeRanks.html">plotBarcodeRanks()</a></code> </p>
        </td>
        <td><p>Plot Barcode Ranks</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCellCounts.html">plotCellCounts()</a></code> </p>
        </td>
        <td><p>Plot Cell Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCountsPerBiotype.html">plotCountsPerBiotype()</a></code> <code><a href="plotCountsPerBiotype.html">plotCountsPerBroadClass()</a></code> </p>
        </td>
        <td><p>Plot Counts per Biotype</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCountsPerGene.html">plotCountsPerGene()</a></code> </p>
        </td>
        <td><p>Plot Counts Per Gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGHeatmap.html">plotDEGHeatmap()</a></code> </p>
        </td>
        <td><p>Differentially Expressed Gene Heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotDEGPCA.html">plotDEGPCA()</a></code> </p>
        </td>
        <td><p>Plot DEG PCA</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotExonicMappingRate.html">plotExonicMappingRate()</a></code> </p>
        </td>
        <td><p>Plot Exonic Mapping Rate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGenderMarkers.html">plotGenderMarkers()</a></code> </p>
        </td>
        <td><p>Plot Sexually Dimorphic Gender Marker Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGene.html">plotGene()</a></code> </p>
        </td>
        <td><p>Plot Gene Expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGeneSaturation.html">plotGeneSaturation()</a></code> </p>
        </td>
        <td><p>Plot Gene Detection Saturation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGenesDetected.html">plotGenesDetected()</a></code> </p>
        </td>
        <td><p>Plot Genes Detected</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGenesPerCell.html">plotGenesPerCell()</a></code> </p>
        </td>
        <td><p>Plot Genes per Cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotHeatmap.html">plotHeatmap()</a></code> <code><a href="plotHeatmap.html">plotCorrelationHeatmap()</a></code> <code><a href="plotHeatmap.html">plotQuantileHeatmap()</a></code> </p>
        </td>
        <td><p>Plot Heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotIntronicMappingRate.html">plotIntronicMappingRate()</a></code> </p>
        </td>
        <td><p>Plot Intronic Mapping Rate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMappedReads.html">plotMappedReads()</a></code> </p>
        </td>
        <td><p>Plot Mapped Reads</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMappingRate.html">plotMappingRate()</a></code> </p>
        </td>
        <td><p>Plot Mapping Rate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMeanSD.html">plotMeanSD()</a></code> </p>
        </td>
        <td><p>Plot Row Standard Deviations vs. Row Means</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMitoRatio.html">plotMitoRatio()</a></code> </p>
        </td>
        <td><p>Plot Mitochondrial Transcript Abundance</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotMitoVsCoding.html">plotMitoVsCoding()</a></code> </p>
        </td>
        <td><p>Plot Mitochondrial vs. Coding Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotNovelty.html">plotNovelty()</a></code> </p>
        </td>
        <td><p>Plot Novelty Score</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotQC.html">plotQC()</a></code> </p>
        </td>
        <td><p>Plot Quality Control</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotRRNAMappingRate.html">plotRRNAMappingRate()</a></code> </p>
        </td>
        <td><p>Plot Ribosomal RNA (rRNA) Mapping Rate</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotReadsPerCell.html">plotReadsPerCell()</a></code> </p>
        </td>
        <td><p>Plot Read Counts per Cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTotalCounts.html">plotTotalCounts()</a></code> </p>
        </td>
        <td><p>Plot Total Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTotalReads.html">plotTotalReads()</a></code> </p>
        </td>
        <td><p>Plot Total Reads</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotUMIsPerCell.html">plotUMIsPerCell()</a></code> </p>
        </td>
        <td><p>Plot UMIs per Cell</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotUMIsVsGenes.html">plotUMIsVsGenes()</a></code> </p>
        </td>
        <td><p>Plot UMI and Gene Correlation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotVolcano.html">plotVolcano()</a></code> </p>
        </td>
        <td><p>Plot Volcano</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotZerosVsDepth.html">plotZerosVsDepth()</a></code> </p>
        </td>
        <td><p>Plot Percentage of Zeros vs. Library Depth</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepareTemplate.html">prepareTemplate()</a></code> </p>
        </td>
        <td><p>Prepare R Markdown Template File</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="printString.html">printString()</a></code> </p>
        </td>
        <td><p>Print String</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rdataExtPattern.html">rdataExtPattern</a></code> </p>
        </td>
        <td><p>R Data Extension Pattern</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rdataLoadError.html">rdataLoadError</a></code> </p>
        </td>
        <td><p>R Data Load Error</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="realpath.html">realpath()</a></code> </p>
        </td>
        <td><p>Express File Paths in Canonical Form</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="relativeLogExpression.html">relativeLogExpression()</a></code> </p>
        </td>
        <td><p>Relative Log Expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeNA.html">removeNA()</a></code> </p>
        </td>
        <td><p>Remove Rows and Columns Containing Only NA Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rse.html">rse</a></code> </p>
        </td>
        <td><p>Example RNA-Seq Data Set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleData.html">sampleData()</a></code> <code><a href="sampleData.html">`sampleData<-`()</a></code> <code><a href="sampleData.html">sampleNames(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleNames.html">`sampleNames<-`(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;character&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sample Names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeNA.html">sanitizeNA()</a></code> </p>
        </td>
        <td><p>Sanitize NA Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizePercent.html">sanitizePercent()</a></code> </p>
        </td>
        <td><p>Sanitize Percent</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeRowData.html">sanitizeRowData()</a></code> </p>
        </td>
        <td><p>Sanitize Row Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeSampleData.html">sanitizeSampleData()</a></code> </p>
        </td>
        <td><p>Sanitize Sample Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="saveData.html">saveData()</a></code> </p>
        </td>
        <td><p>Save Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="saveRdExamples.html">saveRdExamples()</a></code> </p>
        </td>
        <td><p>Save R Documentation Examples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sce.html">sce</a></code> </p>
        </td>
        <td><p>Example Single-Cell RNA-Seq Data Set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="selectSamples.html">selectSamples()</a></code> </p>
        </td>
        <td><p>Select Samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="separator.html">separator()</a></code> </p>
        </td>
        <td><p>Separator Bar</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="show.html">show(<i>&lt;EggNOG&gt;</i>)</a></code> <code><a href="show.html">show(<i>&lt;PANTHER&gt;</i>)</a></code> </p>
        </td>
        <td><p>Show an Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="showSlotInfo.html">showSlotInfo()</a></code> </p>
        </td>
        <td><p>Show Slot Information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sortUnique.html">sortUnique()</a></code> </p>
        </td>
        <td><p>Sort and Make Unique</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sparse.html">sparse</a></code> </p>
        </td>
        <td><p>Example Sparse Matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="standardizeCall.html">standardizeCall()</a></code> </p>
        </td>
        <td><p>Standardize Call</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stripTranscriptVersions.html">stripTranscriptVersions()</a></code> </p>
        </td>
        <td><p>Strip Transcript Versions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetPerSample.html">subsetPerSample()</a></code> </p>
        </td>
        <td><p>Subset Per Sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summary.html">summary(<i>&lt;Gene2Symbol&gt;</i>)</a></code> <code><a href="summary.html">summary(<i>&lt;Tx2Gene&gt;</i>)</a></code> </p>
        </td>
        <td><p>Object Summaries</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tabular.html">tabular()</a></code> </p>
        </td>
        <td><p>R Documentation Table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tbl.html">tbl</a></code> </p>
        </td>
        <td><p>Example Tibble</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tmm.html">tmm()</a></code> </p>
        </td>
        <td><p>Trimmed Mean of M-Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toStringUnique.html">toStringUnique()</a></code> </p>
        </td>
        <td><p>Convert to a Unique Character String</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="topCellsPerSample.html">topCellsPerSample()</a></code> </p>
        </td>
        <td><p>Top Cells per Sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="topTables.html">topTables()</a></code> </p>
        </td>
        <td><p>Top Tables of Differential Expression Results</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tpm.html">tpm(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Transcripts per Million</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="transmit.html">transmit()</a></code> </p>
        </td>
        <td><p>Transmit Files from a Remote FTP Server</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tx_se.html">tx_se</a></code> </p>
        </td>
        <td><p>Example Transcript-Level Data Set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="uniteInterestingGroups.html">uniteInterestingGroups()</a></code> </p>
        </td>
        <td><p>Unite Interesting Groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="validateClasses.html">validateClasses()</a></code> </p>
        </td>
        <td><p>Validate Classes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="writeCounts.html">writeCounts()</a></code> </p>
        </td>
        <td><p>Write Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="zerosVsDepth.html">zerosVsDepth()</a></code> </p>
        </td>
        <td><p>Percentage of Zeros vs. Library Depth</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-run-length-encoding" class="hasAnchor"><a href="#section-run-length-encoding" class="anchor"></a>Run-length encoding</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="decode.html">decode(<i>&lt;DataFrame&gt;</i>)</a></code> </p>
        </td>
        <td><p>Rle objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="encode.html">encode()</a></code> </p>
        </td>
        <td><p>Encode Column Data Using Run-Length Encoding</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-documentation" class="hasAnchor"><a href="#section-documentation" class="anchor"></a>Documentation</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="RdTags.html">RdTags()</a></code> </p>
        </td>
        <td><p>R Documentation Tags</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="parseRd.html">parseRd()</a></code> </p>
        </td>
        <td><p>Parse R Documentation</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="saveRdExamples.html">saveRdExamples()</a></code> </p>
        </td>
        <td><p>Save R Documentation Examples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tabular.html">tabular()</a></code> </p>
        </td>
        <td><p>R Documentation Table</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-developer" class="hasAnchor"><a href="#section-developer" class="anchor"></a>Developer</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="deprecated.html">ensembl()</a></code> <code><a href="deprecated.html">gene2symbolFromGFF()</a></code> <code><a href="deprecated.html">gene2symbolFromGTF()</a></code> <code><a href="deprecated.html">sanitizeAnnotable()</a></code> <code><a href="deprecated.html">tx2geneFromGFF()</a></code> <code><a href="deprecated.html">tx2geneFromGTF()</a></code> <code><a href="deprecated.html">multiassignAsNewEnvir()</a></code> <code><a href="deprecated.html">geomean()</a></code> <code><a href="deprecated.html">midnightTheme()</a></code> <code><a href="deprecated.html">paperwhiteTheme()</a></code> <code><a href="deprecated.html">aggregateFeatures()</a></code> <code><a href="deprecated.html">aggregateReplicates()</a></code> <code><a href="deprecated.html">aggregateSamples()</a></code> <code><a href="deprecated.html">assertFormalGene2symbol()</a></code> <code><a href="deprecated.html">checkClasses()</a></code> <code><a href="deprecated.html">eggnog()</a></code> <code><a href="deprecated.html">ensembl2entrez()</a></code> <code><a href="deprecated.html">fixNA()</a></code> <code><a href="deprecated.html">gene2symbol()</a></code> <code><a href="deprecated.html">grepString()</a></code> <code><a href="deprecated.html">hgnc2gene()</a></code> <code><a href="deprecated.html">hgnc2ensembl()</a></code> <code><a href="deprecated.html">initializeDirectory()</a></code> <code><a href="deprecated.html">kables()</a></code> <code><a href="deprecated.html">makeGene2symbolFromEnsembl()</a></code> <code><a href="deprecated.html">makeGene2symbolFromGFF()</a></code> <code><a href="deprecated.html">makeGene2symbolFromGTF()</a></code> <code><a href="deprecated.html">makeTx2geneFromEnsembl()</a></code> <code><a href="deprecated.html">makeTx2geneFromGFF()</a></code> <code><a href="deprecated.html">makeTx2geneFromGTF()</a></code> <code><a href="deprecated.html">mgi2gene()</a></code> <code><a href="deprecated.html">mgi2ensembl()</a></code> <code><a href="deprecated.html">panther()</a></code> <code><a href="deprecated.html">readFileByExtension()</a></code> <code><a href="deprecated.html">readGFF()</a></code> <code><a href="deprecated.html">readGTF()</a></code> <code><a href="deprecated.html">readJSON()</a></code> <code><a href="deprecated.html">readYAML()</a></code> <code><a href="deprecated.html">tx2gene()</a></code> </p>
        </td>
        <td><p>Deprecated Functions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cleanSystemLibrary.html">cleanSystemLibrary()</a></code> </p>
        </td>
        <td><p>Check for Clean System Library</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="dots.html">dots()</a></code> </p>
        </td>
        <td><p>Extract Dots from Function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matchArgsToDoCall.html">matchArgsToDoCall()</a></code> </p>
        </td>
        <td><p>Match Arguments to Do Call</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matchInterestingGroups.html">matchInterestingGroups()</a></code> </p>
        </td>
        <td><p>Match Interesting Groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MethodDefinition.html">methodFunction()</a></code> <code><a href="MethodDefinition.html">methodFormals()</a></code> <code><a href="MethodDefinition.html">hasLocal()</a></code> <code><a href="MethodDefinition.html">extractLocal()</a></code> </p>
        </td>
        <td><p>Access to the Data Inside an S4 Method Definition</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multiassignAsEnvir.html">multiassignAsEnvir()</a></code> </p>
        </td>
        <td><p>Assign Multiple Objects as an Environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="printString.html">printString()</a></code> </p>
        </td>
        <td><p>Print String</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="showSlotInfo.html">showSlotInfo()</a></code> </p>
        </td>
        <td><p>Show Slot Information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="standardizeCall.html">standardizeCall()</a></code> </p>
        </td>
        <td><p>Standardize Call</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="validateClasses.html">validateClasses()</a></code> </p>
        </td>
        <td><p>Validate Classes</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-miscellaneous" class="hasAnchor"><a href="#section-miscellaneous" class="anchor"></a>Miscellaneous</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="detectHPC.html">detectHPC()</a></code> </p>
        </td>
        <td><p>Detect HPC Environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectLanes.html">detectLanes()</a></code> </p>
        </td>
        <td><p>Detect Sequencing Lanes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectOrganism.html">detectOrganism()</a></code> </p>
        </td>
        <td><p>Detect Organism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="headtail.html">headtail()</a></code> </p>
        </td>
        <td><p>Return the First and Last Part of an Object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="initDir.html">initDir()</a></code> </p>
        </td>
        <td><p>Initialize Directory</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="microplate.html">microplate()</a></code> </p>
        </td>
        <td><p>Microtiter Plate Well Identifiers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="realpath.html">realpath()</a></code> </p>
        </td>
        <td><p>Express File Paths in Canonical Form</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="separator.html">separator()</a></code> </p>
        </td>
        <td><p>Separator Bar</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-assert-checks" class="hasAnchor"><a href="#section-assert-checks" class="anchor"></a>Assert checks</h2>
          <p class="section-desc"><p>Readable check functions to ensure code integrity.</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="assertFormalGene2Symbol.html">assertFormalGene2Symbol()</a></code> </p>
        </td>
        <td><p>Check Gene-to-Symbol Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assertFormalInterestingGroups.html">assertFormalInterestingGroups()</a></code> </p>
        </td>
        <td><p>Check Interesting Groups</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-s-generics-without-methods" class="hasAnchor"><a href="#section-s-generics-without-methods" class="anchor"></a>S4 generics without methods</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="alphaSummary.html">alphaSummary()</a></code> </p>
        </td>
        <td><p>Alpha Level Cutoff Summary Statistics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="barcodePattern.html">barcodePattern</a></code> </p>
        </td>
        <td><p>Single-Cell Barcode Pattern
Trailing number is to match cellranger output.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="barcodeRanksPerSample.html">barcodeRanksPerSample()</a></code> </p>
        </td>
        <td><p>Barcode Ranks per Sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="contrastName.html">contrastName()</a></code> </p>
        </td>
        <td><p>Contrast Name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterCells.html">filterCells()</a></code> </p>
        </td>
        <td><p>Filter Cells</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="relativeLogExpression.html">relativeLogExpression()</a></code> </p>
        </td>
        <td><p>Relative Log Expression</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tmm.html">tmm()</a></code> </p>
        </td>
        <td><p>Trimmed Mean of M-Values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="topTables.html">topTables()</a></code> </p>
        </td>
        <td><p>Top Tables of Differential Expression Results</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-globals" class="hasAnchor"><a href="#section-globals" class="anchor"></a>Globals</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="extPattern.html">extPattern</a></code> </p>
        </td>
        <td><p>File Extension Pattern
Note optional matching of gzip.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genomeMetadataNames.html">genomeMetadataNames</a></code> </p>
        </td>
        <td><p>Slot Names in Metadata Containing Genome Information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lanePattern.html">lanePattern</a></code> </p>
        </td>
        <td><p>Sequencing Lane Grep Pattern</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="naStrings.html">naStrings</a></code> </p>
        </td>
        <td><p>NA Strings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rdataExtPattern.html">rdataExtPattern</a></code> </p>
        </td>
        <td><p>R Data Extension Pattern</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rdataLoadError.html">rdataLoadError</a></code> </p>
        </td>
        <td><p>R Data Load Error</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-example-data" class="hasAnchor"><a href="#section-example-data" class="anchor"></a>Example data</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="df.html">df</a></code> </p>
        </td>
        <td><p>Example S4 DataFrame</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gender_markers.html">gender_markers</a></code> </p>
        </td>
        <td><p>Sexually Dimorphic Gender Marker Genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mat.html">mat</a></code> </p>
        </td>
        <td><p>Example Dense Matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organism_mappings.html">organism_mappings</a></code> </p>
        </td>
        <td><p>Organism Mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rse.html">rse</a></code> </p>
        </td>
        <td><p>Example RNA-Seq Data Set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sce.html">sce</a></code> </p>
        </td>
        <td><p>Example Single-Cell RNA-Seq Data Set</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sparse.html">sparse</a></code> </p>
        </td>
        <td><p>Example Sparse Matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tbl.html">tbl</a></code> </p>
        </td>
        <td><p>Example Tibble</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tx_se.html">tx_se</a></code> </p>
        </td>
        <td><p>Example Transcript-Level Data Set</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-s-classes">S4 classes</a></li>
      <li><a href="#section-import-export">Import/export</a></li>
      <li><a href="#section-genome-annotations">Genome annotations</a></li>
      <li><a href="#section-gene-identifier-mapping">Gene identifier mapping</a></li>
      <li><a href="#section-counts">Counts</a></li>
      <li><a href="#section-sample-level">Sample-level</a></li>
      <li><a href="#section-cell-level">Cell-level</a></li>
      <li><a href="#section-metadata">Metadata</a></li>
      <li><a href="#section-sanitization">Sanitization</a></li>
      <li><a href="#section-aggregation">Aggregation</a></li>
      <li><a href="#section-math">Math</a></li>
      <li><a href="#section-plots">Plots</a></li>
      <li><a href="#section-ggplot-">ggplot2</a></li>
      <li><a href="#section-markdown">Markdown</a></li>
      <li><a href="#section-coercion">Coercion</a></li>
      <li><a href="#section-run-length-encoding">Run-length encoding</a></li>
      <li><a href="#section-documentation">Documentation</a></li>
      <li><a href="#section-developer">Developer</a></li>
      <li><a href="#section-miscellaneous">Miscellaneous</a></li>
      <li><a href="#section-assert-checks">Assert checks</a></li>
      <li><a href="#section-s-generics-without-methods">S4 generics without methods</a></li>
      <li><a href="#section-globals">Globals</a></li>
      <li><a href="#section-example-data">Example data</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

