<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Function reference" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-s-classes" class="hasAnchor"><a href="#section-s-classes" class="anchor"></a>S4 classes</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="EggNOG-class.html">EggNOG()</a></code> </p>
        </td>
        <td><p>EggNOG database annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Ensembl2Entrez-class.html">Ensembl2Entrez()</a></code> </p>
        </td>
        <td><p>Ensembl-to-Entrez gene identifier mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Gene2Symbol-class.html">Gene2Symbol()</a></code> </p>
        </td>
        <td><p>Gene-to-symbol mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="HGNC2Ensembl-class.html">HGNC2Ensembl()</a></code> </p>
        </td>
        <td><p>HGNC-to-Ensembl gene identifier mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="MGI2Ensembl-class.html">MGI2Ensembl()</a></code> </p>
        </td>
        <td><p>MGI-to-Ensembl gene identifier mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="PANTHER-class.html">PANTHER()</a></code> </p>
        </td>
        <td><p>PANTHER database annotations</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="Tx2Gene-class.html">Tx2Gene()</a></code> </p>
        </td>
        <td><p>Transcript-to-gene identifier mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="show.html">show(<i>&lt;EggNOG&gt;</i>)</a></code> <code><a href="show.html">show(<i>&lt;PANTHER&gt;</i>)</a></code> </p>
        </td>
        <td><p>Show an object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="summary.html">summary(<i>&lt;Gene2Symbol&gt;</i>)</a></code> <code><a href="summary.html">summary(<i>&lt;Tx2Gene&gt;</i>)</a></code> </p>
        </td>
        <td><p>Object summaries</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-genome-annotations" class="hasAnchor"><a href="#section-genome-annotations" class="anchor"></a>Genome annotations</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="emptyRanges.html">emptyRanges()</a></code> </p>
        </td>
        <td><p>Generate empty genomic ranges</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneSynonyms.html">geneSynonyms()</a></code> </p>
        </td>
        <td><p>Gene synonyms</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGRangesFromEnsDb.html">makeGRangesFromEnsDb()</a></code> </p>
        </td>
        <td><p>Make <code>GRanges</code> from <code>EnsDb</code> object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGRangesFromEnsembl.html">makeGRangesFromEnsembl()</a></code> <code><a href="makeGRangesFromEnsembl.html">annotable()</a></code> </p>
        </td>
        <td><p>Make <code>GRanges</code> from Ensembl</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGRangesFromGFF.html">makeGRangesFromGFF()</a></code> <code><a href="makeGRangesFromGFF.html">makeGRangesFromGTF()</a></code> </p>
        </td>
        <td><p>Make <code>GRanges</code> from a GFF/GTF file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeGene2Symbol.html">makeGene2SymbolFromEnsembl()</a></code> <code><a href="makeGene2Symbol.html">makeGene2SymbolFromEnsDb()</a></code> <code><a href="makeGene2Symbol.html">makeGene2SymbolFromGFF()</a></code> </p>
        </td>
        <td><p>Make a <code>Gene2Symbol</code> object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeTx2Gene.html">makeTx2GeneFromEnsembl()</a></code> <code><a href="makeTx2Gene.html">makeTx2GeneFromEnsDb()</a></code> <code><a href="makeTx2Gene.html">makeTx2GeneFromGFF()</a></code> </p>
        </td>
        <td><p>Make a <code>Tx2Gene</code> object</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-gene-identifier-mapping" class="hasAnchor"><a href="#section-gene-identifier-mapping" class="anchor"></a>Gene identifier mapping</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="convertGenesToSymbols.html">convertGenesToSymbols(<i>&lt;character&gt;</i>)</a></code> <code><a href="convertGenesToSymbols.html">convertGenesToSymbols(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="convertGenesToSymbols.html">convertGenesToSymbols(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="convertGenesToSymbols.html">convertGenesToSymbols(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="convertGenesToSymbols.html">convertSymbolsToGenes(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Convert genes to symbols</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertSampleIDsToNames.html">convertSampleIDsToNames(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="convertSampleIDsToNames.html">convertSampleIDsToNames(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Convert sample identifiers to names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertTranscriptsToGenes.html">convertTranscriptsToGenes(<i>&lt;character&gt;</i>)</a></code> <code><a href="convertTranscriptsToGenes.html">convertTranscriptsToGenes(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="convertTranscriptsToGenes.html">convertTranscriptsToGenes(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="convertTranscriptsToGenes.html">convertTranscriptsToGenes(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Convert transcripts to genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertUCSCBuildToEnsembl.html">convertUCSCBuildToEnsembl()</a></code> </p>
        </td>
        <td><p>Convert UCSC build to Ensembl</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapGenes.html">mapGenesToRownames(<i>&lt;Gene2Symbol&gt;</i>)</a></code> <code><a href="mapGenes.html">mapGenesToRownames(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="mapGenes.html">mapGenesToIDs(<i>&lt;Gene2Symbol&gt;</i>)</a></code> <code><a href="mapGenes.html">mapGenesToIDs(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="mapGenes.html">mapGenesToSymbols(<i>&lt;Gene2Symbol&gt;</i>)</a></code> <code><a href="mapGenes.html">mapGenesToSymbols(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Map genes</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-counts" class="hasAnchor"><a href="#section-counts" class="anchor"></a>Counts</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="counts.html">counts(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="counts.html">`counts<-`(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meltCounts.html">meltCounts(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Melt count matrix into long format</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tpm.html">tpm(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Transcripts per million</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-metadata" class="hasAnchor"><a href="#section-metadata" class="anchor"></a>Metadata</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="interestingGroups.html">interestingGroups(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="interestingGroups.html">`interestingGroups<-`(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;character&gt;</i>)</a></code> <code><a href="interestingGroups.html">`interestingGroups<-`(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;NULL&gt;</i>)</a></code> </p>
        </td>
        <td><p>Interesting groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organism.html">organism(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;DataFrame&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;GRanges&gt;</i>)</a></code> <code><a href="organism.html">organism(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Organism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="uniteInterestingGroups.html">uniteInterestingGroups(<i>&lt;DataFrame&gt;</i>)</a></code> </p>
        </td>
        <td><p>Unite interesting groups into a single column</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-sample-level-metadata" class="hasAnchor"><a href="#section-sample-level-metadata" class="anchor"></a>Sample-level metadata</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="makeSampleData.html">makeSampleData()</a></code> </p>
        </td>
        <td><p>Make sample data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeSummarizedExperiment.html">makeSummarizedExperiment()</a></code> </p>
        </td>
        <td><p>Make a <code>SummarizedExperiment</code> object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="metrics.html">metrics(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="metrics.html">metrics(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> <code><a href="metrics.html">metricsPerSample(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Quality control metrics</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="minimalSampleData.html">minimalSampleData()</a></code> </p>
        </td>
        <td><p>Minimal sample data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleData.html">sampleData(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="sampleData.html">sampleData(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> <code><a href="sampleData.html">`sampleData<-`(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;DataFrame&gt;</i>)</a></code> <code><a href="sampleData.html">`sampleData<-`(<i>&lt;SingleCellExperiment&gt;</i>,<i>&lt;DataFrame&gt;</i>)</a></code> <code><a href="sampleData.html">sampleNames(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sample data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sampleNames.html">`sampleNames<-`(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;character&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sample names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="selectSamples.html">selectSamples(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="selectSamples.html">selectSamples(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Select samples</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-cell-level-metadata" class="hasAnchor"><a href="#section-cell-level-metadata" class="anchor"></a>Cell-level metadata</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="aggregateCellsToSamples.html">aggregateCellsToSamples(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Aggregate cells to samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cell2sample.html">cell2sample(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Cell-to-sample mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeSingleCellExperiment.html">makeSingleCellExperiment()</a></code> </p>
        </td>
        <td><p>Make a <code>SingleCellExperiment</code> object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mapCellsToSamples.html">mapCellsToSamples()</a></code> </p>
        </td>
        <td><p>Map cells to samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subsetPerSample.html">subsetPerSample(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Subset per sample</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="topCellsPerSample.html">topCellsPerSample(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Top cells per sample</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-sanitization" class="hasAnchor"><a href="#section-sanitization" class="anchor"></a>Sanitization</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="autopadZeros.html">autopadZeros(<i>&lt;character&gt;</i>)</a></code> <code><a href="autopadZeros.html">autopadZeros(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="autopadZeros.html">autopadZeros(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Autopad zeros</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="collapseToString.html">collapseToString(<i>&lt;atomic&gt;</i>)</a></code> <code><a href="collapseToString.html">collapseToString(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="collapseToString.html">collapseToString(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="collapseToString.html">collapseToString(<i>&lt;DataFrame&gt;</i>)</a></code> </p>
        </td>
        <td><p>Collapse to string</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="relevel.html">relevelRowRanges()</a></code> <code><a href="relevel.html">relevelColData()</a></code> </p>
        </td>
        <td><p>Relevel row or column data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeNA.html">removeNA(<i>&lt;atomic&gt;</i>)</a></code> <code><a href="removeNA.html">removeNA(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="removeNA.html">removeNA(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="removeNA.html">removeNA(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="removeNA.html">removeNA(<i>&lt;DataFrame&gt;</i>)</a></code> </p>
        </td>
        <td><p>Remove rows and columns containing only <code>NA</code> values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeNA.html">sanitizeNA(<i>&lt;atomic&gt;</i>)</a></code> <code><a href="sanitizeNA.html">sanitizeNA(<i>&lt;character&gt;</i>)</a></code> <code><a href="sanitizeNA.html">sanitizeNA(<i>&lt;factor&gt;</i>)</a></code> <code><a href="sanitizeNA.html">sanitizeNA(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="sanitizeNA.html">sanitizeNA(<i>&lt;DataFrame&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sanitize <code>NA</code> values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizePercent.html">sanitizePercent(<i>&lt;atomic&gt;</i>)</a></code> <code><a href="sanitizePercent.html">sanitizePercent(<i>&lt;character&gt;</i>)</a></code> </p>
        </td>
        <td><p>Sanitize percentage</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sanitizeSampleData.html">sanitizeSampleData()</a></code> </p>
        </td>
        <td><p>Sanitize sample data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sortUnique.html">sortUnique()</a></code> </p>
        </td>
        <td><p>Sort and make unique</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stripTranscriptVersions.html">stripTranscriptVersions(<i>&lt;character&gt;</i>)</a></code> <code><a href="stripTranscriptVersions.html">stripTranscriptVersions(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="stripTranscriptVersions.html">stripTranscriptVersions(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="stripTranscriptVersions.html">stripTranscriptVersions(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Strip transcript versions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toStringUnique.html">toStringUnique()</a></code> </p>
        </td>
        <td><p>Convert to a unique character string</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-aggregation" class="hasAnchor"><a href="#section-aggregation" class="anchor"></a>Aggregation</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="aggregate.html">aggregateRows(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="aggregate.html">aggregateRows(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="aggregate.html">aggregateRows(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="aggregate.html">aggregateCols(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="aggregate.html">aggregateCols(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="aggregate.html">aggregateCols(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="aggregate.html">aggregateCols(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Aggregate rows or columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="aggregateCellsToSamples.html">aggregateCellsToSamples(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Aggregate cells to samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="combine.html">combine(<i>&lt;SummarizedExperiment&gt;</i>,<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="combine.html">combine(<i>&lt;SingleCellExperiment&gt;</i>,<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Combining or merging different Bioconductor data structures</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-math" class="hasAnchor"><a href="#section-math" class="anchor"></a>Math</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="geometricMean.html">geometricMean(<i>&lt;numeric&gt;</i>)</a></code> <code><a href="geometricMean.html">geometricMean(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="geometricMean.html">geometricMean(<i>&lt;sparseMatrix&gt;</i>)</a></code> </p>
        </td>
        <td><p>Geometric mean</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="logRatio.html">foldChangeToLogRatio(<i>&lt;numeric&gt;</i>)</a></code> <code><a href="logRatio.html">logRatioToFoldChange(<i>&lt;numeric&gt;</i>)</a></code> </p>
        </td>
        <td><p>Interconvert log ratio and fold change values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="zerosVsDepth.html">zerosVsDepth(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="zerosVsDepth.html">zerosVsDepth(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="zerosVsDepth.html">zerosVsDepth(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="zerosVsDepth.html">zerosVsDepth(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Percentage of zeros vs. library depth</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-plots" class="hasAnchor"><a href="#section-plots" class="anchor"></a>Plots</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="plotCorrelationHeatmap.html">plotCorrelationHeatmap(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="plotCorrelationHeatmap.html">plotCorrelationHeatmap(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Correlation heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCounts.html">plotCounts(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCountsPerBiotype.html">plotCountsPerBiotype(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="plotCountsPerBiotype.html">plotCountsPerBroadClass(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot counts per biotype</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCountsPerBroadClass.html">plotCountsPerBroadClass</a></code> </p>
        </td>
        <td><p>Plot counts per broad class definition</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotCountsPerGene.html">plotCountsPerGene(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="plotCountsPerGene.html">plotCountsPerGene(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot counts per gene</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGenderMarkers.html">plotGenderMarkers(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot sexually dimorphic gender marker genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotGenesDetected.html">plotGenesDetected(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="plotGenesDetected.html">plotGenesDetected(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot genes detected</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotHeatmap.html">plotHeatmap(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="plotHeatmap.html">plotHeatmap(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotPCA.html">plotPCA(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Principal component analysis plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotQC.html">plotQC(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Quality control</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotQuantileHeatmap.html">plotQuantileHeatmap(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="plotQuantileHeatmap.html">plotQuantileHeatmap(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Quantile heatmap</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotTotalCounts.html">plotTotalCounts(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> <code><a href="plotTotalCounts.html">plotTotalCounts(<i>&lt;SingleCellExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot total read counts</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plotZerosVsDepth.html">plotZerosVsDepth(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot percentage of zeros vs. library depth</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-ggplot-" class="hasAnchor"><a href="#section-ggplot-" class="anchor"></a>ggplot2</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="ggplot2-geoms.html">basejump_geom_abline()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label_average()</a></code> <code><a href="ggplot2-geoms.html">basejump_geom_label_repel()</a></code> </p>
        </td>
        <td><p>ggplot2 geometric objects</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ggplot2-themes.html">theme_paperwhite()</a></code> <code><a href="ggplot2-themes.html">theme_midnight()</a></code> </p>
        </td>
        <td><p>ggplot2 themes</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-markdown" class="hasAnchor"><a href="#section-markdown" class="anchor"></a>Markdown</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="markdown.html">markdown(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Markdown</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownHeader.html">markdownHeader()</a></code> </p>
        </td>
        <td><p>Markdown header</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownLink.html">markdownLink()</a></code> </p>
        </td>
        <td><p>Markdown hyperlink</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownList.html">markdownList()</a></code> </p>
        </td>
        <td><p>Markdown list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownPlots.html">markdownPlots()</a></code> </p>
        </td>
        <td><p>Multiple Markdown plots</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="markdownTables.html">markdownTables()</a></code> </p>
        </td>
        <td><p>Multiple Markdown tables</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="prepareTemplate.html">prepareTemplate()</a></code> </p>
        </td>
        <td><p>Prepare R Markdown template</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-developer" class="hasAnchor"><a href="#section-developer" class="anchor"></a>Developer</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="formalsList.html">formalsList</a></code> </p>
        </td>
        <td><p>Shared list of optional default formals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matchInterestingGroups.html">matchInterestingGroups()</a></code> </p>
        </td>
        <td><p>Match interesting groups</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="multiassignAsEnvir.html">multiassignAsEnvir()</a></code> </p>
        </td>
        <td><p>Assign multiple objects into a new environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="showSlotInfo.html">showSlotInfo()</a></code> </p>
        </td>
        <td><p>Show slot information</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-miscellaneous" class="hasAnchor"><a href="#section-miscellaneous" class="anchor"></a>Miscellaneous</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="detectHPC.html">detectHPC()</a></code> </p>
        </td>
        <td><p>Detect HPC environment</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectLanes.html">detectLanes()</a></code> </p>
        </td>
        <td><p>Detect sequencing lanes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="detectOrganism.html">detectOrganism()</a></code> </p>
        </td>
        <td><p>Detect organism</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneNames.html">geneNames(<i>&lt;GRanges&gt;</i>)</a></code> <code><a href="geneNames.html">geneNames(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Gene names</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="headtail.html">headtail(<i>&lt;atomic&gt;</i>)</a></code> <code><a href="headtail.html">headtail(<i>&lt;matrix&gt;</i>)</a></code> <code><a href="headtail.html">headtail(<i>&lt;sparseMatrix&gt;</i>)</a></code> <code><a href="headtail.html">headtail(<i>&lt;data.frame&gt;</i>)</a></code> <code><a href="headtail.html">headtail(<i>&lt;DataFrame&gt;</i>)</a></code> <code><a href="headtail.html">headtail(<i>&lt;GRanges&gt;</i>)</a></code> <code><a href="headtail.html">headtail(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Return the first and last parts of an object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="humanize.html">humanize(<i>&lt;SummarizedExperiment&gt;</i>)</a></code> </p>
        </td>
        <td><p>Humanize an R object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matchesGene2Symbol.html">matchesGene2Symbol()</a></code> </p>
        </td>
        <td><p>Check that user-defined gene input matches expected values</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="matchesInterestingGroups.html">matchesInterestingGroups()</a></code> </p>
        </td>
        <td><p>Check that interesting groups match a defined value</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="microplate.html">microplate()</a></code> </p>
        </td>
        <td><p>Microtiter plate well identifiers</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="separator.html">separator()</a></code> </p>
        </td>
        <td><p>Separator bar</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-global-variables" class="hasAnchor"><a href="#section-global-variables" class="anchor"></a>Global variables</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="barcodePattern.html">barcodePattern</a></code> </p>
        </td>
        <td><p>Single-sell barcode pattern</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="genomeMetadataNames.html">genomeMetadataNames</a></code> </p>
        </td>
        <td><p>Slot names in metadata containing genome information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lanePattern.html">lanePattern</a></code> </p>
        </td>
        <td><p>Sequencing lane grep pattern</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="metadataBlacklist.html">metadataBlacklist</a></code> </p>
        </td>
        <td><p>Sample metadata blacklist</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-example-data" class="hasAnchor"><a href="#section-example-data" class="anchor"></a>Example data</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="df.html">df</a></code> </p>
        </td>
        <td><p>Example <code>DataFrame</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gender_markers.html">gender_markers</a></code> </p>
        </td>
        <td><p>Sexually dimorphic gender marker genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="mat.html">mat</a></code> </p>
        </td>
        <td><p>Example <code>matrix</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="organism_mappings.html">organism_mappings</a></code> </p>
        </td>
        <td><p>Organism mappings</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rse.html">rse</a></code> </p>
        </td>
        <td><p>Example <code>SummarizedExperiment</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sce.html">sce</a></code> </p>
        </td>
        <td><p>Example <code>SingleCellExperiment</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sparse.html">sparse</a></code> </p>
        </td>
        <td><p>Example <code>sparseMatrix</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tbl.html">tbl</a></code> </p>
        </td>
        <td><p>Example <code>tibble</code> (<code>tbl_df</code>)</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tx_se.html">tx_se</a></code> </p>
        </td>
        <td><p>Example transcript-level <code>SummarizedExperiment</code></p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-s-classes">S4 classes</a></li>
      <li><a href="#section-genome-annotations">Genome annotations</a></li>
      <li><a href="#section-gene-identifier-mapping">Gene identifier mapping</a></li>
      <li><a href="#section-counts">Counts</a></li>
      <li><a href="#section-metadata">Metadata</a></li>
      <li><a href="#section-sample-level-metadata">Sample-level metadata</a></li>
      <li><a href="#section-cell-level-metadata">Cell-level metadata</a></li>
      <li><a href="#section-sanitization">Sanitization</a></li>
      <li><a href="#section-aggregation">Aggregation</a></li>
      <li><a href="#section-math">Math</a></li>
      <li><a href="#section-plots">Plots</a></li>
      <li><a href="#section-ggplot-">ggplot2</a></li>
      <li><a href="#section-markdown">Markdown</a></li>
      <li><a href="#section-developer">Developer</a></li>
      <li><a href="#section-miscellaneous">Miscellaneous</a></li>
      <li><a href="#section-global-variables">Global variables</a></li>
      <li><a href="#section-example-data">Example data</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

