<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make a Gene2Symbol object — makeGene2Symbol • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Make a Gene2Symbol object — makeGene2Symbol" />

<meta property="og:description" content="Make a Gene2Symbol object" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.11.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make a Gene2Symbol object</h1>
    <small class="dont-index">Source: <a href='https://github.com/acidgenomics/basejump/blob/master/R/makeGene2Symbol.R'><code>R/makeGene2Symbol.R</code></a></small>
    <div class="hidden name"><code>makeGene2Symbol.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Make a Gene2Symbol object</p>
    
    </div>

    <pre class="usage"><span class='fu'>makeGene2SymbolFromEnsembl</span>(<span class='no'>organism</span>, <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"makeUnique"</span>, <span class='st'>"unmodified"</span>, <span class='st'>"1:1"</span>))

<span class='fu'>makeGene2SymbolFromEnsDb</span>(<span class='no'>object</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"makeUnique"</span>, <span class='st'>"unmodified"</span>,
  <span class='st'>"1:1"</span>))

<span class='fu'>makeGene2SymbolFromGFF</span>(<span class='no'>file</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"makeUnique"</span>, <span class='st'>"unmodified"</span>,
  <span class='st'>"1:1"</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>organism</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Full Latin organism name (e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Ensembl genome build assembly name (e.g. <code>"GRCh38"</code>). If set <code>NULL</code>,
defaults to the most recent build available.
Note: don't pass in UCSC build IDs (e.g. <code>"hg38"</code>).</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p><code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code>.
Ensembl release version (e.g. <code>90</code>). We recommend setting this value if
possible, for improved reproducibility. When left unset, the latest release
available via AnnotationHub/ensembldb is used. Note that the latest version
available can vary, depending on the versions of AnnotationHub and
ensembldb in use.</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Formatting method to apply:</p><ul>
<li><p><code>"makeUnique"</code>: <em>Recommended.</em> Apply <code><a href='https://rdrr.io/r/base/make.unique.html'>make.unique()</a></code>
to the <code>geneName</code> column. Gene symbols are made unique, while the gene
IDs remain unmodified.</p></li>
<li><p><code>"unmodified"</code>: Return <code>geneID</code> and <code>geneName</code> columns unmodified, in
long format.</p></li>
<li><p><code>"1:1"</code>: For gene symbols that map to multiple gene IDs, select only the
first annotated gene ID.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
File path.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>Gene2Symbol</code>.</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Updated 2019-07-28.</p>
    
    <h2 class="hasAnchor" id="gff-gtf-file"><a class="anchor" href="#gff-gtf-file"></a>GFF/GTF file</h2>

    


<p>Remote URLs and compressed files are supported.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## makeGene2SymbolFromEnsembl ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGene2SymbolFromEnsembl</span>(<span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Homo sapiens"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>Matching EnsDb from AnnotationHub 2.16.0 (2019-05-02).</span></div><div class='output co'>#&gt; <span class='message'>AH73986: Ensembl 79 EnsDb for Homo sapiens</span>
#&gt; <span class='message'>Run this code to download EnsDb manually:</span>
#&gt; <span class='message'>  &gt; library(AnnotationHub)</span>
#&gt; <span class='message'>  &gt; ah &lt;- AnnotationHub()</span>
#&gt; <span class='message'>  &gt; edb &lt;- ah[["AH73986"]]</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Homo sapiens</span>
#&gt; <span class='message'>  - Genome Build: GRCh38</span>
#&gt; <span class='message'>  - Ensembl Release: 79</span>
#&gt; <span class='message'>  - Level: genes</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneBiotype, geneName, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>65774 genes detected.</span></div><div class='output co'>#&gt; <span class='message'>2403 non-unique gene symbol(s) detected.</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Gene2Symbol with 65774 rows and 2 columns
#&gt;                          geneID    geneName
#&gt;                     &lt;character&gt; &lt;character&gt;
#&gt; ENSG00000000003 ENSG00000000003      TSPAN6
#&gt; ENSG00000000005 ENSG00000000005        TNMD
#&gt; ENSG00000000419 ENSG00000000419        DPM1
#&gt; ENSG00000000457 ENSG00000000457       SCYL3
#&gt; ENSG00000000460 ENSG00000000460    C1orf112
#&gt; ...                         ...         ...
#&gt; LRG_94                   LRG_94      PRF1.1
#&gt; LRG_96                   LRG_96    RAB27A.1
#&gt; LRG_97                   LRG_97      RAC2.1
#&gt; LRG_98                   LRG_98      RAG1.1
#&gt; LRG_99                   LRG_99      RAG2.1</div><div class='input'>
<span class='co'>## makeTx2GeneFromEnsDb ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGene2SymbolFromEnsDb</span>(<span class='st'>"EnsDb.Hsapiens.v75"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>Loading required namespace: EnsDb.Hsapiens.v75</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Homo sapiens</span>
#&gt; <span class='message'>  - Genome Build: GRCh37</span>
#&gt; <span class='message'>  - Ensembl Release: 75</span>
#&gt; <span class='message'>  - Level: genes</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneBiotype, geneName, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>64102 genes detected.</span></div><div class='output co'>#&gt; <span class='message'>3075 non-unique gene symbol(s) detected.</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Gene2Symbol with 64102 rows and 2 columns
#&gt;                          geneID    geneName
#&gt;                     &lt;character&gt; &lt;character&gt;
#&gt; ENSG00000000003 ENSG00000000003      TSPAN6
#&gt; ENSG00000000005 ENSG00000000005        TNMD
#&gt; ENSG00000000419 ENSG00000000419        DPM1
#&gt; ENSG00000000457 ENSG00000000457       SCYL3
#&gt; ENSG00000000460 ENSG00000000460    C1orf112
#&gt; ...                         ...         ...
#&gt; LRG_94                   LRG_94      LRG_94
#&gt; LRG_96                   LRG_96      LRG_96
#&gt; LRG_97                   LRG_97      LRG_97
#&gt; LRG_98                   LRG_98      LRG_98
#&gt; LRG_99                   LRG_99      LRG_99</div><div class='input'>
<span class='co'>## makeGene2SymbolFromGFF ====</span>
<span class='co'>## GTF</span>
<span class='no'>file</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span>(<span class='no'>basejumpTestsURL</span>, <span class='st'>"example.gtf"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGene2SymbolFromGFF</span>(<span class='no'>file</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from GFF file.</span></div><div class='output co'>#&gt; <span class='message'>Importing example.gtf using rtracklayer::import().</span></div><div class='output co'>#&gt; <span class='message'>Ensembl GTF detected.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneBiotype, geneName, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>17 genes detected.</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Gene2Symbol with 17 rows and 2 columns
#&gt;                                geneID    geneName
#&gt;                           &lt;character&gt; &lt;character&gt;
#&gt; ENSMUSG00000025900 ENSMUSG00000025900         Rp1
#&gt; ENSMUSG00000051951 ENSMUSG00000051951        Xkr4
#&gt; ENSMUSG00000064842 ENSMUSG00000064842     Gm26206
#&gt; ENSMUSG00000088333 ENSMUSG00000088333     Gm27396
#&gt; ENSMUSG00000089699 ENSMUSG00000089699      Gm1992
#&gt; ...                               ...         ...
#&gt; ENSMUSG00000103147 ENSMUSG00000103147      Gm7341
#&gt; ENSMUSG00000103161 ENSMUSG00000103161     Gm38148
#&gt; ENSMUSG00000103201 ENSMUSG00000103201     Gm37329
#&gt; ENSMUSG00000103377 ENSMUSG00000103377     Gm37180
#&gt; ENSMUSG00000104017 ENSMUSG00000104017     Gm37363</div><div class='input'>
<span class='co'>## GFF3</span>
<span class='no'>file</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span>(<span class='no'>basejumpTestsURL</span>, <span class='st'>"example.gff3"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGene2SymbolFromGFF</span>(<span class='no'>file</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from GFF file.</span></div><div class='output co'>#&gt; <span class='message'>Importing example.gff3 using rtracklayer::import().</span></div><div class='output co'>#&gt; <span class='message'>Ensembl GFF3 detected.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneBiotype, geneName, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>20 genes detected.</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Gene2Symbol with 20 rows and 2 columns
#&gt;                                geneID    geneName
#&gt;                           &lt;character&gt; &lt;character&gt;
#&gt; ENSMUSG00000025900 ENSMUSG00000025900         Rp1
#&gt; ENSMUSG00000025902 ENSMUSG00000025902       Sox17
#&gt; ENSMUSG00000051951 ENSMUSG00000051951        Xkr4
#&gt; ENSMUSG00000064842 ENSMUSG00000064842     Gm26206
#&gt; ENSMUSG00000088333 ENSMUSG00000088333     Gm27396
#&gt; ...                               ...         ...
#&gt; ENSMUSG00000103161 ENSMUSG00000103161     Gm38148
#&gt; ENSMUSG00000103201 ENSMUSG00000103201     Gm37329
#&gt; ENSMUSG00000103377 ENSMUSG00000103377     Gm37180
#&gt; ENSMUSG00000104017 ENSMUSG00000104017     Gm37363
#&gt; ENSMUSG00000104123 ENSMUSG00000104123     Gm37483</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#gff-gtf-file">GFF/GTF file</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.9100.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


