<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make Gene-to-Symbol Mappings — makeGene2Symbol • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Make Gene-to-Symbol Mappings — makeGene2Symbol" />

<meta property="og:description" content="Make Gene-to-Symbol Mappings" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.8.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make Gene-to-Symbol Mappings</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump//blob/master/R/makeGene2Symbol-methods.R'><code>R/makeGene2Symbol-methods.R</code></a></small>
    <div class="hidden name"><code>makeGene2Symbol.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Make Gene-to-Symbol Mappings</p>
    
    </div>

    <pre class="usage"><span class='fu'>makeGene2SymbolFromEnsembl</span>(<span class='no'>organism</span>, <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='fu'>makeGene2SymbolFromEnsDb</span>(<span class='no'>object</span>)

<span class='fu'>makeGene2SymbolFromGFF</span>(<span class='no'>file</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>organism</th>
      <td><p><code>string</code>. Full Latin organism name (e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code>string</code> or <code>NULL</code>. Ensembl genome build assembly name
(e.g. <code>"GRCh38"</code>). If set <code>NULL</code>, defaults to the most recent build
available. Note: don't pass in UCSC build IDs (e.g. <code>"hg38"</code>).</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p><code>scalar integer</code> or <code>NULL</code>. Ensembl release version (e.g.
<code>90</code>). If set <code>NULL</code>, defaults to the most recent release available.</p></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p><code>string</code>. File path.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>Gene2Symbol</code>.</p>
    
    <h2 class="hasAnchor" id="gff-gtf-file"><a class="anchor" href="#gff-gtf-file"></a>GFF/GTF file</h2>

    
    <p>Remote URLs and compressed files are supported.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><a href='makeGRanges.html'>makeGRanges</a>.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## makeGene2SymbolFromEnsembl ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGene2SymbolFromEnsembl</span>(<span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Homo sapiens"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>Matching EnsDb from AnnotationHub 2.14.1 (2018-10-24).</span></div><div class='output co'>#&gt; <span class='message'>AH64923: Ensembl 94 EnsDb for Homo sapiens</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Homo sapiens</span>
#&gt; <span class='message'>  - Genome Build: GRCh38</span>
#&gt; <span class='message'>  - Ensembl Release: 94</span>
#&gt; <span class='message'>  - Level: genes</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, geneBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>3059 non-unique gene symbol(s) detected.</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Gene2Symbol with 65687 rows and 2 columns
#&gt;                          geneID    geneName
#&gt;                     &lt;character&gt; &lt;character&gt;
#&gt; ENSG00000000003 ENSG00000000003      TSPAN6
#&gt; ENSG00000000005 ENSG00000000005        TNMD
#&gt; ENSG00000000419 ENSG00000000419        DPM1
#&gt; ENSG00000000457 ENSG00000000457       SCYL3
#&gt; ENSG00000000460 ENSG00000000460    C1orf112
#&gt; ...                         ...         ...
#&gt; LRG_995                 LRG_995     FUBP1.1
#&gt; LRG_996                 LRG_996     ERBB3.1
#&gt; LRG_997                 LRG_997      ROS1.1
#&gt; LRG_998                 LRG_998     CCND3.1
#&gt; LRG_999                 LRG_999       CIC.1</div><div class='input'>
<span class='co'>## makeTx2GeneFromEnsDb ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGene2SymbolFromEnsDb</span>(<span class='st'>"EnsDb.Hsapiens.v75"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Homo sapiens</span>
#&gt; <span class='message'>  - Genome Build: GRCh37</span>
#&gt; <span class='message'>  - Ensembl Release: 75</span>
#&gt; <span class='message'>  - Level: genes</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, geneBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>3075 non-unique gene symbol(s) detected.</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Gene2Symbol with 64102 rows and 2 columns
#&gt;                          geneID    geneName
#&gt;                     &lt;character&gt; &lt;character&gt;
#&gt; ENSG00000000003 ENSG00000000003      TSPAN6
#&gt; ENSG00000000005 ENSG00000000005        TNMD
#&gt; ENSG00000000419 ENSG00000000419        DPM1
#&gt; ENSG00000000457 ENSG00000000457       SCYL3
#&gt; ENSG00000000460 ENSG00000000460    C1orf112
#&gt; ...                         ...         ...
#&gt; LRG_94                   LRG_94      LRG_94
#&gt; LRG_96                   LRG_96      LRG_96
#&gt; LRG_97                   LRG_97      LRG_97
#&gt; LRG_98                   LRG_98      LRG_98
#&gt; LRG_99                   LRG_99      LRG_99</div><div class='input'>
<span class='co'>## makeGene2SymbolFromGFF ====</span>
<span class='co'>## GTF</span>
<span class='no'>file</span> <span class='kw'>&lt;-</span> <span class='fu'>file.path</span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"example.gtf"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGene2SymbolFromGFF</span>(<span class='no'>file</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from GFF/GTF file.</span></div><div class='output co'>#&gt; <span class='message'>Importing example.gtf using rtracklayer::import().</span></div><div class='output co'>#&gt; <span class='message'>Ensembl GTF detected.</span></div><div class='output co'>#&gt; <span class='message'>17 gene annotations detected.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, geneBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Gene2Symbol with 17 rows and 2 columns
#&gt;                                geneID    geneName
#&gt;                           &lt;character&gt; &lt;character&gt;
#&gt; ENSMUSG00000025900 ENSMUSG00000025900         Rp1
#&gt; ENSMUSG00000051951 ENSMUSG00000051951        Xkr4
#&gt; ENSMUSG00000064842 ENSMUSG00000064842     Gm26206
#&gt; ENSMUSG00000088333 ENSMUSG00000088333     Gm27396
#&gt; ENSMUSG00000089699 ENSMUSG00000089699      Gm1992
#&gt; ...                               ...         ...
#&gt; ENSMUSG00000103147 ENSMUSG00000103147      Gm7341
#&gt; ENSMUSG00000103161 ENSMUSG00000103161     Gm38148
#&gt; ENSMUSG00000103201 ENSMUSG00000103201     Gm37329
#&gt; ENSMUSG00000103377 ENSMUSG00000103377     Gm37180
#&gt; ENSMUSG00000104017 ENSMUSG00000104017     Gm37363</div><div class='input'>
<span class='co'>## GFF3</span>
<span class='no'>file</span> <span class='kw'>&lt;-</span> <span class='fu'>file.path</span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"example.gff3"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGene2SymbolFromGFF</span>(<span class='no'>file</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from GFF/GTF file.</span></div><div class='output co'>#&gt; <span class='message'>Importing example.gff3 using rtracklayer::import().</span></div><div class='output co'>#&gt; <span class='message'>Ensembl GFF detected.</span></div><div class='output co'>#&gt; <span class='message'>20 gene annotations detected.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, geneBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Gene2Symbol with 20 rows and 2 columns
#&gt;                                geneID    geneName
#&gt;                           &lt;character&gt; &lt;character&gt;
#&gt; ENSMUSG00000025900 ENSMUSG00000025900         Rp1
#&gt; ENSMUSG00000025902 ENSMUSG00000025902       Sox17
#&gt; ENSMUSG00000051951 ENSMUSG00000051951        Xkr4
#&gt; ENSMUSG00000064842 ENSMUSG00000064842     Gm26206
#&gt; ENSMUSG00000088333 ENSMUSG00000088333     Gm27396
#&gt; ...                               ...         ...
#&gt; ENSMUSG00000103161 ENSMUSG00000103161     Gm38148
#&gt; ENSMUSG00000103201 ENSMUSG00000103201     Gm37329
#&gt; ENSMUSG00000103377 ENSMUSG00000103377     Gm37180
#&gt; ENSMUSG00000104017 ENSMUSG00000104017     Gm37363
#&gt; ENSMUSG00000104123 ENSMUSG00000104123     Gm37483</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#gff-gtf-file">GFF/GTF file</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

