<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make Summarized Experiment — makeSummarizedExperiment • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Make Summarized Experiment — makeSummarizedExperiment" />

<meta property="og:description" content="This function is a utility wrapper for
SummarizedExperiment::SummarizedExperiment() that provides automatic
subsetting for row and column data, as well as automatic handling of
transgenes and spike-ins. Additionally, it improves upon the standard
constructor by slotting useful session information into the
S4Vectors::metadata() slot:" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make Summarized Experiment</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/makeSummarizedExperiment.R'><code>R/makeSummarizedExperiment.R</code></a></small>
    <div class="hidden name"><code>makeSummarizedExperiment.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function is a utility wrapper for
<code><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class'>SummarizedExperiment::SummarizedExperiment()</a></code> that provides automatic
subsetting for row and column data, as well as automatic handling of
transgenes and spike-ins. Additionally, it improves upon the standard
constructor by slotting useful session information into the
<code><a href='http://www.rdocumentation.org/packages/S4Vectors/topics/Annotated-class'>S4Vectors::metadata()</a></code> slot:</p>
    
    </div>

    <pre class="usage"><span class='fu'>makeSummarizedExperiment</span>(<span class='no'>assays</span>, <span class='kw'>rowRanges</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>rowData</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>colData</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>spikeNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>assays</th>
      <td><p><code>list</code>. RNA-seq count matrices, which must have matching
dimensions. Counts can be passed in either a dense matrix (<code>matrix</code>) or
sparse matrix (<code>sparseMatrix</code>).</p></td>
    </tr>
    <tr>
      <th>rowRanges</th>
      <td><p><code>GRanges</code> or <code>NULL</code>. <em>Recommended</em>. Metadata describing the
assay rows. If defined, must contain genomic ranges. Can be left <code>NULL</code> if
the genome is poorly annotated and/or ranges aren't available from
AnnotationHub.</p></td>
    </tr>
    <tr>
      <th>rowData</th>
      <td><p><code>DataFrame</code> or <code>NULL</code>. <em>Not recommended</em>. Metadata describing
the assay rows, if genomic ranges are not available.</p></td>
    </tr>
    <tr>
      <th>colData</th>
      <td><p><code>DataFrame</code> <code>data.frame</code>, or <code>matrix</code>. Metadata
describing the assay columns. For bulk RNA-seq, this data describes the
samples. For single-cell RNA-seq, this data describes the cells.</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p><code>list</code> or <code>NULL</code>. Metadata.</p></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code> or <code>NULL</code>. Vector indicating which
<code>assay()</code> rows denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>spikeNames</th>
      <td><p><code>character</code> or <code>NULL</code>. Vector indicating which <code>assay()</code>
rows denote spike-in sequences (e.g. ERCCs).</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    
<ul>
<li><p>Providing <code>rowRanges</code>: <code>RangedSummarizedExperiment</code>.</p></li>
<li><p>Providing <code>rowData</code>: <code>SummarizedExperiment</code>.</p></li>
</ul>

    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <ul>
<li><p><code>date</code>: Today's date, returned from <code><a href='http://www.rdocumentation.org/packages/base/topics/Sys.time'>base::Sys.Date()</a></code>.</p></li>
<li><p><code>wd</code>: Working directory, returned from <code><a href='http://www.rdocumentation.org/packages/base/topics/getwd'>base::getwd()</a></code>.</p></li>
<li><p><code>sessionInfo</code>: <code><a href='http://www.rdocumentation.org/packages/sessioninfo/topics/session_info'>sessioninfo::session_info()</a></code> return.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Column and rows always return sorted alphabetically.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class'>SummarizedExperiment::SummarizedExperiment()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/SingleCellExperiment/topics/SingleCellExperiment'>SingleCellExperiment::SingleCellExperiment()</a></code>.</p></li>
<li><p><code>help("RangedSummarizedExperiment-class", "SummarizedExperiment")</code>.</p></li>
<li><p><code>help("SummarizedExperiment-class", "SummarizedExperiment")</code>.</p></li>
<li><p><code>help("SingleCellExperiment-class", "SingleCellExperiment")</code>.</p></li>
</ul></div>
    <div class='dont-index'><p>Other SummarizedExperiment Functions: <code><a href='counts.html'>counts</a></code>,
  <code><a href='interestingGroups.html'>interestingGroups</a></code>, <code><a href='mapGenes.html'>mapGenes</a></code>,
  <code><a href='meltCounts.html'>meltCounts</a></code>, <code><a href='metrics.html'>metrics</a></code>,
  <code><a href='sampleData.html'>sampleData</a></code>, <code><a href='sampleNames.html'>sampleNames</a></code>,
  <code><a href='selectSamples.html'>selectSamples</a></code>, <code><a href='tpm.html'>tpm</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Rows (genes)</span>
<span class='no'>genes</span> <span class='kw'>&lt;-</span> <span class='fu'>c</span>(
    <span class='st'>"EGFP"</span>,  <span class='co'># transgene (see below)</span>
    <span class='fu'>paste0</span>(<span class='st'>"gene"</span>, <span class='fu'>seq_len</span>(<span class='fl'>3L</span>))
)
<span class='fu'>print</span>(<span class='no'>genes</span>)</div><div class='output co'>#&gt; [1] "EGFP"  "gene1" "gene2" "gene3"</div><div class='input'>
<span class='co'>## Columns (samples)</span>
<span class='no'>samples</span> <span class='kw'>&lt;-</span> <span class='fu'>paste0</span>(<span class='st'>"sample"</span>, <span class='fu'>seq_len</span>(<span class='fl'>4L</span>))
<span class='fu'>print</span>(<span class='no'>samples</span>)</div><div class='output co'>#&gt; [1] "sample1" "sample2" "sample3" "sample4"</div><div class='input'>
<span class='co'>## Counts (assay)</span>
<span class='no'>counts</span> <span class='kw'>&lt;-</span> <span class='fu'>matrix</span>(
    <span class='kw'>data</span> <span class='kw'>=</span> <span class='fu'>seq_len</span>(<span class='fl'>16L</span>),
    <span class='kw'>nrow</span> <span class='kw'>=</span> <span class='fl'>4L</span>,
    <span class='kw'>ncol</span> <span class='kw'>=</span> <span class='fl'>4L</span>,
    <span class='kw'>dimnames</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='no'>genes</span>, <span class='no'>samples</span>)
)
<span class='co'>## Primary assay must be named "counts".</span>
<span class='no'>assays</span> <span class='kw'>&lt;-</span> <span class='fu'>list</span>(<span class='kw'>counts</span> <span class='kw'>=</span> <span class='no'>counts</span>)
<span class='fu'>print</span>(<span class='no'>assays</span>)</div><div class='output co'>#&gt; $counts
#&gt;       sample1 sample2 sample3 sample4
#&gt; EGFP        1       5       9      13
#&gt; gene1       2       6      10      14
#&gt; gene2       3       7      11      15
#&gt; gene3       4       8      12      16
#&gt; </div><div class='input'>
<span class='co'>## Row data (genomic ranges)</span>
<span class='no'>rowRanges</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>GenomicRanges</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/GenomicRanges/topics/GRanges-class'>GRanges</a></span>(
    <span class='kw'>seqnames</span> <span class='kw'>=</span> <span class='fu'>rep</span>(<span class='st'>"1"</span>, <span class='kw'>times</span> <span class='kw'>=</span> <span class='fl'>3L</span>),
    <span class='kw'>ranges</span> <span class='kw'>=</span> <span class='kw pkg'>IRanges</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/IRanges/topics/IRanges-constructor'>IRanges</a></span>(
        <span class='kw'>start</span> <span class='kw'>=</span> <span class='fu'>seq</span>(<span class='kw'>from</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='kw'>to</span> <span class='kw'>=</span> <span class='fl'>201L</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='fl'>100L</span>),
        <span class='kw'>end</span> <span class='kw'>=</span> <span class='fu'>seq</span>(<span class='kw'>from</span> <span class='kw'>=</span> <span class='fl'>100L</span>, <span class='kw'>to</span> <span class='kw'>=</span> <span class='fl'>300L</span>, <span class='kw'>by</span> <span class='kw'>=</span> <span class='fl'>100L</span>)
    )
)
<span class='co'>## Note that we haven't defined the transgene here.</span>
<span class='co'>## It will be handled automatically in the function call.</span>
<span class='fu'>names</span>(<span class='no'>rowRanges</span>) <span class='kw'>&lt;-</span> <span class='fu'>paste0</span>(<span class='st'>"gene"</span>, <span class='fu'>seq_len</span>(<span class='fl'>3L</span>))
<span class='fu'>print</span>(<span class='no'>rowRanges</span>)</div><div class='output co'>#&gt; GRanges object with 3 ranges and 0 metadata columns:
#&gt;         seqnames    ranges strand
#&gt;            &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
#&gt;   gene1        1     1-100      *
#&gt;   gene2        1   101-200      *
#&gt;   gene3        1   201-300      *
#&gt;   -------
#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</div><div class='input'>
<span class='co'>## Column data</span>
<span class='no'>colData</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>S4Vectors</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/S4Vectors/topics/DataFrame-class'>DataFrame</a></span>(
    <span class='kw'>genotype</span> <span class='kw'>=</span> <span class='fu'>rep</span>(<span class='fu'>c</span>(<span class='st'>"wildtype"</span>, <span class='st'>"knockout"</span>), <span class='kw'>times</span> <span class='kw'>=</span> <span class='fl'>1L</span>, <span class='kw'>each</span> <span class='kw'>=</span> <span class='fl'>2L</span>),
    <span class='kw'>age</span> <span class='kw'>=</span> <span class='fu'>rep</span>(<span class='fu'>c</span>(<span class='fl'>3L</span>, <span class='fl'>6L</span>), <span class='fl'>2L</span>),
    <span class='kw'>row.names</span> <span class='kw'>=</span> <span class='no'>samples</span>
)
<span class='fu'>print</span>(<span class='no'>colData</span>)</div><div class='output co'>#&gt; DataFrame with 4 rows and 2 columns
#&gt;            genotype       age
#&gt;         &lt;character&gt; &lt;integer&gt;
#&gt; sample1    wildtype         3
#&gt; sample2    wildtype         6
#&gt; sample3    knockout         3
#&gt; sample4    knockout         6</div><div class='input'>
<span class='co'>## Note that this returns with the dimnames sorted.</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeSummarizedExperiment</span>(
    <span class='kw'>assays</span> <span class='kw'>=</span> <span class='no'>assays</span>,
    <span class='kw'>rowRanges</span> <span class='kw'>=</span> <span class='no'>rowRanges</span>,
    <span class='kw'>colData</span> <span class='kw'>=</span> <span class='no'>colData</span>,
    <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='st'>"EGFP"</span>
)</div><div class='output co'>#&gt; <span class='warning'>Warning: 'DESCRIPTION' file has an 'Encoding' field and re-encoding is not possible</span></div><div class='input'><span class='fu'>print</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: RangedSummarizedExperiment 
#&gt; dim: 4 4 
#&gt; metadata(3): date wd sessionInfo
#&gt; assays(1): counts
#&gt; rownames(4): EGFP gene1 gene2 gene3
#&gt; rowData names(0):
#&gt; colnames(4): sample1 sample2 sample3 sample4
#&gt; colData names(2): genotype age</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

