<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make <code>GRanges</code> from a GFF/GTF file — makeGRangesFromGFF • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Make <code>GRanges</code> from a GFF/GTF file — makeGRangesFromGFF" />

<meta property="og:description" content="Make GRanges from a GFF/GTF file" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.16</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make <code>GRanges</code> from a GFF/GTF file</h1>
    <small class="dont-index">Source: <a href='https://github.com/acidgenomics/basejump/blob/master/R/makeGRangesFromGFF.R'><code>R/makeGRangesFromGFF.R</code></a></small>
    <div class="hidden name"><code>makeGRangesFromGFF.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Make <code>GRanges</code> from a GFF/GTF file</p>
    
    </div>

    <pre class="usage"><span class='fu'>makeGRangesFromGFF</span>(<span class='no'>file</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>.checkAgainstTxDb</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='fu'>makeGRangesFromGTF</span>(<span class='no'>file</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>.checkAgainstTxDb</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>file</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
File path.</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Return as genes or transcripts.</p></td>
    </tr>
    <tr>
      <th>.checkAgainstTxDb</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Enable strict mode, intended for development and unit testing only.
Generate an internal <code>TxDb</code> using <code><a href='https://www.rdocumentation.org/packages/GenomicFeatures/topics/makeTxDbFromGRanges'>GenomicFeatures::makeTxDbFromGRanges()</a></code>
and check that the <code><a href='https://www.rdocumentation.org/packages/IRanges/topics/range-squeezers'>ranges()</a></code>,
<code><a href='https://www.rdocumentation.org/packages/GenomeInfoDb/topics/seqinfo'>seqnames()</a></code>, and identifiers defined in
<code><a href='https://www.rdocumentation.org/packages/base/topics/names'>names()</a></code> are identical. Doesn't work for all GFF/GTF files
due to some current limitations in the GenomicFeatures package, so this is
disabled by default. Generally, GenomicFeatures parses GTF files better
than GFF files. However, it's a useful sanity check and should be enabled
if possible.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Remote URLs and compressed files are supported.</p>
    
    <h2 class="hasAnchor" id="functions"><a class="anchor" href="#functions"></a>Functions</h2>

    <ul>
<li><p><code>makeGRangesFromGTF</code>: GTF file extension alias.
Runs the same internal code as <code>makeGRangesFromGFF()</code>.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="recommendations"><a class="anchor" href="#recommendations"></a>Recommendations</h2>

    
    <ul>
<li><p><strong>Use GTF over GFF3.</strong> We recommend using a GTF file instead of a GFF3
file, when possible. The file format is more compact and easier to parse.</p></li>
<li><p><strong>Use Ensembl over RefSeq.</strong> We generally recommend using Ensembl over
RefSeq, if possible. It's better supported in R and generally used by most
NGS vendors.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="gff-gtf-specification"><a class="anchor" href="#gff-gtf-specification"></a>GFF/GTF specification</h2>

    
    <p>The GFF (General Feature Format) format consists of one line per feature,
each containing 9 columns of data, plus optional track definition lines.</p>
<p>The GTF (General Transfer Format) format is identical to GFF version 2.</p>
<p>The UCSC website has detailed conventions on the GFF3 format, including
the metadata columns.</p>
    
    <h2 class="hasAnchor" id="supported-sources"><a class="anchor" href="#supported-sources"></a>Supported sources</h2>

    
    <p>Currently <code>makeGRangesFromGFF()</code> supports genomes from these sources:</p>
<table class='table'>
<tr><td></td><td>GTF</td><td>GFF3</td><td></td></tr>
<tr><td></td><td>Ensembl</td><td>TRUE</td><td>TRUE</td></tr>
<tr><td>Also available via ensembldb.</td><td></td><td>GENCODE</td><td>TRUE</td></tr>
<tr><td>TRUE</td><td>Similar to Ensembl; duplicate PAR_Y.</td><td></td><td>RefSeq</td></tr>
<tr><td>NA</td><td>TRUE</td><td>Experimental support.</td><td></td></tr>
<tr><td>UCSC</td><td>FALSE</td><td>FALSE</td><td>Use pre-built Bioconductor packages.</td></tr>
<tr><td></td><td>FlyBase</td><td>TRUE</td><td>FALSE</td></tr>
<tr><td>Newer than Ensembl.</td><td></td><td>WormBase</td><td>TRUE</td></tr>
<tr><td>FALSE</td><td>Very similar to Ensembl.</td><td></td><td></td></tr>
</table>
    
    <h2 class="hasAnchor" id="ensembl"><a class="anchor" href="#ensembl"></a>Ensembl</h2>

    
    <p>Note that <code><a href='makeGRangesFromEnsembl.html'>makeGRangesFromEnsembl()</a></code> offers native support for Ensembl genome
builds and returns additional useful metadata that isn't defined inside a
GFF/GTF file.</p>
<p>If you must load a GFF/GTF file directly, then use <code>makeGRangesFromGFF()</code>.</p>
    
    <h2 class="hasAnchor" id="gencode-vs-ensembl"><a class="anchor" href="#gencode-vs-ensembl"></a>GENCODE vs. Ensembl</h2>

    
    <p>Annotations available from Ensembl and GENCODE are very similar.</p>
<p>The GENCODE annotation is made by merging the manual gene annotation produced
by the Ensembl-Havana team and the Ensembl-genebuild automated gene
annotation. The GENCODE annotation is the default gene annotation displayed
in the Ensembl browser. The GENCODE releases coincide with the Ensembl
releases, although GENCODE can skip an Ensembl release if there is no update
to the annotation with respect to the previous release. In practical terms,
the GENCODE annotation is essentially identical to the Ensembl annotation.</p>
<p>However, GENCODE handles pseudoautosomal regions (PAR) differently than
Ensembl. The Ensembl GTF file only includes this annotation once, for
chromosome X. However, GENCODE GTF/GFF3 files include the annotation in the
PAR regions of both chromosomes. You'll see these genes contain a "_PAR_Y"
suffix.</p>
<p>Additionally, GENCODE GFF/GTF files import with a gene identifier containing
a suffix, which differs slightly from the Ensembl GFF/GTF spec
(e.g. GENCODE: ENSG00000000003.14; Ensembl: ENSG00000000003).</p>
<p>The <a href='https://www.gencodegenes.org/pages/faq.html'>GENCODE FAQ</a> has additional
details.</p>
    
    <h2 class="hasAnchor" id="refseq"><a class="anchor" href="#refseq"></a>RefSeq</h2>

    
    <p>Refer to the
<a href='ftp://ftp.ncbi.nlm.nih.gov/genomes/README_GFF3.txt'>current RefSeq spec</a>
for details.</p>
<p>See also:</p><ul>
<li><p><a href='https://www.ncbi.nlm.nih.gov/books/NBK50679/'>RefSeq FAQ</a></p></li>
<li><p>ftp://ftp.ncbi.nih.gov/gene/DATA/gene2refseq.gz</p></li>
</ul>
    
    <h2 class="hasAnchor" id="ucsc"><a class="anchor" href="#ucsc"></a>UCSC</h2>

    
    <p>Loading UCSC genome annotations from a GFF/GTF file are
<em>intentionally not supported</em> by this function.</p>
<p>We recommend using a pre-built <code>TxDb</code> package from Bioconductor instead.
For example, load <code>TxDb.Hsapiens.UCSC.hg38.knownGene</code> for hg38.</p>
<p>For reference, note that UCSC doesn't provide direct GFF/GTF file downloads.
Use of the <a href='https://genome.ucsc.edu/cgi-bin/hgTables'>hgTables</a> table
browser is required in a web browser.</p>
<p>Select the following options to download hg38:</p><ul>
<li><p>clade: <code>Mammal</code></p></li>
<li><p>genome: <code>Human</code></p></li>
<li><p>assembly: <code>Dec. 2013 (GRCh38/hg38)</code></p></li>
<li><p>group: <code>Genes and Gene Predictions</code></p></li>
<li><p>track: <code>GENCODE v29</code></p></li>
<li><p>table: <code>knownGene</code></p></li>
<li><p>region: <code>genome</code></p></li>
<li><p>output format: <code>GTF - gene transfer format</code></p></li>
<li><p>output file: <code>&lt;Enter a file name&gt;</code></p></li>
</ul>
    <p>See also:</p><ul>
<li><p>http://genome.ucsc.edu/cgi-bin/hgTables</p></li>
<li><p>http://hgdownload.soe.ucsc.edu/downloads.html</p></li>
<li><p>ftp://hgdownload.soe.ucsc.edu/goldenPath/hg38/</p></li>
</ul>
    
    <h2 class="hasAnchor" id="example-urls"><a class="anchor" href="#example-urls"></a>Example URLs</h2>

    
    <ul>
<li><p>Ensembl GTF:<br />
ftp://ftp.ensembl.org/pub/release-95/gtf/homo_sapiens/Homo_sapiens.GRCh38.95.gtf.gz</p></li>
<li><p>Ensembl GFF3:<br />
ftp://ftp.ensembl.org/pub/release-95/gff3/homo_sapiens/Homo_sapiens.GRCh38.95.gff3.gz</p></li>
<li><p>GENCODE GTF:<br />
ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_29/gencode.v29.annotation.gtf.gz</p></li>
<li><p>GENCODE GFF3:<br />
ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_29/gencode.v29.annotation.gff3.gz</p></li>
<li><p>RefSeq GFF3:<br />
ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Homo_sapiens/reference/GCF_000001405.38_GRCh38.p12/GCF_000001405.38_GRCh38.p12_genomic.gff.gz</p></li>
<li><p>FlyBase GTF:<br />
ftp://ftp.flybase.net/releases/FB2018_05/dmel_r6.24/gtf/dmel-all-r6.24.gtf.gz</p></li>
<li><p>WormBase GTF:<br />
ftp://ftp.wormbase.org/pub/wormbase/releases/WS267/species/c_elegans/PRJNA13758/c_elegans.PRJNA13758.WS267.canonical_geneset.gtf.gz</p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code><a href='https://www.rdocumentation.org/packages/rtracklayer/topics/export'>rtracklayer::import()</a></code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/GenomicFeatures/topics/makeTxDbFromGRanges'>GenomicFeatures::makeTxDbFromGRanges()</a></code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/GenomicFeatures/topics/makeTxDbFromGFF'>GenomicFeatures::makeTxDbFromGFF()</a></code>.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>file</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/brio/topics/pasteURL'>pasteURL</a></span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"ensembl.gtf"</span>, <span class='kw'>protocol</span> <span class='kw'>=</span> <span class='st'>"none"</span>)

<span class='co'>## Genes</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGRangesFromGFF</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='no'>file</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"genes"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from GFF file.</span></div><div class='output co'>#&gt; <span class='message'>Importing ensembl.gtf using rtracklayer::import().</span></div><div class='output co'>#&gt; <span class='message'>ensembl.gtf does not return syntactically valid names.</span></div><div class='output co'>#&gt; <span class='message'>Ensembl GTF detected.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneBiotype, geneName, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>60 genes detected.</span></div><div class='input'><span class='fu'><a href='summary.html'>summary</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "GRanges object with 60 ranges and 8 metadata columns"</div><div class='input'>
<span class='co'>## Transcripts</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGRangesFromGFF</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='no'>file</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"transcripts"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from GFF file.</span></div><div class='output co'>#&gt; <span class='message'>Importing ensembl.gtf using rtracklayer::import().</span></div><div class='output co'>#&gt; <span class='message'>ensembl.gtf does not return syntactically valid names.</span></div><div class='output co'>#&gt; <span class='message'>Ensembl GTF detected.</span></div><div class='output co'>#&gt; <span class='message'>Merging gene-level annotations into transcript-level object.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, seqnames, transcriptBiotype</span></div><div class='output co'>#&gt; <span class='message'>Arranging by transcriptID.</span></div><div class='output co'>#&gt; <span class='message'>167 transcripts detected.</span></div><div class='input'><span class='fu'><a href='summary.html'>summary</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "GRanges object with 167 ranges and 16 metadata columns"</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#details">Details</a></li>

      <li><a href="#functions">Functions</a></li>

      <li><a href="#recommendations">Recommendations</a></li>

      <li><a href="#gff-gtf-specification">GFF/GTF specification</a></li>

      <li><a href="#supported-sources">Supported sources</a></li>

      <li><a href="#ensembl">Ensembl</a></li>

      <li><a href="#gencode-vs-ensembl">GENCODE vs. Ensembl</a></li>

      <li><a href="#refseq">RefSeq</a></li>

      <li><a href="#ucsc">UCSC</a></li>

      <li><a href="#example-urls">Example URLs</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

