<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make <code>GRanges</code> from Ensembl — makeGRangesFromEnsembl • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Make <code>GRanges</code> from Ensembl — makeGRangesFromEnsembl" />

<meta property="og:description" content="Quickly obtain gene and transcript annotations from
Ensembl using
AnnotationHub and
ensembldb." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.15</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make <code>GRanges</code> from Ensembl</h1>
    <small class="dont-index">Source: <a href='https://github.com/acidgenomics/basejump/blob/master/R/makeGRangesFromEnsembl.R'><code>R/makeGRangesFromEnsembl.R</code></a></small>
    <div class="hidden name"><code>makeGRangesFromEnsembl.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Quickly obtain gene and transcript annotations from
<a href='https://www.ensembl.org/'>Ensembl</a> using
<a href='https://bioconductor.org/packages/AnnotationHub/'>AnnotationHub</a> and
<a href='https://bioconductor.org/packages/ensembldb/'>ensembldb</a>.</p>
    
    </div>

    <pre class="usage"><span class='fu'>makeGRangesFromEnsembl</span>(<span class='no'>organism</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='fu'>annotable</span>(<span class='no'>organism</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>organism</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Full Latin organism name (e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Return as genes or transcripts.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Ensembl genome build assembly name (e.g. <code>"GRCh38"</code>). If set <code>NULL</code>,
defaults to the most recent build available. Note: don't pass in UCSC build
IDs (e.g. <code>"hg38"</code>).</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer(1)</a></code>.
Ensembl release version (e.g. <code>90</code>). If set <code>NULL</code>, defaults to the most
recent release available.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>GRanges</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Simply specify the desired organism, using the full latin name. For example,
we can obtain human annotations with <code>Homo sapiens</code>. Optionally, specific
Ensembl genome builds (e.g. <code>GRCh38</code>) and release versions (e.g. <code>87</code>) are
supported.</p>
<p>Under the hood, this function fetches annotations from AnnotationHub using
the ensembldb package. AnnotationHub supports versioned Ensembl releases,
back to version 87.</p>
<p>Genome build: use "<code>GRCh38</code>" instead of "<code>hg38</code>" for the genome build, since
we're querying Ensembl and not UCSC.</p>
    
    <h2 class="hasAnchor" id="functions"><a class="anchor" href="#functions"></a>Functions</h2>

    <ul>
<li><p><code>annotable</code>: Legacy convenience function that calls <code>makeGRangesFromEnsembl()</code> and
returns a <code>tibble</code> instead of <code>GRanges</code>. Note that <code>GRanges</code> can also be
coercing using <code><a href='https://www.rdocumentation.org/packages/BiocGenerics/topics/as.data.frame'>as.data.frame()</a></code>.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="broad-class-definitions"><a class="anchor" href="#broad-class-definitions"></a>Broad class definitions</h2>

    
    <p>For gene and transcript annotations, a <code>broadClass</code> column is added, which
generalizes the gene types into a smaller number of semantically-meaningful
groups:</p><ul>
<li><p><code>coding</code>.</p></li>
<li><p><code>noncoding</code>.</p></li>
<li><p><code>pseudo</code>.</p></li>
<li><p><code>small</code>.</p></li>
<li><p><code>decaying</code>.</p></li>
<li><p><code>ig</code> (immunoglobulin).</p></li>
<li><p><code>tcr</code> (T cell receptor).</p></li>
<li><p><code>other</code>.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="grch-hg-legacy-annotations"><a class="anchor" href="#grch-hg-legacy-annotations"></a>GRCh37 (hg19) legacy annotations</h2>

    
    <p><code>makeGRangesFromEnsembl()</code> supports the legacy <em>Homo sapiens</em> GRCh37 (release
75) build by internally querying the <a href='https://bioconductor.org/packages/EnsDb.Hsapiens.v75/'>EnsDb.Hsapiens.v75</a> package.
Alternatively, the corresponding GTF/GFF file can be loaded directly from
GENCODE or Ensembl.</p>
    
    <h2 class="hasAnchor" id="annotationhub-queries"><a class="anchor" href="#annotationhub-queries"></a>AnnotationHub queries</h2>

    
    <p>Here's how to perform manual, customized AnnotationHub queries.</p><pre>library(AnnotationHub)
library(ensembldb)
ah &lt;- AnnotationHub()
    # Human ensembldb (EnsDb) records.
ahs &lt;- query(
    x = ah,
    pattern = c(
        "Homo sapiens",
        "GRCh38",
        "Ensembl",
        "EnsDb"
    )
)
mcols(ahs)
print(ahs)
# EnsDb (Ensembl GRCh38 94; 2018-10-11)
ah[["AH64923"]]
    # Human UCSC TxDb records.
ahs &lt;- query(
    x = ah,
    pattern = c(
        "Homo sapiens",
        "UCSC",
        "TxDb",
        "knownGene"
    )
)
mcols(ahs)
print(ahs)
# TxDb (UCSC hg38 GENCODE 24; 2016-12-22)
ah[["AH52260"]]
</pre>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><a href='https://bioconductor.org/packages/AnnotationHub/'>AnnotationHub</a>.</p></li>
<li><p><a href='https://bioconductor.org/packages/ensembldb/'>ensembldb</a>.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Genes</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGRangesFromEnsembl</span>(<span class='st'>"Homo sapiens"</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"genes"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>Matching EnsDb from AnnotationHub 2.14.5 (2018-10-24).</span></div><div class='output co'>#&gt; <span class='message'>AH64923: Ensembl 94 EnsDb for Homo sapiens</span>
#&gt; <span class='message'>Run this code to download EnsDb manually:</span>
#&gt; <span class='message'>  &gt; library(AnnotationHub)</span>
#&gt; <span class='message'>  &gt; ah &lt;- AnnotationHub()</span>
#&gt; <span class='message'>  &gt; edb &lt;- ah[["AH64923"]]</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Homo sapiens</span>
#&gt; <span class='message'>  - Genome Build: GRCh38</span>
#&gt; <span class='message'>  - Ensembl Release: 94</span>
#&gt; <span class='message'>  - Level: genes</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneBiotype, geneName, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by geneID.</span></div><div class='output co'>#&gt; <span class='message'>65687 genes detected.</span></div><div class='input'><span class='fu'><a href='summary.html'>summary</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "GRanges object with 65687 ranges and 8 metadata columns"</div><div class='input'>
<span class='co'>## Transcripts</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGRangesFromEnsembl</span>(<span class='st'>"Homo sapiens"</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"transcripts"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>Matching EnsDb from AnnotationHub 2.14.5 (2018-10-24).</span></div><div class='output co'>#&gt; <span class='message'>AH64923: Ensembl 94 EnsDb for Homo sapiens</span>
#&gt; <span class='message'>Run this code to download EnsDb manually:</span>
#&gt; <span class='message'>  &gt; library(AnnotationHub)</span>
#&gt; <span class='message'>  &gt; ah &lt;- AnnotationHub()</span>
#&gt; <span class='message'>  &gt; edb &lt;- ah[["AH64923"]]</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Homo sapiens</span>
#&gt; <span class='message'>  - Genome Build: GRCh38</span>
#&gt; <span class='message'>  - Ensembl Release: 94</span>
#&gt; <span class='message'>  - Level: transcripts</span></div><div class='output co'>#&gt; <span class='message'>Merging gene-level annotations into transcript-level object.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, seqnames, transcriptBiotype</span></div><div class='output co'>#&gt; <span class='message'>Arranging by transcriptID.</span></div><div class='output co'>#&gt; <span class='message'>14 invalid names: LRG_321t1-1, LRG_321t1-2, LRG_321t2-1, LRG_321t2-2, LRG_321t3-1, LRG_321t3-2, LRG_321t4-1, LRG_321t4-2, LRG_321t5-1, LRG_321t5-2, LRG_321t6-1, LRG_321t6-2, LRG_321t7-1, LRG_321t7-2</span></div><div class='output co'>#&gt; <span class='message'>228432 transcripts detected.</span></div><div class='input'><span class='fu'><a href='summary.html'>summary</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "GRanges object with 228432 ranges and 15 metadata columns"</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#functions">Functions</a></li>

      <li><a href="#broad-class-definitions">Broad class definitions</a></li>

      <li><a href="#grch-hg-legacy-annotations">GRCh37 (hg19) legacy annotations</a></li>

      <li><a href="#annotationhub-queries">AnnotationHub queries</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

