<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make GRanges from Ensembl — makeGRangesFromEnsembl • basejump</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Make GRanges from Ensembl — makeGRangesFromEnsembl" />
<meta property="og:description" content="Quickly obtain gene and transcript annotations from
Ensembl using
AnnotationHub and
ensembldb." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.12.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make GRanges from Ensembl</h1>
    <small class="dont-index">Source: <a href='https://github.com/acidgenomics/basejump/blob/master/R/makeGRangesFromEnsembl.R'><code>R/makeGRangesFromEnsembl.R</code></a></small>
    <div class="hidden name"><code>makeGRangesFromEnsembl.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Quickly obtain gene and transcript annotations from
<a href='https://www.ensembl.org/'>Ensembl</a> using
<a href='https://bioconductor.org/packages/AnnotationHub/'>AnnotationHub</a> and
<a href='https://bioconductor.org/packages/ensembldb/'>ensembldb</a>.</p>
    </div>

    <pre class="usage"><span class='fu'>makeGRangesFromEnsembl</span>(
  <span class='no'>organism</span>,
  <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ignoreTxVersion</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)

<span class='fu'>annotable</span>(
  <span class='no'>organism</span>,
  <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ignoreTxVersion</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>organism</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Full Latin organism name (e.g. <code>"Homo sapiens"</code>).</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Return as genes or transcripts.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Ensembl genome build assembly name (e.g. <code>"GRCh38"</code>). If set <code>NULL</code>,
defaults to the most recent build available.
Note: don't pass in UCSC build IDs (e.g. <code>"hg38"</code>).</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p><code><a href='https://rdrr.io/r/base/integer.html'>integer(1)</a></code>.
Ensembl release version (e.g. <code>90</code>). If set <code>NULL</code>, defaults to the most
recent release available.</p></td>
    </tr>
    <tr>
      <th>ignoreTxVersion</th>
      <td><p><code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code>. Don't the include the transcript version
in the identifier. Only applies when <code>level = "transcripts"</code>. This
simplifies identifier matching when generating a <code>tx2gene</code> file.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>GRanges</code>.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Simply specify the desired organism, using the full latin name. For example,
we can obtain human annotations with <code>Homo sapiens</code>. Optionally, specific
Ensembl genome builds (e.g. <code>GRCh38</code>) and release versions (e.g. <code>87</code>) are
supported.</p>
<p>Under the hood, this function fetches annotations from AnnotationHub using
the ensembldb package. AnnotationHub supports versioned Ensembl releases,
back to version 87.</p>
<p>Genome build: use <code>"GRCh38"</code> instead of <code>"hg38"</code> for the genome build, since
we're querying Ensembl and not UCSC.</p>
    <h2 class="hasAnchor" id="functions"><a class="anchor" href="#functions"></a>Functions</h2>

    
<ul>
<li><p><code>annotable</code>: Legacy convenience function that calls <code>makeGRangesFromEnsembl()</code> and returns
a <code>tbl_df</code> (tibble) instead of <code>GRanges</code>. Note that <code>GRanges</code> can also be
coercing using <code><a href='https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html'>as.data.frame()</a></code>.</p></li>
</ul>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Updated 2020-01-20.</p>
    <h2 class="hasAnchor" id="broad-class-definitions"><a class="anchor" href="#broad-class-definitions"></a>Broad class definitions</h2>

    


<p>For gene and transcript annotations, a <code>broadClass</code> column is added, which
generalizes the gene types into a smaller number of semantically-meaningful
groups:</p><ul>
<li><p><code>coding</code>.</p></li>
<li><p><code>noncoding</code>.</p></li>
<li><p><code>pseudo</code>.</p></li>
<li><p><code>small</code>.</p></li>
<li><p><code>decaying</code>.</p></li>
<li><p><code>ig</code> (immunoglobulin).</p></li>
<li><p><code>tcr</code> (T cell receptor).</p></li>
<li><p><code>other</code>.</p></li>
</ul>

    <h2 class="hasAnchor" id="grch-hg-legacy-annotations"><a class="anchor" href="#grch-hg-legacy-annotations"></a>GRCh37 (hg19) legacy annotations</h2>

    


<p><code>makeGRangesFromEnsembl()</code> supports the legacy <em>Homo sapiens</em> GRCh37 (release
75) build by internally querying the <a href='https://bioconductor.org/packages/EnsDb.Hsapiens.v75/'>EnsDb.Hsapiens.v75</a> package.
Alternatively, the corresponding GTF/GFF file can be loaded directly from
GENCODE or Ensembl.</p>
    <h2 class="hasAnchor" id="annotationhub-queries"><a class="anchor" href="#annotationhub-queries"></a>AnnotationHub queries</h2>

    


<p>Here's how to perform manual, customized AnnotationHub queries.</p><pre><span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>AnnotationHub</span>)
<span class='fu'><a href='https://rdrr.io/r/base/library.html'>library</a></span>(<span class='no'>ensembldb</span>)
<span class='no'>ah</span> <span class='kw'>&amp;</span><span class='no'>lt</span>;- <span class='fu'><a href='https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html'>AnnotationHub</a></span>()

<span class='co'># Human ensembldb (EnsDb) records.</span>
<span class='no'>ahs</span> <span class='kw'>&amp;</span><span class='no'>lt</span>;- <span class='fu'><a href='https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html'>query</a></span>(
    <span class='kw'>x</span> <span class='kw'>=</span> <span class='no'>ah</span>,
    <span class='kw'>pattern</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(
        <span class='st'>"Homo sapiens"</span>,
        <span class='st'>"GRCh38"</span>,
        <span class='st'>"Ensembl"</span>,
        <span class='st'>"EnsDb"</span>
    )
)
<span class='fu'><a href='https://rdrr.io/pkg/S4Vectors/man/Vector-class.html'>mcols</a></span>(<span class='no'>ahs</span>)
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span>(<span class='no'>ahs</span>)
<span class='co'># EnsDb (Ensembl GRCh38 94; 2018-10-11)</span>
<span class='no'>ah</span><span class='kw'>[[</span><span class='st'>"AH64923"</span>]]

<span class='co'># Human UCSC TxDb records.</span>
<span class='no'>ahs</span> <span class='kw'>&amp;</span><span class='no'>lt</span>;- <span class='fu'><a href='https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html'>query</a></span>(
    <span class='kw'>x</span> <span class='kw'>=</span> <span class='no'>ah</span>,
    <span class='kw'>pattern</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(
        <span class='st'>"Homo sapiens"</span>,
        <span class='st'>"UCSC"</span>,
        <span class='st'>"TxDb"</span>,
        <span class='st'>"knownGene"</span>
    )
)
<span class='fu'><a href='https://rdrr.io/pkg/S4Vectors/man/Vector-class.html'>mcols</a></span>(<span class='no'>ahs</span>)
<span class='fu'><a href='https://rdrr.io/r/base/print.html'>print</a></span>(<span class='no'>ahs</span>)
<span class='co'># TxDb (UCSC hg38 GENCODE 24; 2016-12-22)</span>
<span class='no'>ah</span><span class='kw'>[[</span><span class='st'>"AH52260"</span>]]</pre>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<ul>
<li><p><a href='https://bioconductor.org/packages/AnnotationHub/'>AnnotationHub</a>.</p></li>
<li><p><a href='https://bioconductor.org/packages/ensembldb/'>ensembldb</a>.</p></li>
</ul>
</div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## Genes</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGRangesFromEnsembl</span>(<span class='st'>"Homo sapiens"</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"genes"</span>)</div><div class='output co'>#&gt; <span class='message'>→ Making <span style='color: #000000;background-color: #BBBBBB;'>`GRanges`</span><span> from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>→ Matching </span><span style='color: #000000;background-color: #BBBBBB;'>`EnsDb`</span><span> from </span><span style='color: #0000BB;'>AnnotationHub</span><span> 2.20.0 (2020-04-27).</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='color: #0000BB;'>AH78783</span><span>: Ensembl 99 EnsDb for Homo sapiens.</span></div><div class='output co'>#&gt; <span class='message'>→ Making </span><span style='color: #000000;background-color: #BBBBBB;'>`GRanges`</span><span> from </span><span style='color: #000000;background-color: #BBBBBB;'>`EnsDb`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'>    Organism: Homo sapiens</span></div><div class='output co'>#&gt; <span class='message'>    Genome build: GRCh38</span></div><div class='output co'>#&gt; <span class='message'>    Ensembl release: 99</span></div><div class='output co'>#&gt; <span class='message'>    Level: genes</span></div><div class='output co'>#&gt; <span class='message'>→ Defining </span><span style='color: #000000;background-color: #BBBBBB;'>`broadClass`</span><span> using: </span><span style='color: #000000;background-color: #BBBBBB;'>`geneBiotype, geneName, seqnames`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'>→ Arranging by </span><span style='color: #000000;background-color: #BBBBBB;'>`geneID`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> 67998 genes detected.</span></div><div class='input'><span class='fu'><a href='summary.html'>summary</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "GRanges object with 67998 ranges and 8 metadata columns"</div><div class='input'>
<span class='co'>## Transcripts</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGRangesFromEnsembl</span>(<span class='st'>"Homo sapiens"</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"transcripts"</span>)</div><div class='output co'>#&gt; <span class='message'>→ Making </span><span style='color: #000000;background-color: #BBBBBB;'>`GRanges`</span><span> from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>→ Matching </span><span style='color: #000000;background-color: #BBBBBB;'>`EnsDb`</span><span> from </span><span style='color: #0000BB;'>AnnotationHub</span><span> 2.20.0 (2020-04-27).</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> </span><span style='color: #0000BB;'>AH78783</span><span>: Ensembl 99 EnsDb for Homo sapiens.</span></div><div class='output co'>#&gt; <span class='message'>→ Making </span><span style='color: #000000;background-color: #BBBBBB;'>`GRanges`</span><span> from </span><span style='color: #000000;background-color: #BBBBBB;'>`EnsDb`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'>    Organism: Homo sapiens</span></div><div class='output co'>#&gt; <span class='message'>    Genome build: GRCh38</span></div><div class='output co'>#&gt; <span class='message'>    Ensembl release: 99</span></div><div class='output co'>#&gt; <span class='message'>    Level: transcripts</span></div><div class='output co'>#&gt; <span class='message'>→ Merging gene-level annotations: </span><span style='color: #000000;background-color: #BBBBBB;'>`geneName, geneBiotype, seqCoordSystem, description, geneIDVersion, symbol, entrezid`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'>→ Defining </span><span style='color: #000000;background-color: #BBBBBB;'>`broadClass`</span><span> using: </span><span style='color: #000000;background-color: #BBBBBB;'>`geneName, seqnames, transcriptBiotype`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'>→ Arranging by </span><span style='color: #000000;background-color: #BBBBBB;'>`transcriptID`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #BBBB00;'>!</span><span> 14 invalid names: </span><span style='color: #000000;background-color: #BBBBBB;'>`LRG_321t1-1, LRG_321t1-2, LRG_321t2-1, LRG_321t2-2, LRG_321t3-1, LRG_321t3-2, LRG_321t4-1, LRG_3....`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> 250641 transcripts detected.</span></div><div class='input'><span class='fu'><a href='summary.html'>summary</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "GRanges object with 250641 ranges and 16 metadata columns"</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


