<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make Single-Cell Experiment — makeSingleCellExperiment • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Make Single-Cell Experiment — makeSingleCellExperiment" />

<meta property="og:description" content="This function is a utility wrapper for
SummarizedExperiment::SummarizedExperiment() that provides automatic
subsetting for row and column data, as well as automatic handling of
transgenes and spike-ins. Additionally, it improves upon the standard
constructor by slotting useful session information into the metadata()
slot:" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.8.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make Single-Cell Experiment</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/makeSingleCellExperiment.R'><code>R/makeSingleCellExperiment.R</code></a></small>
    <div class="hidden name"><code>makeSingleCellExperiment.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function is a utility wrapper for
<code><a href='https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class'>SummarizedExperiment::SummarizedExperiment()</a></code> that provides automatic
subsetting for row and column data, as well as automatic handling of
transgenes and spike-ins. Additionally, it improves upon the standard
constructor by slotting useful session information into the <code>metadata()</code>
slot:</p>
    
    </div>

    <pre class="usage"><span class='fu'>makeSingleCellExperiment</span>(<span class='no'>assays</span>, <span class='no'>rowRanges</span>, <span class='no'>colData</span>, <span class='no'>metadata</span>,
  <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>spikeNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>assays</th>
      <td><p><code>list</code>. Count matrices, which must have matching dimensions.
Counts can be passed in as either a dense matrix (<code>matrix</code>) or sparse
matrix (<code>sparseMatrix</code>).</p></td>
    </tr>
    <tr>
      <th>rowRanges</th>
      <td><p><code>GRanges</code>. Genomic ranges (e.g. genome annotations).
Metadata describing the assay rows.</p></td>
    </tr>
    <tr>
      <th>colData</th>
      <td><p><code>DataFrame</code>. Metadata describing the assay columns. For bulk
RNA-seq, this data describes the samples. For single-cell RNA-seq, this
data describes the cells.</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p><code>list</code>. Metadata.</p></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code>. Vector indicating which assay
rows denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>spikeNames</th>
      <td><p><code>character</code>. Vector indicating which assay rows denote
spike-in sequences (e.g. ERCCs).</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>SingleCellExperiment</code>.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <ul>
<li><p><code>date</code>: Today's date, returned from <code><a href='https://www.rdocumentation.org/packages/base/topics/Sys.time'>Sys.Date()</a></code>.</p></li>
<li><p><code>wd</code>: Working directory, returned from <code><a href='https://www.rdocumentation.org/packages/base/topics/getwd'>getwd()</a></code>.</p></li>
<li><p><code>sessionInfo</code>: <code><a href='https://www.rdocumentation.org/packages/sessioninfo/topics/session_info'>sessioninfo::session_info()</a></code> return.</p></li>
</ul>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code><a href='https://www.rdocumentation.org/packages/SummarizedExperiment/topics/RangedSummarizedExperiment-class'>SummarizedExperiment::SummarizedExperiment()</a></code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/SingleCellExperiment/topics/SingleCellExperiment'>SingleCellExperiment::SingleCellExperiment()</a></code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/utils/topics/help'>help("RangedSummarizedExperiment-class", "SummarizedExperiment")</a></code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/utils/topics/help'>help("SummarizedExperiment-class", "SummarizedExperiment")</a></code>.</p></li>
<li><p><code><a href='https://www.rdocumentation.org/packages/utils/topics/help'>help("SingleCellExperiment-class", "SingleCellExperiment")</a></code>.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>library</a></span>(<span class='no'>SingleCellExperiment</span>)
<span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/data'>data</a></span>(<span class='no'>sce</span>)
<span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='no'>sce</span>
<span class='no'>assays</span> <span class='kw'>&lt;-</span> <span class='fu'>assays</span>(<span class='no'>object</span>)
<span class='no'>rowRanges</span> <span class='kw'>&lt;-</span> <span class='fu'>rowRanges</span>(<span class='no'>object</span>)
<span class='no'>colData</span> <span class='kw'>&lt;-</span> <span class='fu'>colData</span>(<span class='no'>object</span>)
<span class='no'>metadata</span> <span class='kw'>&lt;-</span> <span class='fu'>metadata</span>(<span class='no'>object</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeSingleCellExperiment</span>(
    <span class='kw'>assays</span> <span class='kw'>=</span> <span class='no'>assays</span>,
    <span class='kw'>rowRanges</span> <span class='kw'>=</span> <span class='no'>rowRanges</span>,
    <span class='kw'>colData</span> <span class='kw'>=</span> <span class='no'>colData</span>,
    <span class='kw'>metadata</span> <span class='kw'>=</span> <span class='no'>metadata</span>
)
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; class: SingleCellExperiment 
#&gt; dim: 500 100 
#&gt; metadata(3): date wd sessionInfo
#&gt; assays(1): counts
#&gt; rownames(500): gene1 gene10 ... gene98 gene99
#&gt; rowData names(5): geneID geneName geneBiotype broadClass entrezID
#&gt; colnames(100): cell001 cell002 ... cell099 cell100
#&gt; colData names(2): expLibSize sampleID
#&gt; reducedDimNames(0):
#&gt; spikeNames(0):</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.2.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

