<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make GRanges from a GFF/GTF file — makeGRangesFromGFF • basejump</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  

<meta property="og:title" content="Make GRanges from a GFF/GTF file — makeGRangesFromGFF" />
<meta property="og:description" content="Make GRanges from a GFF/GTF file" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.12.13</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/acidgenomics/basejump/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make GRanges from a GFF/GTF file</h1>
    <small class="dont-index">Source: <a href='https://github.com/acidgenomics/basejump/blob/master/R/makeGRangesFromGFF.R'><code>R/makeGRangesFromGFF.R</code></a></small>
    <div class="hidden name"><code>makeGRangesFromGFF.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Make GRanges from a GFF/GTF file</p>
    </div>

    <pre class="usage"><span class='fu'>makeGRangesFromGFF</span>(
  <span class='no'>file</span>,
  <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>ignoreTxVersion</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>.checkAgainstTxDb</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)

<span class='fu'>makeGRangesFromGTF</span>(
  <span class='no'>file</span>,
  <span class='kw'>level</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"genes"</span>, <span class='st'>"transcripts"</span>),
  <span class='kw'>ignoreTxVersion</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>.checkAgainstTxDb</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>file</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
File path.</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p><code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code>.
Return as genes or transcripts.</p></td>
    </tr>
    <tr>
      <th>ignoreTxVersion</th>
      <td><p><code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code>. Don't the include the transcript version
in the identifier. Only applies when <code>level = "transcripts"</code>. This
simplifies identifier matching when generating a <code>tx2gene</code> file.</p></td>
    </tr>
    <tr>
      <th>.checkAgainstTxDb</th>
      <td><p><code><a href='https://rdrr.io/r/base/logical.html'>logical(1)</a></code>.
Enable strict mode, intended for development and unit testing only.
Generate an internal <code>TxDb</code> using <code><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGRanges.html'>GenomicFeatures::makeTxDbFromGRanges()</a></code>
and check that the <code><a href='https://rdrr.io/pkg/IRanges/man/range-squeezers.html'>ranges()</a></code>,
<code><a href='https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html'>seqnames()</a></code>, and identifiers defined in
<code><a href='https://rdrr.io/r/base/names.html'>names()</a></code> are identical. Doesn't work for all GFF/GTF files
due to some current limitations in the GenomicFeatures package, so this is
disabled by default. Generally, GenomicFeatures parses GTF files better
than GFF files. However, it's a useful sanity check and should be enabled
if possible.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>GRanges</code>.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Remote URLs and compressed files are supported.</p>
    <h2 class="hasAnchor" id="functions"><a class="anchor" href="#functions"></a>Functions</h2>

    
<ul>
<li><p><code>makeGRangesFromGTF</code>: GTF file extension alias.
Runs the same internal code as <code>makeGRangesFromGFF()</code>.</p></li>
</ul>
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Updated 2020-05-11.</p>
    <h2 class="hasAnchor" id="recommendations"><a class="anchor" href="#recommendations"></a>Recommendations</h2>

    

<ul>
<li><p><strong>Use GTF over GFF3.</strong> We recommend using a GTF file instead of a GFF3
file, when possible. The file format is more compact and easier to parse.</p></li>
<li><p><strong>Use Ensembl over RefSeq.</strong> We generally recommend using Ensembl over
RefSeq, if possible. It's better supported in R and generally used by most
NGS vendors.</p></li>
</ul>

    <h2 class="hasAnchor" id="gff-gtf-specification"><a class="anchor" href="#gff-gtf-specification"></a>GFF/GTF specification</h2>

    


<p>The GFF (General Feature Format) format consists of one line per feature,
each containing 9 columns of data, plus optional track definition lines.</p>
<p>The GTF (General Transfer Format) format is identical to GFF version 2.</p>
<p>The UCSC website has detailed conventions on the GFF3 format, including
the metadata columns.</p>
    <h2 class="hasAnchor" id="feature-type"><a class="anchor" href="#feature-type"></a>Feature type</h2>

    

<ul>
<li><p><code>CDS</code>: <strong>C</strong>o<strong>D</strong>ing <strong>S</strong>sequence. A contiguous sequence that contains a
genomic interval bounded by start and stop codons. CDS refers to the
portion of a genomic DNA sequence that is translated, from the start codon
to the stop codon.</p></li>
<li><p><code>exon</code>: Genomic interval containing 5' UTR (<code>five_prime_UTR</code>), CDS, and
3' UTR (<code>three_prime_UTR</code>).</p></li>
<li><p><code>mRNA</code>: Processed (spliced) mRNA transcript.</p></li>
</ul>

<p>See also:</p><ul>
<li><p><a href='https://pythonhosted.org/gffutils/'>gffutils documentation</a></p></li>
<li><p><a href='https://www.ncbi.nlm.nih.gov/genbank/genomes_gff/'>GenBank GFF documentation</a></p></li>
<li><p><a href='https://ccb.jhu.edu/software/stringtie/gff.shtml'>stringtie GFF documentation</a></p></li>
<li><p><a href='http://gmod.org/wiki/GFF'>gmod.org GFF wiki</a></p></li>
<li><p><a href='https://mblab.wustl.edu/GTF2.html'>Brent Lab GTF2 spec notes</a></p></li>
<li><p><a href='https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md'>Sequence Ontology GFF3 spec notes</a></p></li>
</ul>

    <h2 class="hasAnchor" id="supported-sources"><a class="anchor" href="#supported-sources"></a>Supported sources</h2>

    


<p>Currently <code>makeGRangesFromGFF()</code> supports genomes from these sources:</p><ul>
<li><p>Ensembl (GTF, GFF3).</p></li>
<li><p>GENCODE (GTF, GFF3).</p></li>
<li><p>RefSeq (GTF, GFF3).</p></li>
<li><p>FlyBase (GTF).</p></li>
<li><p>WormBase (GTF).</p></li>
</ul>

    <h2 class="hasAnchor" id="ensembl"><a class="anchor" href="#ensembl"></a>Ensembl</h2>

    


<p>Note that <code><a href='makeGRangesFromEnsembl.html'>makeGRangesFromEnsembl()</a></code> offers native support for Ensembl genome
builds and returns additional useful metadata that isn't defined inside a
GFF/GTF file.</p>
<p>If you must load a GFF/GTF file directly, then use <code>makeGRangesFromGFF()</code>.</p>
    <h2 class="hasAnchor" id="gencode-vs-ensembl"><a class="anchor" href="#gencode-vs-ensembl"></a>GENCODE vs. Ensembl</h2>

    


<p>Annotations available from Ensembl and GENCODE are very similar.</p>
<p>The GENCODE annotation is made by merging the manual gene annotation produced
by the Ensembl-Havana team and the Ensembl-genebuild automated gene
annotation. The GENCODE annotation is the default gene annotation displayed
in the Ensembl browser. The GENCODE releases coincide with the Ensembl
releases, although GENCODE can skip an Ensembl release if there is no update
to the annotation with respect to the previous release. In practical terms,
the GENCODE annotation is essentially identical to the Ensembl annotation.</p>
<p>However, GENCODE handles pseudoautosomal regions (PAR) differently than
Ensembl. The Ensembl GTF file only includes this annotation once, for
chromosome X. However, GENCODE GTF/GFF3 files include the annotation in the
PAR regions of both chromosomes. You'll see these genes contain a "_PAR_Y"
suffix.</p>
<p>Additionally, GENCODE GFF/GTF files import with a gene identifier containing
a suffix, which differs slightly from the Ensembl GFF/GTF spec
(e.g. GENCODE: <code>ENSG00000000003.14</code>; Ensembl: <code>ENSG00000000003</code>).</p>
<p>The <a href='https://www.gencodegenes.org/pages/faq.html'>GENCODE FAQ</a> has additional
details.</p>
    <h2 class="hasAnchor" id="refseq"><a class="anchor" href="#refseq"></a>RefSeq</h2>

    


<p>Refer to the
<a href='ftp://ftp.ncbi.nlm.nih.gov/genomes/README_GFF3.txt'>current RefSeq spec</a>
for details.</p>
<p>See also:</p><ul>
<li><p><a href='https://www.ncbi.nlm.nih.gov/books/NBK50679/'>RefSeq FAQ</a></p></li>
<li><p>ftp://ftp.ncbi.nih.gov/gene/DATA/gene2refseq.gz</p></li>
</ul>

    <h2 class="hasAnchor" id="ucsc"><a class="anchor" href="#ucsc"></a>UCSC</h2>

    


<p>Loading UCSC genome annotations from a GFF/GTF file are
<em>intentionally not supported</em> by this function.</p>
<p>We recommend using a pre-built <code>TxDb</code> package from Bioconductor instead.
For example, load <code>TxDb.Hsapiens.UCSC.hg38.knownGene</code> for hg38.</p>
<p>For reference, note that UCSC doesn't provide direct GFF/GTF file downloads.
Use of the <a href='https://genome.ucsc.edu/cgi-bin/hgTables'>hgTables</a> table
browser is required in a web browser.</p>
<p>Select the following options to download hg38:</p><ul>
<li><p>clade: <code>Mammal</code></p></li>
<li><p>genome: <code>Human</code></p></li>
<li><p>assembly: <code>Dec. 2013 (GRCh38/hg38)</code></p></li>
<li><p>group: <code>Genes and Gene Predictions</code></p></li>
<li><p>track: <code>GENCODE v29</code></p></li>
<li><p>table: <code>knownGene</code></p></li>
<li><p>region: <code>genome</code></p></li>
<li><p>output format: <code>GTF - gene transfer format</code></p></li>
<li><p>output file: <code>&lt;Enter a file name&gt;</code></p></li>
</ul>

<p>Related URLs:</p><ul>
<li><p><a href='http://genome.ucsc.edu/cgi-bin/hgTables'>UCSC hgTables</a></p></li>
<li><p><a href='http://hgdownload.soe.ucsc.edu/downloads.html'>UCSC downloads</a></p></li>
<li><p><a href='ftp://hgdownload.soe.ucsc.edu/goldenPath/hg38/'>UCSC hg38 FTP</a></p></li>
</ul>

    <h2 class="hasAnchor" id="example-urls"><a class="anchor" href="#example-urls"></a>Example URLs</h2>

    

<ul>
<li><p>Ensembl <em>Homo sapiens</em> 95
<a href='ftp://ftp.ensembl.org/pub/release-95/gtf/homo_sapiens/Homo_sapiens.GRCh38.95.gtf.gz'>GTF</a>,
<a href='ftp://ftp.ensembl.org/pub/release-95/gff3/homo_sapiens/Homo_sapiens.GRCh38.95.gff3.gz'>GFF3</a></p></li>
<li><p>GENCODE <em>Homo sapiens</em> 29
<a href='ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_29/gencode.v29.annotation.gtf.gz'>GTF</a>,
<a href='ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_29/gencode.v29.annotation.gff3.gz'>GFF3</a></p></li>
<li><p>RefSeq <em>Homo sapiens</em>
<a href='ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/vertebrate_mammalian/Homo_sapiens/reference/'>GTF and GFF3</a></p></li>
<li><p>FlyBase <em>Drosophila melanogaster</em> r6.24
<a href='ftp://ftp.flybase.net/releases/FB2018_05/dmel_r6.24/gtf/dmel-all-r6.24.gtf.gz'>GTF</a></p></li>
<li><p>WormBase <em>Caenorhabditis elegans</em> WS267
<a href='ftp://ftp.wormbase.org/pub/wormbase/releases/WS267/species/c_elegans/PRJNA13758/c_elegans.PRJNA13758.WS267.canonical_geneset.gtf.gz'>GTF</a></p></li>
</ul>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<ul>
<li><p><code><a href='https://rdrr.io/pkg/rtracklayer/man/export.html'>rtracklayer::import()</a></code>.</p></li>
<li><p><code><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGRanges.html'>GenomicFeatures::makeTxDbFromGRanges()</a></code>.</p></li>
<li><p><code><a href='https://rdrr.io/pkg/GenomicFeatures/man/makeTxDbFromGFF.html'>GenomicFeatures::makeTxDbFromGFF()</a></code>.</p></li>
</ul>
</div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>file</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/pkg/acidbase/man/pasteURL.html'>pasteURL</a></span>(<span class='no'>basejumpTestsURL</span>, <span class='st'>"ensembl.gtf"</span>, <span class='kw'>protocol</span> <span class='kw'>=</span> <span class='st'>"none"</span>)

<span class='co'>## Genes</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGRangesFromGFF</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='no'>file</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"genes"</span>)</div><div class='output co'>#&gt; <span class='message'>→ Making <span style='color: #000000;background-color: #BBBBBB;'>`GRanges`</span><span> from GFF file (</span><span style='color: #0000BB;'>ensembl.gtf</span><span>).</span></div><div class='output co'>#&gt; <span class='message'>→ Importing </span><span style='color: #0000BB;'>ensembl.gtf</span><span> at </span><span style='color: #0000BB;'>https://tests.acidgenomics.com/basejump/v0.12</span><span> using </span><span style='color: #0000BB;'>rtracklayer</span><span>::</span><span style='color: #000000;background-color: #BBBBBB;'>`import()`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> Ensembl GTF detected.</span></div><div class='output co'>#&gt; <span class='message'>→ Defining </span><span style='color: #000000;background-color: #BBBBBB;'>`broadClass`</span><span> using: </span><span style='color: #000000;background-color: #BBBBBB;'>`geneBiotype, geneName, seqnames`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'>→ Arranging by </span><span style='color: #000000;background-color: #BBBBBB;'>`geneID`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> 60 genes detected.</span></div><div class='input'><span class='fu'><a href='summary.html'>summary</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "GRanges object with 60 ranges and 8 metadata columns"</div><div class='input'>
<span class='co'>## Transcripts</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGRangesFromGFF</span>(<span class='kw'>file</span> <span class='kw'>=</span> <span class='no'>file</span>, <span class='kw'>level</span> <span class='kw'>=</span> <span class='st'>"transcripts"</span>)</div><div class='output co'>#&gt; <span class='message'>→ Making </span><span style='color: #000000;background-color: #BBBBBB;'>`GRanges`</span><span> from GFF file (</span><span style='color: #0000BB;'>ensembl.gtf</span><span>).</span></div><div class='output co'>#&gt; <span class='message'>→ Importing </span><span style='color: #0000BB;'>ensembl.gtf</span><span> at </span><span style='color: #0000BB;'>https://tests.acidgenomics.com/basejump/v0.12</span><span> using </span><span style='color: #0000BB;'>rtracklayer</span><span>::</span><span style='color: #000000;background-color: #BBBBBB;'>`import()`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> Ensembl GTF detected.</span></div><div class='output co'>#&gt; <span class='message'>→ Defining </span><span style='color: #000000;background-color: #BBBBBB;'>`broadClass`</span><span> using: </span><span style='color: #000000;background-color: #BBBBBB;'>`geneName, seqnames, transcriptBiotype`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'>→ Arranging by </span><span style='color: #000000;background-color: #BBBBBB;'>`transcriptID`</span><span>.</span></div><div class='output co'>#&gt; <span class='message'></span><span style='color: #00BBBB;'>ℹ</span><span> 167 transcripts detected.</span></div><div class='input'><span class='fu'><a href='summary.html'>summary</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "GRanges object with 167 ranges and 16 metadata columns"</div></span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://acidgenomics.com/'>Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


