<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Make Transcript-to-Gene Mappings — makeTx2Gene • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Make Transcript-to-Gene Mappings — makeTx2Gene" />

<meta property="og:description" content="Make Transcript-to-Gene Mappings" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make Transcript-to-Gene Mappings</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/makeTx2Gene-methods.R'><code>R/makeTx2Gene-methods.R</code></a></small>
    <div class="hidden name"><code>makeTx2Gene.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Make Transcript-to-Gene Mappings</p>
    
    </div>

    <pre class="usage"><span class='fu'>makeTx2GeneFromEnsembl</span>(<span class='no'>organism</span>, <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>release</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='fu'>makeTx2GeneFromEnsDb</span>(<span class='no'>object</span>)

<span class='fu'>makeTx2GeneFromGFF</span>(<span class='no'>file</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>organism</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>. Full Latin organism name
(e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>. Ensembl genome build assembly name (e.g.
<code>"GRCh38"</code>). If set <code>NULL</code>, defaults to the most recent build available.
Note: don't pass in UCSC build IDs (e.g. <code>"hg38"</code>).</p></td>
    </tr>
    <tr>
      <th>release</th>
      <td><p><code>scalar integer</code>. Ensembl release version (e.g. <code>90</code>). If set
<code>NULL</code>, defaults to the most recent release available.</p></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>. File path.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>Tx2Gene</code>.</p>
    
    <h2 class="hasAnchor" id="gff-gtf-file"><a class="anchor" href="#gff-gtf-file"></a>GFF/GTF file</h2>

    
    <p>Remote URLs and compressed files are supported.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code>makeGRanges</code>.</p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## makeTx2GeneFromEnsembl ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeTx2GeneFromEnsembl</span>(<span class='kw'>organism</span> <span class='kw'>=</span> <span class='st'>"Homo sapiens"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from Ensembl.</span></div><div class='output co'>#&gt; <span class='message'>Matching EnsDb from AnnotationHub 2.14.2 (2018-10-24).</span></div><div class='output co'>#&gt; <span class='message'>AH64923: Ensembl 94 EnsDb for Homo sapiens</span></div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Homo sapiens</span>
#&gt; <span class='message'>  - Genome Build: GRCh38</span>
#&gt; <span class='message'>  - Ensembl Release: 94</span>
#&gt; <span class='message'>  - Level: transcripts</span></div><div class='output co'>#&gt; <span class='message'>Using `geneName` column instead of `symbol`.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, transcriptBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by transcriptID.</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Tx2Gene with 228432 rows and 2 columns
#&gt;                    transcriptID          geneID
#&gt;                     &lt;character&gt;     &lt;character&gt;
#&gt; ENST00000000233 ENST00000000233 ENSG00000004059
#&gt; ENST00000000412 ENST00000000412 ENSG00000003056
#&gt; ENST00000000442 ENST00000000442 ENSG00000173153
#&gt; ENST00000001008 ENST00000001008 ENSG00000004478
#&gt; ENST00000001146 ENST00000001146 ENSG00000003137
#&gt; ...                         ...             ...
#&gt; LRG_998t1             LRG_998t1         LRG_998
#&gt; LRG_998t2             LRG_998t2         LRG_998
#&gt; LRG_999t1             LRG_999t1         LRG_999
#&gt; LRG_999t2             LRG_999t2         LRG_999
#&gt; LRG_99t1               LRG_99t1          LRG_99</div><div class='input'>
<span class='co'>## makeTx2GeneFromEnsDb ====</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeTx2GeneFromEnsDb</span>(<span class='st'>"EnsDb.Hsapiens.v75"</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from EnsDb object.</span></div><div class='output co'>#&gt; <span class='message'>  - Organism: Homo sapiens</span>
#&gt; <span class='message'>  - Genome Build: GRCh37</span>
#&gt; <span class='message'>  - Ensembl Release: 75</span>
#&gt; <span class='message'>  - Level: transcripts</span></div><div class='output co'>#&gt; <span class='message'>Using `geneName` column instead of `symbol`.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, transcriptBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by transcriptID.</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Tx2Gene with 215647 rows and 2 columns
#&gt;                    transcriptID          geneID
#&gt;                     &lt;character&gt;     &lt;character&gt;
#&gt; ENST00000000233 ENST00000000233 ENSG00000004059
#&gt; ENST00000000412 ENST00000000412 ENSG00000003056
#&gt; ENST00000000442 ENST00000000442 ENSG00000173153
#&gt; ENST00000001008 ENST00000001008 ENSG00000004478
#&gt; ENST00000001146 ENST00000001146 ENSG00000003137
#&gt; ...                         ...             ...
#&gt; LRG_94t1               LRG_94t1          LRG_94
#&gt; LRG_96t1               LRG_96t1          LRG_96
#&gt; LRG_97t1               LRG_97t1          LRG_97
#&gt; LRG_98t1               LRG_98t1          LRG_98
#&gt; LRG_99t1               LRG_99t1          LRG_99</div><div class='input'>
<span class='co'>## makeTx2GeneFromGFF ====</span>
<span class='co'>## GTF</span>
<span class='no'>file</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/file.path'>file.path</a></span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"example.gtf"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeTx2GeneFromGFF</span>(<span class='no'>file</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from GFF/GTF file.</span></div><div class='output co'>#&gt; <span class='message'>Importing example.gtf using rtracklayer::import().</span></div><div class='output co'>#&gt; <span class='message'>Ensembl GTF detected.</span></div><div class='output co'>#&gt; <span class='message'>20 transcript annotations detected.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, geneBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by transcriptID.</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Tx2Gene with 20 rows and 2 columns
#&gt;                          transcriptID             geneID
#&gt;                           &lt;character&gt;        &lt;character&gt;
#&gt; ENSMUST00000070533 ENSMUST00000070533 ENSMUSG00000051951
#&gt; ENSMUST00000082908 ENSMUST00000082908 ENSMUSG00000064842
#&gt; ENSMUST00000157708 ENSMUST00000157708 ENSMUSG00000088333
#&gt; ENSMUST00000159265 ENSMUST00000159265 ENSMUSG00000051951
#&gt; ENSMUST00000161581 ENSMUST00000161581 ENSMUSG00000089699
#&gt; ...                               ...                ...
#&gt; ENSMUST00000194454 ENSMUST00000194454 ENSMUSG00000102592
#&gt; ENSMUST00000194643 ENSMUST00000194643 ENSMUSG00000102343
#&gt; ENSMUST00000195166 ENSMUST00000195166 ENSMUSG00000103161
#&gt; ENSMUST00000195335 ENSMUST00000195335 ENSMUSG00000103377
#&gt; ENSMUST00000208660 ENSMUST00000208660 ENSMUSG00000025900</div><div class='input'>
<span class='co'>## GFF3</span>
<span class='no'>file</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/file.path'>file.path</a></span>(<span class='no'>basejumpCacheURL</span>, <span class='st'>"example.gff3"</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeTx2GeneFromGFF</span>(<span class='no'>file</span>)</div><div class='output co'>#&gt; <span class='message'>Making GRanges from GFF/GTF file.</span></div><div class='output co'>#&gt; <span class='message'>Importing example.gff3 using rtracklayer::import().</span></div><div class='output co'>#&gt; <span class='message'>Ensembl GFF detected.</span></div><div class='output co'>#&gt; <span class='message'>26 transcript annotations detected.</span></div><div class='output co'>#&gt; <span class='message'>Defining broadClass using: geneName, transcriptBiotype, seqnames</span></div><div class='output co'>#&gt; <span class='message'>Arranging by transcriptID.</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; Tx2Gene with 26 rows and 2 columns
#&gt;                          transcriptID             geneID
#&gt;                           &lt;character&gt;        &lt;character&gt;
#&gt; ENSMUST00000027032 ENSMUST00000027032 ENSMUSG00000025900
#&gt; ENSMUST00000027035 ENSMUST00000027035 ENSMUSG00000025902
#&gt; ENSMUST00000070533 ENSMUST00000070533 ENSMUSG00000051951
#&gt; ENSMUST00000082908 ENSMUST00000082908 ENSMUSG00000064842
#&gt; ENSMUST00000157708 ENSMUST00000157708 ENSMUSG00000088333
#&gt; ...                               ...                ...
#&gt; ENSMUST00000195166 ENSMUST00000195166 ENSMUSG00000103161
#&gt; ENSMUST00000195335 ENSMUST00000195335 ENSMUSG00000103377
#&gt; ENSMUST00000195384 ENSMUST00000195384 ENSMUSG00000102948
#&gt; ENSMUST00000208660 ENSMUST00000208660 ENSMUSG00000025900
#&gt; ENSMUST00000208793 ENSMUST00000208793 ENSMUSG00000025900</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#gff-gtf-file">GFF/GTF file</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

