<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Map Genes — mapGenesToIDs • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Map Genes — mapGenesToIDs" />

<meta property="og:description" content="Take a user-defined gene vector and dynamically map the input to either the
object rownames or the gene names (symbols). These functions are useful for
writing code that needs to handle either gene identifier or gene name input
dynamically (e.g. for single-cell RNA-seq marker analysis)." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Map Genes</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/steinbaugh/basejump/blob/master/R/mapGenes-methods.R'><code>R/mapGenes-methods.R</code></a></small>
    <div class="hidden name"><code>mapGenes.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Take a user-defined gene vector and dynamically map the input to either the
object rownames or the gene names (symbols). These functions are useful for
writing code that needs to handle either gene identifier or gene name input
dynamically (e.g. for single-cell RNA-seq marker analysis).</p>
    
    </div>

    <pre class="usage"><span class='fu'>mapGenesToIDs</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>mapGenesToRownames</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>mapGenesToSymbols</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>mapGenesToRownames</span>(<span class='no'>object</span>, <span class='no'>genes</span>)

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>mapGenesToIDs</span>(<span class='no'>object</span>, <span class='no'>genes</span>)

<span class='co'># S4 method for SummarizedExperiment</span>
<span class='fu'>mapGenesToSymbols</span>(<span class='no'>object</span>, <span class='no'>genes</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>character</code>.</p>
    
    <h2 class="hasAnchor" id="ambiguous-gene-names"><a class="anchor" href="#ambiguous-gene-names"></a>Ambiguous gene names</h2>

    
    <p>Some genomes (e.g. Homo sapiens, Mus musculus) contain duplicated gene names
for multiple gene identifiers. Normally we handle these ambiguous gene names
by sanitizing them with <code><a href='http://www.rdocumentation.org/packages/base/topics/make.names'>base::make.names()</a></code>. If a user requests a gene name
that is duplicated, these functions will return a warning.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Data Functions: <code><a href='aggregate.html'>aggregateCols</a></code>,
  <code><a href='counts.html'>counts</a></code>, <code><a href='gene2symbol.html'>gene2symbol</a></code>,
  <code><a href='makeSummarizedExperiment.html'>makeSummarizedExperiment</a></code>,
  <code>metrics</code>, <code><a href='sampleData.html'>sampleData</a></code>,
  <code><a href='sampleNames.html'>sampleNames</a></code>, <code><a href='selectSamples.html'>selectSamples</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='no'>rse_small</span>
<span class='fu'>head</span>(<span class='fu'>rownames</span>(<span class='no'>object</span>))</div><div class='output co'>#&gt; [1] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419"
#&gt; [4] "ENSG00000000457" "ENSG00000000460" "ENSG00000000938"</div><div class='input'>
<span class='no'>object_symbols</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='convertGenesToSymbols.html'>convertGenesToSymbols</a></span>(<span class='no'>object</span>)
<span class='fu'>head</span>(<span class='fu'>rownames</span>(<span class='no'>object_symbols</span>))</div><div class='output co'>#&gt; [1] "TSPAN6"   "TNMD"     "DPM1"     "SCYL3"    "C1orf112" "FGR"     </div><div class='input'>
<span class='no'>geneIDs</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='fu'>rownames</span>(<span class='no'>object</span>))
<span class='fu'>print</span>(<span class='no'>geneIDs</span>)</div><div class='output co'>#&gt; [1] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419"
#&gt; [4] "ENSG00000000457" "ENSG00000000460" "ENSG00000000938"</div><div class='input'>
<span class='no'>geneNames</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='fu'>as.character</span>(<span class='fu'>rowRanges</span>(<span class='no'>object</span>)$<span class='no'>geneName</span>))
<span class='fu'>print</span>(<span class='no'>geneNames</span>)</div><div class='output co'>#&gt; [1] "TSPAN6"   "TNMD"     "DPM1"     "SCYL3"    "C1orf112" "FGR"     </div><div class='input'>
<span class='co'># Row names</span>
<span class='fu'>mapGenesToRownames</span>(<span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneIDs</span>)</div><div class='output co'>#&gt; [1] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419"
#&gt; [4] "ENSG00000000457" "ENSG00000000460" "ENSG00000000938"</div><div class='input'><span class='fu'>mapGenesToRownames</span>(<span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneNames</span>)</div><div class='output co'>#&gt;            TSPAN6              TNMD              DPM1 
#&gt; "ENSG00000000003" "ENSG00000000005" "ENSG00000000419" 
#&gt;             SCYL3          C1orf112               FGR 
#&gt; "ENSG00000000457" "ENSG00000000460" "ENSG00000000938" </div><div class='input'><span class='fu'>mapGenesToRownames</span>(<span class='no'>object_symbols</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneIDs</span>)</div><div class='output co'>#&gt; ENSG00000000003 ENSG00000000005 ENSG00000000419 ENSG00000000457 
#&gt;        "TSPAN6"          "TNMD"          "DPM1"         "SCYL3" 
#&gt; ENSG00000000460 ENSG00000000938 
#&gt;      "C1orf112"           "FGR" </div><div class='input'><span class='fu'>mapGenesToRownames</span>(<span class='no'>object_symbols</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneNames</span>)</div><div class='output co'>#&gt; [1] "TSPAN6"   "TNMD"     "DPM1"     "SCYL3"    "C1orf112" "FGR"     </div><div class='input'>
<span class='co'># Gene identifiers</span>
<span class='fu'>mapGenesToIDs</span>(<span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneIDs</span>)</div><div class='output co'>#&gt; [1] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419"
#&gt; [4] "ENSG00000000457" "ENSG00000000460" "ENSG00000000938"</div><div class='input'><span class='fu'>mapGenesToIDs</span>(<span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneNames</span>)</div><div class='output co'>#&gt;            TSPAN6              TNMD              DPM1 
#&gt; "ENSG00000000003" "ENSG00000000005" "ENSG00000000419" 
#&gt;             SCYL3          C1orf112               FGR 
#&gt; "ENSG00000000457" "ENSG00000000460" "ENSG00000000938" </div><div class='input'><span class='fu'>mapGenesToIDs</span>(<span class='no'>object_symbols</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneIDs</span>)</div><div class='output co'>#&gt; [1] "ENSG00000000003" "ENSG00000000005" "ENSG00000000419"
#&gt; [4] "ENSG00000000457" "ENSG00000000460" "ENSG00000000938"</div><div class='input'><span class='fu'>mapGenesToIDs</span>(<span class='no'>object_symbols</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneNames</span>)</div><div class='output co'>#&gt;            TSPAN6              TNMD              DPM1 
#&gt; "ENSG00000000003" "ENSG00000000005" "ENSG00000000419" 
#&gt;             SCYL3          C1orf112               FGR 
#&gt; "ENSG00000000457" "ENSG00000000460" "ENSG00000000938" </div><div class='input'>
<span class='co'># Gene names (symbols)</span>
<span class='fu'>mapGenesToSymbols</span>(<span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneIDs</span>)</div><div class='output co'>#&gt; ENSG00000000003 ENSG00000000005 ENSG00000000419 ENSG00000000457 
#&gt;        "TSPAN6"          "TNMD"          "DPM1"         "SCYL3" 
#&gt; ENSG00000000460 ENSG00000000938 
#&gt;      "C1orf112"           "FGR" </div><div class='input'><span class='fu'>mapGenesToSymbols</span>(<span class='no'>object</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneNames</span>)</div><div class='output co'>#&gt; [1] "TSPAN6"   "TNMD"     "DPM1"     "SCYL3"    "C1orf112" "FGR"     </div><div class='input'><span class='fu'>mapGenesToSymbols</span>(<span class='no'>object_symbols</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneIDs</span>)</div><div class='output co'>#&gt; ENSG00000000003 ENSG00000000005 ENSG00000000419 ENSG00000000457 
#&gt;        "TSPAN6"          "TNMD"          "DPM1"         "SCYL3" 
#&gt; ENSG00000000460 ENSG00000000938 
#&gt;      "C1orf112"           "FGR" </div><div class='input'><span class='fu'>mapGenesToSymbols</span>(<span class='no'>object_symbols</span>, <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>geneNames</span>)</div><div class='output co'>#&gt; [1] "TSPAN6"   "TNMD"     "DPM1"     "SCYL3"    "C1orf112" "FGR"     </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#ambiguous-gene-names">Ambiguous gene names</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

