<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Parameters — params • basejump</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Parameters — params" />

<meta property="og:description" content="Parameters" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Parameters</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump/blob/master/R/params.R'><code>R/params.R</code></a></small>
    <div class="hidden name"><code>params.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Parameters</p>
    
    </div>

    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>alpha</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/numeric'>numeric(1)</a></code>.
Adjusted P value ("alpha") cutoff. If left <code>NULL</code>, will use the cutoff
defined in the object.</p></td>
    </tr>
    <tr>
      <th>aspectRatio</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer(1)</a></code>.
Aspect ratio.</p></td>
    </tr>
    <tr>
      <th>assay</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/vector'>vector(1)</a></code>.
Name or index of count matrix slotted in
<a href='https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class'>assays</a>. When passing in a string, the name
must be defined in <a href='https://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class'>assayNames</a>.</p></td>
    </tr>
    <tr>
      <th>assays</th>
      <td><p><code>list</code>.
Count matrices, which must have matching dimensions. Counts can be passed
in as either a dense matrix (<code>matrix</code>) or sparse matrix (<code>sparseMatrix</code>).</p></td>
    </tr>
    <tr>
      <th>bpparam</th>
      <td><p><code>bpparamClass</code>.
BiocParallel parameter specifying the back-end to be used for computations.</p>
<p>We recommend using one of the following:</p><ul>
<li><p><a href='https://www.rdocumentation.org/packages/BiocParallel/topics/register'>bpparam</a>.</p></li>
<li><p><a href='https://www.rdocumentation.org/packages/BiocParallel/topics/SerialParam-class'>SerialParam</a>.</p></li>
<li><p><a href='https://www.rdocumentation.org/packages/BiocParallel/topics/MulticoreParam-class'>MulticoreParam</a>.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>censorSamples</th>
      <td><p><code>character</code>.
Specify a subset of samples to censor.</p></td>
    </tr>
    <tr>
      <th>check</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Perform assert checks.</p></td>
    </tr>
    <tr>
      <th>colData</th>
      <td><p><code>DataFrame</code>.
Metadata describing the assay columns. For bulk RNA-seq, this data
describes the samples. For single-cell RNA-seq, this data describes the
cells.</p></td>
    </tr>
    <tr>
      <th>color</th>
      <td><p><code>ScaleDiscrete</code>.
Desired ggplot2 color scale. Must supply discrete values. When set <code>NULL</code>,
the default ggplot2 color palette will be used. If manual color definitions
are desired, we recommend using <code><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>ggplot2::scale_color_manual()</a></code>.</p>
<p>To set the discrete color palette globally, use:</p><pre>options(basejump.color.discrete = ggplot2::scale_color_viridis_d())
</pre></td>
    </tr>
    <tr>
      <th>counts</th>
      <td><p><code>matrix</code>.
Count matrix. Normalized counts are recommended.</p></td>
    </tr>
    <tr>
      <th>countsAxisLabel</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Counts axis label.</p></td>
    </tr>
    <tr>
      <th>dark</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Plot against a dark background using the <code>basejump::theme_midnight()</code>
ggplot2 theme.</p></td>
    </tr>
    <tr>
      <th>dimsUse</th>
      <td><p><code>integer</code>.
Vector of length 2 that denotes the columns from the reduced dimension
matrix to use for <code>centerX</code> and <code>centerY</code> column calculations. Defaults the
first and second dimensions.</p></td>
    </tr>
    <tr>
      <th>dir</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Directory path.</p></td>
    </tr>
    <tr>
      <th>direction</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Plot "<code>both</code>", "<code>up</code>", or "<code>down</code>" directions.</p></td>
    </tr>
    <tr>
      <th>ensemblRelease</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer(1)</a></code>.
Ensembl release version (e.g. <code>90</code>). We recommend setting this value if
possible, for improved reproducibility. When left unset, the latest release
available via AnnotationHub/ensembldb is used. Note that the latest version
available can vary, depending on the versions of AnnotationHub and
ensembldb in use.</p></td>
    </tr>
    <tr>
      <th>envir</th>
      <td><p><code>environment</code>.
Environment to use for assignment. Defaults to
<a href='https://www.rdocumentation.org/packages/base/topics/sys.parent'>parent.frame()</a>, which will assign into the calling
environment.</p></td>
    </tr>
    <tr>
      <th>expression</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Calculation to apply.
Uses <a href='https://www.rdocumentation.org/packages/base/topics/match.arg'>match.arg()</a> internally and defaults to the first
argument in the <code>character</code> vector.</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
File path.</p></td>
    </tr>
    <tr>
      <th>fill</th>
      <td><p><code>ggproto</code>/<code>ScaleDiscrete</code>.
Desired ggplot2 fill scale. Must supply discrete values. When set to
<code>NULL</code>, the default ggplot2 color palette will be used. If manual color
definitions are desired, we recommend using <code><a href='https://ggplot2.tidyverse.org/reference/scale_manual.html'>ggplot2::scale_fill_manual()</a></code>.</p>
<p>To set the discrete fill palette globally, use:</p><pre>options(basejump.fill.discrete = ggplot2::scale_fill_viridis_d())
</pre></td>
    </tr>
    <tr>
      <th>flip</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Flip x and y axes. Recommended for plots containing many samples.</p></td>
    </tr>
    <tr>
      <th>gene2symbol</th>
      <td><p><code>Gene2Symbol</code>.
Gene-to-symbol mappings. Must contain <code>geneID</code> and <code>geneName</code> columns. See
<code>Gene2Symbol</code> for more information.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p><code>character</code>.
Gene identifiers. It is considered better practice to input the stable gene
identifiers from Ensembl (e.g. "ENSG00000000003") and not the (HGNC) gene
symbols (e.g. "TSPN6"), if possible.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Ensembl genome build assembly name (e.g. <code>"GRCh38"</code>). If set <code>NULL</code>,
defaults to the most recent build available. Note: don't pass in UCSC build
IDs (e.g. <code>"hg38"</code>).</p></td>
    </tr>
    <tr>
      <th>geom</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Plot type.
Uses <a href='https://www.rdocumentation.org/packages/base/topics/match.arg'>match.arg()</a> internally and defaults to the first
argument in the <code>character</code> vector.</p></td>
    </tr>
    <tr>
      <th>gffFile</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
GFF/GTF (General Feature Format) file. Generally, we recommend using a GTF
(GFFv2) instead of a GFFv3 file if possible.</p></td>
    </tr>
    <tr>
      <th>headerLevel</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer(1)</a></code> (<code>1</code>-<code>7</code>).
Markdown header level.</p></td>
    </tr>
    <tr>
      <th>i</th>
      <td><p>Indices specifying elements to extract or replace. Indices are <code>numeric</code> or
<code>character</code> vectors, empty (<code>missing</code>), or <code>NULL</code>.</p>
<p>For more information:</p><pre>help(topic = "Extract", package = "base")
</pre></td>
    </tr>
    <tr>
      <th>inherits</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Should the enclosing frames of the <code>environment</code> be searched?</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code>.
Groups of interest that define the samples. If left unset, defaults to
<code>sampleName</code>.</p></td>
    </tr>
    <tr>
      <th>label</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Superimpose sample text labels on the plot.</p></td>
    </tr>
    <tr>
      <th>labelSize</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer(1)</a></code>.
Size of the text label.</p></td>
    </tr>
    <tr>
      <th>lfcThreshold</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/numeric'>numeric(1)</a></code>.
Log fold change ratio (base 2) cutoff threshold. If left <code>NULL</code>, will use
the cutoff defined in the object.</p></td>
    </tr>
    <tr>
      <th>legend</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Show plot legend.</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Return ranges as "<code>genes</code>" or "<code>transcripts</code>".</p></td>
    </tr>
    <tr>
      <th>limit</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/numeric'>numeric(1)</a></code>.
Threshold to denote on the plot, using a dashed line.</p></td>
    </tr>
    <tr>
      <th>max</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/numeric'>numeric(1)</a></code>.
Recommended maximum value cutoff.</p></td>
    </tr>
    <tr>
      <th>metadata</th>
      <td><p><code>list</code>.
Metadata.</p></td>
    </tr>
    <tr>
      <th>min</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/numeric'>numeric(1)</a></code>.
Recommended minimum value cutoff.</p></td>
    </tr>
    <tr>
      <th>minCounts</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer(1)</a></code>.
Minimum number of counts per gene in the count matrix.</p></td>
    </tr>
    <tr>
      <th>n</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer(1)</a></code>.
Number to include.</p></td>
    </tr>
    <tr>
      <th>ntop</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer(1)</a></code>.
Number of top genes to label.</p></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Full Latin organism name (e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>perMillion</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Display as counts per million.</p></td>
    </tr>
    <tr>
      <th>perSample</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Visualize the distributions per sample.</p></td>
    </tr>
    <tr>
      <th>plotlist</th>
      <td><p><code>list</code>.
List containing <code>ggplot</code> objects.</p></td>
    </tr>
    <tr>
      <th>pointAlpha</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/numeric'>numeric(1)</a></code> (<code>0</code>-<code>1</code>).
Alpha transparency level. Useful when there many points in the dataset
(e.g. single-cell data), and some points can be masked.</p></td>
    </tr>
    <tr>
      <th>pointColor</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Default point color for the plot.</p></td>
    </tr>
    <tr>
      <th>pointScalar</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/integer'>integer(1)</a></code>.
Default point size for the plot.</p></td>
    </tr>
    <tr>
      <th>pointsAsNumbers</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Plot the points as numbers (<code>TRUE</code>) or dots (<code>FALSE</code>).</p></td>
    </tr>
    <tr>
      <th>pointSize</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/numeric'>numeric(1)</a></code>.
Point size for dots in the plot.</p></td>
    </tr>
    <tr>
      <th>prefilter</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Apply prefiltering to remove zero count genes.</p></td>
    </tr>
    <tr>
      <th>progress</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Show progress, using progress bars.</p></td>
    </tr>
    <tr>
      <th>reducedDim</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Name of reduced dimension matrix slotted in
<a href='https://www.rdocumentation.org/packages/SingleCellExperiment/topics/SCE-reducedDims'>reducedDims</a>. Includes TNSE, UMAP, PCA,
for example.</p></td>
    </tr>
    <tr>
      <th>return</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Return type.
Uses <a href='https://www.rdocumentation.org/packages/base/topics/match.arg'>match.arg()</a> internally and defaults to the first
argument in the <code>character</code> vector.</p></td>
    </tr>
    <tr>
      <th>rowData</th>
      <td><p><code>DataFrame</code>.
Metadata describing the assay rows, if genomic ranges are not available.
<em>Use rowRanges (GRanges) instead, if possible</em>.</p></td>
    </tr>
    <tr>
      <th>rowRanges</th>
      <td><p><code>GRanges</code>.
Genomic ranges (e.g. genome annotations). Metadata describing the assay
rows.</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Sample metadata file path. CSV or TSV is preferred, but Excel worksheets
are also supported. Check the documentation for conventions and required
columns.</p></td>
    </tr>
    <tr>
      <th>samples</th>
      <td><p><code>character</code>.
Specify a subset of samples to include.</p></td>
    </tr>
    <tr>
      <th>sigPointColor</th>
      <td><p><code>character</code>.
Color names for labeling upregulated and downregulated genes. Also supports
a character string for labeling DEGs with the same color, regardless of
direction.</p></td>
    </tr>
    <tr>
      <th>sort</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Resort using <code>sort</code>.</p></td>
    </tr>
    <tr>
      <th>spikeNames</th>
      <td><p><code>character</code>.
Vector indicating which assay rows denote spike-in sequences (e.g. ERCCs).</p></td>
    </tr>
    <tr>
      <th>subtitle</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Plot subtitle.</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Plot title.</p></td>
    </tr>
    <tr>
      <th>trans</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Name of the axis scale transformation to apply.</p>
<p>For more information:</p><pre>help(topic = "scale_x_continuous", package = "ggplot2")
</pre></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code>.
Vector indicating which assay rows denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>trendline</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Include trendline on plot.</p></td>
    </tr>
    <tr>
      <th>tx2gene</th>
      <td><p><code>Tx2Gene</code>.
Transcript-to-gene mappings.</p></td>
    </tr>
    <tr>
      <th>url</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/character'>character(1)</a></code>.
Uniform Resource Locator (URL). HTTP or FTP address.</p></td>
    </tr>
    <tr>
      <th>value</th>
      <td><p>Value to assign.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p><code><a href='https://www.rdocumentation.org/packages/base/topics/logical'>logical(1)</a></code>.
Run the function with verbose messages?
<em>It is only recommended to enable this when debugging.</em></p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>No value.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

