<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Parameters — params • basejump</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Parameters — params" />

<meta property="og:description" content="Parameters" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basejump</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.8.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/basejump.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/steinbaugh/basejump/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Parameters</h1>
    <small class="dont-index">Source: <a href='https://github.com/steinbaugh/basejump//blob/master/R/params.R'><code>R/params.R</code></a></small>
    <div class="hidden name"><code>params.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Parameters</p>
    
    </div>

    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>alpha</th>
      <td><p><code>scalar numeric</code> or <code>NULL</code>. Adjusted P value ("alpha") cutoff.
If left <code>NULL</code>, will use the cutoff defined in the object.</p></td>
    </tr>
    <tr>
      <th>aspectRatio</th>
      <td><p><code>scalar integer</code>. Aspect ratio.</p></td>
    </tr>
    <tr>
      <th>assay</th>
      <td><p><code>scalar</code>. Name or index of count matrix slotted in
<code><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class'>SummarizedExperiment::assays()</a></code>. When passing in a <code>string</code>, the name must
be defined in <code><a href='http://www.rdocumentation.org/packages/SummarizedExperiment/topics/SummarizedExperiment-class'>SummarizedExperiment::assayNames()</a></code>.</p></td>
    </tr>
    <tr>
      <th>bpparam</th>
      <td><p><code>bpparamClass</code>. BiocParallel parameter specifying the back-end
to be used for computations. See <code><a href='http://www.rdocumentation.org/packages/BiocParallel/topics/register'>BiocParallel::bpparam()</a></code> for details.</p>
<p>We recommend one of the following:</p><ul>
<li><p><code><a href='http://www.rdocumentation.org/packages/BiocParallel/topics/register'>BiocParallel::bpparam()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/BiocParallel/topics/SerialParam-class'>BiocParallel::SerialParam()</a></code>.</p></li>
<li><p><code><a href='http://www.rdocumentation.org/packages/BiocParallel/topics/MulticoreParam-class'>BiocParallel::MulticoreParam()</a></code>.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>check</th>
      <td><p><code>boolean</code>. Perform assert checks.</p></td>
    </tr>
    <tr>
      <th>color</th>
      <td><p><code>ggproto</code>/<code>ScaleDiscrete</code> or <code>NULL</code>. Desired ggplot2 color
scale. Must supply discrete values. When set to <code>NULL</code>, the default ggplot2
color palette will be used. If manual color definitions are desired, we
recommend using <code><a href='ggplot2.tidyverse.org/reference/scale_manual.html'>ggplot2::scale_color_manual()</a></code>. To set the discrete color
palette globally, use
<code>options(basejump.color.discrete = ggplot2::scale_color_viridis_d())</code>.</p></td>
    </tr>
    <tr>
      <th>counts</th>
      <td><p><code>matrix</code>. Count matrix. Normalized counts are recommended.</p></td>
    </tr>
    <tr>
      <th>countsAxisLabel</th>
      <td><p><code>string</code>. Counts axis label.</p></td>
    </tr>
    <tr>
      <th>dark</th>
      <td><p><code>boolean</code>. Plot against a dark background using
<code><a href='ggplot2-themes.html'>basejump::theme_midnight()</a></code> ggplot2 theme.</p></td>
    </tr>
    <tr>
      <th>dimsUse</th>
      <td><p><code>integer</code>. Vector of length 2 that denotes the columns from
the reduced dimension matrix to use for <code>centerX</code> and <code>centerY</code> column
calculations. Defaults the first and second dimensions.</p></td>
    </tr>
    <tr>
      <th>dir</th>
      <td><p><code>string</code>. Directory path.</p></td>
    </tr>
    <tr>
      <th>direction</th>
      <td><p><code>string</code>. Plot "<code>both</code>", "<code>up</code>", or "<code>down</code>" directions.</p></td>
    </tr>
    <tr>
      <th>ensemblRelease</th>
      <td><p><code>scalar integer</code>. Ensembl release version (e.g. <code>90</code>).</p></td>
    </tr>
    <tr>
      <th>envir</th>
      <td><p><code>environment</code> to use for assignment. Defaults to
<code>parent.frame()</code>, which will assign into the calling environment.</p></td>
    </tr>
    <tr>
      <th>expression</th>
      <td><p><code>string</code>. Calculation to apply. Uses <code><a href='http://www.rdocumentation.org/packages/base/topics/match.arg'>base::match.arg()</a></code>
and defaults to the first argument in the <code>character</code> vector.</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p><code>string</code>. File path.</p></td>
    </tr>
    <tr>
      <th>fill</th>
      <td><p><code>ggproto</code>/<code>ScaleDiscrete</code> or <code>NULL</code>. Desired ggplot2 fill scale.
Must supply discrete values. When set to <code>NULL</code>, the default ggplot2 color
palette will be used. If manual color definitions are desired, we recommend
using <code><a href='ggplot2.tidyverse.org/reference/scale_manual.html'>ggplot2::scale_fill_manual()</a></code>. To set the discrete fill palette
globally, use <code>options(basejump.fill.discrete = scale_fill_viridis_d())</code>.</p></td>
    </tr>
    <tr>
      <th>flip</th>
      <td><p><code>boolean</code>. Flip x and y axes. Recommended for plots containing
many samples.</p></td>
    </tr>
    <tr>
      <th>gene2symbol</th>
      <td><p><code>Gene2Symbol</code>. Gene-to-symbol mappings. Must contain
<code>geneID</code> and <code>geneName</code> columns. See <code><a href='Gene2Symbol.html'>Gene2Symbol()</a></code> for more information.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p><code>character</code>. Gene identifiers. It is considered better practice
to input the stable gene identifiers from Ensembl (e.g. "ENSG00000000003")
and not the (HGNC) gene symbols (e.g. "TSPN6"), if possible.</p></td>
    </tr>
    <tr>
      <th>geom</th>
      <td><p><code>string</code>. Plot type. Uses <code><a href='http://www.rdocumentation.org/packages/base/topics/match.arg'>base::match.arg()</a></code> and defaults to the
first argument in the <code>character</code> vector.</p></td>
    </tr>
    <tr>
      <th>headerLevel</th>
      <td><p><code>scalar integer</code> (<code>1</code>-<code>7</code>). Markdown header level.</p></td>
    </tr>
    <tr>
      <th>i</th>
      <td><p>Indices specifying elements to extract or replace. Indices are
<code>numeric</code> or <code>character</code> vectors, empty (<code>missing</code>), or <code>NULL</code>. See
<code>help(topic = "Extract", package = "base")</code> for more information.</p></td>
    </tr>
    <tr>
      <th>inherits</th>
      <td><p><code>boolean</code>. Should the enclosing frames of the <code>environment</code>
be searched?</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code> or <code>NULL</code>. Groups of interest that
define the samples. If left unset, defaults to <code>sampleName</code>.</p></td>
    </tr>
    <tr>
      <th>label</th>
      <td><p><code>boolean</code>. Superimpose sample text labels on the plot.</p></td>
    </tr>
    <tr>
      <th>labelSize</th>
      <td><p><code>scalar integer</code>. Size of the text label.</p></td>
    </tr>
    <tr>
      <th>lfcThreshold</th>
      <td><p><code>scalar numeric</code> or <code>NULL</code>. Log fold change ratio (base
2) cutoff threshold. If left <code>NULL</code>, will use the cutoff defined in the
object.</p></td>
    </tr>
    <tr>
      <th>legend</th>
      <td><p><code>boolean</code>. Show plot legend.</p></td>
    </tr>
    <tr>
      <th>level</th>
      <td><p><code>string</code>. Return ranges as "<code>genes</code>" or "<code>transcripts</code>".</p></td>
    </tr>
    <tr>
      <th>limit</th>
      <td><p><code>scalar numeric</code>. Threshold to denote on the plot, using a
dashed line.</p></td>
    </tr>
    <tr>
      <th>max</th>
      <td><p><code>scalar numeric</code>. Recommended maximum value cutoff.</p></td>
    </tr>
    <tr>
      <th>min</th>
      <td><p><code>scalar numeric</code>. Recommended minimum value cutoff.</p></td>
    </tr>
    <tr>
      <th>minCounts</th>
      <td><p><code>scalar integer</code>. Minimum number of counts per gene in the
count matrix.</p></td>
    </tr>
    <tr>
      <th>n</th>
      <td><p><code>scalar integer</code>. Number to include.</p></td>
    </tr>
    <tr>
      <th>ntop</th>
      <td><p><code>scalar integer</code>. Number of top genes to label.</p></td>
    </tr>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code>string</code>. Full Latin organism name (e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>perMillion</th>
      <td><p><code>boolean</code>. Display as counts per million.</p></td>
    </tr>
    <tr>
      <th>perSample</th>
      <td><p><code>boolean</code>. Visualize the distributions per sample.</p></td>
    </tr>
    <tr>
      <th>plotlist</th>
      <td><p><code>list</code> containing <code>ggplot</code> objects.</p></td>
    </tr>
    <tr>
      <th>pointAlpha</th>
      <td><p><code>scalar numeric</code> (<code>0</code>-<code>1</code>). Alpha transparency level.
Useful when there many points in the dataset (e.g. single-cell data), and
some points can be masked.</p></td>
    </tr>
    <tr>
      <th>pointColor</th>
      <td><p><code>string</code>. Default point color for the plot.</p></td>
    </tr>
    <tr>
      <th>pointScalar</th>
      <td><p><code>scalar integer</code>. Default point size for the plot.</p></td>
    </tr>
    <tr>
      <th>pointsAsNumbers</th>
      <td><p><code>boolean</code>. Plot the points as numbers (<code>TRUE</code>) or
dots (<code>FALSE</code>).</p></td>
    </tr>
    <tr>
      <th>pointSize</th>
      <td><p><code>scalar numeric</code>. Point size for dots in the plot.</p></td>
    </tr>
    <tr>
      <th>prefilter</th>
      <td><p><code>boolean</code>. Apply prefiltering to remove zero count genes.</p></td>
    </tr>
    <tr>
      <th>progress</th>
      <td><p><code>boolean</code>. Show progress, using progress bars.</p></td>
    </tr>
    <tr>
      <th>reducedDim</th>
      <td><p><code>string</code>. Name of reduced dimension matrix slotted in
<code><a href='http://www.rdocumentation.org/packages/SingleCellExperiment/topics/SCE-reducedDims'>SingleCellExperiment::reducedDims()</a></code>. Includes TNSE, UMAP, PCA, for
example.</p></td>
    </tr>
    <tr>
      <th>return</th>
      <td><p><code>string</code>. Return type. Uses <code><a href='http://www.rdocumentation.org/packages/base/topics/match.arg'>base::match.arg()</a></code>
internally and defaults to the first argument in the <code>character</code> vector.</p></td>
    </tr>
    <tr>
      <th>rowRanges</th>
      <td><p><code>GRanges</code>. Genomic ranges (e.g. genome annotations).</p></td>
    </tr>
    <tr>
      <th>samples</th>
      <td><p><code>character</code> or <code>NULL</code>. Samples to include.</p></td>
    </tr>
    <tr>
      <th>sigPointColor</th>
      <td><p><code>character</code>. Color names for labeling upregulated and
downregulated genes. Also supports a character string for labeling DEGs
with the same color, regardless of direction.</p></td>
    </tr>
    <tr>
      <th>subtitle</th>
      <td><p><code>string</code> or <code>NULL</code>. Plot subtitle.</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p><code>string</code> or <code>NULL</code>. Plot title.</p></td>
    </tr>
    <tr>
      <th>trans</th>
      <td><p><code>string</code>. Name of the axis scale transformation to apply. See
<code>help("scale_x_continuous", "ggplot2")</code> for more information.</p></td>
    </tr>
    <tr>
      <th>trendline</th>
      <td><p><code>boolean</code>. Include trendline on plot.</p></td>
    </tr>
    <tr>
      <th>tx2gene</th>
      <td><p><code>Tx2Gene</code>. Transcript-to-gene mappings.</p></td>
    </tr>
    <tr>
      <th>url</th>
      <td><p><code>string</code>. Uniform Resource Locator (URL). HTTP or FTP address.</p></td>
    </tr>
    <tr>
      <th>value</th>
      <td><p>Value to assign.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p><code>boolean</code>. Run the function with verbose messages? It is only
recommended to enable this when debugging.</p></td>
    </tr>
    <tr>
      <th>x</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>No value.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='/'>Michael Steinbaugh</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

