# EggNOG =======================================================================
#' EggNOG Database Annotations
#'
#' [EggNOG](http://eggnogdb.embl.de) is a database of biological information
#' hosted by the EMBL. It is based on the original idea of COGs (**c**lusters of
#' **o**rthologous **g**roups) and expands that idea to non-supervised
#' orthologous groups constructed from numerous organisms. eggNOG stands for
#' **e**volutionary **g**enealogy of **g**enes: **N**on-supervised
#' **O**rthologous **G**roups.
#'
#' @family S4 classes
#' @author Michael Steinbaugh
#' @inherit EggNOG
#' @export
#'
#' @return This class extends `list` and contains:
#'
#' 1. "`cogFunctionalCategories`": **C**luster of **O**rthologous **G**roups
#'    (COG) functional category information.
#' 2. "`annotations`": up-to-date functional descriptions and categories
#'    for **Eu**karyotes **N**on-supervised **O**rthologous **G**roups (euNOG)
#'    and **N**on-supervised **O**rthologous **G**roups (NOG) protein
#'    identifiers.
#'
#' @seealso [EggNOG()].
#'
#' The [EggNOG README file](http://eggnogdb.embl.de/download/latest/README.txt)
#' contains additional useful reference information.
setClass(Class = "EggNOG", contains = "SimpleDataFrameList")

setValidity(
    Class = "EggNOG",
    method = function(object) {
        assert_are_identical(
            x = names(object),
            y = c("cogFunctionalCategories", "annotations")
        )
        assert_are_identical(
            x = colnames(object[["cogFunctionalCategories"]]),
            y = c("letter", "description")
        )
        assert_are_identical(
            x = colnames(object[["annotations"]]),
            y = c(
                "eggnogID",
                "consensusFunctionalDescription",
                "cogFunctionalCategory"
            )
        )
        .assertHasPrototypeMetadata(object)
        TRUE
    }
)



# Ensembl2Entrez ===============================================================
#' Ensembl-to-Entrez Gene Identifier Mappings
#'
#' Defines 1:1 mappings from Ensembl gene IDs to Entrez IDs. Uses the oldest
#' Entrez ID if there are multiple identifiers that map to an Ensembl gene ID.
#'
#' @family S4 classes
#' @author Michael Steinbaugh
#' @inherit Ensembl2Entrez
#' @export
#'
#' @return Contains a `DataFrame` with `geneID` and `entrezID` columns.
#'
#' @seealso [Ensembl2Entrez()].
setClass(Class = "Ensembl2Entrez", contains = "DataFrame")

setValidity(
    Class = "Ensembl2Entrez",
    method = function(object) {
        assert_are_identical(
            x = colnames(object),
            y = c("geneID", "entrezID")
        )
        assert_are_identical(
            x = rownames(object),
            y = as.character(object[["geneID"]])
        )
        assert_has_no_duplicates(object[["geneID"]])
        assert_is_integer(object[["entrezID"]])
        # Require genome metadata.
        .assertHasGenomeMetadata(object)
        TRUE
    }
)



# Gene2Symbol ==================================================================
#' Gene-to-Symbol Mappings
#'
#' @section Genome metadata:
#'
#' We recommend slotting `organism`, `genomeBuild`, and `ensemblRelease` into
#' [S4Vectors::metadata()].
#'
#' @family S4 classes
#' @author Michael Steinbaugh
#' @inherit Gene2Symbol
#' @export
#'
#' @return Contains a `DataFrame` with `geneID` and `geneName` columns.
#'
#' @seealso [Gene2Symbol()], [makeGene2Symbol].
setClass(Class = "Gene2Symbol", contains = "DataFrame")

setValidity(
    Class = "Gene2Symbol",
    method = function(object) {
        assert_are_identical(
            x = colnames(object),
            y = c("geneID", "geneName")
        )
        assert_has_rows(object)
        # Assert that all columns are character.
        invisible(lapply(object, assert_is_character))
        # Assert that neither column has duplicates.
        invisible(lapply(object, assert_has_no_duplicates))
        stopifnot(all(complete.cases(object)))
        # Require genome metadata.
        .assertHasGenomeMetadata(object)
        TRUE
    }
)



# HGNC2Ensembl =================================================================
#' HGNC-to-Ensembl Gene Identifier Mappings
#'
#' @family S4 classes
#' @author Michael Steinbaugh
#' @inherit HGNC2Ensembl
#' @export
#'
#' @return Contains a `DataFrame` with `hgncID` and `geneID` columns.
#'
#' @seealso [HGNC2Ensembl()].
setClass(Class = "HGNC2Ensembl", contains = "DataFrame")

setValidity(
    Class = "HGNC2Ensembl",
    method = function(object) {
        assert_are_identical(
            x = lapply(object, class),
            y = list(
                hgncID = "integer",
                geneID = "character"
            )
        )
        assert_are_identical(
            x = rownames(object),
            y = NULL
        )
        # Require simple metadata.
        .assertHasPrototypeMetadata(object)
        TRUE
    }
)



# MGI2Ensembl ==================================================================
#' MGI-to-Ensembl Gene Identifier Mappings
#'
#' @family S4 classes
#' @author Michael Steinbaugh
#' @inherit MGI2Ensembl
#' @export
#'
#' @return Contains a `DataFrame` with `mgiID` and `geneID` columns.
#'
#' @seealso [MGI2Ensembl()].
setClass(Class = "MGI2Ensembl", contains = "DataFrame")

setValidity(
    Class = "MGI2Ensembl",
    method = function(object) {
        assert_are_identical(
            x = colnames(object),
            y = c("mgiID", "geneID")
        )
        assert_are_identical(
            x = rownames(object),
            y = NULL
        )
        # Require simple metadata.
        .assertHasPrototypeMetadata(object)
        TRUE
    }
)



# PANTHER ======================================================================
#' PANTHER Database Annotations
#'
#' [PANTHER](http://www.pantherdb.org) gene ontology definitions. PANTHER stands
#' for **P**rotein **AN**alysis **TH**rough **E**volutionary **R**elationships.
#'
#' @family S4 classes
#' @author Michael Steinbaugh
#' @inherit PANTHER
#' @export
#'
#' @return Contains a `DataFrame`.
#'
#' @seealso [PANTHER()].
setClass(Class = "PANTHER", contains = "DataFrame")

setValidity(
    Class = "PANTHER",
    method = function(object) {
        assert_are_identical(
            x = colnames(object),
            y = c(
                "geneID",
                "goBP",
                "goCC",
                "goMF",
                "pantherClass",
                "pantherFamilyName",
                "pantherPathway",
                "pantherSubfamilyID",
                "pantherSubfamilyName"
            )
        )
        .assertHasPrototypeMetadata(object)
        assert_is_subset(
            x = c("organism", "release"),
            y = names(metadata(object))
        )
        TRUE
    }
)



# Tx2Gene ======================================================================
#' Transcript-to-Gene Identifier Mappings
#'
#' @section Genome metadata:
#'
#' We recommend slotting `organism`, `genomeBuild`, and `ensemblRelease` into
#' [S4Vectors::metadata()].
#'
#' @family S4 classes
#' @author Michael Steinbaugh
#' @inherit Tx2Gene
#' @export
#'
#' @return Contains a `DataFrame` with `transcriptID` and `geneID` columns.
#'
#' @seealso [Tx2Gene()], [makeTx2Gene].
setClass(Class = "Tx2Gene", contains = "DataFrame")

setValidity(
    Class = "Tx2Gene",
    method = function(object) {
        assert_has_rows(object)
        assert_are_identical(
            x = colnames(object),
            y = c("transcriptID", "geneID")
        )
        # Assert that all columns are character.
        invisible(lapply(object, assert_is_character))
        # Assert that there are no duplicate transcripts.
        assert_has_no_duplicates(object[["transcriptID"]])
        stopifnot(all(complete.cases(object)))
        # Require genome metadata.
        .assertHasGenomeMetadata(object)
        TRUE
    }
)



# Internal =====================================================================
.prototypeMetadata <- list(
    version = packageVersion("basejump"),
    date = Sys.Date()
)



.assertHasPrototypeMetadata <- function(object) {
    assert_is_subset(
        x = names(.prototypeMetadata),
        y = names(metadata(object))
    )
}



.prototypeGenomeMetadata <- c(
    .prototypeMetadata,
    list(
        organism = character(),
        genomeBuild = character(),
        ensemblRelease = integer()
    )
)



.assertHasGenomeMetadata <- function(object) {
    if (!all(genomeInfo %in% names(metadata(object)))) {
        stop(paste0(
            "Object does not contain genome information.\n",
            "Requires: ", toString(genomeInfo)
        ))
    }
    .assertHasPrototypeMetadata(object)
}
